%% Generated by Sphinx.
\def\sphinxdocclass{report}
\documentclass[letterpaper,10pt,english]{sphinxmanual}
\ifdefined\pdfpxdimen
   \let\sphinxpxdimen\pdfpxdimen\else\newdimen\sphinxpxdimen
\fi \sphinxpxdimen=.75bp\relax
\ifdefined\pdfimageresolution
    \pdfimageresolution= \numexpr \dimexpr1in\relax/\sphinxpxdimen\relax
\fi
%% let collapsible pdf bookmarks panel have high depth per default
\PassOptionsToPackage{bookmarksdepth=5}{hyperref}

\PassOptionsToPackage{booktabs}{sphinx}
\PassOptionsToPackage{colorrows}{sphinx}

\PassOptionsToPackage{warn}{textcomp}
\usepackage[utf8]{inputenc}
\ifdefined\DeclareUnicodeCharacter
% support both utf8 and utf8x syntaxes
  \ifdefined\DeclareUnicodeCharacterAsOptional
    \def\sphinxDUC#1{\DeclareUnicodeCharacter{"#1}}
  \else
    \let\sphinxDUC\DeclareUnicodeCharacter
  \fi
  \sphinxDUC{00A0}{\nobreakspace}
  \sphinxDUC{2500}{\sphinxunichar{2500}}
  \sphinxDUC{2502}{\sphinxunichar{2502}}
  \sphinxDUC{2514}{\sphinxunichar{2514}}
  \sphinxDUC{251C}{\sphinxunichar{251C}}
  \sphinxDUC{2572}{\textbackslash}
\fi
\usepackage{cmap}
\usepackage[T1]{fontenc}
\usepackage{amsmath,amssymb,amstext}
\usepackage{babel}



\usepackage{tgtermes}
\usepackage{tgheros}
\renewcommand{\ttdefault}{txtt}



\usepackage[Bjarne]{fncychap}
\usepackage{sphinx}

\fvset{fontsize=auto}
\usepackage{geometry}


% Include hyperref last.
\usepackage{hyperref}
% Fix anchor placement for figures with captions.
\usepackage{hypcap}% it must be loaded after hyperref.
% Set up styles of URL: it should be placed after hyperref.
\urlstyle{same}


\usepackage{sphinxmessages}
\setcounter{tocdepth}{1}



\title{CBMPy User Guide}
\date{November 01, 2023}
\release{0.8.8}
\author{Brett G. Olivier PhD}
\newcommand{\sphinxlogo}{\sphinxincludegraphics{pysces_cbm1_head.pdf}\par}
\renewcommand{\releasename}{Release}
\makeindex
\begin{document}

\ifdefined\shorthandoff
  \ifnum\catcode`\=\string=\active\shorthandoff{=}\fi
  \ifnum\catcode`\"=\active\shorthandoff{"}\fi
\fi

\pagestyle{empty}
\sphinxmaketitle
\pagestyle{plain}
\sphinxtableofcontents
\pagestyle{normal}
\phantomsection\label{\detokenize{cbmpy::doc}}


\sphinxAtStartPar
Contents:

\sphinxstepscope


\chapter{CBMPy: Installation Guide}
\label{\detokenize{install_doc:cbmpy-installation-guide}}\label{\detokenize{install_doc::doc}}

\section{Introduction}
\label{\detokenize{install_doc:introduction}}

\subsection{Support}
\label{\detokenize{install_doc:support}}
\sphinxAtStartPar
PySCeS CBMPy is Open Source software released under the GNU GPL 3 licence (included with the source code)
and is in constant development. All the latest downloads, documentation and development information
is available on \sphinxstylestrong{SourceForge}: \sphinxurl{http://cbmpy.sourceforge.net}

\sphinxAtStartPar
CBMPy is developed in the Department of Systems Bioinformatiocs at the Vrije Universiteit Amsterdam, as part
of the BeBasic Metatoolkit project, by Brett Olivier (\sphinxhref{mailto:bgoli@users.sourceforge.net}{bgoli@users.sourceforge.net})


\subsection{Python standard library modules}
\label{\detokenize{install_doc:python-standard-library-modules}}
\sphinxAtStartPar
CBMPy is developed and tested on Python 2.7.x. but should also work on Python 3.
The following Python Standard Library modules are used in CBMPy and should be available as part of any CPython
distribution and not require additional installation:

\begin{sphinxVerbatim}[commandchars=\\\{\}]
\PYG{l+s+s1}{\PYGZsq{}}\PYG{l+s+s1}{cPickle}\PYG{l+s+s1}{\PYGZsq{}}\PYG{p}{,} \PYG{l+s+s1}{\PYGZsq{}}\PYG{l+s+s1}{cStringIO}\PYG{l+s+s1}{\PYGZsq{}}\PYG{p}{,} \PYG{l+s+s1}{\PYGZsq{}}\PYG{l+s+s1}{cgi}\PYG{l+s+s1}{\PYGZsq{}}\PYG{p}{,} \PYG{l+s+s1}{\PYGZsq{}}\PYG{l+s+s1}{copy}\PYG{l+s+s1}{\PYGZsq{}}\PYG{p}{,} \PYG{l+s+s1}{\PYGZsq{}}\PYG{l+s+s1}{gc}\PYG{l+s+s1}{\PYGZsq{}}\PYG{p}{,} \PYG{l+s+s1}{\PYGZsq{}}\PYG{l+s+s1}{itertools}\PYG{l+s+s1}{\PYGZsq{}}\PYG{p}{,} \PYG{l+s+s1}{\PYGZsq{}}\PYG{l+s+s1}{locale}\PYG{l+s+s1}{\PYGZsq{}}\PYG{p}{,} \PYG{l+s+s1}{\PYGZsq{}}\PYG{l+s+s1}{math}\PYG{l+s+s1}{\PYGZsq{}}\PYG{p}{,}
\PYG{l+s+s1}{\PYGZsq{}}\PYG{l+s+s1}{multiprocessing}\PYG{l+s+s1}{\PYGZsq{}}\PYG{p}{,} \PYG{l+s+s1}{\PYGZsq{}}\PYG{l+s+s1}{os}\PYG{l+s+s1}{\PYGZsq{}}\PYG{p}{,} \PYG{l+s+s1}{\PYGZsq{}}\PYG{l+s+s1}{pprint}\PYG{l+s+s1}{\PYGZsq{}}\PYG{p}{,} \PYG{l+s+s1}{\PYGZsq{}}\PYG{l+s+s1}{random}\PYG{l+s+s1}{\PYGZsq{}}\PYG{p}{,} \PYG{l+s+s1}{\PYGZsq{}}\PYG{l+s+s1}{re}\PYG{l+s+s1}{\PYGZsq{}}\PYG{p}{,} \PYG{l+s+s1}{\PYGZsq{}}\PYG{l+s+s1}{shutil}\PYG{l+s+s1}{\PYGZsq{}}\PYG{p}{,} \PYG{l+s+s1}{\PYGZsq{}}\PYG{l+s+s1}{subprocess}\PYG{l+s+s1}{\PYGZsq{}}\PYG{p}{,}
\PYG{l+s+s1}{\PYGZsq{}}\PYG{l+s+s1}{time}\PYG{l+s+s1}{\PYGZsq{}}\PYG{p}{,} \PYG{l+s+s1}{\PYGZsq{}}\PYG{l+s+s1}{urllib2}\PYG{l+s+s1}{\PYGZsq{}}\PYG{p}{,} \PYG{l+s+s1}{\PYGZsq{}}\PYG{l+s+s1}{webbrowser}\PYG{l+s+s1}{\PYGZsq{}}\PYG{p}{,} \PYG{l+s+s1}{\PYGZsq{}}\PYG{l+s+s1}{xml}\PYG{l+s+s1}{\PYGZsq{}}
\end{sphinxVerbatim}


\subsection{Required libraries (Python bindings)}
\label{\detokenize{install_doc:required-libraries-python-bindings}}
\sphinxAtStartPar
Besides those mentioned above, the following packages are required for CBMPy’s
core functionality. Note that it is possible to install CBMPy using only
\sphinxstyleemphasis{numpy} but that only very limited subset of functionality is then available.
CBMPy is primarily developed on Microsoft Windows and Ubuntu Linux and where possible the package
name is provided such that can be used with the software center or package manager
\sphinxcode{\sphinxupquote{sudo apt\sphinxhyphen{}get install \textless{}package\textgreater{}}} (please see the man pages for sudo and apt\sphinxhyphen{}get
if you don’t know what this). A comprehensive list of modules are listed at the end of this
document. In the case of external C/C++ libraries the Python bindings should be installed (e.g. libSBML). Many
of these are available in \sphinxstyleemphasis{batteries included} Python distributions and via the pip installer.


\section{Overview}
\label{\detokenize{install_doc:overview}}
\sphinxAtStartPar
CBMPy has been designed to be used as a framework and can be used in different contexts. Here are the
minimal requirements for each role. The fastest way to install the latest official version of CBMPy is
using PyPI: \sphinxurl{https://pypi.python.org/pypi/cbmpy} while development source, binaries, tools and utilities are available from
SourceForge: \sphinxurl{https://sourceforge.net/projects/cbmpy/files}


\subsection{PyPI}
\label{\detokenize{install_doc:pypi}}
\sphinxAtStartPar
Try one of:

\begin{sphinxVerbatim}[commandchars=\\\{\}]
\PYG{n}{pip} \PYG{n}{install} \PYG{n}{cbmpy}
\PYG{n}{easy\PYGZus{}install} \PYG{n}{cbmpy}
\end{sphinxVerbatim}


\subsection{Minimal}
\label{\detokenize{install_doc:minimal}}
\sphinxAtStartPar
Read, write and convert model files as well as view, create and edit model components and annotation.
\begin{itemize}
\item {} 
\sphinxAtStartPar
\sphinxstyleemphasis{numpy}

\item {} 
\sphinxAtStartPar
\sphinxstyleemphasis{libSBML} (Python bindings)

\item {} 
\sphinxAtStartPar
\sphinxstyleemphasis{biopython}

\end{itemize}


\subsection{Normal}
\label{\detokenize{install_doc:normal}}
\sphinxAtStartPar
In addition to the minimal requirements a linear optimizer is required
\begin{quote}
\begin{itemize}
\item {} 
\sphinxAtStartPar
Optimization libraries (one or more of):

\end{itemize}
\begin{itemize}
\item {} 
\sphinxAtStartPar
CPLEX (LP, MILP): \sphinxurl{http:://www.ibm.com}

\item {} 
\sphinxAtStartPar
GLPK  (LP): \sphinxurl{http://tfinley.net/software/pyglpk/}

\end{itemize}
\begin{itemize}
\item {} 
\sphinxAtStartPar
\sphinxstyleemphasis{matplotlib}

\item {} 
\sphinxAtStartPar
\sphinxstyleemphasis{sympy}

\item {} 
\sphinxAtStartPar
\sphinxstyleemphasis{xlwt}

\item {} 
\sphinxAtStartPar
\sphinxstyleemphasis{xlrd}

\end{itemize}
\end{quote}


\subsection{Full}
\label{\detokenize{install_doc:full}}
\sphinxAtStartPar
This includes graphical interface support
\begin{itemize}
\item {} 
\sphinxAtStartPar
\sphinxstyleemphasis{pyqt4}

\item {} 
\sphinxAtStartPar
\sphinxstyleemphasis{wxPython}

\item {} 
\sphinxAtStartPar
\sphinxstyleemphasis{suds}

\item {} 
\sphinxAtStartPar
\sphinxstyleemphasis{scipy}

\item {} 
\sphinxAtStartPar
\sphinxstyleemphasis{h5py}

\item {} 
\sphinxAtStartPar
\sphinxstyleemphasis{networkx}

\end{itemize}


\subsection{User tools}
\label{\detokenize{install_doc:user-tools}}
\sphinxAtStartPar
These are highly recommended user tools but are not required for using CBMPy.
\begin{itemize}
\item {} 
\sphinxAtStartPar
iPython

\item {} 
\sphinxAtStartPar
iPython\sphinxhyphen{}notebook

\item {} 
\sphinxAtStartPar
SCiTE

\end{itemize}


\section{Installing on Ubuntu Linux}
\label{\detokenize{install_doc:installing-on-ubuntu-linux}}

\subsection{Python 2.7 (full install)}
\label{\detokenize{install_doc:python-2-7-full-install}}
\sphinxAtStartPar
First we create a scientific Python workbench using the Ubuntu package manager:

\begin{sphinxVerbatim}[commandchars=\\\{\}]
\PYG{n}{sudo} \PYG{n}{apt}\PYG{o}{\PYGZhy{}}\PYG{n}{get} \PYG{n}{install} \PYG{n}{python}\PYG{o}{\PYGZhy{}}\PYG{n}{dev} \PYG{n}{python}\PYG{o}{\PYGZhy{}}\PYG{n}{numpy} \PYG{n}{python}\PYG{o}{\PYGZhy{}}\PYG{n}{scipy} \PYG{n}{python}\PYG{o}{\PYGZhy{}}\PYG{n}{matplotlib}  \PYG{n}{python}\PYG{o}{\PYGZhy{}}\PYG{n}{pip}
\PYG{n}{sudo} \PYG{n}{apt}\PYG{o}{\PYGZhy{}}\PYG{n}{get} \PYG{n}{install} \PYG{n}{python}\PYG{o}{\PYGZhy{}}\PYG{n}{sympy} \PYG{n}{python}\PYG{o}{\PYGZhy{}}\PYG{n}{suds} \PYG{n}{python}\PYG{o}{\PYGZhy{}}\PYG{n}{xlrd} \PYG{n}{python}\PYG{o}{\PYGZhy{}}\PYG{n}{xlwt} \PYG{n}{python}\PYG{o}{\PYGZhy{}}\PYG{n}{h5py}
\PYG{n}{sudo} \PYG{n}{apt}\PYG{o}{\PYGZhy{}}\PYG{n}{get} \PYG{n}{install} \PYG{n}{python}\PYG{o}{\PYGZhy{}}\PYG{n}{biopython} \PYG{n}{python}\PYG{o}{\PYGZhy{}}\PYG{n}{wxgtk2}\PYG{l+m+mf}{.8} \PYG{n}{python}\PYG{o}{\PYGZhy{}}\PYG{n}{qt4}
\PYG{n}{sudo} \PYG{n}{apt}\PYG{o}{\PYGZhy{}}\PYG{n}{get} \PYG{n}{install} \PYG{n}{ipython} \PYG{n}{ipython}\PYG{o}{\PYGZhy{}}\PYG{n}{notebook}
\end{sphinxVerbatim}


\subsection{libSBML}
\label{\detokenize{install_doc:libsbml}}
\sphinxAtStartPar
Installing libSBML is now easy using PiP and PyPI, first we need some dependencies:

\begin{sphinxVerbatim}[commandchars=\\\{\}]
\PYG{n}{sudo} \PYG{n}{apt}\PYG{o}{\PYGZhy{}}\PYG{n}{get} \PYG{n}{install} \PYG{n}{libxml2} \PYG{n}{libxml2}\PYG{o}{\PYGZhy{}}\PYG{n}{dev}
\PYG{n}{sudo} \PYG{n}{apt}\PYG{o}{\PYGZhy{}}\PYG{n}{get} \PYG{n}{install} \PYG{n}{zlib1g} \PYG{n}{zlib1g}\PYG{o}{\PYGZhy{}}\PYG{n}{dev}
\PYG{n}{sudo} \PYG{n}{apt}\PYG{o}{\PYGZhy{}}\PYG{n}{get} \PYG{n}{install} \PYG{n}{bzip2} \PYG{n}{libbz2}\PYG{o}{\PYGZhy{}}\PYG{n}{dev}
\end{sphinxVerbatim}

\sphinxAtStartPar
Then we can install the latest and greatest with:

\begin{sphinxVerbatim}[commandchars=\\\{\}]
\PYG{n}{sudo} \PYG{n}{pip} \PYG{n}{install} \PYG{n}{python}\PYG{o}{\PYGZhy{}}\PYG{n}{libsbml}
\end{sphinxVerbatim}


\subsection{glpk/python\sphinxhyphen{}glpk}
\label{\detokenize{install_doc:glpk-python-glpk}}
\sphinxAtStartPar
Currently, CBMPy still requires GLPK version 4.47 to work with glpk\sphinxhyphen{}0.3 so it is important \sphinxstyleemphasis{not} to install the latest
using the ubuntu package manager but rather use these instructions. First download the GLPK source and bindings:

\begin{sphinxVerbatim}[commandchars=\\\{\}]
\PYG{n}{https}\PYG{p}{:}\PYG{o}{/}\PYG{o}{/}\PYG{n}{sourceforge}\PYG{o}{.}\PYG{n}{net}\PYG{o}{/}\PYG{n}{projects}\PYG{o}{/}\PYG{n}{cbmpy}\PYG{o}{/}\PYG{n}{files}\PYG{o}{/}\PYG{n}{tools}\PYG{o}{/}\PYG{n}{glpk}\PYG{o}{/}\PYG{n}{glpk}\PYG{o}{\PYGZhy{}}\PYG{l+m+mf}{4.47}\PYG{o}{.}\PYG{n}{tar}\PYG{o}{.}\PYG{n}{gz}
\PYG{n}{https}\PYG{p}{:}\PYG{o}{/}\PYG{o}{/}\PYG{n}{sourceforge}\PYG{o}{.}\PYG{n}{net}\PYG{o}{/}\PYG{n}{projects}\PYG{o}{/}\PYG{n}{cbmpy}\PYG{o}{/}\PYG{n}{files}\PYG{o}{/}\PYG{n}{tools}\PYG{o}{/}\PYG{n}{glpk}\PYG{o}{/}\PYG{n}{glpk}\PYG{o}{\PYGZhy{}}\PYG{l+m+mf}{0.3}\PYG{o}{.}\PYG{n}{tar}\PYG{o}{.}\PYG{n}{bz2}
\end{sphinxVerbatim}

\sphinxAtStartPar
and then install the the following dependency:

\begin{sphinxVerbatim}[commandchars=\\\{\}]
\PYG{n}{sudo} \PYG{n}{apt}\PYG{o}{\PYGZhy{}}\PYG{n}{get} \PYG{n}{install} \PYG{n}{libgmp}\PYG{o}{\PYGZhy{}}\PYG{n}{dev}
\end{sphinxVerbatim}

\sphinxAtStartPar
Unpack the glpk\sphinxhyphen{}4.47 source and make it your current directory:

\begin{sphinxVerbatim}[commandchars=\\\{\}]
\PYG{n}{cd} \PYG{n}{GLPK} \PYG{n}{source} \PYG{p}{(}\PYG{n}{e}\PYG{o}{.}\PYG{n}{g}\PYG{o}{.} \PYG{n}{glpk}\PYG{o}{\PYGZhy{}}\PYG{l+m+mf}{4.47}\PYG{p}{)}
\end{sphinxVerbatim}

\sphinxAtStartPar
Build and install GLPK:

\begin{sphinxVerbatim}[commandchars=\\\{\}]
\PYG{o}{.}\PYG{o}{/}\PYG{n}{configure} \PYG{o}{\PYGZhy{}}\PYG{o}{\PYGZhy{}}\PYG{k}{with}\PYG{o}{\PYGZhy{}}\PYG{n}{gmp}
\PYG{n}{make}
\PYG{n}{make} \PYG{n}{check}
\PYG{n}{sudo} \PYG{n}{make} \PYG{n}{install}
\PYG{n}{sudo} \PYG{n}{ldconfig}
\end{sphinxVerbatim}

\sphinxAtStartPar
Change to the python\sphinxhyphen{}glpk source (glpk\sphinxhyphen{}0.3) and make it the current directory:

\begin{sphinxVerbatim}[commandchars=\\\{\}]
\PYG{n}{cd} \PYG{n}{to} \PYG{n}{python}\PYG{o}{\PYGZhy{}}\PYG{n}{glpk} \PYG{n}{source} \PYG{p}{(}\PYG{n}{glpk}\PYG{o}{\PYGZhy{}}\PYG{l+m+mf}{0.3}\PYG{p}{)}\PYG{p}{:}
\end{sphinxVerbatim}

\sphinxAtStartPar
Then build and install it with the following commands:

\begin{sphinxVerbatim}[commandchars=\\\{\}]
\PYG{n}{make}
\PYG{n}{sudo} \PYG{n}{make} \PYG{n}{install}
\end{sphinxVerbatim}

\sphinxAtStartPar
Finally, install CBMPy either using PyPI:

\begin{sphinxVerbatim}[commandchars=\\\{\}]
\PYG{n}{pip} \PYG{n}{install} \PYG{n}{cbmpy}
     \PYG{o+ow}{or}
\PYG{n}{easy\PYGZus{}install} \PYG{n}{cbmpy}
\end{sphinxVerbatim}

\sphinxAtStartPar
Or download a source file from either:

\begin{sphinxVerbatim}[commandchars=\\\{\}]
\PYG{n}{https}\PYG{p}{:}\PYG{o}{/}\PYG{o}{/}\PYG{n}{pypi}\PYG{o}{.}\PYG{n}{python}\PYG{o}{.}\PYG{n}{org}\PYG{o}{/}\PYG{n}{pypi}\PYG{o}{/}\PYG{n}{cbmpy}
\PYG{n}{https}\PYG{p}{:}\PYG{o}{/}\PYG{o}{/}\PYG{n}{sourceforge}\PYG{o}{.}\PYG{n}{net}\PYG{o}{/}\PYG{n}{projects}\PYG{o}{/}\PYG{n}{cbmpy}\PYG{o}{/}\PYG{n}{files}
\end{sphinxVerbatim}

\sphinxAtStartPar
and install manually:

\begin{sphinxVerbatim}[commandchars=\\\{\}]
\PYG{n}{python} \PYG{n}{setup}\PYG{o}{.}\PYG{n}{py} \PYG{n}{build} \PYG{n}{sdist}
\PYG{n}{sudo} \PYG{n}{python} \PYG{n}{setup}\PYG{o}{.}\PYG{n}{py} \PYG{n}{install}
\end{sphinxVerbatim}


\section{Installing on Microsoft Windows}
\label{\detokenize{install_doc:installing-on-microsoft-windows}}
\sphinxAtStartPar
For the modeller that does not want to customize his installation and install
all of the individual packages by hand there are some \sphinxstyleemphasis{batteries included} Python
distributions which have many (if not all) of the required packages required.

\sphinxAtStartPar
CBMPy is developed using a 64 bit version of \sphinxstyleemphasis{Anaconda Python 2.7} (\sphinxurl{http://continuum.io}) has also been installed on
32 bit \sphinxstyleemphasis{Python(x,y)} and Enthought Python Distribution (EPD) \sphinxurl{http://www.enthought.com}


\subsection{Anaconda}
\label{\detokenize{install_doc:anaconda}}
\sphinxAtStartPar
Most of the required packages can be installed using the \sphinxstyleemphasis{conda} package manager for example:

\begin{sphinxVerbatim}[commandchars=\\\{\}]
\PYG{n}{conda} \PYG{n}{install} \PYG{n}{sympy} \PYG{n}{pyqt4}
\end{sphinxVerbatim}

\sphinxAtStartPar
or using pip (libSBML) or setuptools


\subsection{Python(x,y)}
\label{\detokenize{install_doc:python-x-y}}
\sphinxAtStartPar
Select the following packages from the \sphinxstyleemphasis{Python} branch of the Python(x,y) installation directory:
\begin{itemize}
\item {} 
\sphinxAtStartPar
WxPython

\item {} 
\sphinxAtStartPar
Sympy

\item {} 
\sphinxAtStartPar
NetworkX

\item {} 
\sphinxAtStartPar
xlrd

\item {} 
\sphinxAtStartPar
xlwt

\item {} 
\sphinxAtStartPar
h5py

\item {} 
\sphinxAtStartPar
wxPython

\item {} 
\sphinxAtStartPar
PyQT4

\end{itemize}


\subsection{Installing CBMPy}
\label{\detokenize{install_doc:installing-cbmpy}}
\sphinxAtStartPar
There are two ways to install CBMPy either download the latest release as
source bundle or binary from \sphinxurl{http://cbmpy.sourceforge.net} and unzip or execute from a
a temporary directory (recommended). Or, if you want the latest
(greatest and potentially broken) version grab the latest revision from the
the CBMPy Subversion repository:

\begin{sphinxVerbatim}[commandchars=\\\{\}]
\PYG{n}{svn} \PYG{n}{co} \PYG{n}{http}\PYG{p}{:}\PYG{o}{/}\PYG{o}{/}\PYG{n}{sourceforge}\PYG{o}{.}\PYG{n}{net}\PYG{o}{/}\PYG{n}{p}\PYG{o}{/}\PYG{n}{cbmpy}\PYG{o}{/}\PYG{n}{code}\PYG{o}{/}\PYG{n}{HEAD}\PYG{o}{/}\PYG{n}{tree}\PYG{o}{/}\PYG{n}{trunk}\PYG{o}{/}\PYG{n}{cbmpy} \PYG{n}{cbmpy}
\end{sphinxVerbatim}

\sphinxAtStartPar
In both cases you should should now have a directory that contains a file
\sphinxstyleemphasis{setup.py} which can install by simply typing the following into a Windows shell
(command line):

\begin{sphinxVerbatim}[commandchars=\\\{\}]
\PYG{n}{python} \PYG{n}{setup}\PYG{o}{.}\PYG{n}{py} \PYG{n}{build}
\PYG{n}{python} \PYG{n}{setup}\PYG{o}{.}\PYG{n}{py} \PYG{n}{install}
\end{sphinxVerbatim}


\subsection{Installing libSBML with Python bindings}
\label{\detokenize{install_doc:installing-libsbml-with-python-bindings}}
\sphinxAtStartPar
It is highly recommended to install libSBML which CBMPy uses to provide support
for the Systems Biology Markup Language (SBML). First go to the libSBML download
page \sphinxurl{http://sbml.org/Software/libSBML} page follow the \sphinxstyleemphasis{Download libSBML} \textendash{}\textgreater{} \sphinxstyleemphasis{Stable} \textendash{}\textgreater{}
\sphinxstyleemphasis{Windows} \textendash{}\textgreater{} \sphinxstyleemphasis{32bit} path and download libSBML (e.g. libSBML\sphinxhyphen{}5.10.0\sphinxhyphen{}win\sphinxhyphen{}x86.exe). The latest
stable version can be found at \sphinxurl{http://sbml.org/Software/libSBML}
\begin{quote}

\sphinxAtStartPar
\sphinxurl{http://sourceforge.net/projects/sbml/files/libsbml/5.10.0/stable/Windows/32-bit/libSBML-5.10.0-win-x86.exe/download}
\end{quote}

\sphinxAtStartPar
Run the installer and make sure you select the Python Bindings during installation
or install the appropriate Python bindings that match your Python(x,y) version directly e.g.
(libSBML\sphinxhyphen{}5.10.0\sphinxhyphen{}win\sphinxhyphen{}py2.7\sphinxhyphen{}x86.exe)


\subsection{Optmization (1): IBM cplex optimization studio (Academic)}
\label{\detokenize{install_doc:optmization-1-ibm-cplex-optimization-studio-academic}}
\sphinxAtStartPar
If you have access to the the IBM CPLEX solver. It is a a good idea to use the latest available version.
Again choose the appropriate 32 or 64 bit version and an installation path that suites your setup.
\begin{itemize}
\item {} 
\sphinxAtStartPar
Run \sphinxstylestrong{cplex\_studio126.win\sphinxhyphen{}x86\sphinxhyphen{}32.exe}

\item {} 
\sphinxAtStartPar
Select English language and accept licence

\item {} 
\sphinxAtStartPar
Set “Program” install directory to C:\textbackslash{}ILOG\textbackslash{}CPLEX\_Studio126

\item {} 
\sphinxAtStartPar
Allow default associations to be set and PATH update

\end{itemize}

\sphinxAtStartPar
Once installation is complete we need to install the Python bindings
\begin{itemize}
\item {} 
\sphinxAtStartPar
Open a terminal

\item {} 
\sphinxAtStartPar
Execute \sphinxcode{\sphinxupquote{cd c:\textbackslash{}\textbackslash{}ILOG\textbackslash{}\textbackslash{}CPLEX\_Studio126\textbackslash{}\textbackslash{}cplex\textbackslash{}\textbackslash{}python\textbackslash{}\textbackslash{}x86\_win32}}

\item {} 
\sphinxAtStartPar
Execute \sphinxcode{\sphinxupquote{python setup.py install}}

\end{itemize}


\subsection{Optmization (2): GLPK}
\label{\detokenize{install_doc:optmization-2-glpk}}
\sphinxAtStartPar
CBMPy 0.7.4 includes support for the free, Open Source GLPK solver. This allows access
to CBMPy’s LP functionality (MILP’s requires CPLEX). A port of PyGLPK 0.3
is maintained by the OpenCOBRA project which is mirrored here:
\begin{quote}

\sphinxAtStartPar
\sphinxurl{https://sourceforge.net/projects/cbmpy/files/tools/glpk/}
\end{quote}

\sphinxAtStartPar
Select the binary or source distribution you require and either execute the binary:
\begin{itemize}
\item {} 
\sphinxAtStartPar
Execute \sphinxcode{\sphinxupquote{glpk\sphinxhyphen{}0.3.win32\sphinxhyphen{}py2.7.exe}}

\end{itemize}


\subsection{Testing your new installation}
\label{\detokenize{install_doc:testing-your-new-installation}}
\sphinxAtStartPar
If everything has gone according to plan you can test your installation:
\begin{itemize}
\item {} 
\sphinxAtStartPar
Open a terminal

\item {} 
\sphinxAtStartPar
Execute \sphinxcode{\sphinxupquote{ipython}}

\item {} 
\sphinxAtStartPar
In ipython shell, execute \sphinxcode{\sphinxupquote{import numpy, h5py, xlrd, xlwt}}

\end{itemize}

\sphinxAtStartPar
No import errors should occur.
\begin{itemize}
\item {} 
\sphinxAtStartPar
Execute \sphinxcode{\sphinxupquote{import libsbml}}

\item {} 
\sphinxAtStartPar
Execute \sphinxcode{\sphinxupquote{libsbml.LIBSBML\_VERSION\_STRING}}

\end{itemize}

\sphinxAtStartPar
A successful test should return (for example):

\begin{sphinxVerbatim}[commandchars=\\\{\}]
\PYG{n}{In} \PYG{p}{:} \PYG{n}{libsbml}\PYG{o}{.}\PYG{n}{LIBSBML\PYGZus{}VERSION\PYGZus{}STRING}
\PYG{n}{Out}\PYG{p}{:} \PYG{l+s+s1}{\PYGZsq{}}\PYG{l+s+s1}{51000}\PYG{l+s+s1}{\PYGZsq{}}
\end{sphinxVerbatim}
\begin{itemize}
\item {} 
\sphinxAtStartPar
Execute \sphinxcode{\sphinxupquote{import cbmpy as cbm}}

\end{itemize}

\sphinxAtStartPar
This should return:

\begin{sphinxVerbatim}[commandchars=\\\{\}]
\PYG{n}{In} \PYG{p}{[}\PYG{l+m+mi}{1}\PYG{p}{]}\PYG{p}{:} \PYG{k+kn}{import} \PYG{n+nn}{cbmpy} \PYG{k}{as} \PYG{n+nn}{cbm}

\PYG{o}{*}\PYG{o}{*}\PYG{o}{*}\PYG{o}{*}\PYG{o}{*}\PYG{o}{*}\PYG{o}{*}\PYG{o}{*}\PYG{o}{*}\PYG{o}{*}\PYG{o}{*}\PYG{o}{*}\PYG{o}{*}\PYG{o}{*}\PYG{o}{*}\PYG{o}{*}\PYG{o}{*}\PYG{o}{*}\PYG{o}{*}\PYG{o}{*}\PYG{o}{*}\PYG{o}{*}\PYG{o}{*}\PYG{o}{*}\PYG{o}{*}\PYG{o}{*}\PYG{o}{*}\PYG{o}{*}\PYG{o}{*}\PYG{o}{*}\PYG{o}{*}\PYG{o}{*}\PYG{o}{*}\PYG{o}{*}\PYG{o}{*}\PYG{o}{*}\PYG{o}{*}\PYG{o}{*}\PYG{o}{*}\PYG{o}{*}\PYG{o}{*}\PYG{o}{*}\PYG{o}{*}\PYG{o}{*}\PYG{o}{*}\PYG{o}{*}\PYG{o}{*}\PYG{o}{*}\PYG{o}{*}\PYG{o}{*}\PYG{o}{*}\PYG{o}{*}\PYG{o}{*}\PYG{o}{*}\PYG{o}{*}\PYG{o}{*}\PYG{o}{*}\PYG{o}{*}\PYG{o}{*}\PYG{o}{*}\PYG{o}{*}\PYG{o}{*}\PYG{o}{*}\PYG{o}{*}\PYG{o}{*}\PYG{o}{*}\PYG{o}{*}
\PYG{o}{*} \PYG{n}{Welcome} \PYG{n}{to} \PYG{n}{CBMPy} \PYG{p}{(}\PYG{l+m+mf}{0.7}\PYG{l+m+mf}{.4}\PYG{p}{)} \PYG{o}{\PYGZhy{}} \PYG{n}{PySCeS} \PYG{n}{Constraint} \PYG{n}{Based} \PYG{n}{Modelling}    \PYG{o}{*}
\PYG{o}{*}                \PYG{n}{http}\PYG{p}{:}\PYG{o}{/}\PYG{o}{/}\PYG{n}{cbmpy}\PYG{o}{.}\PYG{n}{sourceforge}\PYG{o}{.}\PYG{n}{net}                     \PYG{o}{*}
\PYG{o}{*} \PYG{n}{Copyright}\PYG{p}{(}\PYG{n}{C}\PYG{p}{)} \PYG{n}{Brett} \PYG{n}{G}\PYG{o}{.} \PYG{n}{Olivier} \PYG{l+m+mi}{2010} \PYG{o}{\PYGZhy{}} \PYG{l+m+mi}{2015}                       \PYG{o}{*}
\PYG{o}{*} \PYG{n}{Dept}\PYG{o}{.} \PYG{n}{of} \PYG{n}{Systems} \PYG{n}{Bioinformatics}                                 \PYG{o}{*}
\PYG{o}{*} \PYG{n}{Vrije} \PYG{n}{Universiteit} \PYG{n}{Amsterdam}\PYG{p}{,} \PYG{n}{Amsterdam}\PYG{p}{,} \PYG{n}{The} \PYG{n}{Netherlands}        \PYG{o}{*}
\PYG{o}{*} \PYG{n}{CBMPy} \PYG{o+ow}{is} \PYG{n}{distributed} \PYG{n}{under} \PYG{n}{the} \PYG{n}{GNU} \PYG{n}{GPL} \PYG{n}{v} \PYG{l+m+mf}{3.0} \PYG{n}{licence}\PYG{p}{,} \PYG{n}{see}       \PYG{o}{*}
\PYG{o}{*} \PYG{n}{LICENCE} \PYG{p}{(}\PYG{n}{supplied} \PYG{k}{with} \PYG{n}{this} \PYG{n}{release}\PYG{p}{)} \PYG{k}{for} \PYG{n}{details}                \PYG{o}{*}
\PYG{o}{*}\PYG{o}{*}\PYG{o}{*}\PYG{o}{*}\PYG{o}{*}\PYG{o}{*}\PYG{o}{*}\PYG{o}{*}\PYG{o}{*}\PYG{o}{*}\PYG{o}{*}\PYG{o}{*}\PYG{o}{*}\PYG{o}{*}\PYG{o}{*}\PYG{o}{*}\PYG{o}{*}\PYG{o}{*}\PYG{o}{*}\PYG{o}{*}\PYG{o}{*}\PYG{o}{*}\PYG{o}{*}\PYG{o}{*}\PYG{o}{*}\PYG{o}{*}\PYG{o}{*}\PYG{o}{*}\PYG{o}{*}\PYG{o}{*}\PYG{o}{*}\PYG{o}{*}\PYG{o}{*}\PYG{o}{*}\PYG{o}{*}\PYG{o}{*}\PYG{o}{*}\PYG{o}{*}\PYG{o}{*}\PYG{o}{*}\PYG{o}{*}\PYG{o}{*}\PYG{o}{*}\PYG{o}{*}\PYG{o}{*}\PYG{o}{*}\PYG{o}{*}\PYG{o}{*}\PYG{o}{*}\PYG{o}{*}\PYG{o}{*}\PYG{o}{*}\PYG{o}{*}\PYG{o}{*}\PYG{o}{*}\PYG{o}{*}\PYG{o}{*}\PYG{o}{*}\PYG{o}{*}\PYG{o}{*}\PYG{o}{*}\PYG{o}{*}\PYG{o}{*}\PYG{o}{*}\PYG{o}{*}\PYG{o}{*}\PYG{o}{*}
\end{sphinxVerbatim}

\sphinxAtStartPar
Exit ipython with CTRL\sphinxhyphen{}D

\sphinxAtStartPar
If you installed CPLEX then try:
\begin{itemize}
\item {} 
\sphinxAtStartPar
Open a terminal

\item {} 
\sphinxAtStartPar
Execute \sphinxcode{\sphinxupquote{ipython}}

\item {} 
\sphinxAtStartPar
Execute \sphinxcode{\sphinxupquote{import cplex}}

\item {} 
\sphinxAtStartPar
Execute \sphinxcode{\sphinxupquote{lp = cplex.Cplex()}}

\item {} 
\sphinxAtStartPar
Execute \sphinxcode{\sphinxupquote{lp.solve()}}

\end{itemize}

\sphinxAtStartPar
A succesful test should return:

\begin{sphinxVerbatim}[commandchars=\\\{\}]
\PYG{n}{In} \PYG{p}{:} \PYG{n}{lp}\PYG{o}{.}\PYG{n}{solve}\PYG{p}{(}\PYG{p}{)}
\PYG{n}{Tried} \PYG{n}{aggregator} \PYG{l+m+mi}{1} \PYG{n}{time}\PYG{o}{.}
\PYG{n}{No} \PYG{n}{LP} \PYG{n}{presolve} \PYG{o+ow}{or} \PYG{n}{aggregator} \PYG{n}{reductions}\PYG{o}{.}
\PYG{n}{Presolve} \PYG{n}{time} \PYG{o}{=}    \PYG{l+m+mf}{0.00} \PYG{n}{sec}\PYG{o}{.}
\end{sphinxVerbatim}

\sphinxAtStartPar
Exit ipython with CTRL\sphinxhyphen{}D

\sphinxAtStartPar
If you installed GLPK then try:
\begin{itemize}
\item {} 
\sphinxAtStartPar
Open a terminal

\item {} 
\sphinxAtStartPar
Execute \sphinxcode{\sphinxupquote{ipython}}

\item {} 
\sphinxAtStartPar
Execute \sphinxcode{\sphinxupquote{import glpk}}

\item {} 
\sphinxAtStartPar
Execute \sphinxcode{\sphinxupquote{lp = glpk.LPX()}}

\end{itemize}

\sphinxAtStartPar
A succesful test should return:

\begin{sphinxVerbatim}[commandchars=\\\{\}]
\PYG{n}{In} \PYG{p}{:} \PYG{n}{glpk}\PYG{o}{.}\PYG{n}{LPX}\PYG{p}{(}\PYG{p}{)}
\PYG{o}{\PYGZlt{}}\PYG{n}{glpk}\PYG{o}{.}\PYG{n}{LPX} \PYG{l+m+mi}{0}\PYG{o}{\PYGZhy{}}\PYG{n}{by}\PYG{o}{\PYGZhy{}}\PYG{l+m+mi}{0} \PYG{n}{at} \PYG{l+m+mh}{0x036C24C8}\PYG{o}{\PYGZgt{}}
\end{sphinxVerbatim}

\sphinxAtStartPar
Exit ipython with CTRL\sphinxhyphen{}D


\subsection{Install CBMPy (http://cbmpy.sourceforge.net)}
\label{\detokenize{install_doc:install-cbmpy-http-cbmpy-sourceforge-net}}
\sphinxAtStartPar
Download the latest version of CBMPy
\begin{itemize}
\item {} 
\sphinxAtStartPar
Run \sphinxstylestrong{cbmpy\sphinxhyphen{}0.7.x.win32.exe} (or newer for 32 bit Windows)

\item {} 
\sphinxAtStartPar
Run \sphinxstylestrong{cbmpy\sphinxhyphen{}0.7.x.amd64.exe} (or newer for 64 bit Windows)

\end{itemize}

\sphinxAtStartPar
Test installation:
\begin{itemize}
\item {} 
\sphinxAtStartPar
Open a terminal

\item {} 
\sphinxAtStartPar
Execute \sphinxcode{\sphinxupquote{ipython}}

\item {} 
\sphinxAtStartPar
Execute \sphinxcode{\sphinxupquote{import cbmpy as cbm}}

\end{itemize}

\sphinxAtStartPar
This should return:

\begin{sphinxVerbatim}[commandchars=\\\{\}]
\PYG{n}{In} \PYG{p}{[}\PYG{l+m+mi}{1}\PYG{p}{]}\PYG{p}{:} \PYG{k+kn}{import} \PYG{n+nn}{cbmpy} \PYG{k}{as} \PYG{n+nn}{cbm}

\PYG{o}{*}\PYG{o}{*}\PYG{o}{*}\PYG{o}{*}\PYG{o}{*}
\PYG{n}{Using} \PYG{n}{GLPK}
\PYG{o}{*}\PYG{o}{*}\PYG{o}{*}\PYG{o}{*}\PYG{o}{*}

\PYG{n}{WX} \PYG{n}{GUI} \PYG{n}{tools} \PYG{n}{available}\PYG{o}{.}
\PYG{n}{Qt4} \PYG{n}{GUI} \PYG{n}{tools} \PYG{n}{available}

\PYG{n}{CBMPy} \PYG{n}{environment}
\PYG{o}{*}\PYG{o}{*}\PYG{o}{*}\PYG{o}{*}\PYG{o}{*}\PYG{o}{*}\PYG{o}{*}\PYG{o}{*}\PYG{o}{*}\PYG{o}{*}\PYG{o}{*}\PYG{o}{*}\PYG{o}{*}\PYG{o}{*}\PYG{o}{*}\PYG{o}{*}\PYG{o}{*}\PYG{o}{*}
\PYG{n}{Revision}\PYG{p}{:} \PYG{n}{r346}


\PYG{o}{*}\PYG{o}{*}\PYG{o}{*}\PYG{o}{*}\PYG{o}{*}\PYG{o}{*}\PYG{o}{*}\PYG{o}{*}\PYG{o}{*}\PYG{o}{*}\PYG{o}{*}\PYG{o}{*}\PYG{o}{*}\PYG{o}{*}\PYG{o}{*}\PYG{o}{*}\PYG{o}{*}\PYG{o}{*}\PYG{o}{*}\PYG{o}{*}\PYG{o}{*}\PYG{o}{*}\PYG{o}{*}\PYG{o}{*}\PYG{o}{*}\PYG{o}{*}\PYG{o}{*}\PYG{o}{*}\PYG{o}{*}\PYG{o}{*}\PYG{o}{*}\PYG{o}{*}\PYG{o}{*}\PYG{o}{*}\PYG{o}{*}\PYG{o}{*}\PYG{o}{*}\PYG{o}{*}\PYG{o}{*}\PYG{o}{*}\PYG{o}{*}\PYG{o}{*}\PYG{o}{*}\PYG{o}{*}\PYG{o}{*}\PYG{o}{*}\PYG{o}{*}\PYG{o}{*}\PYG{o}{*}\PYG{o}{*}\PYG{o}{*}\PYG{o}{*}\PYG{o}{*}\PYG{o}{*}\PYG{o}{*}\PYG{o}{*}\PYG{o}{*}\PYG{o}{*}\PYG{o}{*}\PYG{o}{*}\PYG{o}{*}\PYG{o}{*}\PYG{o}{*}\PYG{o}{*}\PYG{o}{*}\PYG{o}{*}\PYG{o}{*}
\PYG{o}{*} \PYG{n}{Welcome} \PYG{n}{to} \PYG{n}{CBMPy} \PYG{p}{(}\PYG{l+m+mf}{0.7}\PYG{l+m+mf}{.4}\PYG{p}{)} \PYG{o}{\PYGZhy{}} \PYG{n}{PySCeS} \PYG{n}{Constraint} \PYG{n}{Based} \PYG{n}{Modelling}    \PYG{o}{*}
\PYG{o}{*}                \PYG{n}{http}\PYG{p}{:}\PYG{o}{/}\PYG{o}{/}\PYG{n}{cbmpy}\PYG{o}{.}\PYG{n}{sourceforge}\PYG{o}{.}\PYG{n}{net}                     \PYG{o}{*}
\PYG{o}{*} \PYG{n}{Copyright}\PYG{p}{(}\PYG{n}{C}\PYG{p}{)} \PYG{n}{Brett} \PYG{n}{G}\PYG{o}{.} \PYG{n}{Olivier} \PYG{l+m+mi}{2010} \PYG{o}{\PYGZhy{}} \PYG{l+m+mi}{2015}                       \PYG{o}{*}
\PYG{o}{*} \PYG{n}{Dept}\PYG{o}{.} \PYG{n}{of} \PYG{n}{Systems} \PYG{n}{Bioinformatics}                                 \PYG{o}{*}
\PYG{o}{*} \PYG{n}{Vrije} \PYG{n}{Universiteit} \PYG{n}{Amsterdam}\PYG{p}{,} \PYG{n}{Amsterdam}\PYG{p}{,} \PYG{n}{The} \PYG{n}{Netherlands}        \PYG{o}{*}
\PYG{o}{*} \PYG{n}{CBMPy} \PYG{o+ow}{is} \PYG{n}{distributed} \PYG{n}{under} \PYG{n}{the} \PYG{n}{GNU} \PYG{n}{GPL} \PYG{n}{v} \PYG{l+m+mf}{3.0} \PYG{n}{licence}\PYG{p}{,} \PYG{n}{see}       \PYG{o}{*}
\PYG{o}{*} \PYG{n}{LICENCE} \PYG{p}{(}\PYG{n}{supplied} \PYG{k}{with} \PYG{n}{this} \PYG{n}{release}\PYG{p}{)} \PYG{k}{for} \PYG{n}{details}                \PYG{o}{*}
\PYG{o}{*}\PYG{o}{*}\PYG{o}{*}\PYG{o}{*}\PYG{o}{*}\PYG{o}{*}\PYG{o}{*}\PYG{o}{*}\PYG{o}{*}\PYG{o}{*}\PYG{o}{*}\PYG{o}{*}\PYG{o}{*}\PYG{o}{*}\PYG{o}{*}\PYG{o}{*}\PYG{o}{*}\PYG{o}{*}\PYG{o}{*}\PYG{o}{*}\PYG{o}{*}\PYG{o}{*}\PYG{o}{*}\PYG{o}{*}\PYG{o}{*}\PYG{o}{*}\PYG{o}{*}\PYG{o}{*}\PYG{o}{*}\PYG{o}{*}\PYG{o}{*}\PYG{o}{*}\PYG{o}{*}\PYG{o}{*}\PYG{o}{*}\PYG{o}{*}\PYG{o}{*}\PYG{o}{*}\PYG{o}{*}\PYG{o}{*}\PYG{o}{*}\PYG{o}{*}\PYG{o}{*}\PYG{o}{*}\PYG{o}{*}\PYG{o}{*}\PYG{o}{*}\PYG{o}{*}\PYG{o}{*}\PYG{o}{*}\PYG{o}{*}\PYG{o}{*}\PYG{o}{*}\PYG{o}{*}\PYG{o}{*}\PYG{o}{*}\PYG{o}{*}\PYG{o}{*}\PYG{o}{*}\PYG{o}{*}\PYG{o}{*}\PYG{o}{*}\PYG{o}{*}\PYG{o}{*}\PYG{o}{*}\PYG{o}{*}\PYG{o}{*}
\end{sphinxVerbatim}

\sphinxAtStartPar
Exit ipython with CTRL\sphinxhyphen{}D


\section{Linux: Ubuntu}
\label{\detokenize{install_doc:linux-ubuntu}}
\sphinxAtStartPar
On Linux many of the base dependencies are available as packages or from the Python Cheeseshop (\sphinxurl{http://pypi.python.org/pypi}).
For \sphinxstylestrong{libSBML}, \sphinxstylestrong{CPLEX} and/or \sphinxstylestrong{GLPK} please see the \sphinxstyleemphasis{Generic installation on Microsoft Windows (XP, 7, 2008)} for more details.
For example using \sphinxstylestrong{Ubuntu} the base dependencies can be easily installed (depending on what functionality is required).
If you don’t know what these packages are please look them up before installing.

\sphinxAtStartPar
Required:

\begin{sphinxVerbatim}[commandchars=\\\{\}]
\PYG{n}{sudo} \PYG{n}{apt}\PYG{o}{\PYGZhy{}}\PYG{n}{get} \PYG{n}{install} \PYG{n}{python}\PYG{o}{\PYGZhy{}}\PYG{n}{dev} \PYG{n}{python}\PYG{o}{\PYGZhy{}}\PYG{n}{numpy}

\PYG{o}{\PYGZhy{}} \PYG{n}{libSBML} \PYG{k}{for} \PYG{n}{SBML} \PYG{n}{support}\PYG{o}{.}
\end{sphinxVerbatim}

\sphinxAtStartPar
Please see \sphinxurl{http://sbml.org/Software/libSBML} or try the following. Depending on your configurationyou need to install libxml2, bzip2 and their associated “dev” packages:

\begin{sphinxVerbatim}[commandchars=\\\{\}]
\PYG{n}{apt}\PYG{o}{\PYGZhy{}}\PYG{n}{get} \PYG{n}{install} \PYG{n}{libxml2} \PYG{n}{libxml2}\PYG{o}{\PYGZhy{}}\PYG{n}{dev}
\PYG{n}{apt}\PYG{o}{\PYGZhy{}}\PYG{n}{get} \PYG{n}{install} \PYG{n}{zlib1g} \PYG{n}{zlib1g}\PYG{o}{\PYGZhy{}}\PYG{n}{dev}
\PYG{n}{apt}\PYG{o}{\PYGZhy{}}\PYG{n}{get} \PYG{n}{install} \PYG{n}{bzip2} \PYG{n}{libbz2}\PYG{o}{\PYGZhy{}}\PYG{n}{dev}

\PYG{n}{easy\PYGZus{}install} \PYG{n}{pip}

\PYG{c+c1}{\PYGZsh{} for standard libSBML}
\PYG{n}{pip} \PYG{n}{install} \PYG{n}{python}\PYG{o}{\PYGZhy{}}\PYG{n}{libsbml}

\PYG{c+c1}{\PYGZsh{} for \PYGZdq{}experimental\PYGZdq{} libSBML (for FBC V2 and Groups support)}
\PYG{n}{pip} \PYG{n}{install} \PYG{n}{python}\PYG{o}{\PYGZhy{}}\PYG{n}{libsbml}\PYG{o}{\PYGZhy{}}\PYG{n}{experimental}
\end{sphinxVerbatim}
\begin{itemize}
\item {} 
\sphinxAtStartPar
Optimization (at least one of):
\begin{itemize}
\item {} 
\sphinxAtStartPar
IBM CPLEX: \sphinxurl{http://www.ibm.com}

\item {} 
\sphinxAtStartPar
PyGLPK: \sphinxurl{https://sourceforge.net/projects/cbmpy/files/tools/glpk/}

\end{itemize}

\end{itemize}

\sphinxAtStartPar
Please note that due to changes in the GLPK API the current version of PyGLPK (0.3) \sphinxstylestrong{only supports GLPK up
until version 4.47}. If your system has a newer version of GLPK then the current workaround is to uninstall the newer
version and compile 4.47 from source (also available from the above directory). Dependencies are standard Linux build tools
and GMP etc:

\begin{sphinxVerbatim}[commandchars=\\\{\}]
\PYG{n}{tar} \PYG{n}{xzf} \PYG{n}{glpk}\PYG{o}{\PYGZhy{}}\PYG{l+m+mf}{4.47}\PYG{o}{.}\PYG{n}{tar}\PYG{o}{.}\PYG{n}{gz}
\PYG{n}{cd} \PYG{n}{glpk}\PYG{o}{\PYGZhy{}}\PYG{l+m+mf}{4.47}
\PYG{o}{.}\PYG{o}{/}\PYG{n}{configure} \PYG{o}{\PYGZhy{}}\PYG{o}{\PYGZhy{}}\PYG{k}{with}\PYG{o}{\PYGZhy{}}\PYG{n}{gmp}
\PYG{n}{make}
\PYG{n}{make} \PYG{n}{check}
\PYG{n}{sudo} \PYG{n}{make} \PYG{n}{install}
\end{sphinxVerbatim}

\sphinxAtStartPar
Graphical interfaces (highly recommended):

\begin{sphinxVerbatim}[commandchars=\\\{\}]
\PYG{n}{sudo} \PYG{n}{apt}\PYG{o}{\PYGZhy{}}\PYG{n}{get} \PYG{n}{install} \PYG{n}{python}\PYG{o}{\PYGZhy{}}\PYG{n}{wxgtk2}\PYG{l+m+mf}{.8} \PYG{n}{python}\PYG{o}{\PYGZhy{}}\PYG{n}{qt4} \PYG{n}{python}\PYG{o}{\PYGZhy{}}\PYG{n}{matplotlib}
\end{sphinxVerbatim}

\sphinxAtStartPar
Extended IO (highly recommended):

\begin{sphinxVerbatim}[commandchars=\\\{\}]
\PYG{n}{sudo} \PYG{n}{apt}\PYG{o}{\PYGZhy{}}\PYG{n}{get} \PYG{n}{install} \PYG{n}{python}\PYG{o}{\PYGZhy{}}\PYG{n}{xlrd} \PYG{n}{python}\PYG{o}{\PYGZhy{}}\PYG{n}{xlwt} \PYG{n}{python}\PYG{o}{\PYGZhy{}}\PYG{n}{sympy}
\end{sphinxVerbatim}

\sphinxAtStartPar
Web services and database:

\begin{sphinxVerbatim}[commandchars=\\\{\}]
\PYG{n}{sudo} \PYG{n}{apt}\PYG{o}{\PYGZhy{}}\PYG{n}{get} \PYG{n}{install} \PYG{n}{python}\PYG{o}{\PYGZhy{}}\PYG{n}{suds} \PYG{n}{python}\PYG{o}{\PYGZhy{}}\PYG{n}{pysqlite2}
\end{sphinxVerbatim}

\sphinxAtStartPar
Advanced functionality:

\begin{sphinxVerbatim}[commandchars=\\\{\}]
\PYG{n}{sudo} \PYG{n}{apt}\PYG{o}{\PYGZhy{}}\PYG{n}{get} \PYG{n}{install} \PYG{n}{python}\PYG{o}{\PYGZhy{}}\PYG{n}{scipy} \PYG{n}{python}\PYG{o}{\PYGZhy{}}\PYG{n}{h5py} \PYG{n}{python}\PYG{o}{\PYGZhy{}}\PYG{n}{networkx}
\end{sphinxVerbatim}

\sphinxAtStartPar
User tools (highly recommended):

\begin{sphinxVerbatim}[commandchars=\\\{\}]
\PYG{n}{sudo} \PYG{n}{apt}\PYG{o}{\PYGZhy{}}\PYG{n}{get} \PYG{n}{install} \PYG{n}{ipython} \PYG{n}{ipython}\PYG{o}{\PYGZhy{}}\PYG{n}{notebook} \PYG{n}{scite}
\end{sphinxVerbatim}


\section{Linux: Ubuntu 14.04}
\label{\detokenize{install_doc:linux-ubuntu-14-04}}

\subsection{Python2}
\label{\detokenize{install_doc:python2}}
\sphinxAtStartPar
First we create a scientific Python workbench:

\begin{sphinxVerbatim}[commandchars=\\\{\}]
\PYG{n}{sudo} \PYG{n}{apt}\PYG{o}{\PYGZhy{}}\PYG{n}{get} \PYG{n}{install} \PYG{n}{python}\PYG{o}{\PYGZhy{}}\PYG{n}{dev} \PYG{n}{python}\PYG{o}{\PYGZhy{}}\PYG{n}{numpy} \PYG{n}{python}\PYG{o}{\PYGZhy{}}\PYG{n}{scipy}
\PYG{n}{sudo} \PYG{n}{apt}\PYG{o}{\PYGZhy{}}\PYG{n}{get} \PYG{n}{install} \PYG{n}{python}\PYG{o}{\PYGZhy{}}\PYG{n}{matplotlib}  \PYG{n}{python}\PYG{o}{\PYGZhy{}}\PYG{n}{pip}
\PYG{n}{sudo} \PYG{n}{apt}\PYG{o}{\PYGZhy{}}\PYG{n}{get} \PYG{n}{install} \PYG{n}{python}\PYG{o}{\PYGZhy{}}\PYG{n}{sympy} \PYG{n}{python}\PYG{o}{\PYGZhy{}}\PYG{n}{suds} \PYG{n}{python}\PYG{o}{\PYGZhy{}}\PYG{n}{xlrd}
\PYG{n}{sudo} \PYG{n}{apt}\PYG{o}{\PYGZhy{}}\PYG{n}{get} \PYG{n}{install} \PYG{n}{python}\PYG{o}{\PYGZhy{}}\PYG{n}{xlwt} \PYG{n}{python}\PYG{o}{\PYGZhy{}}\PYG{n}{h5py}
\PYG{n}{sudo} \PYG{n}{apt}\PYG{o}{\PYGZhy{}}\PYG{n}{get} \PYG{n}{install} \PYG{n}{python}\PYG{o}{\PYGZhy{}}\PYG{n}{wxgtk2}\PYG{l+m+mf}{.8} \PYG{n}{python}\PYG{o}{\PYGZhy{}}\PYG{n}{qt4}
\PYG{n}{sudo} \PYG{n}{apt}\PYG{o}{\PYGZhy{}}\PYG{n}{get} \PYG{n}{install} \PYG{n}{ipython} \PYG{n}{ipython}\PYG{o}{\PYGZhy{}}\PYG{n}{notebook}
\end{sphinxVerbatim}


\subsection{libSBML}
\label{\detokenize{install_doc:id1}}
\sphinxAtStartPar
Installing libSBML is now easy using Pip:

\begin{sphinxVerbatim}[commandchars=\\\{\}]
\PYG{n}{sudo} \PYG{n}{apt}\PYG{o}{\PYGZhy{}}\PYG{n}{get} \PYG{n}{install} \PYG{n}{libxml2} \PYG{n}{libxml2}\PYG{o}{\PYGZhy{}}\PYG{n}{dev}
\PYG{n}{sudo} \PYG{n}{apt}\PYG{o}{\PYGZhy{}}\PYG{n}{get} \PYG{n}{install} \PYG{n}{zlib1g} \PYG{n}{zlib1g}\PYG{o}{\PYGZhy{}}\PYG{n}{dev}
\PYG{n}{sudo} \PYG{n}{apt}\PYG{o}{\PYGZhy{}}\PYG{n}{get} \PYG{n}{install} \PYG{n}{bzip2} \PYG{n}{libbz2}\PYG{o}{\PYGZhy{}}\PYG{n}{dev}

\PYG{n}{sudo} \PYG{n}{pip} \PYG{n}{install} \PYG{n}{python}\PYG{o}{\PYGZhy{}}\PYG{n}{libsbml}
\end{sphinxVerbatim}


\subsection{glpk/python\sphinxhyphen{}glpk}
\label{\detokenize{install_doc:id2}}
\sphinxAtStartPar
GLPK needs to be version 4.47 to work with glpk\sphinxhyphen{}0.3:

\begin{sphinxVerbatim}[commandchars=\\\{\}]
\PYG{n}{sudo} \PYG{n}{apt}\PYG{o}{\PYGZhy{}}\PYG{n}{get} \PYG{n}{install} \PYG{n}{libgmp}\PYG{o}{\PYGZhy{}}\PYG{n}{dev}
\end{sphinxVerbatim}

\sphinxAtStartPar
cd GLPK source (e.g. glpk\sphinxhyphen{}4.47):

\begin{sphinxVerbatim}[commandchars=\\\{\}]
\PYG{o}{.}\PYG{o}{/}\PYG{n}{configure} \PYG{o}{\PYGZhy{}}\PYG{o}{\PYGZhy{}}\PYG{k}{with}\PYG{o}{\PYGZhy{}}\PYG{n}{gmp}
\PYG{n}{make}
\PYG{n}{make} \PYG{n}{check}
\PYG{n}{sudo} \PYG{n}{make} \PYG{n}{install}
\PYG{n}{sudo} \PYG{n}{ldconfig}
\end{sphinxVerbatim}

\sphinxAtStartPar
cd to python\sphinxhyphen{}glpk source (glpk\sphinxhyphen{}0.3):

\begin{sphinxVerbatim}[commandchars=\\\{\}]
\PYG{n}{make}
\PYG{n}{sudo} \PYG{n}{make} \PYG{n}{install}
\end{sphinxVerbatim}


\subsection{CBMPy}
\label{\detokenize{install_doc:cbmpy}}
\sphinxAtStartPar
Finally, install CBMPy:

\begin{sphinxVerbatim}[commandchars=\\\{\}]
\PYG{n}{python} \PYG{n}{setup}\PYG{o}{.}\PYG{n}{py} \PYG{n}{build} \PYG{n}{sdist}
\PYG{n}{sudo} \PYG{n}{python} \PYG{n}{setup}\PYG{o}{.}\PYG{n}{py} \PYG{n}{install}
\end{sphinxVerbatim}


\subsection{Installing PyscesMarinerCBM}
\label{\detokenize{install_doc:installing-pyscesmarinercbm}}
\sphinxAtStartPar
This will install PySCeS Mariner that adds SOAP web\sphinxhyphen{}services
capability to CBMPy. First unpack pyscesmariner\sphinxhyphen{}0.7.7.zip and install
the cherrypy webserver:

\begin{sphinxVerbatim}[commandchars=\\\{\}]
\PYG{n}{sudo} \PYG{n}{apt}\PYG{o}{\PYGZhy{}}\PYG{n}{get} \PYG{n}{install} \PYG{n}{python}\PYG{o}{\PYGZhy{}}\PYG{n}{cherrypy}
\end{sphinxVerbatim}


\subsection{Install soaplib}
\label{\detokenize{install_doc:install-soaplib}}
\sphinxAtStartPar
cd \textless{}pysces\_cbm\_mariner\textgreater{}/misc:

\begin{sphinxVerbatim}[commandchars=\\\{\}]
\PYG{n}{tar} \PYG{o}{\PYGZhy{}}\PYG{n}{xf} \PYG{n}{soaplib}\PYG{o}{\PYGZhy{}}\PYG{l+m+mf}{0.8}\PYG{l+m+mf}{.1}\PYG{o}{.}\PYG{n}{tar}\PYG{o}{.}\PYG{n}{gz}
\PYG{n}{cd} \PYG{n}{soaplib}\PYG{o}{\PYGZhy{}}\PYG{l+m+mf}{0.8}\PYG{l+m+mf}{.1}
\PYG{n}{python} \PYG{n}{setup}\PYG{o}{.}\PYG{n}{py} \PYG{n}{build} \PYG{n}{sdist}
\PYG{n}{sudo} \PYG{n}{python} \PYG{n}{setup}\PYG{o}{.}\PYG{n}{py} \PYG{n}{install}
\end{sphinxVerbatim}


\subsection{Install Mariner}
\label{\detokenize{install_doc:install-mariner}}
\sphinxAtStartPar
cd \textless{}pysces\_cbm\_mariner\textgreater{} and set mariner configuration (not needed for Ubuntu, Windows or if the server does not read SBML):

\begin{sphinxVerbatim}[commandchars=\\\{\}]
\PYG{n}{sudo} \PYG{n}{nano} \PYG{o}{/}\PYG{n}{usr}\PYG{o}{/}\PYG{n}{local}\PYG{o}{/}\PYG{n}{lib}\PYG{o}{/}\PYG{n}{python2}\PYG{l+m+mf}{.7}\PYG{o}{/}\PYG{n}{dist}\PYG{o}{\PYGZhy{}}\PYG{n}{packages}\PYG{o}{/}\PYG{n}{pyscesmariner}\PYG{o}{/}\PYG{n}{MarinerConfig}\PYG{o}{.}\PYG{n}{py}
\PYG{n}{PATH\PYGZus{}LIBSBMLTHREAD} \PYG{o}{=} \PYG{l+s+s1}{\PYGZsq{}}\PYG{l+s+s1}{/usr/local/lib/python2.7/dist\PYGZhy{}packages/pyscesmariner/libSBMLthread.pyc}\PYG{l+s+s1}{\PYGZsq{}}
\PYG{n}{PATH\PYGZus{}LIBSBML\PYGZus{}CONVERTTHREAD} \PYG{o}{=} \PYG{l+s+s1}{\PYGZsq{}}\PYG{l+s+s1}{/usr/local/lib/python2.7/dist\PYGZhy{}packages/pyscesmariner/libSBMLConvertThread.py}\PYG{l+s+s1}{\PYGZsq{}}
\end{sphinxVerbatim}

\sphinxAtStartPar
cd to \textless{}pysces\_cbm\_mariner\textgreater{}:

\begin{sphinxVerbatim}[commandchars=\\\{\}]
\PYG{n}{python} \PYG{n}{setup}\PYG{o}{.}\PYG{n}{py} \PYG{n}{build} \PYG{n}{sdist}
\PYG{n}{sudo} \PYG{n}{python} \PYG{n}{setup}\PYG{o}{.}\PYG{n}{py} \PYG{n}{install}
\end{sphinxVerbatim}


\subsection{Test installation}
\label{\detokenize{install_doc:test-installation}}
\sphinxAtStartPar
Open a new terminal window:

\begin{sphinxVerbatim}[commandchars=\\\{\}]
\PYG{c+c1}{\PYGZsh{} cd \PYGZlt{}pysces\PYGZus{}cbm\PYGZus{}mariner\PYGZgt{}/demo}
\PYG{n}{python} \PYG{n}{cbm\PYGZus{}server\PYGZus{}demo}\PYG{o}{.}\PYG{n}{py}
\end{sphinxVerbatim}

\sphinxAtStartPar
Open another terminal and run the client demo:

\begin{sphinxVerbatim}[commandchars=\\\{\}]
\PYG{n}{python} \PYG{n}{cbm\PYGZus{}client\PYGZus{}demo}\PYG{o}{.}\PYG{n}{py}
\end{sphinxVerbatim}

\sphinxAtStartPar
Kill the server by closing the terminal window.


\subsection{Python3}
\label{\detokenize{install_doc:python3}}
\sphinxAtStartPar
Not all dependencies are available for Python3:

\begin{sphinxVerbatim}[commandchars=\\\{\}]
\PYG{n}{sudo} \PYG{n}{apt}\PYG{o}{\PYGZhy{}}\PYG{n}{get} \PYG{n}{install} \PYG{n}{python3}\PYG{o}{\PYGZhy{}}\PYG{n}{dev} \PYG{n}{python3}\PYG{o}{\PYGZhy{}}\PYG{n}{numpy} \PYG{n}{python3}\PYG{o}{\PYGZhy{}}\PYG{n}{scipy}
\PYG{n}{sudo} \PYG{n}{apt}\PYG{o}{\PYGZhy{}}\PYG{n}{get} \PYG{n}{install} \PYG{n}{python3}\PYG{o}{\PYGZhy{}}\PYG{n}{matplotlib}  \PYG{n}{python3}\PYG{o}{\PYGZhy{}}\PYG{n}{pip}
\PYG{n}{sudo} \PYG{n}{apt}\PYG{o}{\PYGZhy{}}\PYG{n}{get} \PYG{n}{install} \PYG{n}{python3}\PYG{o}{\PYGZhy{}}\PYG{n}{xlrd} \PYG{n}{python3}\PYG{o}{\PYGZhy{}}\PYG{n}{h5py}

\PYG{c+c1}{\PYGZsh{} need to find out what is going on with Python3 and xlwt suds}
\PYG{c+c1}{\PYGZsh{} easy\PYGZus{}install3 sympy ???}
\PYG{c+c1}{\PYGZsh{} wxPython and PyQt4 not in Ubuntu P3 builds yet}

\PYG{n}{sudo} \PYG{n}{apt}\PYG{o}{\PYGZhy{}}\PYG{n}{get} \PYG{n}{install} \PYG{n}{ipython3} \PYG{n}{ipython3}\PYG{o}{\PYGZhy{}}\PYG{n}{notebook}

\PYG{n}{sudo} \PYG{n}{apt}\PYG{o}{\PYGZhy{}}\PYG{n}{get} \PYG{n}{install} \PYG{n}{libxml2} \PYG{n}{libxml2}\PYG{o}{\PYGZhy{}}\PYG{n}{dev}
\PYG{n}{sudo} \PYG{n}{apt}\PYG{o}{\PYGZhy{}}\PYG{n}{get} \PYG{n}{install} \PYG{n}{zlib1g} \PYG{n}{zlib1g}\PYG{o}{\PYGZhy{}}\PYG{n}{dev}
\PYG{n}{sudo} \PYG{n}{apt}\PYG{o}{\PYGZhy{}}\PYG{n}{get} \PYG{n}{install} \PYG{n}{bzip2} \PYG{n}{libbz2}\PYG{o}{\PYGZhy{}}\PYG{n}{dev}

\PYG{n}{sudo} \PYG{n}{pip3} \PYG{n}{install} \PYG{n}{python}\PYG{o}{\PYGZhy{}}\PYG{n}{libsbml}

\PYG{n}{sudo} \PYG{n}{apt}\PYG{o}{\PYGZhy{}}\PYG{n}{get} \PYG{n}{install} \PYG{n}{python}\PYG{o}{\PYGZhy{}}\PYG{n}{qt4} \PYG{n}{python}\PYG{o}{\PYGZhy{}}\PYG{n}{qt4}\PYG{o}{\PYGZhy{}}\PYG{n}{dev} \PYG{n}{python}\PYG{o}{\PYGZhy{}}\PYG{n}{sip}
\PYG{n}{sudo} \PYG{n}{apt}\PYG{o}{\PYGZhy{}}\PYG{n}{get} \PYG{n}{install} \PYG{n}{python}\PYG{o}{\PYGZhy{}}\PYG{n}{sip}\PYG{o}{\PYGZhy{}}\PYG{n}{dev} \PYG{n}{build}\PYG{o}{\PYGZhy{}}\PYG{n}{essential}
\end{sphinxVerbatim}


\section{Apple Macintosh: OS X}
\label{\detokenize{install_doc:apple-macintosh-os-x}}
\sphinxAtStartPar
Installation is similar to Linux except packages are installed using distutils and pip. The first step is to install the Mac development tools \sphinxcode{\sphinxupquote{xcode}}

\sphinxAtStartPar
Install \sphinxcode{\sphinxupquote{Python}} packages:

\begin{sphinxVerbatim}[commandchars=\\\{\}]
\PYG{n}{sudo} \PYG{n}{easy\PYGZus{}install} \PYG{n}{numpy} \PYG{n}{ipython} \PYG{n}{scipy} \PYG{n}{matplotlib}
\PYG{n}{sudo} \PYG{n}{easy\PYGZus{}install} \PYG{n}{xlrd} \PYG{n}{xlwt} \PYG{n}{sympy} \PYG{n}{suds} \PYG{n}{pyparsing} \PYG{n}{pip}
\end{sphinxVerbatim}

\sphinxAtStartPar
Use pip to install advanced Ipython and libsbml:

\begin{sphinxVerbatim}[commandchars=\\\{\}]
\PYG{n}{sudo} \PYG{n}{pip} \PYG{n}{install} \PYG{n}{ipython}\PYG{p}{[}\PYG{n}{notebook}\PYG{p}{]}
\PYG{n}{ARCHFLAGS}\PYG{o}{=}\PYG{o}{\PYGZhy{}}\PYG{n}{Wno}\PYG{o}{\PYGZhy{}}\PYG{n}{error}\PYG{o}{=}\PYG{n}{unused}\PYG{o}{\PYGZhy{}}\PYG{n}{command}\PYG{o}{\PYGZhy{}}\PYG{n}{line}\PYG{o}{\PYGZhy{}}\PYG{n}{argument}\PYG{o}{\PYGZhy{}}\PYG{n}{hard}\PYG{o}{\PYGZhy{}}\PYG{n}{error}\PYG{o}{\PYGZhy{}}\PYG{o+ow}{in}\PYG{o}{\PYGZhy{}}\PYG{n}{future}  \PYG{n}{pip} \PYG{n}{install} \PYG{n}{python}\PYG{o}{\PYGZhy{}}\PYG{n}{libsbml}
\end{sphinxVerbatim}

\sphinxAtStartPar
For \sphinxcode{\sphinxupquote{solvers}}, either install your own copy of CPLEX or build PyGLPK which requires building both the GMP and GLPK libraries.

\sphinxAtStartPar
\sphinxcode{\sphinxupquote{GMP}} (\sphinxurl{https://gmplib.org/}):

\begin{sphinxVerbatim}[commandchars=\\\{\}]
\PYG{n}{download} \PYG{n}{gmp}
\PYG{o}{.}\PYG{o}{/}\PYG{n}{configure} \PYG{o}{\PYGZhy{}}\PYG{o}{\PYGZhy{}}\PYG{n}{prefix}\PYG{o}{=}\PYG{o}{/}\PYG{n}{usr}\PYG{o}{/}\PYG{n}{local}
\PYG{n}{make}
\PYG{n}{make} \PYG{n}{check}
\PYG{n}{sudo} \PYG{n}{make} \PYG{n}{install}
\end{sphinxVerbatim}

\sphinxAtStartPar
\sphinxcode{\sphinxupquote{GLPK}}  (\sphinxurl{http://sourceforge.net/projects/cbmpy/files/tools/glpk}):

\begin{sphinxVerbatim}[commandchars=\\\{\}]
\PYG{n}{download} \PYG{n}{glpk}\PYG{o}{\PYGZhy{}}\PYG{l+m+mf}{4.47}\PYG{o}{.}\PYG{n}{tar}\PYG{o}{.}\PYG{n}{gz}
\PYG{o}{.}\PYG{o}{/}\PYG{n}{configure} \PYG{o}{\PYGZhy{}}\PYG{o}{\PYGZhy{}}\PYG{n}{prefix}\PYG{o}{=}\PYG{o}{/}\PYG{n}{usr}\PYG{o}{/}\PYG{n}{local} \PYG{o}{\PYGZhy{}}\PYG{o}{\PYGZhy{}}\PYG{k}{with}\PYG{o}{\PYGZhy{}}\PYG{n}{gmp}
\PYG{n}{make}
\PYG{n}{sudo} \PYG{n}{make} \PYG{n}{install}
\end{sphinxVerbatim}

\sphinxAtStartPar
\sphinxcode{\sphinxupquote{PyGLPK}} (\sphinxurl{http://sourceforge.net/projects/cbmpy/files/tools/glpk}):

\begin{sphinxVerbatim}[commandchars=\\\{\}]
\PYG{n}{download} \PYG{n}{python}\PYG{o}{\PYGZhy{}}\PYG{n}{glpk}\PYG{o}{\PYGZhy{}}\PYG{l+m+mf}{0.3}
\PYG{n}{python} \PYG{n}{setup}\PYG{o}{.}\PYG{n}{py} \PYG{n}{build}
\PYG{n}{sudo} \PYG{n}{python} \PYG{n}{setup}\PYG{o}{.}\PYG{n}{py} \PYG{n}{install}
\end{sphinxVerbatim}


\section{Installing PySCeS\sphinxhyphen{}CBM Mariner (Microsoft Windows and Linux)}
\label{\detokenize{install_doc:installing-pysces-cbm-mariner-microsoft-windows-and-linux}}
\sphinxAtStartPar
The PySCeS Mariner module exposes the CBMPy functionality as SOAP
web services (e.g. as a backend to FAME (\sphinxurl{http://F-A-M-E.org})). It is available for download from SourceForge:
\begin{itemize}
\item {} 
\sphinxAtStartPar
PySCeS\sphinxhyphen{}CBM Mariner: \sphinxurl{http://sourceforge.net/projects/cbmpy/files/release/pysces\_mariner/}

\end{itemize}


\subsection{Dependencies: CherryPy, libXML and SOAPlib}
\label{\detokenize{install_doc:dependencies-cherrypy-libxml-and-soaplib}}
\sphinxAtStartPar
PySCeS\sphinxhyphen{}CBM Mariner requires (pure python) soaplib 0.8.1 (supplied with it) or
downloadable from:

\begin{sphinxVerbatim}[commandchars=\\\{\}]
\PYG{n}{https}\PYG{o}{/}\PYG{o}{/}\PYG{n}{sourceforge}\PYG{o}{.}\PYG{n}{net}\PYG{o}{/}\PYG{n}{projects}\PYG{o}{/}\PYG{n}{cbmpy}\PYG{o}{/}\PYG{n}{files}\PYG{o}{/}\PYG{n}{tools}\PYG{o}{/}\PYG{n}{soaplib}\PYG{o}{/}
\end{sphinxVerbatim}

\sphinxAtStartPar
Soaplib itself has two dependencies which should be installed first:
\begin{itemize}
\item {} 
\sphinxAtStartPar
LXML (\sphinxurl{http://lxml.de})
\begin{itemize}
\item {} 
\sphinxAtStartPar
Windows: install with \sphinxcode{\sphinxupquote{easy\_install lxml}}

\item {} 
\sphinxAtStartPar
Linux (Ubuntu) use \sphinxcode{\sphinxupquote{sudo apt\sphinxhyphen{}get install python\sphinxhyphen{}lxml}}

\end{itemize}

\item {} 
\sphinxAtStartPar
CherryPy (\sphinxurl{http://www.cherrypy.org})
\begin{itemize}
\item {} 
\sphinxAtStartPar
Windows: install with \sphinxcode{\sphinxupquote{easy\_install cherrypy}}

\item {} 
\sphinxAtStartPar
Linux (Ubuntu) use \sphinxcode{\sphinxupquote{sudo apt\sphinxhyphen{}get install python\sphinxhyphen{}cherrypy}}

\end{itemize}

\item {} 
\sphinxAtStartPar
SOAPLIB 0.8.1:
\begin{itemize}
\item {} 
\sphinxAtStartPar
Windows: \sphinxcode{\sphinxupquote{Execute soaplib\sphinxhyphen{}0.8.1.win32.exe}}

\item {} 
\sphinxAtStartPar
Linux: Unpack the zip archive and run \sphinxcode{\sphinxupquote{sudo python setup.py install}}

\end{itemize}

\end{itemize}

\sphinxAtStartPar
Test installation:
\begin{itemize}
\item {} 
\sphinxAtStartPar
Open a terminal

\item {} 
\sphinxAtStartPar
Execute “ipython”

\item {} 
\sphinxAtStartPar
Execute “import cherrypy, lxml, soaplib” no errors or warnings should be generated

\item {} 
\sphinxAtStartPar
Exit ipython with CTRL\sphinxhyphen{}D

\item {} 
\sphinxAtStartPar
change directory to supplied soaplib tests e.g. “cd e:\textbackslash{}cbmpy\textbackslash{}tests\textbackslash{}soaplib”

\item {} 
\sphinxAtStartPar
Execute “python binary\_test.py”

\item {} 
\sphinxAtStartPar
Execute “python primitive\_test.py”

\end{itemize}

\sphinxAtStartPar
All tests should pass.


\subsection{PySCeS\sphinxhyphen{}CBM Mariner (http://cbmpy.sourceforge.net)}
\label{\detokenize{install_doc:pysces-cbm-mariner-http-cbmpy-sourceforge-net}}
\sphinxAtStartPar
Download and install the latest version (0.7.4 or newer is required for CBMPy 0.7+):
\begin{itemize}
\item {} 
\sphinxAtStartPar
Windows: \sphinxcode{\sphinxupquote{Execute pyscesmariner\sphinxhyphen{}0.7.7.zip}}

\item {} 
\sphinxAtStartPar
Linux: unpack the archive and run \sphinxcode{\sphinxupquote{sudo python setup.py install}}

\end{itemize}

\sphinxAtStartPar
To test installation, on Linux execute the commands in \sphinxstyleemphasis{run\_server.bat} from the terminal directly.
\begin{itemize}
\item {} 
\sphinxAtStartPar
Open two terminals and in both

\item {} 
\sphinxAtStartPar
Change directory to supplied PySCeS\sphinxhyphen{}CBM Mariner tests e.g. \sphinxcode{\sphinxupquote{cd e:\textbackslash{}\textbackslash{}cbmpy\textbackslash{}\textbackslash{}tests\textbackslash{}\textbackslash{}pyscesmariner}}

\item {} 
\sphinxAtStartPar
In terminal one \sphinxcode{\sphinxupquote{Execute run\_server.bat}}

\end{itemize}

\sphinxAtStartPar
Which should now display:

\begin{sphinxVerbatim}[commandchars=\\\{\}]
\PYG{n}{E}\PYG{p}{:}\PYGZbs{}\PYGZbs{}\PYG{n}{cbmpy}\PYGZbs{}\PYGZbs{}\PYG{n}{tests}\PYGZbs{}\PYGZbs{}\PYG{n}{pyscesmariner}\PYG{o}{\PYGZgt{}}\PYG{n}{python} \PYG{n}{cbm\PYGZus{}server\PYGZus{}demo}\PYG{o}{.}\PYG{n}{py}
\PYG{n}{Mariner} \PYG{n}{using} \PYG{n}{E}\PYG{p}{:}\PYGZbs{}\PYGZbs{}\PYG{n}{cbmpy}\PYGZbs{}\PYGZbs{}\PYG{n}{tests}\PYGZbs{}\PYGZbs{}\PYG{n}{pyscesmariner} \PYG{k}{as} \PYG{n}{a} \PYG{n}{working} \PYG{n}{directory}
\PYG{n}{Mariner} \PYG{n}{server} \PYG{n}{name}\PYG{p}{:} \PYG{l+m+mf}{10.0}\PYG{l+m+mf}{.2}\PYG{l+m+mf}{.15}
\PYG{n}{Mariner} \PYG{n}{using} \PYG{n}{port}\PYG{p}{:} \PYG{l+m+mi}{31313}

\PYG{n}{Welcome} \PYG{n}{to} \PYG{n}{the} \PYG{n}{PySCeS} \PYG{n}{Constraint} \PYG{n}{Based} \PYG{n}{Modelling} \PYG{n}{Toolkit} \PYG{p}{(}\PYG{l+m+mf}{0.7}\PYG{l+m+mf}{.4}\PYG{p}{)}

\PYG{o}{\PYGZlt{}}\PYG{n}{snipped}\PYG{o}{\PYGZgt{}}

\PYG{n}{Multiple} \PYG{n}{Environment} \PYG{n}{Module} \PYG{p}{(}\PYG{l+m+mf}{0.6}\PYG{l+m+mf}{.2} \PYG{p}{[}\PYG{n}{r1147}\PYG{p}{]}\PYG{p}{)}

\PYG{n}{PySCeSCBM}\PYG{o}{/}\PYG{n}{Mariner} \PYG{n}{initialising} \PYG{o}{.}\PYG{o}{.}\PYG{o}{.} \PYG{n}{this} \PYG{n}{console} \PYG{o+ow}{is} \PYG{n}{now} \PYG{n}{blocked}
\end{sphinxVerbatim}

\sphinxAtStartPar
In terminal two:
\begin{itemize}
\item {} 
\sphinxAtStartPar
Execute \sphinxcode{\sphinxupquote{python cbm\_client\_demo.py}}

\end{itemize}

\sphinxAtStartPar
This should end without errors and display \sphinxcode{\sphinxupquote{done.}} Congratulations
you have successfully installed CBMPy and PySCeS\sphinxhyphen{}CBM Mariner!

\sphinxstepscope


\chapter{Introduction}
\label{\detokenize{manual_doc:introduction}}\label{\detokenize{manual_doc:introducing-cbmpy}}\label{\detokenize{manual_doc::doc}}
\sphinxAtStartPar
PySCeS CBMPy is a new platform for constraint based modelling and analysis. It has been designed using principles
developed in the PySCeS simulation software project: usability, flexibility and accessibility. Its architecture
is both extensible and flexible using data structures that are intuitive to the biologist (metabolites, reactions, compartments)
while transparently translating these into the underlying mathematical structures used in advanced analysis (LP’s, MILP’s).

\sphinxAtStartPar
PySCeS CBMPy implements popular analyses such as FBA, FVA, element/charge balancing, network analysis and model editing as
well as advanced methods developed specifically for the ecosystem modelling: minimal distance methods, flux minimization and input selection.

\sphinxAtStartPar
To cater for a diverse range of modelling needs PySCeS CBMPy supports user interaction via:
\begin{itemize}
\item {} 
\sphinxAtStartPar
interactive console, scripting for advanced use or as a library for software development

\item {} 
\sphinxAtStartPar
GUI, for quick access to a visual representation of the model, analysis methods and annotation tools

\item {} 
\sphinxAtStartPar
SOAP based web services: using the Mariner framework much high level functionality is exposed for integration into web tools

\end{itemize}

\sphinxAtStartPar
For more information on the development and use of PySCeS CBMPy visit the website (\sphinxurl{http:cbmpy.sourceforge.net}) for up to date information and
feel free to contact the development team (\sphinxhref{mailto:bgoli@users.sourceforge.net}{bgoli@users.sourceforge.net}).

\sphinxstepscope


\chapter{CBMPy Module Reference}
\label{\detokenize{modules_doc:module-cbmpy.CBCommon}}\label{\detokenize{modules_doc:cbmpy-module-reference}}\label{\detokenize{modules_doc::doc}}\index{module@\spxentry{module}!cbmpy.CBCommon@\spxentry{cbmpy.CBCommon}}\index{cbmpy.CBCommon@\spxentry{cbmpy.CBCommon}!module@\spxentry{module}}

\section{CBMPy: CBCommon module}
\label{\detokenize{modules_doc:cbmpy-cbcommon-module}}
\sphinxAtStartPar
PySCeS Constraint Based Modelling (\sphinxurl{http://cbmpy.sourceforge.net})
Copyright (C) 2009\sphinxhyphen{}2024 Brett G. Olivier, VU University Amsterdam, Amsterdam, The Netherlands

\sphinxAtStartPar
This program is free software: you can redistribute it and/or modify
it under the terms of the GNU General Public License as published by
the Free Software Foundation, either version 3 of the License, or
(at your option) any later version.

\sphinxAtStartPar
This program is distributed in the hope that it will be useful,
but WITHOUT ANY WARRANTY; without even the implied warranty of
MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
GNU General Public License for more details.

\sphinxAtStartPar
You should have received a copy of the GNU General Public License
along with this program.  If not, see \textless{}\sphinxurl{http://www.gnu.org/licenses/}\textgreater{}

\sphinxAtStartPar
Author: Brett G. Olivier PhD
Contact developers: \sphinxurl{https://github.com/SystemsBioinformatics/cbmpy/issues}
Last edit: \$Author: bgoli \$ (\$Id: CBCommon.py 710 2020\sphinxhyphen{}04\sphinxhyphen{}27 14:22:34Z bgoli \$)
\index{ComboGen (class in cbmpy.CBCommon)@\spxentry{ComboGen}\spxextra{class in cbmpy.CBCommon}}

\begin{fulllineitems}
\phantomsection\label{\detokenize{modules_doc:cbmpy.CBCommon.ComboGen}}
\pysigstartsignatures
\pysigline{\sphinxbfcode{\sphinxupquote{class\DUrole{w,w}{  }}}\sphinxbfcode{\sphinxupquote{ComboGen}}}
\pysigstopsignatures
\sphinxAtStartPar
Generate sets of unique combinations

\end{fulllineitems}

\index{MIRIAMannotation (class in cbmpy.CBCommon)@\spxentry{MIRIAMannotation}\spxextra{class in cbmpy.CBCommon}}

\begin{fulllineitems}
\phantomsection\label{\detokenize{modules_doc:cbmpy.CBCommon.MIRIAMannotation}}
\pysigstartsignatures
\pysigline{\sphinxbfcode{\sphinxupquote{class\DUrole{w,w}{  }}}\sphinxbfcode{\sphinxupquote{MIRIAMannotation}}}
\pysigstopsignatures
\sphinxAtStartPar
The MIRIAMannotation class MIRIAM annotations: Biological Qualifiers
\index{addIDorgURI() (MIRIAMannotation method)@\spxentry{addIDorgURI()}\spxextra{MIRIAMannotation method}}

\begin{fulllineitems}
\phantomsection\label{\detokenize{modules_doc:cbmpy.CBCommon.MIRIAMannotation.addIDorgURI}}
\pysigstartsignatures
\pysiglinewithargsret{\sphinxbfcode{\sphinxupquote{addIDorgURI}}}{\sphinxparam{\DUrole{n,n}{qual}}\sphinxparamcomma \sphinxparam{\DUrole{n,n}{uri}}}{}
\pysigstopsignatures
\sphinxAtStartPar
Add a URI directly into a qualifier collection:
\begin{itemize}
\item {} 
\sphinxAtStartPar
\sphinxstyleemphasis{qual} a Biomodels biological qualifier e.g. “is” “isEncodedBy”

\item {} 
\sphinxAtStartPar
\sphinxstyleemphasis{uri} the complete identifiers.org uri e.g. \sphinxurl{http://identifiers.org/chebi/CHEBI:58088}

\end{itemize}

\end{fulllineitems}

\index{addMIRIAMannotation() (MIRIAMannotation method)@\spxentry{addMIRIAMannotation()}\spxextra{MIRIAMannotation method}}

\begin{fulllineitems}
\phantomsection\label{\detokenize{modules_doc:cbmpy.CBCommon.MIRIAMannotation.addMIRIAMannotation}}
\pysigstartsignatures
\pysiglinewithargsret{\sphinxbfcode{\sphinxupquote{addMIRIAMannotation}}}{\sphinxparam{\DUrole{n,n}{qual}}\sphinxparamcomma \sphinxparam{\DUrole{n,n}{entity}}\sphinxparamcomma \sphinxparam{\DUrole{n,n}{mid}}}{}
\pysigstopsignatures
\sphinxAtStartPar
Add a qualified MIRIAM annotation or entity:
\begin{itemize}
\item {} 
\sphinxAtStartPar
\sphinxstyleemphasis{qual} a Biomodels biological qualifier e.g. “is” “isEncodedBy”

\item {} 
\sphinxAtStartPar
\sphinxstyleemphasis{entity} a MIRIAM resource entity e.g. “ChEBI”

\item {} 
\sphinxAtStartPar
\sphinxstyleemphasis{mid} the entity id e.g. CHEBI:17158

\end{itemize}

\end{fulllineitems}

\index{checkEntity() (MIRIAMannotation method)@\spxentry{checkEntity()}\spxextra{MIRIAMannotation method}}

\begin{fulllineitems}
\phantomsection\label{\detokenize{modules_doc:cbmpy.CBCommon.MIRIAMannotation.checkEntity}}
\pysigstartsignatures
\pysiglinewithargsret{\sphinxbfcode{\sphinxupquote{checkEntity}}}{\sphinxparam{\DUrole{n,n}{entity}}}{}
\pysigstopsignatures
\sphinxAtStartPar
Check an entity entry, this is a MIRIAM resource name: “chEBI”. The test is case insensitive and will correct the case
of wrongly capitalised entities automatically. If the entity is not recognised then a list of possible candidates
based on the first letters of the input is displayed.
\begin{itemize}
\item {} 
\sphinxAtStartPar
\sphinxstyleemphasis{entity} a MIRIAM resource entity e.g. “ChEBI”

\end{itemize}

\end{fulllineitems}

\index{checkEntityPattern() (MIRIAMannotation method)@\spxentry{checkEntityPattern()}\spxextra{MIRIAMannotation method}}

\begin{fulllineitems}
\phantomsection\label{\detokenize{modules_doc:cbmpy.CBCommon.MIRIAMannotation.checkEntityPattern}}
\pysigstartsignatures
\pysiglinewithargsret{\sphinxbfcode{\sphinxupquote{checkEntityPattern}}}{\sphinxparam{\DUrole{n,n}{entity}}}{}
\pysigstopsignatures
\sphinxAtStartPar
For an entity key compile the pattern to a regex, if necessary.
\begin{itemize}
\item {} 
\sphinxAtStartPar
\sphinxstyleemphasis{entity} a MIRIAM resource entity

\end{itemize}

\end{fulllineitems}

\index{checkId() (MIRIAMannotation method)@\spxentry{checkId()}\spxextra{MIRIAMannotation method}}

\begin{fulllineitems}
\phantomsection\label{\detokenize{modules_doc:cbmpy.CBCommon.MIRIAMannotation.checkId}}
\pysigstartsignatures
\pysiglinewithargsret{\sphinxbfcode{\sphinxupquote{checkId}}}{\sphinxparam{\DUrole{n,n}{entity}}\sphinxparamcomma \sphinxparam{\DUrole{n,n}{mid}}}{}
\pysigstopsignatures
\sphinxAtStartPar
Check that a entity id e.g. CHEBI:17158
\begin{itemize}
\item {} 
\sphinxAtStartPar
\sphinxstyleemphasis{mid} the entity id e.g. CHEBI:17158

\end{itemize}

\end{fulllineitems}

\index{deleteMIRIAMannotation() (MIRIAMannotation method)@\spxentry{deleteMIRIAMannotation()}\spxextra{MIRIAMannotation method}}

\begin{fulllineitems}
\phantomsection\label{\detokenize{modules_doc:cbmpy.CBCommon.MIRIAMannotation.deleteMIRIAMannotation}}
\pysigstartsignatures
\pysiglinewithargsret{\sphinxbfcode{\sphinxupquote{deleteMIRIAMannotation}}}{\sphinxparam{\DUrole{n,n}{qual}}\sphinxparamcomma \sphinxparam{\DUrole{n,n}{entity}}\sphinxparamcomma \sphinxparam{\DUrole{n,n}{mid}}}{}
\pysigstopsignatures
\sphinxAtStartPar
Deletes a qualified MIRIAM annotation or entity:
\begin{itemize}
\item {} 
\sphinxAtStartPar
\sphinxstyleemphasis{qual} a Biomodels biological qualifier e.g. “is” “isEncodedBy”

\item {} 
\sphinxAtStartPar
\sphinxstyleemphasis{entity} a MIRIAM resource entity e.g. “ChEBI”

\item {} 
\sphinxAtStartPar
\sphinxstyleemphasis{mid} the entity id e.g. CHEBI:17158

\end{itemize}

\end{fulllineitems}

\index{getAllMIRIAMUris() (MIRIAMannotation method)@\spxentry{getAllMIRIAMUris()}\spxextra{MIRIAMannotation method}}

\begin{fulllineitems}
\phantomsection\label{\detokenize{modules_doc:cbmpy.CBCommon.MIRIAMannotation.getAllMIRIAMUris}}
\pysigstartsignatures
\pysiglinewithargsret{\sphinxbfcode{\sphinxupquote{getAllMIRIAMUris}}}{}{}
\pysigstopsignatures
\sphinxAtStartPar
Return a dictionary of qualifiers that contain ID.org URL’S

\end{fulllineitems}

\index{getAndViewUrisForQualifier() (MIRIAMannotation method)@\spxentry{getAndViewUrisForQualifier()}\spxextra{MIRIAMannotation method}}

\begin{fulllineitems}
\phantomsection\label{\detokenize{modules_doc:cbmpy.CBCommon.MIRIAMannotation.getAndViewUrisForQualifier}}
\pysigstartsignatures
\pysiglinewithargsret{\sphinxbfcode{\sphinxupquote{getAndViewUrisForQualifier}}}{\sphinxparam{\DUrole{n,n}{qual}}}{}
\pysigstopsignatures
\sphinxAtStartPar
Retrieve all url’s associated with qualifier and attempt to open them all in a new browser tab
\begin{itemize}
\item {} 
\sphinxAtStartPar
\sphinxstyleemphasis{qual} the qualifier e.g. “is” or “isEncoded”

\end{itemize}

\end{fulllineitems}

\index{getMIRIAMUrisForQualifier() (MIRIAMannotation method)@\spxentry{getMIRIAMUrisForQualifier()}\spxextra{MIRIAMannotation method}}

\begin{fulllineitems}
\phantomsection\label{\detokenize{modules_doc:cbmpy.CBCommon.MIRIAMannotation.getMIRIAMUrisForQualifier}}
\pysigstartsignatures
\pysiglinewithargsret{\sphinxbfcode{\sphinxupquote{getMIRIAMUrisForQualifier}}}{\sphinxparam{\DUrole{n,n}{qual}}}{}
\pysigstopsignatures
\sphinxAtStartPar
Return all list of urls associated with qualifier:
\begin{itemize}
\item {} 
\sphinxAtStartPar
\sphinxstyleemphasis{qual} the qualifier e.g. “is” or “isEncoded”

\end{itemize}

\end{fulllineitems}

\index{viewURL() (MIRIAMannotation method)@\spxentry{viewURL()}\spxextra{MIRIAMannotation method}}

\begin{fulllineitems}
\phantomsection\label{\detokenize{modules_doc:cbmpy.CBCommon.MIRIAMannotation.viewURL}}
\pysigstartsignatures
\pysiglinewithargsret{\sphinxbfcode{\sphinxupquote{viewURL}}}{\sphinxparam{\DUrole{n,n}{url}}}{}
\pysigstopsignatures
\sphinxAtStartPar
This will try to open the URL in a new tab of the default webbrowser
\begin{itemize}
\item {} 
\sphinxAtStartPar
\sphinxstyleemphasis{url} the url

\end{itemize}

\end{fulllineitems}


\end{fulllineitems}

\index{StructMatrix (class in cbmpy.CBCommon)@\spxentry{StructMatrix}\spxextra{class in cbmpy.CBCommon}}

\begin{fulllineitems}
\phantomsection\label{\detokenize{modules_doc:cbmpy.CBCommon.StructMatrix}}
\pysigstartsignatures
\pysiglinewithargsret{\sphinxbfcode{\sphinxupquote{class\DUrole{w,w}{  }}}\sphinxbfcode{\sphinxupquote{StructMatrix}}}{\sphinxparam{\DUrole{n,n}{array}}\sphinxparamcomma \sphinxparam{\DUrole{n,n}{ridx}}\sphinxparamcomma \sphinxparam{\DUrole{n,n}{cidx}}\sphinxparamcomma \sphinxparam{\DUrole{n,n}{row}\DUrole{o,o}{=}\DUrole{default_value}{None}}\sphinxparamcomma \sphinxparam{\DUrole{n,n}{col}\DUrole{o,o}{=}\DUrole{default_value}{None}}}{}
\pysigstopsignatures
\sphinxAtStartPar
This class is specifically designed to store structural matrix information
give it an array and row/col index permutations it can generate its own
row/col labels given the label src.
\index{getColsByIdx() (StructMatrix method)@\spxentry{getColsByIdx()}\spxextra{StructMatrix method}}

\begin{fulllineitems}
\phantomsection\label{\detokenize{modules_doc:cbmpy.CBCommon.StructMatrix.getColsByIdx}}
\pysigstartsignatures
\pysiglinewithargsret{\sphinxbfcode{\sphinxupquote{getColsByIdx}}}{\sphinxparam{\DUrole{o,o}{*}\DUrole{n,n}{args}}}{}
\pysigstopsignatures
\sphinxAtStartPar
Return the columns referenced by index (1,3,5)

\end{fulllineitems}

\index{getColsByName() (StructMatrix method)@\spxentry{getColsByName()}\spxextra{StructMatrix method}}

\begin{fulllineitems}
\phantomsection\label{\detokenize{modules_doc:cbmpy.CBCommon.StructMatrix.getColsByName}}
\pysigstartsignatures
\pysiglinewithargsret{\sphinxbfcode{\sphinxupquote{getColsByName}}}{\sphinxparam{\DUrole{o,o}{*}\DUrole{n,n}{args}}}{}
\pysigstopsignatures
\sphinxAtStartPar
Return the columns referenced by label (‘s’,’x’,’d’)

\end{fulllineitems}

\index{getIndexes() (StructMatrix method)@\spxentry{getIndexes()}\spxextra{StructMatrix method}}

\begin{fulllineitems}
\phantomsection\label{\detokenize{modules_doc:cbmpy.CBCommon.StructMatrix.getIndexes}}
\pysigstartsignatures
\pysiglinewithargsret{\sphinxbfcode{\sphinxupquote{getIndexes}}}{\sphinxparam{\DUrole{n,n}{axis}\DUrole{o,o}{=}\DUrole{default_value}{\textquotesingle{}all\textquotesingle{}}}}{}
\pysigstopsignatures
\sphinxAtStartPar
Return the matrix indexes ({[}rows{]},{[}cols{]}) where axis=’row’/’col’/’all’

\end{fulllineitems}

\index{getLabels() (StructMatrix method)@\spxentry{getLabels()}\spxextra{StructMatrix method}}

\begin{fulllineitems}
\phantomsection\label{\detokenize{modules_doc:cbmpy.CBCommon.StructMatrix.getLabels}}
\pysigstartsignatures
\pysiglinewithargsret{\sphinxbfcode{\sphinxupquote{getLabels}}}{\sphinxparam{\DUrole{n,n}{axis}\DUrole{o,o}{=}\DUrole{default_value}{\textquotesingle{}all\textquotesingle{}}}}{}
\pysigstopsignatures
\sphinxAtStartPar
Return the matrix labels ({[}rows{]},{[}cols{]}) where axis=’row’/’col’/’all’

\end{fulllineitems}

\index{getRowsByIdx() (StructMatrix method)@\spxentry{getRowsByIdx()}\spxextra{StructMatrix method}}

\begin{fulllineitems}
\phantomsection\label{\detokenize{modules_doc:cbmpy.CBCommon.StructMatrix.getRowsByIdx}}
\pysigstartsignatures
\pysiglinewithargsret{\sphinxbfcode{\sphinxupquote{getRowsByIdx}}}{\sphinxparam{\DUrole{o,o}{*}\DUrole{n,n}{args}}}{}
\pysigstopsignatures
\sphinxAtStartPar
Return the rows referenced by index (1,3,5)

\end{fulllineitems}

\index{getRowsByName() (StructMatrix method)@\spxentry{getRowsByName()}\spxextra{StructMatrix method}}

\begin{fulllineitems}
\phantomsection\label{\detokenize{modules_doc:cbmpy.CBCommon.StructMatrix.getRowsByName}}
\pysigstartsignatures
\pysiglinewithargsret{\sphinxbfcode{\sphinxupquote{getRowsByName}}}{\sphinxparam{\DUrole{o,o}{*}\DUrole{n,n}{args}}}{}
\pysigstopsignatures
\sphinxAtStartPar
Return the rows referenced by label (‘s’,’x’,’d’)

\end{fulllineitems}

\index{setCol() (StructMatrix method)@\spxentry{setCol()}\spxextra{StructMatrix method}}

\begin{fulllineitems}
\phantomsection\label{\detokenize{modules_doc:cbmpy.CBCommon.StructMatrix.setCol}}
\pysigstartsignatures
\pysiglinewithargsret{\sphinxbfcode{\sphinxupquote{setCol}}}{\sphinxparam{\DUrole{n,n}{src}}}{}
\pysigstopsignatures
\sphinxAtStartPar
Assuming that the col index array is a permutation (full/subset)
of a source label array by supplying that src to setCol
maps the row labels to cidx and creates self.col (col label list)

\end{fulllineitems}

\index{setRow() (StructMatrix method)@\spxentry{setRow()}\spxextra{StructMatrix method}}

\begin{fulllineitems}
\phantomsection\label{\detokenize{modules_doc:cbmpy.CBCommon.StructMatrix.setRow}}
\pysigstartsignatures
\pysiglinewithargsret{\sphinxbfcode{\sphinxupquote{setRow}}}{\sphinxparam{\DUrole{n,n}{src}}}{}
\pysigstopsignatures
\sphinxAtStartPar
Assuming that the row index array is a permutation (full/subset)
of a source label array by supplying that source to setRow it
maps the row labels to ridx and creates self.row (row label list)

\end{fulllineitems}


\end{fulllineitems}

\index{StructMatrixLP (class in cbmpy.CBCommon)@\spxentry{StructMatrixLP}\spxextra{class in cbmpy.CBCommon}}

\begin{fulllineitems}
\phantomsection\label{\detokenize{modules_doc:cbmpy.CBCommon.StructMatrixLP}}
\pysigstartsignatures
\pysiglinewithargsret{\sphinxbfcode{\sphinxupquote{class\DUrole{w,w}{  }}}\sphinxbfcode{\sphinxupquote{StructMatrixLP}}}{\sphinxparam{\DUrole{n,n}{array}}\sphinxparamcomma \sphinxparam{\DUrole{n,n}{ridx}}\sphinxparamcomma \sphinxparam{\DUrole{n,n}{cidx}}\sphinxparamcomma \sphinxparam{\DUrole{n,n}{row}\DUrole{o,o}{=}\DUrole{default_value}{None}}\sphinxparamcomma \sphinxparam{\DUrole{n,n}{col}\DUrole{o,o}{=}\DUrole{default_value}{None}}\sphinxparamcomma \sphinxparam{\DUrole{n,n}{rhs}\DUrole{o,o}{=}\DUrole{default_value}{None}}\sphinxparamcomma \sphinxparam{\DUrole{n,n}{operators}\DUrole{o,o}{=}\DUrole{default_value}{None}}}{}
\pysigstopsignatures
\sphinxAtStartPar
Adds some stuff to StructMatrix that makes it LP friendly
\index{getCopy() (StructMatrixLP method)@\spxentry{getCopy()}\spxextra{StructMatrixLP method}}

\begin{fulllineitems}
\phantomsection\label{\detokenize{modules_doc:cbmpy.CBCommon.StructMatrixLP.getCopy}}
\pysigstartsignatures
\pysiglinewithargsret{\sphinxbfcode{\sphinxupquote{getCopy}}}{\sphinxparam{\DUrole{n,n}{attr\_str}}\sphinxparamcomma \sphinxparam{\DUrole{n,n}{deep}\DUrole{o,o}{=}\DUrole{default_value}{False}}}{}
\pysigstopsignatures
\sphinxAtStartPar
Return a copy of the attribute with name attr\_str. Uses the copy module \sphinxtitleref{copy.copy} or \sphinxtitleref{copy.deepcopy}
\begin{itemize}
\item {} 
\sphinxAtStartPar
\sphinxstyleemphasis{attr\_str} a string of the attribute name: ‘row’, ‘col’

\item {} 
\sphinxAtStartPar
\sphinxstyleemphasis{deep} {[}default=False{]} try to do a deepcopy. Use with caution see copy module docstring for details

\end{itemize}

\end{fulllineitems}


\end{fulllineitems}

\index{checkChemFormula() (in module cbmpy.CBCommon)@\spxentry{checkChemFormula()}\spxextra{in module cbmpy.CBCommon}}

\begin{fulllineitems}
\phantomsection\label{\detokenize{modules_doc:cbmpy.CBCommon.checkChemFormula}}
\pysigstartsignatures
\pysiglinewithargsret{\sphinxbfcode{\sphinxupquote{checkChemFormula}}}{\sphinxparam{\DUrole{n,n}{cf}}\sphinxparamcomma \sphinxparam{\DUrole{n,n}{quiet}\DUrole{o,o}{=}\DUrole{default_value}{False}}}{}
\pysigstopsignatures
\sphinxAtStartPar
Checks whether a string conforms to a Chemical Formula C3Br5 etc, returns True/False. Please see the SBML
Level 3 specification and \sphinxurl{http://wikipedia.org/wiki/Hill\_system} for more information.
\begin{itemize}
\item {} 
\sphinxAtStartPar
\sphinxstyleemphasis{cf} a string that contains a formula to check

\item {} 
\sphinxAtStartPar
\sphinxstyleemphasis{quiet} {[}default=False{]} do not print error messages

\end{itemize}

\end{fulllineitems}

\index{checkId() (in module cbmpy.CBCommon)@\spxentry{checkId()}\spxextra{in module cbmpy.CBCommon}}

\begin{fulllineitems}
\phantomsection\label{\detokenize{modules_doc:cbmpy.CBCommon.checkId}}
\pysigstartsignatures
\pysiglinewithargsret{\sphinxbfcode{\sphinxupquote{checkId}}}{\sphinxparam{\DUrole{n,n}{s}}}{}
\pysigstopsignatures
\sphinxAtStartPar
Checks the validity of the string to see if it conforms to a C variable. Returns true/false
\begin{itemize}
\item {} 
\sphinxAtStartPar
\sphinxstyleemphasis{s} a string

\end{itemize}

\end{fulllineitems}

\index{createAssociationDictFromNode() (in module cbmpy.CBCommon)@\spxentry{createAssociationDictFromNode()}\spxextra{in module cbmpy.CBCommon}}

\begin{fulllineitems}
\phantomsection\label{\detokenize{modules_doc:cbmpy.CBCommon.createAssociationDictFromNode}}
\pysigstartsignatures
\pysiglinewithargsret{\sphinxbfcode{\sphinxupquote{createAssociationDictFromNode}}}{\sphinxparam{\DUrole{n,n}{node}}\sphinxparamcomma \sphinxparam{\DUrole{n,n}{out}}\sphinxparamcomma \sphinxparam{\DUrole{n,n}{model}}\sphinxparamcomma \sphinxparam{\DUrole{n,n}{useweakref}\DUrole{o,o}{=}\DUrole{default_value}{True}}\sphinxparamcomma \sphinxparam{\DUrole{n,n}{cntr}\DUrole{o,o}{=}\DUrole{default_value}{0}}}{}
\pysigstopsignatures
\sphinxAtStartPar
Converts a GPR string ‘((g1 and g2) or g3)’ to a dictionary via a Python AST.
In future I will get rid of all the string elements and work only with AST’s.
\begin{itemize}
\item {} 
\sphinxAtStartPar
\sphinxstyleemphasis{node} a Python AST node (e.g. body)

\item {} 
\sphinxAtStartPar
\sphinxstyleemphasis{out} a gpr dictionary

\item {} 
\sphinxAtStartPar
\sphinxstyleemphasis{model} a CBMPy model instance

\item {} 
\sphinxAtStartPar
\sphinxstyleemphasis{useweakref} {[}default=True{]} use a weakref as the gene object or alternatively the label

\end{itemize}

\end{fulllineitems}

\index{extractGeneIdsFromString() (in module cbmpy.CBCommon)@\spxentry{extractGeneIdsFromString()}\spxextra{in module cbmpy.CBCommon}}

\begin{fulllineitems}
\phantomsection\label{\detokenize{modules_doc:cbmpy.CBCommon.extractGeneIdsFromString}}
\pysigstartsignatures
\pysiglinewithargsret{\sphinxbfcode{\sphinxupquote{extractGeneIdsFromString}}}{\sphinxparam{\DUrole{n,n}{g}}\sphinxparamcomma \sphinxparam{\DUrole{n,n}{return\_clean\_gpr}\DUrole{o,o}{=}\DUrole{default_value}{False}}}{}
\pysigstopsignatures
\sphinxAtStartPar
Extract and return a list of gene names from a gene association string formulation
\begin{itemize}
\item {} 
\sphinxAtStartPar
\sphinxstyleemphasis{g} a COBRA style gene association string

\item {} 
\sphinxAtStartPar
\sphinxstyleemphasis{return\_clean\_gpr} {[}default=False{]} in addition to the list returns the “cleaned” GPR string

\end{itemize}

\end{fulllineitems}

\index{fixId() (in module cbmpy.CBCommon)@\spxentry{fixId()}\spxextra{in module cbmpy.CBCommon}}

\begin{fulllineitems}
\phantomsection\label{\detokenize{modules_doc:cbmpy.CBCommon.fixId}}
\pysigstartsignatures
\pysiglinewithargsret{\sphinxbfcode{\sphinxupquote{fixId}}}{\sphinxparam{\DUrole{n,n}{s}}\sphinxparamcomma \sphinxparam{\DUrole{n,n}{replace}\DUrole{o,o}{=}\DUrole{default_value}{None}}}{}
\pysigstopsignatures
\sphinxAtStartPar
Checks a string (Sid) to see if it is a valid C style variable. first letter must be an underscore or letter,
the rest should be alphanumeric or underscore.
\begin{itemize}
\item {} 
\sphinxAtStartPar
\sphinxstyleemphasis{s} the string to test

\item {} 
\sphinxAtStartPar
\sphinxstyleemphasis{replace} {[}None{]} default is to leave out offensive character, otherwise replace with this one

\end{itemize}

\end{fulllineitems}

\index{func\_getAssociationStrFromGprDict() (in module cbmpy.CBCommon)@\spxentry{func\_getAssociationStrFromGprDict()}\spxextra{in module cbmpy.CBCommon}}

\begin{fulllineitems}
\phantomsection\label{\detokenize{modules_doc:cbmpy.CBCommon.func_getAssociationStrFromGprDict}}
\pysigstartsignatures
\pysiglinewithargsret{\sphinxbfcode{\sphinxupquote{func\_getAssociationStrFromGprDict}}}{\sphinxparam{\DUrole{n,n}{gprd}}\sphinxparamcomma \sphinxparam{\DUrole{n,n}{out}}\sphinxparamcomma \sphinxparam{\DUrole{n,n}{parent}\DUrole{o,o}{=}\DUrole{default_value}{\textquotesingle{}\textquotesingle{}}}}{}
\pysigstopsignatures
\sphinxAtStartPar
Get a old school GPR association string from a CBMPy gprDict, e.g. obtained from gpr.getTree()
\begin{itemize}
\item {} 
\sphinxAtStartPar
\sphinxstyleemphasis{gprd} the gprDictionary

\item {} 
\sphinxAtStartPar
\sphinxstyleemphasis{out} the output string

\item {} 
\sphinxAtStartPar
\sphinxstyleemphasis{parent} {[}default=’’{]} the string representing the current nodes parent relationship, used for recursion

\end{itemize}

\end{fulllineitems}

\index{getGPRasDictFromString() (in module cbmpy.CBCommon)@\spxentry{getGPRasDictFromString()}\spxextra{in module cbmpy.CBCommon}}

\begin{fulllineitems}
\phantomsection\label{\detokenize{modules_doc:cbmpy.CBCommon.getGPRasDictFromString}}
\pysigstartsignatures
\pysiglinewithargsret{\sphinxbfcode{\sphinxupquote{getGPRasDictFromString}}}{\sphinxparam{\DUrole{n,n}{node}}\sphinxparamcomma \sphinxparam{\DUrole{n,n}{out}}}{}
\pysigstopsignatures
\sphinxAtStartPar
Converts a GPR string ‘((g1 and g2) or g3)’ to a gprDict which is returned
\begin{itemize}
\item {} 
\sphinxAtStartPar
\sphinxstyleemphasis{node} a Python AST note (e.g. \sphinxtitleref{ast.parse(gprstring).body{[}0{]}})

\item {} 
\sphinxAtStartPar
\sphinxstyleemphasis{out} a new dictionary that will be be created in place

\end{itemize}

\end{fulllineitems}

\index{parseGeneAssociation() (in module cbmpy.CBCommon)@\spxentry{parseGeneAssociation()}\spxextra{in module cbmpy.CBCommon}}

\begin{fulllineitems}
\phantomsection\label{\detokenize{modules_doc:cbmpy.CBCommon.parseGeneAssociation}}
\pysigstartsignatures
\pysiglinewithargsret{\sphinxbfcode{\sphinxupquote{parseGeneAssociation}}}{\sphinxparam{\DUrole{n,n}{gs}}}{}
\pysigstopsignatures
\sphinxAtStartPar
Parse a COBRA style gene association into a nested list.
\begin{itemize}
\item {} 
\sphinxAtStartPar
\sphinxstyleemphasis{gs} a string containing a gene association

\end{itemize}

\end{fulllineitems}

\index{processSpeciesChargeChemFormulaAnnot() (in module cbmpy.CBCommon)@\spxentry{processSpeciesChargeChemFormulaAnnot()}\spxextra{in module cbmpy.CBCommon}}

\begin{fulllineitems}
\phantomsection\label{\detokenize{modules_doc:cbmpy.CBCommon.processSpeciesChargeChemFormulaAnnot}}
\pysigstartsignatures
\pysiglinewithargsret{\sphinxbfcode{\sphinxupquote{processSpeciesChargeChemFormulaAnnot}}}{\sphinxparam{\DUrole{n,n}{s}}\sphinxparamcomma \sphinxparam{\DUrole{n,n}{getFromName}\DUrole{o,o}{=}\DUrole{default_value}{False}}\sphinxparamcomma \sphinxparam{\DUrole{n,n}{overwriteChemFormula}\DUrole{o,o}{=}\DUrole{default_value}{False}}\sphinxparamcomma \sphinxparam{\DUrole{n,n}{overwriteCharge}\DUrole{o,o}{=}\DUrole{default_value}{False}}}{}
\pysigstopsignatures
\sphinxAtStartPar
Disambiguate the chemical formula from either the Notes or the overloaded name
\begin{itemize}
\item {} 
\sphinxAtStartPar
\sphinxstyleemphasis{s} a species object

\item {} 
\sphinxAtStartPar
\sphinxstyleemphasis{getFromName} {[}default=False{]} whether to try strip the chemical formula from the name (old COBRA style)

\item {} 
\sphinxAtStartPar
\sphinxstyleemphasis{overwriteChemFormula} {[}default=False{]}

\item {} 
\sphinxAtStartPar
\sphinxstyleemphasis{overwriteCharge} {[}default=False{]}

\end{itemize}

\end{fulllineitems}

\phantomsection\label{\detokenize{modules_doc:module-cbmpy.CBConfig}}\index{module@\spxentry{module}!cbmpy.CBConfig@\spxentry{cbmpy.CBConfig}}\index{cbmpy.CBConfig@\spxentry{cbmpy.CBConfig}!module@\spxentry{module}}

\section{CBMPy: CBConfig module}
\label{\detokenize{modules_doc:cbmpy-cbconfig-module}}
\sphinxAtStartPar
PySCeS Constraint Based Modelling (\sphinxurl{http://cbmpy.sourceforge.net})
Copyright (C) 2009\sphinxhyphen{}2024 Brett G. Olivier, VU University Amsterdam, Amsterdam, The Netherlands

\sphinxAtStartPar
This program is free software: you can redistribute it and/or modify
it under the terms of the GNU General Public License as published by
the Free Software Foundation, either version 3 of the License, or
(at your option) any later version.

\sphinxAtStartPar
This program is distributed in the hope that it will be useful,
but WITHOUT ANY WARRANTY; without even the implied warranty of
MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
GNU General Public License for more details.

\sphinxAtStartPar
You should have received a copy of the GNU General Public License
along with this program.  If not, see \textless{}\sphinxurl{http://www.gnu.org/licenses/}\textgreater{}

\sphinxAtStartPar
Author: Brett G. Olivier PhD
Contact developers: \sphinxurl{https://github.com/SystemsBioinformatics/cbmpy/issues}
Last edit: \$Author: bgoli \$ (\$Id: CBConfig.py 711 2020\sphinxhyphen{}04\sphinxhyphen{}27 14:22:34Z bgoli \$)
\index{current\_version() (in module cbmpy.CBConfig)@\spxentry{current\_version()}\spxextra{in module cbmpy.CBConfig}}

\begin{fulllineitems}
\phantomsection\label{\detokenize{modules_doc:cbmpy.CBConfig.current_version}}
\pysigstartsignatures
\pysiglinewithargsret{\sphinxbfcode{\sphinxupquote{current\_version}}}{}{}
\pysigstopsignatures
\sphinxAtStartPar
Return the current CBMPy version as a string

\end{fulllineitems}

\index{current\_version\_tuple() (in module cbmpy.CBConfig)@\spxentry{current\_version\_tuple()}\spxextra{in module cbmpy.CBConfig}}

\begin{fulllineitems}
\phantomsection\label{\detokenize{modules_doc:cbmpy.CBConfig.current_version_tuple}}
\pysigstartsignatures
\pysiglinewithargsret{\sphinxbfcode{\sphinxupquote{current\_version\_tuple}}}{}{}
\pysigstopsignatures
\sphinxAtStartPar
Return the current CBMPy version as a tuple (x, y, z)

\end{fulllineitems}

\phantomsection\label{\detokenize{modules_doc:module-cbmpy.CBCPLEX}}\index{module@\spxentry{module}!cbmpy.CBCPLEX@\spxentry{cbmpy.CBCPLEX}}\index{cbmpy.CBCPLEX@\spxentry{cbmpy.CBCPLEX}!module@\spxentry{module}}

\section{CBMPy: CBCPLEX module}
\label{\detokenize{modules_doc:cbmpy-cbcplex-module}}
\sphinxAtStartPar
PySCeS Constraint Based Modelling (\sphinxurl{http://cbmpy.sourceforge.net})
Copyright (C) 2009\sphinxhyphen{}2024 Brett G. Olivier, VU University Amsterdam, Amsterdam, The Netherlands

\sphinxAtStartPar
This program is free software: you can redistribute it and/or modify
it under the terms of the GNU General Public License as published by
the Free Software Foundation, either version 3 of the License, or
(at your option) any later version.

\sphinxAtStartPar
This program is distributed in the hope that it will be useful,
but WITHOUT ANY WARRANTY; without even the implied warranty of
MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
GNU General Public License for more details.

\sphinxAtStartPar
You should have received a copy of the GNU General Public License
along with this program.  If not, see \textless{}\sphinxurl{http://www.gnu.org/licenses/}\textgreater{}

\sphinxAtStartPar
Author: Brett G. Olivier PhD
Contact developers: \sphinxurl{https://github.com/SystemsBioinformatics/cbmpy/issues}
Last edit: \$Author: bgoli \$ (\$Id: CBCPLEX.py 710 2020\sphinxhyphen{}04\sphinxhyphen{}27 14:22:34Z bgoli \$)
\index{cplx\_FluxVariabilityAnalysis() (in module cbmpy.CBCPLEX)@\spxentry{cplx\_FluxVariabilityAnalysis()}\spxextra{in module cbmpy.CBCPLEX}}

\begin{fulllineitems}
\phantomsection\label{\detokenize{modules_doc:cbmpy.CBCPLEX.cplx_FluxVariabilityAnalysis}}
\pysigstartsignatures
\pysiglinewithargsret{\sphinxbfcode{\sphinxupquote{cplx\_FluxVariabilityAnalysis}}}{\sphinxparam{\DUrole{n,n}{fba}}\sphinxparamcomma \sphinxparam{\DUrole{n,n}{selected\_reactions}\DUrole{o,o}{=}\DUrole{default_value}{None}}\sphinxparamcomma \sphinxparam{\DUrole{n,n}{pre\_opt}\DUrole{o,o}{=}\DUrole{default_value}{True}}\sphinxparamcomma \sphinxparam{\DUrole{n,n}{tol}\DUrole{o,o}{=}\DUrole{default_value}{None}}\sphinxparamcomma \sphinxparam{\DUrole{n,n}{objF2constr}\DUrole{o,o}{=}\DUrole{default_value}{True}}\sphinxparamcomma \sphinxparam{\DUrole{n,n}{rhs\_sense}\DUrole{o,o}{=}\DUrole{default_value}{\textquotesingle{}lower\textquotesingle{}}}\sphinxparamcomma \sphinxparam{\DUrole{n,n}{optPercentage}\DUrole{o,o}{=}\DUrole{default_value}{100.0}}\sphinxparamcomma \sphinxparam{\DUrole{n,n}{work\_dir}\DUrole{o,o}{=}\DUrole{default_value}{None}}\sphinxparamcomma \sphinxparam{\DUrole{n,n}{quiet}\DUrole{o,o}{=}\DUrole{default_value}{True}}\sphinxparamcomma \sphinxparam{\DUrole{n,n}{debug}\DUrole{o,o}{=}\DUrole{default_value}{False}}\sphinxparamcomma \sphinxparam{\DUrole{n,n}{oldlpgen}\DUrole{o,o}{=}\DUrole{default_value}{False}}\sphinxparamcomma \sphinxparam{\DUrole{n,n}{markupmodel}\DUrole{o,o}{=}\DUrole{default_value}{True}}\sphinxparamcomma \sphinxparam{\DUrole{n,n}{default\_on\_fail}\DUrole{o,o}{=}\DUrole{default_value}{False}}\sphinxparamcomma \sphinxparam{\DUrole{n,n}{roundoff\_span}\DUrole{o,o}{=}\DUrole{default_value}{10}}\sphinxparamcomma \sphinxparam{\DUrole{n,n}{method}\DUrole{o,o}{=}\DUrole{default_value}{\textquotesingle{}o\textquotesingle{}}}}{}
\pysigstopsignatures
\sphinxAtStartPar
Perform a flux variability analysis on an fba model:
\begin{itemize}
\item {} 
\sphinxAtStartPar
\sphinxstyleemphasis{fba} an FBA model object

\item {} 
\sphinxAtStartPar
\sphinxstyleemphasis{selected reactions} {[}default=None{]} means use all reactions otherwise use the reactions listed here

\item {} 
\sphinxAtStartPar
\sphinxstyleemphasis{pre\_opt} {[}default=True{]} attempt to presolve the FBA and report its results in the ouput, if this is disabled and \sphinxstyleemphasis{objF2constr} is True then the rid/value of the current active objective is used

\item {} 
\sphinxAtStartPar
\sphinxstyleemphasis{tol}  {[}default=None{]} do not floor/ceiling the objective function constraint, otherwise round of to \sphinxstyleemphasis{tol}

\item {} 
\sphinxAtStartPar
\sphinxstyleemphasis{rhs\_sense} {[}default=’lower’{]} means objC \textgreater{}= objVal the inequality to use for the objective constraint can also be \sphinxstyleemphasis{upper} or \sphinxstyleemphasis{equal}

\item {} 
\sphinxAtStartPar
\sphinxstyleemphasis{optPercentage} {[}default=100.0{]} means the percentage optimal value to use for the RHS of the objective constraint: optimal\_value*(optPercentage/100.0)

\item {} 
\sphinxAtStartPar
\sphinxstyleemphasis{work\_dir} {[}default=None{]} the FVA working directory for temporary files default = cwd+fva

\item {} 
\sphinxAtStartPar
\sphinxstyleemphasis{debug} {[}default=False{]} if True write out all the intermediate FVA LP’s into work\_dir

\item {} 
\sphinxAtStartPar
\sphinxstyleemphasis{quiet} {[}default=False{]} if enabled, supress CPLEX output

\item {} 
\sphinxAtStartPar
\sphinxstyleemphasis{objF2constr} {[}default=True{]} add the model objective function as a constraint using rhs\_sense etc. If
this is True with pre\_opt=False then the id/value of the active objective is used to form the constraint

\item {} 
\sphinxAtStartPar
\sphinxstyleemphasis{markupmodel} {[}default=True{]} add the values returned by the fva to the reaction.fva\_min and reaction.fva\_max

\item {} 
\sphinxAtStartPar
\sphinxstyleemphasis{default\_on\_fail} {[}default=False{]} if \sphinxstyleemphasis{pre\_opt} is enabled replace a failed minimum/maximum with the solution value

\item {} 
\sphinxAtStartPar
\sphinxstyleemphasis{roundoff\_span} {[}default=10{]} number of digits is round off (not individual min/max values)

\item {} 
\sphinxAtStartPar
\sphinxstyleemphasis{method} {[}default=’o’{]} choose the CPLEX method to use for solution, default is automatic. See CPLEX reference manual for details
\begin{itemize}
\item {} 
\sphinxAtStartPar
‘o’: auto

\item {} 
\sphinxAtStartPar
‘p’: primal

\item {} 
\sphinxAtStartPar
‘d’: dual

\item {} 
\sphinxAtStartPar
‘b’: barrier (no crossover)

\item {} 
\sphinxAtStartPar
‘h’: barrier

\item {} 
\sphinxAtStartPar
‘s’: sifting

\item {} 
\sphinxAtStartPar
‘c’: concurrent

\end{itemize}

\end{itemize}

\sphinxAtStartPar
Returns an array with columns: Reaction, Reduced Costs, Variability Min, Variability Max, abs(Max\sphinxhyphen{}Min), MinStatus, MaxStatus and a list containing the row names.

\end{fulllineitems}

\index{cplx\_MinimizeNumActiveFluxes() (in module cbmpy.CBCPLEX)@\spxentry{cplx\_MinimizeNumActiveFluxes()}\spxextra{in module cbmpy.CBCPLEX}}

\begin{fulllineitems}
\phantomsection\label{\detokenize{modules_doc:cbmpy.CBCPLEX.cplx_MinimizeNumActiveFluxes}}
\pysigstartsignatures
\pysiglinewithargsret{\sphinxbfcode{\sphinxupquote{cplx\_MinimizeNumActiveFluxes}}}{\sphinxparam{\DUrole{n,n}{fba}}\sphinxparamcomma \sphinxparam{\DUrole{n,n}{selected\_reactions}\DUrole{o,o}{=}\DUrole{default_value}{None}}\sphinxparamcomma \sphinxparam{\DUrole{n,n}{pre\_opt}\DUrole{o,o}{=}\DUrole{default_value}{True}}\sphinxparamcomma \sphinxparam{\DUrole{n,n}{tol}\DUrole{o,o}{=}\DUrole{default_value}{None}}\sphinxparamcomma \sphinxparam{\DUrole{n,n}{objF2constr}\DUrole{o,o}{=}\DUrole{default_value}{True}}\sphinxparamcomma \sphinxparam{\DUrole{n,n}{rhs\_sense}\DUrole{o,o}{=}\DUrole{default_value}{\textquotesingle{}lower\textquotesingle{}}}\sphinxparamcomma \sphinxparam{\DUrole{n,n}{optPercentage}\DUrole{o,o}{=}\DUrole{default_value}{100.0}}\sphinxparamcomma \sphinxparam{\DUrole{n,n}{work\_dir}\DUrole{o,o}{=}\DUrole{default_value}{None}}\sphinxparamcomma \sphinxparam{\DUrole{n,n}{quiet}\DUrole{o,o}{=}\DUrole{default_value}{False}}\sphinxparamcomma \sphinxparam{\DUrole{n,n}{debug}\DUrole{o,o}{=}\DUrole{default_value}{False}}\sphinxparamcomma \sphinxparam{\DUrole{n,n}{objective\_coefficients}\DUrole{o,o}{=}\DUrole{default_value}{None}}\sphinxparamcomma \sphinxparam{\DUrole{n,n}{return\_lp\_obj}\DUrole{o,o}{=}\DUrole{default_value}{False}}\sphinxparamcomma \sphinxparam{\DUrole{n,n}{populate}\DUrole{o,o}{=}\DUrole{default_value}{None}}\sphinxparamcomma \sphinxparam{\DUrole{n,n}{oldlpgen}\DUrole{o,o}{=}\DUrole{default_value}{False}}}{}
\pysigstopsignatures
\sphinxAtStartPar
Minimize the sum of active fluxes, updates the model with the values of the solution and returns the value
of the MILP objective function (not the model objective function which remains unchanged). If population mode is activated
output is as described below:
\begin{quote}
\begin{description}
\sphinxlineitem{Min: sum(Bi)}
\sphinxAtStartPar
Bi = 0 \sphinxhyphen{}\textgreater{} Ci Ji = 0

\sphinxlineitem{Such that:}
\sphinxAtStartPar
NJi = 0
Jbio = opt

\sphinxlineitem{where:}
\sphinxAtStartPar
Binary Bi

\end{description}
\end{quote}

\sphinxAtStartPar
Arguments:
\begin{itemize}
\item {} 
\sphinxAtStartPar
\sphinxstyleemphasis{fba} an FBA model object

\item {} 
\sphinxAtStartPar
\sphinxstyleemphasis{selected reactions} {[}default=None{]} means use all reactions otherwise use the reactions listed here

\item {} 
\sphinxAtStartPar
\sphinxstyleemphasis{pre\_opt} {[}default=True{]} attempt to presolve the FBA and report its results in the ouput, if this is diabled and \sphinxstyleemphasis{objF2constr} is True then the vid/value of the current active objective is used

\item {} 
\sphinxAtStartPar
\sphinxstyleemphasis{tol}  {[}default=None{]} do not floor/ceiling the objective function constraint, otherwise round of to \sphinxstyleemphasis{tol}

\item {} 
\sphinxAtStartPar
\sphinxstyleemphasis{rhs\_sense} {[}default=’lower’{]} means objC \textgreater{}= objVal the inequality to use for the objective constraint can also be \sphinxstyleemphasis{upper} or \sphinxstyleemphasis{equal}
Note this does not necessarily mean the upper or lower bound, although practically it will. If in doubt use \sphinxstyleemphasis{equal}

\item {} 
\sphinxAtStartPar
\sphinxstyleemphasis{optPercentage} {[}default=100.0{]} means the percentage optimal value to use for the RHS of the objective constraint: optimal\_value * (optPercentage/100.0)

\item {} 
\sphinxAtStartPar
\sphinxstyleemphasis{work\_dir} {[}default=None{]} the MSAF working directory for temporary files default = cwd+fva

\item {} 
\sphinxAtStartPar
\sphinxstyleemphasis{debug} {[}default=False{]} if True write out all the intermediate MSAF LP’s into work\_dir

\item {} 
\sphinxAtStartPar
\sphinxstyleemphasis{quiet} {[}default=False{]} if enabled supress CPLEX output

\item {} 
\sphinxAtStartPar
\sphinxstyleemphasis{objF2constr} {[}default=True{]} add the model objective function as a constraint using rhs\_sense etc. If
this is True with pre\_opt=False then the id/value of the active objective is used to form the constraint

\item {} 
\sphinxAtStartPar
\sphinxstyleemphasis{objective\_coefficients} {[}default=None{]} a dictionary of (reaction\_id : float) pairs that provide the are introduced as objective coefficients to the absolute flux value. Note that the default value of the coefficient (non\sphinxhyphen{}specified) is +1.

\item {} 
\sphinxAtStartPar
\sphinxstyleemphasis{return\_lp\_obj} {[}default=False{]} off by default when enabled it returns the CPLEX LP object

\item {} 
\sphinxAtStartPar
\sphinxstyleemphasis{populate} {[}default=None{]} enable search algorithm to find multiple (sub)optimal solutions. Set with a tuple of (RELGAP=0.0, POPULATE\_LIMIT=20, TIME\_LIMIT=300) suggested values only.
\sphinxhyphen{} \sphinxstyleemphasis{RELGAP} {[}default=0.0{]} relative gap to optimal solution
\sphinxhyphen{} \sphinxstyleemphasis{POPULATE\_LIMIT} {[}default=20{]} terminate when so many solutions have been found
\sphinxhyphen{} \sphinxstyleemphasis{TIME\_LIMIT} {[}default=300{]} terminate search after so many seconds important with higher values of \sphinxstyleemphasis{POPULATION\_LIMIT}

\item {} 
\sphinxAtStartPar
\sphinxstyleemphasis{with\_reduced\_costs} {[}default=’uncsaled’{]} can be ‘scaled’, ‘unscaled’ or anything else which is None

\end{itemize}

\sphinxAtStartPar
With outputs:
\begin{itemize}
\item {} 
\sphinxAtStartPar
\sphinxstyleemphasis{mincnt} the objective function value OR

\item {} 
\sphinxAtStartPar
\sphinxstyleemphasis{mincnt, cpx} the objective function and cplex model OR

\item {} 
\sphinxAtStartPar
\sphinxstyleemphasis{populate\_data, mincnt} a population data set OR

\item {} 
\sphinxAtStartPar
\sphinxstyleemphasis{populate\_data, mincnt, cpx} both the cps object and population data set

\end{itemize}

\sphinxAtStartPar
depending on selected flags.

\end{fulllineitems}

\index{cplx\_MinimizeSumOfAbsFluxes() (in module cbmpy.CBCPLEX)@\spxentry{cplx\_MinimizeSumOfAbsFluxes()}\spxextra{in module cbmpy.CBCPLEX}}

\begin{fulllineitems}
\phantomsection\label{\detokenize{modules_doc:cbmpy.CBCPLEX.cplx_MinimizeSumOfAbsFluxes}}
\pysigstartsignatures
\pysiglinewithargsret{\sphinxbfcode{\sphinxupquote{cplx\_MinimizeSumOfAbsFluxes}}}{\sphinxparam{\DUrole{n,n}{fba}}\sphinxparamcomma \sphinxparam{\DUrole{n,n}{selected\_reactions}\DUrole{o,o}{=}\DUrole{default_value}{None}}\sphinxparamcomma \sphinxparam{\DUrole{n,n}{pre\_opt}\DUrole{o,o}{=}\DUrole{default_value}{True}}\sphinxparamcomma \sphinxparam{\DUrole{n,n}{tol}\DUrole{o,o}{=}\DUrole{default_value}{None}}\sphinxparamcomma \sphinxparam{\DUrole{n,n}{objF2constr}\DUrole{o,o}{=}\DUrole{default_value}{True}}\sphinxparamcomma \sphinxparam{\DUrole{n,n}{rhs\_sense}\DUrole{o,o}{=}\DUrole{default_value}{\textquotesingle{}lower\textquotesingle{}}}\sphinxparamcomma \sphinxparam{\DUrole{n,n}{optPercentage}\DUrole{o,o}{=}\DUrole{default_value}{100.0}}\sphinxparamcomma \sphinxparam{\DUrole{n,n}{work\_dir}\DUrole{o,o}{=}\DUrole{default_value}{None}}\sphinxparamcomma \sphinxparam{\DUrole{n,n}{quiet}\DUrole{o,o}{=}\DUrole{default_value}{False}}\sphinxparamcomma \sphinxparam{\DUrole{n,n}{debug}\DUrole{o,o}{=}\DUrole{default_value}{False}}\sphinxparamcomma \sphinxparam{\DUrole{n,n}{objective\_coefficients}\DUrole{o,o}{=}\DUrole{default_value}{None}}\sphinxparamcomma \sphinxparam{\DUrole{n,n}{return\_lp\_obj}\DUrole{o,o}{=}\DUrole{default_value}{False}}\sphinxparamcomma \sphinxparam{\DUrole{n,n}{oldlpgen}\DUrole{o,o}{=}\DUrole{default_value}{False}}\sphinxparamcomma \sphinxparam{\DUrole{n,n}{with\_reduced\_costs}\DUrole{o,o}{=}\DUrole{default_value}{None}}\sphinxparamcomma \sphinxparam{\DUrole{n,n}{method}\DUrole{o,o}{=}\DUrole{default_value}{\textquotesingle{}o\textquotesingle{}}}}{}
\pysigstopsignatures
\sphinxAtStartPar
Minimize the sum of absolute fluxes sum(abs(J1) + abs(J2) + abs(J3) … abs(Jn)) by adding two constraints per flux
and a variable representing the absolute value:
\begin{quote}
\begin{description}
\sphinxlineitem{Min: Ci abs\_Ji}
\sphinxAtStartPar
Ji \sphinxhyphen{} abs\_Ji \textless{}= 0
Ji + abs\_Ji \textgreater{}= 0

\sphinxlineitem{Such that:}
\sphinxAtStartPar
NJi = 0
Jopt = opt

\end{description}
\end{quote}

\sphinxAtStartPar
returns the value of the flux minimization objective function (not the model objective function which remains unchanged from)

\sphinxAtStartPar
Arguments:
\begin{itemize}
\item {} 
\sphinxAtStartPar
\sphinxstyleemphasis{fba} an FBA model object

\item {} 
\sphinxAtStartPar
\sphinxstyleemphasis{selected reactions} {[}default=None{]} means use all reactions otherwise use the reactions listed here

\item {} 
\sphinxAtStartPar
\sphinxstyleemphasis{pre\_opt} {[}default=True{]} attempt to presolve the FBA and report its results in the ouput, if this is disabled and \sphinxstyleemphasis{objF2constr} is True then the vid/value of the current active objective is used

\item {} 
\sphinxAtStartPar
\sphinxstyleemphasis{tol}  {[}default=None{]} do not floor/ceiling the objective function constraint, otherwise round of to \sphinxstyleemphasis{tol}

\item {} 
\sphinxAtStartPar
\sphinxstyleemphasis{rhs\_sense} {[}default=’lower’{]} means objC \textgreater{}= objVal the inequality to use for the objective constraint can also be \sphinxstyleemphasis{upper} or \sphinxstyleemphasis{equal}

\item {} 
\sphinxAtStartPar
\sphinxstyleemphasis{optPercentage} {[}default=100.0{]} means the percentage optimal value to use for the RHS of the objective constraint: optimal\_value*(optPercentage/100.0)

\item {} 
\sphinxAtStartPar
\sphinxstyleemphasis{work\_dir} {[}default=None{]} the MSAF working directory for temporary files default = cwd+fva

\item {} 
\sphinxAtStartPar
\sphinxstyleemphasis{debug} {[}default=False{]} if True write out all the intermediate MSAF LP’s into work\_dir

\item {} 
\sphinxAtStartPar
\sphinxstyleemphasis{quiet} {[}default=False{]} if enabled supress CPLEX output

\item {} 
\sphinxAtStartPar
\sphinxstyleemphasis{objF2constr} {[}default=True{]} add the model objective function as a constraint using rhs\_sense etc. If
this is True with pre\_opt=False then the id/value of the active objective is used to form the constraint

\item {} 
\sphinxAtStartPar
\sphinxstyleemphasis{objective\_coefficients} {[}default=None{]} a dictionary of (reaction\_id : float) pairs that provide the are introduced as objective coefficients to the absolute flux value. Note that the default value of the coefficient (non\sphinxhyphen{}specified) is +1.

\item {} 
\sphinxAtStartPar
\sphinxstyleemphasis{return\_lp\_obj} {[}default=False{]} off by default when enabled it returns the CPLEX LP object

\item {} 
\sphinxAtStartPar
\sphinxstyleemphasis{with\_reduced\_costs} {[}default=None{]} if not None should be ‘scaled’ or ‘unscaled’

\item {} 
\sphinxAtStartPar
\sphinxstyleemphasis{method} {[}default=’o’{]} choose the CPLEX method to use for solution, default is automatic. See CPLEX reference manual for details
\begin{itemize}
\item {} 
\sphinxAtStartPar
‘o’: auto

\item {} 
\sphinxAtStartPar
‘p’: primal

\item {} 
\sphinxAtStartPar
‘d’: dual

\item {} 
\sphinxAtStartPar
‘b’: barrier (no crossover)

\item {} 
\sphinxAtStartPar
‘h’: barrier

\item {} 
\sphinxAtStartPar
‘s’: sifting

\item {} 
\sphinxAtStartPar
‘c’: concurrent

\end{itemize}

\end{itemize}

\sphinxAtStartPar
With outputs:
\begin{itemize}
\item {} 
\sphinxAtStartPar
\sphinxstyleemphasis{fba} an update instance of a CBModel. Note that the FBA model objective function value is the original value set as a constraint

\end{itemize}

\end{fulllineitems}

\index{cplx\_MultiFluxVariabilityAnalysis() (in module cbmpy.CBCPLEX)@\spxentry{cplx\_MultiFluxVariabilityAnalysis()}\spxextra{in module cbmpy.CBCPLEX}}

\begin{fulllineitems}
\phantomsection\label{\detokenize{modules_doc:cbmpy.CBCPLEX.cplx_MultiFluxVariabilityAnalysis}}
\pysigstartsignatures
\pysiglinewithargsret{\sphinxbfcode{\sphinxupquote{cplx\_MultiFluxVariabilityAnalysis}}}{\sphinxparam{\DUrole{n,n}{lp}}\sphinxparamcomma \sphinxparam{\DUrole{n,n}{selected\_reactions}\DUrole{o,o}{=}\DUrole{default_value}{None}}\sphinxparamcomma \sphinxparam{\DUrole{n,n}{tol}\DUrole{o,o}{=}\DUrole{default_value}{1e\sphinxhyphen{}10}}\sphinxparamcomma \sphinxparam{\DUrole{n,n}{rhs\_sense}\DUrole{o,o}{=}\DUrole{default_value}{\textquotesingle{}lower\textquotesingle{}}}\sphinxparamcomma \sphinxparam{\DUrole{n,n}{optPercentage}\DUrole{o,o}{=}\DUrole{default_value}{100.0}}\sphinxparamcomma \sphinxparam{\DUrole{n,n}{work\_dir}\DUrole{o,o}{=}\DUrole{default_value}{None}}\sphinxparamcomma \sphinxparam{\DUrole{n,n}{debug}\DUrole{o,o}{=}\DUrole{default_value}{False}}}{}
\pysigstopsignatures
\sphinxAtStartPar
Perform a flux variability analysis on a multistate LP
\begin{itemize}
\item {} 
\sphinxAtStartPar
\sphinxstyleemphasis{lp} a multistate LP

\item {} 
\sphinxAtStartPar
\sphinxstyleemphasis{selected reactions} {[}default=None{]} means use all reactions otherwise use the reactions listed here

\item {} 
\sphinxAtStartPar
\sphinxstyleemphasis{pre\_opt} {[}default=True{]} attempt to presolve the FBA and report its results in the ouput

\item {} 
\sphinxAtStartPar
\sphinxstyleemphasis{tol}  {[}default=1e\sphinxhyphen{}10{]} do floor/ceiling the objective function constraint, otherwise floor/ceil to \sphinxstyleemphasis{tol}

\item {} 
\sphinxAtStartPar
\sphinxstyleemphasis{rhs\_sense} {[}default=’lower’{]} means objC \textgreater{}= objVal the inequality to use for the objective constraint can also be \sphinxstyleemphasis{upper} or \sphinxstyleemphasis{equal}

\item {} 
\sphinxAtStartPar
\sphinxstyleemphasis{optPercentage} {[}default=100.0{]} means the percentage optimal value to use for the RHS of the objective constraint: optimal\_value*(optPercentage/100.0)

\item {} 
\sphinxAtStartPar
\sphinxstyleemphasis{work\_dir} {[}default=None{]} the FVA working directory for temporary files default = cwd+fva

\item {} 
\sphinxAtStartPar
\sphinxstyleemphasis{debug} {[}default=False{]} if True write out all the intermediate FVA LP’s into work\_dir

\item {} 
\sphinxAtStartPar
\sphinxstyleemphasis{bypass} {[}default=False{]} bypass everything and only run the min/max on lp

\end{itemize}

\sphinxAtStartPar
and returns an array with columns:

\begin{sphinxVerbatim}[commandchars=\\\{\}]
\PYG{n}{Reaction}\PYG{p}{,} \PYG{n}{Reduced} \PYG{n}{Costs}\PYG{p}{,} \PYG{n}{Variability} \PYG{n}{Min}\PYG{p}{,} \PYG{n}{Variability} \PYG{n}{Max}\PYG{p}{,} \PYG{n+nb}{abs}\PYG{p}{(}\PYG{n}{Max}\PYG{o}{\PYGZhy{}}\PYG{n}{Min}\PYG{p}{)}\PYG{p}{,} \PYG{n}{MinStatus}\PYG{p}{,} \PYG{n}{MaxStatus}
\end{sphinxVerbatim}

\sphinxAtStartPar
and a list containing the row names.

\end{fulllineitems}

\index{cplx\_SolveMILP() (in module cbmpy.CBCPLEX)@\spxentry{cplx\_SolveMILP()}\spxextra{in module cbmpy.CBCPLEX}}

\begin{fulllineitems}
\phantomsection\label{\detokenize{modules_doc:cbmpy.CBCPLEX.cplx_SolveMILP}}
\pysigstartsignatures
\pysiglinewithargsret{\sphinxbfcode{\sphinxupquote{cplx\_SolveMILP}}}{\sphinxparam{\DUrole{n,n}{c}}\sphinxparamcomma \sphinxparam{\DUrole{n,n}{auto\_mipgap}\DUrole{o,o}{=}\DUrole{default_value}{False}}}{}
\pysigstopsignatures
\sphinxAtStartPar
Solve and MILP
\begin{itemize}
\item {} 
\sphinxAtStartPar
\sphinxstyleemphasis{auto\_mipgap} auto decrease mipgap until mipgap == absmipgap

\end{itemize}

\end{fulllineitems}

\index{cplx\_WriteFVAtoCSV() (in module cbmpy.CBCPLEX)@\spxentry{cplx\_WriteFVAtoCSV()}\spxextra{in module cbmpy.CBCPLEX}}

\begin{fulllineitems}
\phantomsection\label{\detokenize{modules_doc:cbmpy.CBCPLEX.cplx_WriteFVAtoCSV}}
\pysigstartsignatures
\pysiglinewithargsret{\sphinxbfcode{\sphinxupquote{cplx\_WriteFVAtoCSV}}}{\sphinxparam{\DUrole{n,n}{pid}}\sphinxparamcomma \sphinxparam{\DUrole{n,n}{fva}}\sphinxparamcomma \sphinxparam{\DUrole{n,n}{names}}\sphinxparamcomma \sphinxparam{\DUrole{n,n}{Dir}\DUrole{o,o}{=}\DUrole{default_value}{None}}\sphinxparamcomma \sphinxparam{\DUrole{n,n}{fbaObj}\DUrole{o,o}{=}\DUrole{default_value}{None}}}{}
\pysigstopsignatures
\sphinxAtStartPar
Takes the resuls of a FluxVariabilityAnalysis method and writes it to a nice
csv file. Note this method has been refactored to \sphinxtitleref{CBWrite.WriteFVAtoCSV()}.
\begin{itemize}
\item {} 
\sphinxAtStartPar
\sphinxstyleemphasis{pid} filename\_base for the CSV output

\item {} 
\sphinxAtStartPar
\sphinxstyleemphasis{fva} FluxVariabilityAnalysis() OUTPUT\_ARRAY

\item {} 
\sphinxAtStartPar
\sphinxstyleemphasis{names} FluxVariabilityAnalysis() OUTPUT\_NAMES

\item {} 
\sphinxAtStartPar
\sphinxstyleemphasis{Dir} {[}default=None{]} if set the output directory for the csv files

\item {} 
\sphinxAtStartPar
\sphinxstyleemphasis{fbaObj} {[}default=None{]} if supplied adds extra model information into the output tables

\end{itemize}

\end{fulllineitems}

\index{cplx\_analyzeModel() (in module cbmpy.CBCPLEX)@\spxentry{cplx\_analyzeModel()}\spxextra{in module cbmpy.CBCPLEX}}

\begin{fulllineitems}
\phantomsection\label{\detokenize{modules_doc:cbmpy.CBCPLEX.cplx_analyzeModel}}
\pysigstartsignatures
\pysiglinewithargsret{\sphinxbfcode{\sphinxupquote{cplx\_analyzeModel}}}{\sphinxparam{\DUrole{n,n}{f}}\sphinxparamcomma \sphinxparam{\DUrole{n,n}{lpFname}\DUrole{o,o}{=}\DUrole{default_value}{None}}\sphinxparamcomma \sphinxparam{\DUrole{n,n}{return\_lp\_obj}\DUrole{o,o}{=}\DUrole{default_value}{False}}\sphinxparamcomma \sphinxparam{\DUrole{n,n}{with\_reduced\_costs}\DUrole{o,o}{=}\DUrole{default_value}{\textquotesingle{}unscaled\textquotesingle{}}}\sphinxparamcomma \sphinxparam{\DUrole{n,n}{with\_sensitivity}\DUrole{o,o}{=}\DUrole{default_value}{False}}\sphinxparamcomma \sphinxparam{\DUrole{n,n}{del\_intermediate}\DUrole{o,o}{=}\DUrole{default_value}{False}}\sphinxparamcomma \sphinxparam{\DUrole{n,n}{build\_n}\DUrole{o,o}{=}\DUrole{default_value}{True}}\sphinxparamcomma \sphinxparam{\DUrole{n,n}{quiet}\DUrole{o,o}{=}\DUrole{default_value}{False}}\sphinxparamcomma \sphinxparam{\DUrole{n,n}{oldlpgen}\DUrole{o,o}{=}\DUrole{default_value}{False}}\sphinxparamcomma \sphinxparam{\DUrole{n,n}{method}\DUrole{o,o}{=}\DUrole{default_value}{\textquotesingle{}o\textquotesingle{}}}}{}
\pysigstopsignatures
\sphinxAtStartPar
Optimize a model and add the result of the optimization to the model object
(e.g. \sphinxtitleref{reaction.value}, \sphinxtitleref{objectiveFunction.value}). The stoichiometric
matrix is automatically generated. This is a common function available
in all solver interfaces. By default returns the objective function value
\begin{itemize}
\item {} 
\sphinxAtStartPar
\sphinxstyleemphasis{f} an instantiated PySCeSCBM model object

\item {} 
\sphinxAtStartPar
\sphinxstyleemphasis{lpFname} {[}default=None{]} the name of the intermediate LP file. If not specified no LP file is produced

\item {} 
\sphinxAtStartPar
\sphinxstyleemphasis{return\_lp\_obj} {[}default=False{]} off by default when enabled it returns the CPLEX LP object

\item {} 
\sphinxAtStartPar
\sphinxstyleemphasis{with\_reduced\_costs} {[}default=’unscaled’{]} calculate and add reduced cost information to mode this can be: ‘unscaled’ or ‘scaled’
or anything else which is interpreted as ‘None’. Scaled means s\_rcost = (r.reduced\_cost*rval)/obj\_value

\item {} 
\sphinxAtStartPar
\sphinxstyleemphasis{with\_sensitivity} {[}default=False{]} add solution sensitivity information (not yet implemented)

\item {} 
\sphinxAtStartPar
\sphinxstyleemphasis{del\_intermediate} {[}default=False{]} redundant except if output file is produced and deleted (not useful)

\item {} 
\sphinxAtStartPar
\sphinxstyleemphasis{build\_n} {[}default=True{]} generate stoichiometry from the reaction network (reactions/reagents/species)

\item {} 
\sphinxAtStartPar
\sphinxstyleemphasis{quiet} {[}default=False{]} suppress cplex output

\item {} 
\sphinxAtStartPar
\sphinxstyleemphasis{method} {[}default=’o’{]} choose the CPLEX method to use for solution, default is automatic. See CPLEX reference manual for details
\begin{itemize}
\item {} 
\sphinxAtStartPar
‘o’: auto

\item {} 
\sphinxAtStartPar
‘p’: primal

\item {} 
\sphinxAtStartPar
‘d’: dual

\item {} 
\sphinxAtStartPar
‘b’: barrier (no crossover)

\item {} 
\sphinxAtStartPar
‘h’: barrier

\item {} 
\sphinxAtStartPar
‘s’: sifting

\item {} 
\sphinxAtStartPar
‘c’: concurrent

\end{itemize}

\end{itemize}

\end{fulllineitems}

\index{cplx\_constructLPfromFBA() (in module cbmpy.CBCPLEX)@\spxentry{cplx\_constructLPfromFBA()}\spxextra{in module cbmpy.CBCPLEX}}

\begin{fulllineitems}
\phantomsection\label{\detokenize{modules_doc:cbmpy.CBCPLEX.cplx_constructLPfromFBA}}
\pysigstartsignatures
\pysiglinewithargsret{\sphinxbfcode{\sphinxupquote{cplx\_constructLPfromFBA}}}{\sphinxparam{\DUrole{n,n}{fba}}\sphinxparamcomma \sphinxparam{\DUrole{n,n}{fname}\DUrole{o,o}{=}\DUrole{default_value}{None}}}{}
\pysigstopsignatures
\sphinxAtStartPar
Create a CPLEX LP in memory.
\sphinxhyphen{} \sphinxstyleemphasis{fba} an FBA object
\sphinxhyphen{} \sphinxstyleemphasis{fname} optional filename if defined writes out the constructed lp

\end{fulllineitems}

\index{cplx\_fixConSense() (in module cbmpy.CBCPLEX)@\spxentry{cplx\_fixConSense()}\spxextra{in module cbmpy.CBCPLEX}}

\begin{fulllineitems}
\phantomsection\label{\detokenize{modules_doc:cbmpy.CBCPLEX.cplx_fixConSense}}
\pysigstartsignatures
\pysiglinewithargsret{\sphinxbfcode{\sphinxupquote{cplx\_fixConSense}}}{\sphinxparam{\DUrole{n,n}{operator}}}{}
\pysigstopsignatures
\sphinxAtStartPar
Fixes the sense of inequality operators, returns corrected sense symbol
\begin{itemize}
\item {} 
\sphinxAtStartPar
\sphinxstyleemphasis{operator} the operator to check

\end{itemize}

\end{fulllineitems}

\index{cplx\_func\_GetCPXandPresolve() (in module cbmpy.CBCPLEX)@\spxentry{cplx\_func\_GetCPXandPresolve()}\spxextra{in module cbmpy.CBCPLEX}}

\begin{fulllineitems}
\phantomsection\label{\detokenize{modules_doc:cbmpy.CBCPLEX.cplx_func_GetCPXandPresolve}}
\pysigstartsignatures
\pysiglinewithargsret{\sphinxbfcode{\sphinxupquote{cplx\_func\_GetCPXandPresolve}}}{\sphinxparam{\DUrole{n,n}{fba}}\sphinxparamcomma \sphinxparam{\DUrole{n,n}{pre\_opt}}\sphinxparamcomma \sphinxparam{\DUrole{n,n}{objF2constr}}\sphinxparamcomma \sphinxparam{\DUrole{n,n}{quiet}\DUrole{o,o}{=}\DUrole{default_value}{False}}\sphinxparamcomma \sphinxparam{\DUrole{n,n}{oldlpgen}\DUrole{o,o}{=}\DUrole{default_value}{False}}\sphinxparamcomma \sphinxparam{\DUrole{n,n}{with\_reduced\_costs}\DUrole{o,o}{=}\DUrole{default_value}{\textquotesingle{}unscaled\textquotesingle{}}}\sphinxparamcomma \sphinxparam{\DUrole{n,n}{method}\DUrole{o,o}{=}\DUrole{default_value}{\textquotesingle{}o\textquotesingle{}}}}{}
\pysigstopsignatures
\sphinxAtStartPar
This is a utility function that does a presolve for FVA, MSAF etc. Generates properly formatted
empty objects if pre\_opt == False
\begin{itemize}
\item {} 
\sphinxAtStartPar
\sphinxstyleemphasis{pre\_opt} a boolean

\item {} 
\sphinxAtStartPar
\sphinxstyleemphasis{fba} a CBModel object

\item {} 
\sphinxAtStartPar
\sphinxstyleemphasis{objF2constr} add objective function as constraint

\item {} 
\sphinxAtStartPar
\sphinxstyleemphasis{quiet} {[}default=False{]} supress cplex output

\item {} 
\sphinxAtStartPar
\sphinxstyleemphasis{with\_reduced\_costs} {[}default=’unscaled’{]} can be ‘scaled’ or ‘unscaled’

\item {} 
\sphinxAtStartPar
\sphinxstyleemphasis{method} {[}default=’o’{]} choose the CPLEX method to use for solution, default is automatic. See CPLEX reference manual for details
\begin{itemize}
\item {} 
\sphinxAtStartPar
‘o’: auto

\item {} 
\sphinxAtStartPar
‘p’: primal

\item {} 
\sphinxAtStartPar
‘d’: dual

\item {} 
\sphinxAtStartPar
‘b’: barrier (no crossover)

\item {} 
\sphinxAtStartPar
‘h’: barrier

\item {} 
\sphinxAtStartPar
‘s’: sifting

\item {} 
\sphinxAtStartPar
‘c’: concurrent

\end{itemize}

\end{itemize}

\sphinxAtStartPar
Returns: pre\_sol, pre\_oid, pre\_oval, OPTIMAL\_PRESOLUTION, REDUCED\_COSTS

\end{fulllineitems}

\index{cplx\_func\_SetObjectiveFunctionAsConstraint() (in module cbmpy.CBCPLEX)@\spxentry{cplx\_func\_SetObjectiveFunctionAsConstraint()}\spxextra{in module cbmpy.CBCPLEX}}

\begin{fulllineitems}
\phantomsection\label{\detokenize{modules_doc:cbmpy.CBCPLEX.cplx_func_SetObjectiveFunctionAsConstraint}}
\pysigstartsignatures
\pysiglinewithargsret{\sphinxbfcode{\sphinxupquote{cplx\_func\_SetObjectiveFunctionAsConstraint}}}{\sphinxparam{\DUrole{n,n}{cpx}}\sphinxparamcomma \sphinxparam{\DUrole{n,n}{rhs\_sense}}\sphinxparamcomma \sphinxparam{\DUrole{n,n}{oval}}\sphinxparamcomma \sphinxparam{\DUrole{n,n}{tol}}\sphinxparamcomma \sphinxparam{\DUrole{n,n}{optPercentage}}}{}
\pysigstopsignatures\begin{description}
\sphinxlineitem{Take the objective function and “optimum” value and add it as a constraint}\begin{itemize}
\item {} 
\sphinxAtStartPar
\sphinxstyleemphasis{cpx} a cplex object

\item {} 
\sphinxAtStartPar
\sphinxstyleemphasis{oval} the objective value

\item {} 
\sphinxAtStartPar
\sphinxstyleemphasis{tol}  {[}default=None{]} do not floor/ceiling the objective function constraint, otherwise round of to \sphinxstyleemphasis{tol}

\item {} 
\sphinxAtStartPar
\sphinxstyleemphasis{rhs\_sense} {[}default=’lower’{]} means objC \textgreater{}= objVal the inequality to use for the objective constraint can also be \sphinxstyleemphasis{upper} or \sphinxstyleemphasis{equal}

\item {} 
\sphinxAtStartPar
\sphinxstyleemphasis{optPercentage} {[}default=100.0{]} means the percentage optimal value to use for the RHS of the objective constraint: optimal\_value*(optPercentage/100.0)

\end{itemize}

\end{description}

\end{fulllineitems}

\index{cplx\_getCPLEXModelFromLP() (in module cbmpy.CBCPLEX)@\spxentry{cplx\_getCPLEXModelFromLP()}\spxextra{in module cbmpy.CBCPLEX}}

\begin{fulllineitems}
\phantomsection\label{\detokenize{modules_doc:cbmpy.CBCPLEX.cplx_getCPLEXModelFromLP}}
\pysigstartsignatures
\pysiglinewithargsret{\sphinxbfcode{\sphinxupquote{cplx\_getCPLEXModelFromLP}}}{\sphinxparam{\DUrole{n,n}{lptFile}}\sphinxparamcomma \sphinxparam{\DUrole{n,n}{Dir}\DUrole{o,o}{=}\DUrole{default_value}{None}}}{}
\pysigstopsignatures
\sphinxAtStartPar
Load a LPT (CPLEX format) file and return a CPLX LP model
\begin{itemize}
\item {} 
\sphinxAtStartPar
\sphinxstyleemphasis{lptfile} an CPLEX LP format file

\item {} 
\sphinxAtStartPar
\sphinxstyleemphasis{Dir} an optional directory

\end{itemize}

\end{fulllineitems}

\index{cplx\_getDualValues() (in module cbmpy.CBCPLEX)@\spxentry{cplx\_getDualValues()}\spxextra{in module cbmpy.CBCPLEX}}

\begin{fulllineitems}
\phantomsection\label{\detokenize{modules_doc:cbmpy.CBCPLEX.cplx_getDualValues}}
\pysigstartsignatures
\pysiglinewithargsret{\sphinxbfcode{\sphinxupquote{cplx\_getDualValues}}}{\sphinxparam{\DUrole{n,n}{c}}}{}
\pysigstopsignatures
\sphinxAtStartPar
Get the get the dual values of the solution
\begin{itemize}
\item {} 
\sphinxAtStartPar
\sphinxstyleemphasis{c} a CPLEX LP

\end{itemize}

\sphinxAtStartPar
Output is a dictionary of \{name : value\} pairs

\end{fulllineitems}

\index{cplx\_getModelFromLP() (in module cbmpy.CBCPLEX)@\spxentry{cplx\_getModelFromLP()}\spxextra{in module cbmpy.CBCPLEX}}

\begin{fulllineitems}
\phantomsection\label{\detokenize{modules_doc:cbmpy.CBCPLEX.cplx_getModelFromLP}}
\pysigstartsignatures
\pysiglinewithargsret{\sphinxbfcode{\sphinxupquote{cplx\_getModelFromLP}}}{\sphinxparam{\DUrole{n,n}{lptFile}}\sphinxparamcomma \sphinxparam{\DUrole{n,n}{Dir}\DUrole{o,o}{=}\DUrole{default_value}{None}}}{}
\pysigstopsignatures
\sphinxAtStartPar
Load a LPT (CPLEX format) file and return a CPLX LP model
\begin{itemize}
\item {} 
\sphinxAtStartPar
\sphinxstyleemphasis{lptfile} an CPLEX LP format file

\item {} 
\sphinxAtStartPar
\sphinxstyleemphasis{Dir} an optional directory

\end{itemize}

\end{fulllineitems}

\index{cplx\_getModelFromObj() (in module cbmpy.CBCPLEX)@\spxentry{cplx\_getModelFromObj()}\spxextra{in module cbmpy.CBCPLEX}}

\begin{fulllineitems}
\phantomsection\label{\detokenize{modules_doc:cbmpy.CBCPLEX.cplx_getModelFromObj}}
\pysigstartsignatures
\pysiglinewithargsret{\sphinxbfcode{\sphinxupquote{cplx\_getModelFromObj}}}{\sphinxparam{\DUrole{n,n}{fba}}}{}
\pysigstopsignatures
\sphinxAtStartPar
Return a CPLEX object from a FBA model object (via LP file)

\end{fulllineitems}

\index{cplx\_getOptimalSolution() (in module cbmpy.CBCPLEX)@\spxentry{cplx\_getOptimalSolution()}\spxextra{in module cbmpy.CBCPLEX}}

\begin{fulllineitems}
\phantomsection\label{\detokenize{modules_doc:cbmpy.CBCPLEX.cplx_getOptimalSolution}}
\pysigstartsignatures
\pysiglinewithargsret{\sphinxbfcode{\sphinxupquote{cplx\_getOptimalSolution}}}{\sphinxparam{\DUrole{n,n}{c}}}{}
\pysigstopsignatures
\sphinxAtStartPar
From a CPLX model extract a tuple of solution, ObjFuncName and ObjFuncVal

\end{fulllineitems}

\index{cplx\_getOptimalSolution2() (in module cbmpy.CBCPLEX)@\spxentry{cplx\_getOptimalSolution2()}\spxextra{in module cbmpy.CBCPLEX}}

\begin{fulllineitems}
\phantomsection\label{\detokenize{modules_doc:cbmpy.CBCPLEX.cplx_getOptimalSolution2}}
\pysigstartsignatures
\pysiglinewithargsret{\sphinxbfcode{\sphinxupquote{cplx\_getOptimalSolution2}}}{\sphinxparam{\DUrole{n,n}{c}}\sphinxparamcomma \sphinxparam{\DUrole{n,n}{names}}}{}
\pysigstopsignatures
\sphinxAtStartPar
From a CPLX model extract a tuple of solution, ObjFuncName and ObjFuncVal

\end{fulllineitems}

\index{cplx\_getReducedCosts() (in module cbmpy.CBCPLEX)@\spxentry{cplx\_getReducedCosts()}\spxextra{in module cbmpy.CBCPLEX}}

\begin{fulllineitems}
\phantomsection\label{\detokenize{modules_doc:cbmpy.CBCPLEX.cplx_getReducedCosts}}
\pysigstartsignatures
\pysiglinewithargsret{\sphinxbfcode{\sphinxupquote{cplx\_getReducedCosts}}}{\sphinxparam{\DUrole{n,n}{c}}\sphinxparamcomma \sphinxparam{\DUrole{n,n}{scaled}\DUrole{o,o}{=}\DUrole{default_value}{False}}}{}
\pysigstopsignatures
\sphinxAtStartPar
Extract ReducedCosts from LP and return as a dictionary ‘Rid’ : reduced cost
\begin{itemize}
\item {} 
\sphinxAtStartPar
\sphinxstyleemphasis{c} a cplex LP object

\item {} 
\sphinxAtStartPar
\sphinxstyleemphasis{scaled} scale the reduced cost by the optimal flux value

\end{itemize}

\end{fulllineitems}

\index{cplx\_getSensitivities() (in module cbmpy.CBCPLEX)@\spxentry{cplx\_getSensitivities()}\spxextra{in module cbmpy.CBCPLEX}}

\begin{fulllineitems}
\phantomsection\label{\detokenize{modules_doc:cbmpy.CBCPLEX.cplx_getSensitivities}}
\pysigstartsignatures
\pysiglinewithargsret{\sphinxbfcode{\sphinxupquote{cplx\_getSensitivities}}}{\sphinxparam{\DUrole{n,n}{c}}}{}
\pysigstopsignatures
\sphinxAtStartPar
Get the sensitivities of each constraint on the objective function with inpt
\begin{itemize}
\item {} 
\sphinxAtStartPar
\sphinxstyleemphasis{c} a CPLEX LP

\end{itemize}

\sphinxAtStartPar
Output is a tuple of bound and objective sensitivities where the objective
sensitivity is described in the CPLEX reference manual as:

\begin{sphinxVerbatim}[commandchars=\\\{\}]
\PYG{o}{.}\PYG{o}{.}\PYG{o}{.} \PYG{n}{the} \PYG{n}{objective} \PYG{n}{sensitivity} \PYG{n}{shows} \PYG{n}{each} \PYG{n}{variable}\PYG{p}{,} \PYG{n}{its} \PYG{n}{reduced} \PYG{n}{cost} \PYG{o+ow}{and} \PYG{n}{the} \PYG{n+nb}{range} \PYG{n}{over}
\PYG{n}{which} \PYG{n}{its} \PYG{n}{objective} \PYG{n}{function} \PYG{n}{coefficient} \PYG{n}{can} \PYG{n}{vary} \PYG{n}{without} \PYG{n}{forcing} \PYG{n}{a} \PYG{n}{change}
\PYG{o+ow}{in} \PYG{n}{the} \PYG{n}{optimal} \PYG{n}{basis}\PYG{o}{.} \PYG{n}{The} \PYG{n}{current} \PYG{n}{value} \PYG{n}{of} \PYG{n}{each} \PYG{n}{objective} \PYG{n}{coefficient} \PYG{o+ow}{is}
\PYG{n}{also} \PYG{n}{displayed} \PYG{k}{for} \PYG{n}{reference}\PYG{o}{.}

\PYG{o}{\PYGZhy{}} \PYG{o}{*}\PYG{n}{objective} \PYG{n}{coefficient} \PYG{n}{sensitivity}\PYG{o}{*} \PYG{p}{\PYGZob{}}\PYG{n}{flux} \PYG{p}{:} \PYG{p}{(}\PYG{n}{reduced\PYGZus{}cost}\PYG{p}{,} \PYG{n}{lower\PYGZus{}obj\PYGZus{}sensitivity}\PYG{p}{,} \PYG{n}{coeff\PYGZus{}value}\PYG{p}{,} \PYG{n}{upper\PYGZus{}obj\PYGZus{}sensitivity}\PYG{p}{)}\PYG{p}{\PYGZcb{}}
\PYG{o}{\PYGZhy{}} \PYG{o}{*}\PYG{n}{rhs} \PYG{n}{sensitivity}\PYG{o}{*} \PYG{p}{\PYGZob{}}\PYG{n}{constraint} \PYG{p}{:} \PYG{p}{(}\PYG{n}{low}\PYG{p}{,} \PYG{n}{value}\PYG{p}{,} \PYG{n}{high}\PYG{p}{)}\PYG{p}{\PYGZcb{}}
\PYG{o}{\PYGZhy{}} \PYG{o}{*}\PYG{n}{bound} \PYG{n}{sensitivity} \PYG{n}{ranges}\PYG{o}{*} \PYG{p}{\PYGZob{}}\PYG{n}{flux} \PYG{p}{:} \PYG{p}{(}\PYG{n}{lb\PYGZus{}low}\PYG{p}{,} \PYG{n}{lb\PYGZus{}high}\PYG{p}{,} \PYG{n}{ub\PYGZus{}low}\PYG{p}{,} \PYG{n}{ub\PYGZus{}high}\PYG{p}{)}\PYG{p}{\PYGZcb{}}
\end{sphinxVerbatim}

\end{fulllineitems}

\index{cplx\_getShadowPrices() (in module cbmpy.CBCPLEX)@\spxentry{cplx\_getShadowPrices()}\spxextra{in module cbmpy.CBCPLEX}}

\begin{fulllineitems}
\phantomsection\label{\detokenize{modules_doc:cbmpy.CBCPLEX.cplx_getShadowPrices}}
\pysigstartsignatures
\pysiglinewithargsret{\sphinxbfcode{\sphinxupquote{cplx\_getShadowPrices}}}{\sphinxparam{\DUrole{n,n}{c}}}{}
\pysigstopsignatures
\sphinxAtStartPar
Returns a dictionary of shadow prices containing ‘N\_row\_id’ : (lb, rhs, ub)
\begin{itemize}
\item {} 
\sphinxAtStartPar
\sphinxstyleemphasis{c} a cplex LP object

\end{itemize}

\end{fulllineitems}

\index{cplx\_getSolutionStatus() (in module cbmpy.CBCPLEX)@\spxentry{cplx\_getSolutionStatus()}\spxextra{in module cbmpy.CBCPLEX}}

\begin{fulllineitems}
\phantomsection\label{\detokenize{modules_doc:cbmpy.CBCPLEX.cplx_getSolutionStatus}}
\pysigstartsignatures
\pysiglinewithargsret{\sphinxbfcode{\sphinxupquote{cplx\_getSolutionStatus}}}{\sphinxparam{\DUrole{n,n}{c}}}{}
\pysigstopsignatures
\sphinxAtStartPar
Returns one of:
\begin{itemize}
\item {} 
\sphinxAtStartPar
\sphinxstyleemphasis{LPS\_OPT}: solution is optimal;

\item {} 
\sphinxAtStartPar
\sphinxstyleemphasis{LPS\_FEAS}: solution is feasible;

\item {} 
\sphinxAtStartPar
\sphinxstyleemphasis{LPS\_INFEAS}: solution is infeasible;

\item {} 
\sphinxAtStartPar
\sphinxstyleemphasis{LPS\_NOFEAS}: problem has no feasible solution;

\item {} 
\sphinxAtStartPar
\sphinxstyleemphasis{LPS\_UNBND}: problem has unbounded solution;

\item {} 
\sphinxAtStartPar
\sphinxstyleemphasis{LPS\_UNDEF}: solution is undefined.

\item {} 
\sphinxAtStartPar
\sphinxstyleemphasis{LPS\_NONE}: no solution

\end{itemize}

\end{fulllineitems}

\index{cplx\_runInputScan() (in module cbmpy.CBCPLEX)@\spxentry{cplx\_runInputScan()}\spxextra{in module cbmpy.CBCPLEX}}

\begin{fulllineitems}
\phantomsection\label{\detokenize{modules_doc:cbmpy.CBCPLEX.cplx_runInputScan}}
\pysigstartsignatures
\pysiglinewithargsret{\sphinxbfcode{\sphinxupquote{cplx\_runInputScan}}}{\sphinxparam{\DUrole{n,n}{fba}}\sphinxparamcomma \sphinxparam{\DUrole{n,n}{exDict}}\sphinxparamcomma \sphinxparam{\DUrole{n,n}{wDir}}\sphinxparamcomma \sphinxparam{\DUrole{n,n}{input\_lb}\DUrole{o,o}{=}\DUrole{default_value}{\sphinxhyphen{}10.0}}\sphinxparamcomma \sphinxparam{\DUrole{n,n}{input\_ub}\DUrole{o,o}{=}\DUrole{default_value}{0.0}}\sphinxparamcomma \sphinxparam{\DUrole{n,n}{writeHformat}\DUrole{o,o}{=}\DUrole{default_value}{False}}\sphinxparamcomma \sphinxparam{\DUrole{n,n}{rationalLPout}\DUrole{o,o}{=}\DUrole{default_value}{False}}}{}
\pysigstopsignatures
\sphinxAtStartPar
scans all inputs

\end{fulllineitems}

\index{cplx\_setFBAsolutionToModel() (in module cbmpy.CBCPLEX)@\spxentry{cplx\_setFBAsolutionToModel()}\spxextra{in module cbmpy.CBCPLEX}}

\begin{fulllineitems}
\phantomsection\label{\detokenize{modules_doc:cbmpy.CBCPLEX.cplx_setFBAsolutionToModel}}
\pysigstartsignatures
\pysiglinewithargsret{\sphinxbfcode{\sphinxupquote{cplx\_setFBAsolutionToModel}}}{\sphinxparam{\DUrole{n,n}{fba}}\sphinxparamcomma \sphinxparam{\DUrole{n,n}{lp}}\sphinxparamcomma \sphinxparam{\DUrole{n,n}{with\_reduced\_costs}\DUrole{o,o}{=}\DUrole{default_value}{\textquotesingle{}unscaled\textquotesingle{}}}}{}
\pysigstopsignatures
\sphinxAtStartPar
Sets the FBA solution from a CPLEX solution to an FBA object
\begin{itemize}
\item {} 
\sphinxAtStartPar
\sphinxstyleemphasis{fba} and fba object

\item {} 
\sphinxAtStartPar
\sphinxstyleemphasis{lp} a CPLEX LP object

\item {} 
\sphinxAtStartPar
\sphinxstyleemphasis{with\_reduced\_costs} {[}default=’unscaled’{]} calculate and add reduced cost information to mode this can be: ‘unscaled’ or ‘scaled’
or anything else which is interpreted as None. Scaled is: s\_rcost = (r.reduced\_cost*rval)/obj\_value

\end{itemize}

\end{fulllineitems}

\index{cplx\_setMIPGapTolerance() (in module cbmpy.CBCPLEX)@\spxentry{cplx\_setMIPGapTolerance()}\spxextra{in module cbmpy.CBCPLEX}}

\begin{fulllineitems}
\phantomsection\label{\detokenize{modules_doc:cbmpy.CBCPLEX.cplx_setMIPGapTolerance}}
\pysigstartsignatures
\pysiglinewithargsret{\sphinxbfcode{\sphinxupquote{cplx\_setMIPGapTolerance}}}{\sphinxparam{\DUrole{n,n}{c}}\sphinxparamcomma \sphinxparam{\DUrole{n,n}{tol}}}{}
\pysigstopsignatures
\sphinxAtStartPar
Sets the the relative MIP gap tolerance

\end{fulllineitems}

\index{cplx\_setObjective() (in module cbmpy.CBCPLEX)@\spxentry{cplx\_setObjective()}\spxextra{in module cbmpy.CBCPLEX}}

\begin{fulllineitems}
\phantomsection\label{\detokenize{modules_doc:cbmpy.CBCPLEX.cplx_setObjective}}
\pysigstartsignatures
\pysiglinewithargsret{\sphinxbfcode{\sphinxupquote{cplx\_setObjective}}}{\sphinxparam{\DUrole{n,n}{c}}\sphinxparamcomma \sphinxparam{\DUrole{n,n}{pid}}\sphinxparamcomma \sphinxparam{\DUrole{n,n}{expr}\DUrole{o,o}{=}\DUrole{default_value}{None}}\sphinxparamcomma \sphinxparam{\DUrole{n,n}{sense}\DUrole{o,o}{=}\DUrole{default_value}{\textquotesingle{}maximize\textquotesingle{}}}\sphinxparamcomma \sphinxparam{\DUrole{n,n}{reset}\DUrole{o,o}{=}\DUrole{default_value}{True}}}{}
\pysigstopsignatures
\sphinxAtStartPar
Set a new objective function note that there is a major memory leak in
\sphinxtitleref{c.variables.get\_names()} whch is used when reset=True. If this is a problem
use cplx\_setObjective2 which takes \sphinxstyleemphasis{names} as an input:
\begin{itemize}
\item {} 
\sphinxAtStartPar
\sphinxstyleemphasis{c} a CPLEX LP object

\item {} 
\sphinxAtStartPar
\sphinxstyleemphasis{pid} the r\_id of the flux to be optimized

\item {} 
\sphinxAtStartPar
\sphinxstyleemphasis{expr} a list of (coefficient, flux) pairs

\item {} 
\sphinxAtStartPar
\sphinxstyleemphasis{sense} ‘maximize’/’minimize’

\item {} 
\sphinxAtStartPar
\sphinxstyleemphasis{reset} {[}default=True{]} reset all objective function coefficients to zero

\end{itemize}

\end{fulllineitems}

\index{cplx\_setObjective2() (in module cbmpy.CBCPLEX)@\spxentry{cplx\_setObjective2()}\spxextra{in module cbmpy.CBCPLEX}}

\begin{fulllineitems}
\phantomsection\label{\detokenize{modules_doc:cbmpy.CBCPLEX.cplx_setObjective2}}
\pysigstartsignatures
\pysiglinewithargsret{\sphinxbfcode{\sphinxupquote{cplx\_setObjective2}}}{\sphinxparam{\DUrole{n,n}{c}}\sphinxparamcomma \sphinxparam{\DUrole{n,n}{pid}}\sphinxparamcomma \sphinxparam{\DUrole{n,n}{names}}\sphinxparamcomma \sphinxparam{\DUrole{n,n}{expr}\DUrole{o,o}{=}\DUrole{default_value}{None}}\sphinxparamcomma \sphinxparam{\DUrole{n,n}{sense}\DUrole{o,o}{=}\DUrole{default_value}{\textquotesingle{}maximize\textquotesingle{}}}\sphinxparamcomma \sphinxparam{\DUrole{n,n}{reset}\DUrole{o,o}{=}\DUrole{default_value}{True}}}{}
\pysigstopsignatures
\sphinxAtStartPar
Set a new objective function. This is a workaround function to avoid the
e is a major memory leak in \sphinxtitleref{c.variables.get\_names()} whch is used
in cplx\_setObjective()  when reset=True.

\end{fulllineitems}

\index{cplx\_setOutputStreams() (in module cbmpy.CBCPLEX)@\spxentry{cplx\_setOutputStreams()}\spxextra{in module cbmpy.CBCPLEX}}

\begin{fulllineitems}
\phantomsection\label{\detokenize{modules_doc:cbmpy.CBCPLEX.cplx_setOutputStreams}}
\pysigstartsignatures
\pysiglinewithargsret{\sphinxbfcode{\sphinxupquote{cplx\_setOutputStreams}}}{\sphinxparam{\DUrole{n,n}{lp}}\sphinxparamcomma \sphinxparam{\DUrole{n,n}{mode}\DUrole{o,o}{=}\DUrole{default_value}{\textquotesingle{}default\textquotesingle{}}}}{}
\pysigstopsignatures
\sphinxAtStartPar
Sets the noise level of the solver, mode can be one of:
\begin{itemize}
\item {} 
\sphinxAtStartPar
\sphinxstyleemphasis{None} silent i.e. no output

\item {} 
\sphinxAtStartPar
\sphinxstyleemphasis{‘file’} set solver to silent and output logs to \sphinxstyleemphasis{CPLX\_RESULT\_STREAM\_FILE} cplex\_output.log

\item {} 
\sphinxAtStartPar
\sphinxstyleemphasis{‘iostream’} set solver to silent and output logs to \sphinxstyleemphasis{CPLX\_RESULT\_STREAM\_IO} csio

\item {} 
\sphinxAtStartPar
\sphinxstyleemphasis{‘default’} or anything else noisy with full output closes STREAM\_IO and STREAM\_FILE (default)

\end{itemize}

\end{fulllineitems}

\index{cplx\_setSolutionStatusToModel() (in module cbmpy.CBCPLEX)@\spxentry{cplx\_setSolutionStatusToModel()}\spxextra{in module cbmpy.CBCPLEX}}

\begin{fulllineitems}
\phantomsection\label{\detokenize{modules_doc:cbmpy.CBCPLEX.cplx_setSolutionStatusToModel}}
\pysigstartsignatures
\pysiglinewithargsret{\sphinxbfcode{\sphinxupquote{cplx\_setSolutionStatusToModel}}}{\sphinxparam{\DUrole{n,n}{m}}\sphinxparamcomma \sphinxparam{\DUrole{n,n}{lp}}}{}
\pysigstopsignatures
\sphinxAtStartPar
Sets the lp solutions status to the CBMPy model

\end{fulllineitems}

\index{cplx\_singleGeneScan() (in module cbmpy.CBCPLEX)@\spxentry{cplx\_singleGeneScan()}\spxextra{in module cbmpy.CBCPLEX}}

\begin{fulllineitems}
\phantomsection\label{\detokenize{modules_doc:cbmpy.CBCPLEX.cplx_singleGeneScan}}
\pysigstartsignatures
\pysiglinewithargsret{\sphinxbfcode{\sphinxupquote{cplx\_singleGeneScan}}}{\sphinxparam{\DUrole{n,n}{fba}}\sphinxparamcomma \sphinxparam{\DUrole{n,n}{r\_off\_low}\DUrole{o,o}{=}\DUrole{default_value}{0.0}}\sphinxparamcomma \sphinxparam{\DUrole{n,n}{r\_off\_upp}\DUrole{o,o}{=}\DUrole{default_value}{0.0}}\sphinxparamcomma \sphinxparam{\DUrole{n,n}{optrnd}\DUrole{o,o}{=}\DUrole{default_value}{8}}\sphinxparamcomma \sphinxparam{\DUrole{n,n}{altout}\DUrole{o,o}{=}\DUrole{default_value}{False}}}{}
\pysigstopsignatures
\sphinxAtStartPar
Perform a single gene deletion scan
\begin{itemize}
\item {} 
\sphinxAtStartPar
\sphinxstyleemphasis{fba} a model object

\item {} 
\sphinxAtStartPar
\sphinxstyleemphasis{r\_off\_low} the lower bound of a deactivated reaction

\item {} 
\sphinxAtStartPar
\sphinxstyleemphasis{r\_off\_upp} the upper bound of a deactivated reaction

\item {} 
\sphinxAtStartPar
\sphinxstyleemphasis{optrnd} {[}default=8{]} round off the optimal value

\item {} 
\sphinxAtStartPar
\sphinxstyleemphasis{altout} {[}default=False{]} by default return a list of gene:opt pairs, alternatively (True) return an extended result set including gene groups, optima and effect map

\end{itemize}

\end{fulllineitems}

\index{cplx\_singleReactionDeletionScan() (in module cbmpy.CBCPLEX)@\spxentry{cplx\_singleReactionDeletionScan()}\spxextra{in module cbmpy.CBCPLEX}}

\begin{fulllineitems}
\phantomsection\label{\detokenize{modules_doc:cbmpy.CBCPLEX.cplx_singleReactionDeletionScan}}
\pysigstartsignatures
\pysiglinewithargsret{\sphinxbfcode{\sphinxupquote{cplx\_singleReactionDeletionScan}}}{\sphinxparam{\DUrole{n,n}{fba}}\sphinxparamcomma \sphinxparam{\DUrole{n,n}{r\_off\_low}\DUrole{o,o}{=}\DUrole{default_value}{0.0}}\sphinxparamcomma \sphinxparam{\DUrole{n,n}{r\_off\_upp}\DUrole{o,o}{=}\DUrole{default_value}{0.0}}\sphinxparamcomma \sphinxparam{\DUrole{n,n}{optrnd}\DUrole{o,o}{=}\DUrole{default_value}{8}}}{}
\pysigstopsignatures
\sphinxAtStartPar
Perform a single reaction deletion scan
\begin{itemize}
\item {} 
\sphinxAtStartPar
\sphinxstyleemphasis{fba} a model object

\item {} 
\sphinxAtStartPar
\sphinxstyleemphasis{r\_off\_low} the lower bound of a deactivated reaction

\item {} 
\sphinxAtStartPar
\sphinxstyleemphasis{r\_off\_upp} the upper bound of a deactivated reaction

\item {} 
\sphinxAtStartPar
\sphinxstyleemphasis{optrnd} {[}default=8{]} round off the optimal value

\end{itemize}

\end{fulllineitems}

\index{cplx\_writeLPsolution() (in module cbmpy.CBCPLEX)@\spxentry{cplx\_writeLPsolution()}\spxextra{in module cbmpy.CBCPLEX}}

\begin{fulllineitems}
\phantomsection\label{\detokenize{modules_doc:cbmpy.CBCPLEX.cplx_writeLPsolution}}
\pysigstartsignatures
\pysiglinewithargsret{\sphinxbfcode{\sphinxupquote{cplx\_writeLPsolution}}}{\sphinxparam{\DUrole{n,n}{fba\_sol}}\sphinxparamcomma \sphinxparam{\DUrole{n,n}{objf\_name}}\sphinxparamcomma \sphinxparam{\DUrole{n,n}{fname}}\sphinxparamcomma \sphinxparam{\DUrole{n,n}{Dir}\DUrole{o,o}{=}\DUrole{default_value}{None}}\sphinxparamcomma \sphinxparam{\DUrole{n,n}{separator}\DUrole{o,o}{=}\DUrole{default_value}{\textquotesingle{},\textquotesingle{}}}}{}
\pysigstopsignatures
\sphinxAtStartPar
This function writes the optimal solution, produced wth \sphinxtitleref{cplx\_getOptimalSolution} to file
\begin{itemize}
\item {} 
\sphinxAtStartPar
\sphinxstyleemphasis{fba\_sol} a dictionary of Flux : value pairs

\item {} 
\sphinxAtStartPar
\sphinxstyleemphasis{objf\_name} the objective function flux id

\item {} 
\sphinxAtStartPar
\sphinxstyleemphasis{fname} the output filename

\item {} 
\sphinxAtStartPar
\sphinxstyleemphasis{Dir} {[}default=None{]} use directory if not None

\item {} 
\sphinxAtStartPar
\sphinxstyleemphasis{separator} {[}default=’,’{]} the column separator

\end{itemize}

\end{fulllineitems}

\index{cplx\_writeLPtoLPTfile() (in module cbmpy.CBCPLEX)@\spxentry{cplx\_writeLPtoLPTfile()}\spxextra{in module cbmpy.CBCPLEX}}

\begin{fulllineitems}
\phantomsection\label{\detokenize{modules_doc:cbmpy.CBCPLEX.cplx_writeLPtoLPTfile}}
\pysigstartsignatures
\pysiglinewithargsret{\sphinxbfcode{\sphinxupquote{cplx\_writeLPtoLPTfile}}}{\sphinxparam{\DUrole{n,n}{c}}\sphinxparamcomma \sphinxparam{\DUrole{n,n}{filename}}\sphinxparamcomma \sphinxparam{\DUrole{n,n}{title}\DUrole{o,o}{=}\DUrole{default_value}{None}}\sphinxparamcomma \sphinxparam{\DUrole{n,n}{Dir}\DUrole{o,o}{=}\DUrole{default_value}{None}}}{}
\pysigstopsignatures
\sphinxAtStartPar
Write out a CPLEX model as an LP format file

\end{fulllineitems}

\index{getReducedCosts() (in module cbmpy.CBCPLEX)@\spxentry{getReducedCosts()}\spxextra{in module cbmpy.CBCPLEX}}

\begin{fulllineitems}
\phantomsection\label{\detokenize{modules_doc:cbmpy.CBCPLEX.getReducedCosts}}
\pysigstartsignatures
\pysiglinewithargsret{\sphinxbfcode{\sphinxupquote{getReducedCosts}}}{\sphinxparam{\DUrole{n,n}{fba}}}{}
\pysigstopsignatures
\sphinxAtStartPar
Get a dictionary of reduced costs for each reaction/flux

\end{fulllineitems}

\index{setReducedCosts() (in module cbmpy.CBCPLEX)@\spxentry{setReducedCosts()}\spxextra{in module cbmpy.CBCPLEX}}

\begin{fulllineitems}
\phantomsection\label{\detokenize{modules_doc:cbmpy.CBCPLEX.setReducedCosts}}
\pysigstartsignatures
\pysiglinewithargsret{\sphinxbfcode{\sphinxupquote{setReducedCosts}}}{\sphinxparam{\DUrole{n,n}{fba}}\sphinxparamcomma \sphinxparam{\DUrole{n,n}{reduced\_costs}}}{}
\pysigstopsignatures
\sphinxAtStartPar
For each reaction/flux, sets the attribute “reduced\_cost” from a dictionary of
reduced costs
\begin{itemize}
\item {} 
\sphinxAtStartPar
\sphinxstyleemphasis{fba} an fba object

\item {} 
\sphinxAtStartPar
\sphinxstyleemphasis{reduced\_costs} a dictionary of \{reaction : value\} pairs

\end{itemize}

\end{fulllineitems}

\phantomsection\label{\detokenize{modules_doc:module-cbmpy.CBDataStruct}}\index{module@\spxentry{module}!cbmpy.CBDataStruct@\spxentry{cbmpy.CBDataStruct}}\index{cbmpy.CBDataStruct@\spxentry{cbmpy.CBDataStruct}!module@\spxentry{module}}

\section{CBMPy: CBDataStruct module}
\label{\detokenize{modules_doc:cbmpy-cbdatastruct-module}}
\sphinxAtStartPar
PySCeS Constraint Based Modelling (\sphinxurl{http://cbmpy.sourceforge.net})
Copyright (C) 2009\sphinxhyphen{}2024 Brett G. Olivier, VU University Amsterdam, Amsterdam, The Netherlands

\sphinxAtStartPar
This program is free software: you can redistribute it and/or modify
it under the terms of the GNU General Public License as published by
the Free Software Foundation, either version 3 of the License, or
(at your option) any later version.

\sphinxAtStartPar
This program is distributed in the hope that it will be useful,
but WITHOUT ANY WARRANTY; without even the implied warranty of
MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
GNU General Public License for more details.

\sphinxAtStartPar
You should have received a copy of the GNU General Public License
along with this program.  If not, see \textless{}\sphinxurl{http://www.gnu.org/licenses/}\textgreater{}

\sphinxAtStartPar
Author: Brett G. Olivier PhD
Contact developers: \sphinxurl{https://github.com/SystemsBioinformatics/cbmpy/issues}
Last edit: \$Author: bgoli \$ (\$Id: CBDataStruct.py 710 2020\sphinxhyphen{}04\sphinxhyphen{}27 14:22:34Z bgoli \$)
\phantomsection\label{\detokenize{modules_doc:module-cbmpy.CBGUI}}\index{module@\spxentry{module}!cbmpy.CBGUI@\spxentry{cbmpy.CBGUI}}\index{cbmpy.CBGUI@\spxentry{cbmpy.CBGUI}!module@\spxentry{module}}

\section{CBMPy: CBGUI module}
\label{\detokenize{modules_doc:cbmpy-cbgui-module}}
\sphinxAtStartPar
PySCeS Constraint Based Modelling (\sphinxurl{http://cbmpy.sourceforge.net})
Copyright (C) 2009\sphinxhyphen{}2024 Brett G. Olivier, VU University Amsterdam, Amsterdam, The Netherlands

\sphinxAtStartPar
This program is free software: you can redistribute it and/or modify
it under the terms of the GNU General Public License as published by
the Free Software Foundation, either version 3 of the License, or
(at your option) any later version.

\sphinxAtStartPar
This program is distributed in the hope that it will be useful,
but WITHOUT ANY WARRANTY; without even the implied warranty of
MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
GNU General Public License for more details.

\sphinxAtStartPar
You should have received a copy of the GNU General Public License
along with this program.  If not, see \textless{}\sphinxurl{http://www.gnu.org/licenses/}\textgreater{}

\sphinxAtStartPar
Author: Brett G. Olivier PhD
Contact developers: \sphinxurl{https://github.com/SystemsBioinformatics/cbmpy/issues}
Last edit: \$Author: bgoli \$ (\$Id: CBGUI.py 710 2020\sphinxhyphen{}04\sphinxhyphen{}27 14:22:34Z bgoli \$)
\index{loadCBGUI() (in module cbmpy.CBGUI)@\spxentry{loadCBGUI()}\spxextra{in module cbmpy.CBGUI}}

\begin{fulllineitems}
\phantomsection\label{\detokenize{modules_doc:cbmpy.CBGUI.loadCBGUI}}
\pysigstartsignatures
\pysiglinewithargsret{\sphinxbfcode{\sphinxupquote{loadCBGUI}}}{\sphinxparam{\DUrole{n,n}{mod}}\sphinxparamcomma \sphinxparam{\DUrole{n,n}{version}\DUrole{o,o}{=}\DUrole{default_value}{2}}}{}
\pysigstopsignatures
\sphinxAtStartPar
Load an FBA model instance into the quick editor to view or change basic model properties
\begin{itemize}
\item {} 
\sphinxAtStartPar
\sphinxstyleemphasis{mod} a PySCeS CBMPy model instance

\end{itemize}

\end{fulllineitems}

\phantomsection\label{\detokenize{modules_doc:module-cbmpy.CBModel}}\index{module@\spxentry{module}!cbmpy.CBModel@\spxentry{cbmpy.CBModel}}\index{cbmpy.CBModel@\spxentry{cbmpy.CBModel}!module@\spxentry{module}}

\section{CBMPy: CBModel module}
\label{\detokenize{modules_doc:cbmpy-cbmodel-module}}
\sphinxAtStartPar
PySCeS Constraint Based Modelling (\sphinxurl{http://cbmpy.sourceforge.net})
Copyright (C) 2009\sphinxhyphen{}2024 Brett G. Olivier, VU University Amsterdam, Amsterdam, The Netherlands

\sphinxAtStartPar
This program is free software: you can redistribute it and/or modify
it under the terms of the GNU General Public License as published by
the Free Software Foundation, either version 3 of the License, or
(at your option) any later version.

\sphinxAtStartPar
This program is distributed in the hope that it will be useful,
but WITHOUT ANY WARRANTY; without even the implied warranty of
MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
GNU General Public License for more details.

\sphinxAtStartPar
You should have received a copy of the GNU General Public License
along with this program.  If not, see \textless{}\sphinxurl{http://www.gnu.org/licenses/}\textgreater{}

\sphinxAtStartPar
Author: Brett G. Olivier PhD
Contact developers: \sphinxurl{https://github.com/SystemsBioinformatics/cbmpy/issues}
Last edit: \$Author: bgoli \$ (\$Id: CBModel.py 706 2020\sphinxhyphen{}03\sphinxhyphen{}23 21:31:49Z bgoli \$)
\index{Compartment (class in cbmpy.CBModel)@\spxentry{Compartment}\spxextra{class in cbmpy.CBModel}}

\begin{fulllineitems}
\phantomsection\label{\detokenize{modules_doc:cbmpy.CBModel.Compartment}}
\pysigstartsignatures
\pysiglinewithargsret{\sphinxbfcode{\sphinxupquote{class\DUrole{w,w}{  }}}\sphinxbfcode{\sphinxupquote{Compartment}}}{\sphinxparam{\DUrole{n,n}{pid}}\sphinxparamcomma \sphinxparam{\DUrole{n,n}{name}\DUrole{o,o}{=}\DUrole{default_value}{None}}\sphinxparamcomma \sphinxparam{\DUrole{n,n}{size}\DUrole{o,o}{=}\DUrole{default_value}{1}}\sphinxparamcomma \sphinxparam{\DUrole{n,n}{dimensions}\DUrole{o,o}{=}\DUrole{default_value}{3}}\sphinxparamcomma \sphinxparam{\DUrole{n,n}{volume}\DUrole{o,o}{=}\DUrole{default_value}{None}}}{}
\pysigstopsignatures
\sphinxAtStartPar
A compartment
\index{containsReactions() (Compartment method)@\spxentry{containsReactions()}\spxextra{Compartment method}}

\begin{fulllineitems}
\phantomsection\label{\detokenize{modules_doc:cbmpy.CBModel.Compartment.containsReactions}}
\pysigstartsignatures
\pysiglinewithargsret{\sphinxbfcode{\sphinxupquote{containsReactions}}}{}{}
\pysigstopsignatures
\sphinxAtStartPar
Lists the species contained in this compartment

\end{fulllineitems}

\index{containsSpecies() (Compartment method)@\spxentry{containsSpecies()}\spxextra{Compartment method}}

\begin{fulllineitems}
\phantomsection\label{\detokenize{modules_doc:cbmpy.CBModel.Compartment.containsSpecies}}
\pysigstartsignatures
\pysiglinewithargsret{\sphinxbfcode{\sphinxupquote{containsSpecies}}}{}{}
\pysigstopsignatures
\sphinxAtStartPar
Lists the species contained in this compartment

\end{fulllineitems}

\index{getDimensions() (Compartment method)@\spxentry{getDimensions()}\spxextra{Compartment method}}

\begin{fulllineitems}
\phantomsection\label{\detokenize{modules_doc:cbmpy.CBModel.Compartment.getDimensions}}
\pysigstartsignatures
\pysiglinewithargsret{\sphinxbfcode{\sphinxupquote{getDimensions}}}{}{}
\pysigstopsignatures
\sphinxAtStartPar
Get the compartment dimensions

\end{fulllineitems}

\index{getSize() (Compartment method)@\spxentry{getSize()}\spxextra{Compartment method}}

\begin{fulllineitems}
\phantomsection\label{\detokenize{modules_doc:cbmpy.CBModel.Compartment.getSize}}
\pysigstartsignatures
\pysiglinewithargsret{\sphinxbfcode{\sphinxupquote{getSize}}}{}{}
\pysigstopsignatures
\sphinxAtStartPar
Get the compartment size

\end{fulllineitems}

\index{setDimensions() (Compartment method)@\spxentry{setDimensions()}\spxextra{Compartment method}}

\begin{fulllineitems}
\phantomsection\label{\detokenize{modules_doc:cbmpy.CBModel.Compartment.setDimensions}}
\pysigstartsignatures
\pysiglinewithargsret{\sphinxbfcode{\sphinxupquote{setDimensions}}}{\sphinxparam{\DUrole{n,n}{dimensions}}}{}
\pysigstopsignatures
\sphinxAtStartPar
Get the compartment dimensions
\begin{itemize}
\item {} 
\sphinxAtStartPar
\sphinxstyleemphasis{dimensions} set the new compartment dimensions

\end{itemize}

\end{fulllineitems}

\index{setId() (Compartment method)@\spxentry{setId()}\spxextra{Compartment method}}

\begin{fulllineitems}
\phantomsection\label{\detokenize{modules_doc:cbmpy.CBModel.Compartment.setId}}
\pysigstartsignatures
\pysiglinewithargsret{\sphinxbfcode{\sphinxupquote{setId}}}{\sphinxparam{\DUrole{n,n}{fid}}}{}
\pysigstopsignatures
\sphinxAtStartPar
Sets the object Id
\begin{quote}
\begin{itemize}
\item {} 
\sphinxAtStartPar
\sphinxstyleemphasis{fid} a valid c variable style id string

\end{itemize}

\sphinxAtStartPar
Reimplements @FBase.setId()
\end{quote}

\end{fulllineitems}

\index{setSize() (Compartment method)@\spxentry{setSize()}\spxextra{Compartment method}}

\begin{fulllineitems}
\phantomsection\label{\detokenize{modules_doc:cbmpy.CBModel.Compartment.setSize}}
\pysigstartsignatures
\pysiglinewithargsret{\sphinxbfcode{\sphinxupquote{setSize}}}{\sphinxparam{\DUrole{n,n}{size}}}{}
\pysigstopsignatures
\sphinxAtStartPar
Set the compartment size
\begin{itemize}
\item {} 
\sphinxAtStartPar
\sphinxstyleemphasis{size} the new compartment size

\end{itemize}

\end{fulllineitems}


\end{fulllineitems}

\index{ConstraintComponent (class in cbmpy.CBModel)@\spxentry{ConstraintComponent}\spxextra{class in cbmpy.CBModel}}

\begin{fulllineitems}
\phantomsection\label{\detokenize{modules_doc:cbmpy.CBModel.ConstraintComponent}}
\pysigstartsignatures
\pysiglinewithargsret{\sphinxbfcode{\sphinxupquote{class\DUrole{w,w}{  }}}\sphinxbfcode{\sphinxupquote{ConstraintComponent}}}{\sphinxparam{\DUrole{n,n}{pid}}\sphinxparamcomma \sphinxparam{\DUrole{n,n}{coefficient}}\sphinxparamcomma \sphinxparam{\DUrole{n,n}{variable}}\sphinxparamcomma \sphinxparam{\DUrole{n,n}{ctype}\DUrole{o,o}{=}\DUrole{default_value}{\textquotesingle{}linear\textquotesingle{}}}}{}
\pysigstopsignatures
\sphinxAtStartPar
A weighted flux that appears in an user defined constraint

\end{fulllineitems}

\index{Fbase (class in cbmpy.CBModel)@\spxentry{Fbase}\spxextra{class in cbmpy.CBModel}}

\begin{fulllineitems}
\phantomsection\label{\detokenize{modules_doc:cbmpy.CBModel.Fbase}}
\pysigstartsignatures
\pysigline{\sphinxbfcode{\sphinxupquote{class\DUrole{w,w}{  }}}\sphinxbfcode{\sphinxupquote{Fbase}}}
\pysigstopsignatures
\sphinxAtStartPar
Base class for CB Model objects
\index{addMIRIAMannotation() (Fbase method)@\spxentry{addMIRIAMannotation()}\spxextra{Fbase method}}

\begin{fulllineitems}
\phantomsection\label{\detokenize{modules_doc:cbmpy.CBModel.Fbase.addMIRIAMannotation}}
\pysigstartsignatures
\pysiglinewithargsret{\sphinxbfcode{\sphinxupquote{addMIRIAMannotation}}}{\sphinxparam{\DUrole{n,n}{qual}}\sphinxparamcomma \sphinxparam{\DUrole{n,n}{entity}}\sphinxparamcomma \sphinxparam{\DUrole{n,n}{mid}}}{}
\pysigstopsignatures
\sphinxAtStartPar
Add a qualified MIRIAM annotation or entity:
\begin{itemize}
\item {} 
\sphinxAtStartPar
\sphinxstyleemphasis{qual} a Biomodels biological qualifier e.g. “is” “isEncodedBy”

\item {} 
\sphinxAtStartPar
\sphinxstyleemphasis{entity} a MIRIAM resource entity e.g. “ChEBI”

\item {} 
\sphinxAtStartPar
\sphinxstyleemphasis{mid} the entity id e.g. CHEBI:17158 or fully qualifies url (if only\_qual\_uri)

\end{itemize}

\end{fulllineitems}

\index{addMIRIAMuri() (Fbase method)@\spxentry{addMIRIAMuri()}\spxextra{Fbase method}}

\begin{fulllineitems}
\phantomsection\label{\detokenize{modules_doc:cbmpy.CBModel.Fbase.addMIRIAMuri}}
\pysigstartsignatures
\pysiglinewithargsret{\sphinxbfcode{\sphinxupquote{addMIRIAMuri}}}{\sphinxparam{\DUrole{n,n}{qual}}\sphinxparamcomma \sphinxparam{\DUrole{n,n}{uri}}}{}
\pysigstopsignatures
\sphinxAtStartPar
Add a qualified MIRIAM annotation or entity:
\begin{itemize}
\item {} 
\sphinxAtStartPar
\sphinxstyleemphasis{qual} a Biomodels biological qualifier e.g. “is” “isEncodedBy”

\item {} 
\sphinxAtStartPar
\sphinxstyleemphasis{uri} the fully qualified entity id e.g. \sphinxurl{http://identifiers.org/chebi/CHEBI:12345} (no validity checking is done)

\end{itemize}

\end{fulllineitems}

\index{clone() (Fbase method)@\spxentry{clone()}\spxextra{Fbase method}}

\begin{fulllineitems}
\phantomsection\label{\detokenize{modules_doc:cbmpy.CBModel.Fbase.clone}}
\pysigstartsignatures
\pysiglinewithargsret{\sphinxbfcode{\sphinxupquote{clone}}}{}{}
\pysigstopsignatures
\sphinxAtStartPar
Return a clone of this object. Cloning performs a deepcop on the object which will also clone
any objects that exist as attributes of this object, in other words an independent copy of the
original. If this is not the desired behaviour override this method when subclassing or implement
your own.

\end{fulllineitems}

\index{deleteAnnotation() (Fbase method)@\spxentry{deleteAnnotation()}\spxextra{Fbase method}}

\begin{fulllineitems}
\phantomsection\label{\detokenize{modules_doc:cbmpy.CBModel.Fbase.deleteAnnotation}}
\pysigstartsignatures
\pysiglinewithargsret{\sphinxbfcode{\sphinxupquote{deleteAnnotation}}}{\sphinxparam{\DUrole{n,n}{key}}}{}
\pysigstopsignatures
\sphinxAtStartPar
Unsets (deltes) an objects annotation with key
\begin{itemize}
\item {} 
\sphinxAtStartPar
\sphinxstyleemphasis{key} the annotation key

\end{itemize}

\end{fulllineitems}

\index{deleteMIRIAMannotation() (Fbase method)@\spxentry{deleteMIRIAMannotation()}\spxextra{Fbase method}}

\begin{fulllineitems}
\phantomsection\label{\detokenize{modules_doc:cbmpy.CBModel.Fbase.deleteMIRIAMannotation}}
\pysigstartsignatures
\pysiglinewithargsret{\sphinxbfcode{\sphinxupquote{deleteMIRIAMannotation}}}{\sphinxparam{\DUrole{n,n}{qual}}\sphinxparamcomma \sphinxparam{\DUrole{n,n}{entity}}\sphinxparamcomma \sphinxparam{\DUrole{n,n}{mid}}}{}
\pysigstopsignatures
\sphinxAtStartPar
Deletes a qualified MIRIAM annotation or entity:
\begin{itemize}
\item {} 
\sphinxAtStartPar
\sphinxstyleemphasis{qual} a Biomodels biological qualifier e.g. “is” “isEncodedBy”

\item {} 
\sphinxAtStartPar
\sphinxstyleemphasis{entity} a MIRIAM resource entity e.g. “ChEBI”

\item {} 
\sphinxAtStartPar
\sphinxstyleemphasis{mid} the entity id e.g. CHEBI:17158

\end{itemize}

\end{fulllineitems}

\index{getAnnotation() (Fbase method)@\spxentry{getAnnotation()}\spxextra{Fbase method}}

\begin{fulllineitems}
\phantomsection\label{\detokenize{modules_doc:cbmpy.CBModel.Fbase.getAnnotation}}
\pysigstartsignatures
\pysiglinewithargsret{\sphinxbfcode{\sphinxupquote{getAnnotation}}}{\sphinxparam{\DUrole{n,n}{key}}}{}
\pysigstopsignatures
\sphinxAtStartPar
Return the object annotation associated with:
\begin{itemize}
\item {} 
\sphinxAtStartPar
\sphinxstyleemphasis{key} the annotation key

\end{itemize}

\end{fulllineitems}

\index{getAnnotations() (Fbase method)@\spxentry{getAnnotations()}\spxextra{Fbase method}}

\begin{fulllineitems}
\phantomsection\label{\detokenize{modules_doc:cbmpy.CBModel.Fbase.getAnnotations}}
\pysigstartsignatures
\pysiglinewithargsret{\sphinxbfcode{\sphinxupquote{getAnnotations}}}{}{}
\pysigstopsignatures
\sphinxAtStartPar
Return the object annotation dictionary

\end{fulllineitems}

\index{getCompartmentId() (Fbase method)@\spxentry{getCompartmentId()}\spxextra{Fbase method}}

\begin{fulllineitems}
\phantomsection\label{\detokenize{modules_doc:cbmpy.CBModel.Fbase.getCompartmentId}}
\pysigstartsignatures
\pysiglinewithargsret{\sphinxbfcode{\sphinxupquote{getCompartmentId}}}{}{}
\pysigstopsignatures
\sphinxAtStartPar
Return the compartment id where this element is located

\end{fulllineitems}

\index{getId() (Fbase method)@\spxentry{getId()}\spxextra{Fbase method}}

\begin{fulllineitems}
\phantomsection\label{\detokenize{modules_doc:cbmpy.CBModel.Fbase.getId}}
\pysigstartsignatures
\pysiglinewithargsret{\sphinxbfcode{\sphinxupquote{getId}}}{}{}
\pysigstopsignatures
\sphinxAtStartPar
Return the object ID.

\end{fulllineitems}

\index{getMIRIAMannotations() (Fbase method)@\spxentry{getMIRIAMannotations()}\spxextra{Fbase method}}

\begin{fulllineitems}
\phantomsection\label{\detokenize{modules_doc:cbmpy.CBModel.Fbase.getMIRIAMannotations}}
\pysigstartsignatures
\pysiglinewithargsret{\sphinxbfcode{\sphinxupquote{getMIRIAMannotations}}}{}{}
\pysigstopsignatures
\sphinxAtStartPar
Returns a dictionary of all MIRIAM annotations associated with this object
or None of there are none defined.

\end{fulllineitems}

\index{getMetaId() (Fbase method)@\spxentry{getMetaId()}\spxextra{Fbase method}}

\begin{fulllineitems}
\phantomsection\label{\detokenize{modules_doc:cbmpy.CBModel.Fbase.getMetaId}}
\pysigstartsignatures
\pysiglinewithargsret{\sphinxbfcode{\sphinxupquote{getMetaId}}}{}{}
\pysigstopsignatures
\sphinxAtStartPar
Return the object metaId.

\end{fulllineitems}

\index{getModel() (Fbase method)@\spxentry{getModel()}\spxextra{Fbase method}}

\begin{fulllineitems}
\phantomsection\label{\detokenize{modules_doc:cbmpy.CBModel.Fbase.getModel}}
\pysigstartsignatures
\pysiglinewithargsret{\sphinxbfcode{\sphinxupquote{getModel}}}{}{}
\pysigstopsignatures
\sphinxAtStartPar
Get the parent model object linked to in objref, can return model or None for unlinked object

\sphinxAtStartPar
\# Overwritten by Model

\end{fulllineitems}

\index{getName() (Fbase method)@\spxentry{getName()}\spxextra{Fbase method}}

\begin{fulllineitems}
\phantomsection\label{\detokenize{modules_doc:cbmpy.CBModel.Fbase.getName}}
\pysigstartsignatures
\pysiglinewithargsret{\sphinxbfcode{\sphinxupquote{getName}}}{}{}
\pysigstopsignatures
\sphinxAtStartPar
Return the object name.

\end{fulllineitems}

\index{getNotes() (Fbase method)@\spxentry{getNotes()}\spxextra{Fbase method}}

\begin{fulllineitems}
\phantomsection\label{\detokenize{modules_doc:cbmpy.CBModel.Fbase.getNotes}}
\pysigstartsignatures
\pysiglinewithargsret{\sphinxbfcode{\sphinxupquote{getNotes}}}{}{}
\pysigstopsignatures
\sphinxAtStartPar
Return the object’s notes

\end{fulllineitems}

\index{getPid() (Fbase method)@\spxentry{getPid()}\spxextra{Fbase method}}

\begin{fulllineitems}
\phantomsection\label{\detokenize{modules_doc:cbmpy.CBModel.Fbase.getPid}}
\pysigstartsignatures
\pysiglinewithargsret{\sphinxbfcode{\sphinxupquote{getPid}}}{}{}
\pysigstopsignatures
\sphinxAtStartPar
Return the object ID.

\end{fulllineitems}

\index{getSBOterm() (Fbase method)@\spxentry{getSBOterm()}\spxextra{Fbase method}}

\begin{fulllineitems}
\phantomsection\label{\detokenize{modules_doc:cbmpy.CBModel.Fbase.getSBOterm}}
\pysigstartsignatures
\pysiglinewithargsret{\sphinxbfcode{\sphinxupquote{getSBOterm}}}{}{}
\pysigstopsignatures
\sphinxAtStartPar
Return the SBO term for this object.

\end{fulllineitems}

\index{hasAnnotation() (Fbase method)@\spxentry{hasAnnotation()}\spxextra{Fbase method}}

\begin{fulllineitems}
\phantomsection\label{\detokenize{modules_doc:cbmpy.CBModel.Fbase.hasAnnotation}}
\pysigstartsignatures
\pysiglinewithargsret{\sphinxbfcode{\sphinxupquote{hasAnnotation}}}{\sphinxparam{\DUrole{n,n}{key}}}{}
\pysigstopsignatures
\sphinxAtStartPar
Returns a boolean representing the presence/absence of the key in the objext annotation
\begin{itemize}
\item {} 
\sphinxAtStartPar
\sphinxstyleemphasis{key} the annotation key

\end{itemize}

\end{fulllineitems}

\index{serialize() (Fbase method)@\spxentry{serialize()}\spxextra{Fbase method}}

\begin{fulllineitems}
\phantomsection\label{\detokenize{modules_doc:cbmpy.CBModel.Fbase.serialize}}
\pysigstartsignatures
\pysiglinewithargsret{\sphinxbfcode{\sphinxupquote{serialize}}}{\sphinxparam{\DUrole{n,n}{protocol}\DUrole{o,o}{=}\DUrole{default_value}{0}}}{}
\pysigstopsignatures
\sphinxAtStartPar
Serialize object, returns a string by default
\begin{itemize}
\item {} \begin{description}
\sphinxlineitem{\sphinxstyleemphasis{protocol} {[}default=0{]} serialize to a string or binary if required,}
\sphinxAtStartPar
see pickle module documentation for details

\end{description}

\end{itemize}

\sphinxAtStartPar
\# Reimplemented in Model

\end{fulllineitems}

\index{serializeToDisk() (Fbase method)@\spxentry{serializeToDisk()}\spxextra{Fbase method}}

\begin{fulllineitems}
\phantomsection\label{\detokenize{modules_doc:cbmpy.CBModel.Fbase.serializeToDisk}}
\pysigstartsignatures
\pysiglinewithargsret{\sphinxbfcode{\sphinxupquote{serializeToDisk}}}{\sphinxparam{\DUrole{n,n}{filename}}\sphinxparamcomma \sphinxparam{\DUrole{n,n}{protocol}\DUrole{o,o}{=}\DUrole{default_value}{2}}}{}
\pysigstopsignatures
\sphinxAtStartPar
Serialize to disk using pickle protocol:
\begin{itemize}
\item {} 
\sphinxAtStartPar
\sphinxstyleemphasis{filename} the name of the output file

\item {} \begin{description}
\sphinxlineitem{\sphinxstyleemphasis{protocol} {[}default=2{]} serialize to a string or binary if required,}
\sphinxAtStartPar
see pickle module documentation for details

\end{description}

\end{itemize}

\sphinxAtStartPar
\# Reimplemented in Model

\end{fulllineitems}

\index{setAnnotation() (Fbase method)@\spxentry{setAnnotation()}\spxextra{Fbase method}}

\begin{fulllineitems}
\phantomsection\label{\detokenize{modules_doc:cbmpy.CBModel.Fbase.setAnnotation}}
\pysigstartsignatures
\pysiglinewithargsret{\sphinxbfcode{\sphinxupquote{setAnnotation}}}{\sphinxparam{\DUrole{n,n}{key}}\sphinxparamcomma \sphinxparam{\DUrole{n,n}{value}}\sphinxparamcomma \sphinxparam{\DUrole{n,n}{ext}\DUrole{o,o}{=}\DUrole{default_value}{None}}}{}
\pysigstopsignatures
\sphinxAtStartPar
Set an objects annotation as a key : value pair.
\begin{itemize}
\item {} 
\sphinxAtStartPar
\sphinxstyleemphasis{key} the annotation key

\item {} 
\sphinxAtStartPar
\sphinxstyleemphasis{value} the annotation value

\item {} 
\sphinxAtStartPar
\sphinxstyleemphasis{ext} a dictionary of extended properties in FBCv3 id, name, uri

\end{itemize}

\end{fulllineitems}

\index{setCompartmentId() (Fbase method)@\spxentry{setCompartmentId()}\spxextra{Fbase method}}

\begin{fulllineitems}
\phantomsection\label{\detokenize{modules_doc:cbmpy.CBModel.Fbase.setCompartmentId}}
\pysigstartsignatures
\pysiglinewithargsret{\sphinxbfcode{\sphinxupquote{setCompartmentId}}}{\sphinxparam{\DUrole{n,n}{compartment}}}{}
\pysigstopsignatures
\sphinxAtStartPar
Set the compartment id where this element is located

\end{fulllineitems}

\index{setId() (Fbase method)@\spxentry{setId()}\spxextra{Fbase method}}

\begin{fulllineitems}
\phantomsection\label{\detokenize{modules_doc:cbmpy.CBModel.Fbase.setId}}
\pysigstartsignatures
\pysiglinewithargsret{\sphinxbfcode{\sphinxupquote{setId}}}{\sphinxparam{\DUrole{n,n}{fid}}}{}
\pysigstopsignatures
\sphinxAtStartPar
Sets the object Id
\begin{quote}
\begin{itemize}
\item {} 
\sphinxAtStartPar
\sphinxstyleemphasis{fid} a valid c variable style id string

\end{itemize}

\sphinxAtStartPar
Reimplemented by @Reaction, @Species, @Compartment, @Gene
\end{quote}

\end{fulllineitems}

\index{setMetaId() (Fbase method)@\spxentry{setMetaId()}\spxextra{Fbase method}}

\begin{fulllineitems}
\phantomsection\label{\detokenize{modules_doc:cbmpy.CBModel.Fbase.setMetaId}}
\pysigstartsignatures
\pysiglinewithargsret{\sphinxbfcode{\sphinxupquote{setMetaId}}}{\sphinxparam{\DUrole{n,n}{mid}\DUrole{o,o}{=}\DUrole{default_value}{None}}}{}
\pysigstopsignatures
\sphinxAtStartPar
Sets the object Id
\begin{itemize}
\item {} 
\sphinxAtStartPar
\sphinxstyleemphasis{mid} {[}default=None{]} a valid c variable style metaid string, if None it will be set as meta+id

\end{itemize}

\end{fulllineitems}

\index{setName() (Fbase method)@\spxentry{setName()}\spxextra{Fbase method}}

\begin{fulllineitems}
\phantomsection\label{\detokenize{modules_doc:cbmpy.CBModel.Fbase.setName}}
\pysigstartsignatures
\pysiglinewithargsret{\sphinxbfcode{\sphinxupquote{setName}}}{\sphinxparam{\DUrole{n,n}{name}}}{}
\pysigstopsignatures
\sphinxAtStartPar
Set the object name:
\begin{itemize}
\item {} 
\sphinxAtStartPar
\sphinxstyleemphasis{name} the name string

\end{itemize}

\end{fulllineitems}

\index{setNotes() (Fbase method)@\spxentry{setNotes()}\spxextra{Fbase method}}

\begin{fulllineitems}
\phantomsection\label{\detokenize{modules_doc:cbmpy.CBModel.Fbase.setNotes}}
\pysigstartsignatures
\pysiglinewithargsret{\sphinxbfcode{\sphinxupquote{setNotes}}}{\sphinxparam{\DUrole{n,n}{notes}}}{}
\pysigstopsignatures
\sphinxAtStartPar
Sets the object’s notes:
\begin{itemize}
\item {} 
\sphinxAtStartPar
\sphinxstyleemphasis{notes} the note string, should preferably be (X)HTML for SBML

\end{itemize}

\end{fulllineitems}

\index{setPid() (Fbase method)@\spxentry{setPid()}\spxextra{Fbase method}}

\begin{fulllineitems}
\phantomsection\label{\detokenize{modules_doc:cbmpy.CBModel.Fbase.setPid}}
\pysigstartsignatures
\pysiglinewithargsret{\sphinxbfcode{\sphinxupquote{setPid}}}{\sphinxparam{\DUrole{n,n}{fid}}}{}
\pysigstopsignatures
\sphinxAtStartPar
Sets the object Id
\begin{itemize}
\item {} 
\sphinxAtStartPar
\sphinxstyleemphasis{fid} a valid c variable style id string

\end{itemize}

\end{fulllineitems}

\index{setSBOterm() (Fbase method)@\spxentry{setSBOterm()}\spxextra{Fbase method}}

\begin{fulllineitems}
\phantomsection\label{\detokenize{modules_doc:cbmpy.CBModel.Fbase.setSBOterm}}
\pysigstartsignatures
\pysiglinewithargsret{\sphinxbfcode{\sphinxupquote{setSBOterm}}}{\sphinxparam{\DUrole{n,n}{sbo}}}{}
\pysigstopsignatures
\sphinxAtStartPar
Set the SBO term for this object.
\begin{itemize}
\item {} 
\sphinxAtStartPar
\sphinxstyleemphasis{sbo} the SBOterm with format: SBO:nnnnnnn”

\end{itemize}

\end{fulllineitems}


\end{fulllineitems}

\index{FluxBound (class in cbmpy.CBModel)@\spxentry{FluxBound}\spxextra{class in cbmpy.CBModel}}

\begin{fulllineitems}
\phantomsection\label{\detokenize{modules_doc:cbmpy.CBModel.FluxBound}}
\pysigstartsignatures
\pysiglinewithargsret{\sphinxbfcode{\sphinxupquote{class\DUrole{w,w}{  }}}\sphinxbfcode{\sphinxupquote{FluxBound}}}{\sphinxparam{\DUrole{n,n}{pid}}\sphinxparamcomma \sphinxparam{\DUrole{n,n}{reaction}}\sphinxparamcomma \sphinxparam{\DUrole{n,n}{operation}}\sphinxparamcomma \sphinxparam{\DUrole{n,n}{value}}}{}
\pysigstopsignatures
\sphinxAtStartPar
A reaction fluxbound
\index{getType() (FluxBound method)@\spxentry{getType()}\spxextra{FluxBound method}}

\begin{fulllineitems}
\phantomsection\label{\detokenize{modules_doc:cbmpy.CBModel.FluxBound.getType}}
\pysigstartsignatures
\pysiglinewithargsret{\sphinxbfcode{\sphinxupquote{getType}}}{}{}
\pysigstopsignatures
\sphinxAtStartPar
Returns the \sphinxstyleemphasis{type} of FluxBound: ‘lower’, ‘upper’, ‘equality’ or None

\end{fulllineitems}

\index{getValue() (FluxBound method)@\spxentry{getValue()}\spxextra{FluxBound method}}

\begin{fulllineitems}
\phantomsection\label{\detokenize{modules_doc:cbmpy.CBModel.FluxBound.getValue}}
\pysigstartsignatures
\pysiglinewithargsret{\sphinxbfcode{\sphinxupquote{getValue}}}{}{}
\pysigstopsignatures
\sphinxAtStartPar
Returns the current value of the attribute (input/solution)

\end{fulllineitems}

\index{setReactionId() (FluxBound method)@\spxentry{setReactionId()}\spxextra{FluxBound method}}

\begin{fulllineitems}
\phantomsection\label{\detokenize{modules_doc:cbmpy.CBModel.FluxBound.setReactionId}}
\pysigstartsignatures
\pysiglinewithargsret{\sphinxbfcode{\sphinxupquote{setReactionId}}}{\sphinxparam{\DUrole{n,n}{react}}}{}
\pysigstopsignatures
\sphinxAtStartPar
Sets the reaction attribute of the FluxBound

\end{fulllineitems}

\index{setValue() (FluxBound method)@\spxentry{setValue()}\spxextra{FluxBound method}}

\begin{fulllineitems}
\phantomsection\label{\detokenize{modules_doc:cbmpy.CBModel.FluxBound.setValue}}
\pysigstartsignatures
\pysiglinewithargsret{\sphinxbfcode{\sphinxupquote{setValue}}}{\sphinxparam{\DUrole{n,n}{value}}}{}
\pysigstopsignatures
\sphinxAtStartPar
Sets the attribute ‘’value’’

\end{fulllineitems}


\end{fulllineitems}

\index{FluxBoundBase (class in cbmpy.CBModel)@\spxentry{FluxBoundBase}\spxextra{class in cbmpy.CBModel}}

\begin{fulllineitems}
\phantomsection\label{\detokenize{modules_doc:cbmpy.CBModel.FluxBoundBase}}
\pysigstartsignatures
\pysiglinewithargsret{\sphinxbfcode{\sphinxupquote{class\DUrole{w,w}{  }}}\sphinxbfcode{\sphinxupquote{FluxBoundBase}}}{\sphinxparam{\DUrole{n,n}{pid}}\sphinxparamcomma \sphinxparam{\DUrole{n,n}{operator}}\sphinxparamcomma \sphinxparam{\DUrole{n,n}{value}}\sphinxparamcomma \sphinxparam{\DUrole{n,n}{parent}\DUrole{o,o}{=}\DUrole{default_value}{None}}}{}
\pysigstopsignatures
\sphinxAtStartPar
A refactored and streamlined FluxBound base class that can be a generic bound, superclass to FluxBoundUpper and FluxBoundLower
\index{getType() (FluxBoundBase method)@\spxentry{getType()}\spxextra{FluxBoundBase method}}

\begin{fulllineitems}
\phantomsection\label{\detokenize{modules_doc:cbmpy.CBModel.FluxBoundBase.getType}}
\pysigstartsignatures
\pysiglinewithargsret{\sphinxbfcode{\sphinxupquote{getType}}}{}{}
\pysigstopsignatures
\sphinxAtStartPar
Returns the \sphinxstyleemphasis{type} of FluxBound: ‘lower’, ‘upper’

\end{fulllineitems}

\index{getValue() (FluxBoundBase method)@\spxentry{getValue()}\spxextra{FluxBoundBase method}}

\begin{fulllineitems}
\phantomsection\label{\detokenize{modules_doc:cbmpy.CBModel.FluxBoundBase.getValue}}
\pysigstartsignatures
\pysiglinewithargsret{\sphinxbfcode{\sphinxupquote{getValue}}}{}{}
\pysigstopsignatures
\sphinxAtStartPar
Returns the current value of the attribute (input/solution)

\end{fulllineitems}

\index{setValue() (FluxBoundBase method)@\spxentry{setValue()}\spxextra{FluxBoundBase method}}

\begin{fulllineitems}
\phantomsection\label{\detokenize{modules_doc:cbmpy.CBModel.FluxBoundBase.setValue}}
\pysigstartsignatures
\pysiglinewithargsret{\sphinxbfcode{\sphinxupquote{setValue}}}{\sphinxparam{\DUrole{n,n}{value}}}{}
\pysigstopsignatures
\sphinxAtStartPar
Sets the value attribute:
\begin{itemize}
\item {} 
\sphinxAtStartPar
\sphinxstyleemphasis{value} a float

\end{itemize}

\end{fulllineitems}

\index{value (FluxBoundBase property)@\spxentry{value}\spxextra{FluxBoundBase property}}

\begin{fulllineitems}
\phantomsection\label{\detokenize{modules_doc:cbmpy.CBModel.FluxBoundBase.value}}
\pysigstartsignatures
\pysigline{\sphinxbfcode{\sphinxupquote{property\DUrole{w,w}{  }}}\sphinxbfcode{\sphinxupquote{value}}}
\pysigstopsignatures
\sphinxAtStartPar
Returns the current value of the attribute (input/solution)

\end{fulllineitems}


\end{fulllineitems}

\index{FluxBoundLower (class in cbmpy.CBModel)@\spxentry{FluxBoundLower}\spxextra{class in cbmpy.CBModel}}

\begin{fulllineitems}
\phantomsection\label{\detokenize{modules_doc:cbmpy.CBModel.FluxBoundLower}}
\pysigstartsignatures
\pysiglinewithargsret{\sphinxbfcode{\sphinxupquote{class\DUrole{w,w}{  }}}\sphinxbfcode{\sphinxupquote{FluxBoundLower}}}{\sphinxparam{\DUrole{n,n}{reaction}}\sphinxparamcomma \sphinxparam{\DUrole{n,n}{value}\DUrole{o,o}{=}\DUrole{default_value}{\sphinxhyphen{}inf}}}{}
\pysigstopsignatures
\end{fulllineitems}

\index{FluxBoundUpper (class in cbmpy.CBModel)@\spxentry{FluxBoundUpper}\spxextra{class in cbmpy.CBModel}}

\begin{fulllineitems}
\phantomsection\label{\detokenize{modules_doc:cbmpy.CBModel.FluxBoundUpper}}
\pysigstartsignatures
\pysiglinewithargsret{\sphinxbfcode{\sphinxupquote{class\DUrole{w,w}{  }}}\sphinxbfcode{\sphinxupquote{FluxBoundUpper}}}{\sphinxparam{\DUrole{n,n}{reaction}}\sphinxparamcomma \sphinxparam{\DUrole{n,n}{value}\DUrole{o,o}{=}\DUrole{default_value}{inf}}}{}
\pysigstopsignatures
\end{fulllineitems}

\index{FluxObjective (class in cbmpy.CBModel)@\spxentry{FluxObjective}\spxextra{class in cbmpy.CBModel}}

\begin{fulllineitems}
\phantomsection\label{\detokenize{modules_doc:cbmpy.CBModel.FluxObjective}}
\pysigstartsignatures
\pysiglinewithargsret{\sphinxbfcode{\sphinxupquote{class\DUrole{w,w}{  }}}\sphinxbfcode{\sphinxupquote{FluxObjective}}}{\sphinxparam{\DUrole{n,n}{pid}}\sphinxparamcomma \sphinxparam{\DUrole{n,n}{reaction}}\sphinxparamcomma \sphinxparam{\DUrole{n,n}{coefficient}\DUrole{o,o}{=}\DUrole{default_value}{1}}\sphinxparamcomma \sphinxparam{\DUrole{n,n}{ctype}\DUrole{o,o}{=}\DUrole{default_value}{\textquotesingle{}linear\textquotesingle{}}}}{}
\pysigstopsignatures
\sphinxAtStartPar
A weighted flux that appears in an objective function

\sphinxAtStartPar
NOTE: reaction is a string containing a reaction id

\end{fulllineitems}

\index{FluxObjectiveQuadratic (class in cbmpy.CBModel)@\spxentry{FluxObjectiveQuadratic}\spxextra{class in cbmpy.CBModel}}

\begin{fulllineitems}
\phantomsection\label{\detokenize{modules_doc:cbmpy.CBModel.FluxObjectiveQuadratic}}
\pysigstartsignatures
\pysiglinewithargsret{\sphinxbfcode{\sphinxupquote{class\DUrole{w,w}{  }}}\sphinxbfcode{\sphinxupquote{FluxObjectiveQuadratic}}}{\sphinxparam{\DUrole{n,n}{pid}}\sphinxparamcomma \sphinxparam{\DUrole{n,n}{reaction}}\sphinxparamcomma \sphinxparam{\DUrole{n,n}{reaction2}}\sphinxparamcomma \sphinxparam{\DUrole{n,n}{coefficient}\DUrole{o,o}{=}\DUrole{default_value}{1}}\sphinxparamcomma \sphinxparam{\DUrole{n,n}{ctype}\DUrole{o,o}{=}\DUrole{default_value}{\textquotesingle{}quadratic\textquotesingle{}}}}{}
\pysigstopsignatures
\sphinxAtStartPar
A weighted quadratic flux that appears in an objective function, this fluxobjective contains
two reaction terms to define “quadratic” fluxobjectives of the type \textless{}coefficient\textgreater{}*\textless{}variable1\textgreater{}*\textless{}variable2\textgreater{}
For example 2*R1*R2

\sphinxAtStartPar
NOTE: reaction is a string containing a reaction id

\end{fulllineitems}

\index{Gene (class in cbmpy.CBModel)@\spxentry{Gene}\spxextra{class in cbmpy.CBModel}}

\begin{fulllineitems}
\phantomsection\label{\detokenize{modules_doc:cbmpy.CBModel.Gene}}
\pysigstartsignatures
\pysiglinewithargsret{\sphinxbfcode{\sphinxupquote{class\DUrole{w,w}{  }}}\sphinxbfcode{\sphinxupquote{Gene}}}{\sphinxparam{\DUrole{n,n}{pid}}\sphinxparamcomma \sphinxparam{\DUrole{n,n}{label}\DUrole{o,o}{=}\DUrole{default_value}{None}}\sphinxparamcomma \sphinxparam{\DUrole{n,n}{active}\DUrole{o,o}{=}\DUrole{default_value}{True}}}{}
\pysigstopsignatures
\sphinxAtStartPar
Contains all the information about a gene (or gene+protein construct depending on your philosophy)
\index{getLabel() (Gene method)@\spxentry{getLabel()}\spxextra{Gene method}}

\begin{fulllineitems}
\phantomsection\label{\detokenize{modules_doc:cbmpy.CBModel.Gene.getLabel}}
\pysigstartsignatures
\pysiglinewithargsret{\sphinxbfcode{\sphinxupquote{getLabel}}}{}{}
\pysigstopsignatures
\sphinxAtStartPar
Returns the gene label

\end{fulllineitems}

\index{isActive() (Gene method)@\spxentry{isActive()}\spxextra{Gene method}}

\begin{fulllineitems}
\phantomsection\label{\detokenize{modules_doc:cbmpy.CBModel.Gene.isActive}}
\pysigstartsignatures
\pysiglinewithargsret{\sphinxbfcode{\sphinxupquote{isActive}}}{}{}
\pysigstopsignatures
\sphinxAtStartPar
Returns whether the gene is active or not

\end{fulllineitems}

\index{resetActivity() (Gene method)@\spxentry{resetActivity()}\spxextra{Gene method}}

\begin{fulllineitems}
\phantomsection\label{\detokenize{modules_doc:cbmpy.CBModel.Gene.resetActivity}}
\pysigstartsignatures
\pysiglinewithargsret{\sphinxbfcode{\sphinxupquote{resetActivity}}}{}{}
\pysigstopsignatures
\sphinxAtStartPar
Reset the gene to its default activity state

\end{fulllineitems}

\index{setActive() (Gene method)@\spxentry{setActive()}\spxextra{Gene method}}

\begin{fulllineitems}
\phantomsection\label{\detokenize{modules_doc:cbmpy.CBModel.Gene.setActive}}
\pysigstartsignatures
\pysiglinewithargsret{\sphinxbfcode{\sphinxupquote{setActive}}}{}{}
\pysigstopsignatures
\sphinxAtStartPar
Set the gene to be active

\end{fulllineitems}

\index{setId() (Gene method)@\spxentry{setId()}\spxextra{Gene method}}

\begin{fulllineitems}
\phantomsection\label{\detokenize{modules_doc:cbmpy.CBModel.Gene.setId}}
\pysigstartsignatures
\pysiglinewithargsret{\sphinxbfcode{\sphinxupquote{setId}}}{\sphinxparam{\DUrole{n,n}{fid}}}{}
\pysigstopsignatures
\sphinxAtStartPar
Sets the object Id
\begin{quote}
\begin{itemize}
\item {} 
\sphinxAtStartPar
\sphinxstyleemphasis{fid} a valid c variable style id string

\end{itemize}

\sphinxAtStartPar
Reimplements Fbase method
\end{quote}

\end{fulllineitems}

\index{setInactive() (Gene method)@\spxentry{setInactive()}\spxextra{Gene method}}

\begin{fulllineitems}
\phantomsection\label{\detokenize{modules_doc:cbmpy.CBModel.Gene.setInactive}}
\pysigstartsignatures
\pysiglinewithargsret{\sphinxbfcode{\sphinxupquote{setInactive}}}{}{}
\pysigstopsignatures
\sphinxAtStartPar
Set the gene to be inactive

\end{fulllineitems}

\index{setLabel() (Gene method)@\spxentry{setLabel()}\spxextra{Gene method}}

\begin{fulllineitems}
\phantomsection\label{\detokenize{modules_doc:cbmpy.CBModel.Gene.setLabel}}
\pysigstartsignatures
\pysiglinewithargsret{\sphinxbfcode{\sphinxupquote{setLabel}}}{\sphinxparam{\DUrole{n,n}{label}}}{}
\pysigstopsignatures
\sphinxAtStartPar
Sets the gene label, checks that the new label is unique

\end{fulllineitems}


\end{fulllineitems}

\index{GeneProteinAssociation (class in cbmpy.CBModel)@\spxentry{GeneProteinAssociation}\spxextra{class in cbmpy.CBModel}}

\begin{fulllineitems}
\phantomsection\label{\detokenize{modules_doc:cbmpy.CBModel.GeneProteinAssociation}}
\pysigstartsignatures
\pysiglinewithargsret{\sphinxbfcode{\sphinxupquote{class\DUrole{w,w}{  }}}\sphinxbfcode{\sphinxupquote{GeneProteinAssociation}}}{\sphinxparam{\DUrole{n,n}{pid}}\sphinxparamcomma \sphinxparam{\DUrole{n,n}{protein}}\sphinxparamcomma \sphinxparam{\DUrole{n,n}{use\_compiled}\DUrole{o,o}{=}\DUrole{default_value}{False}}}{}
\pysigstopsignatures
\sphinxAtStartPar
This class associates genes to proteins.
\index{addAssociation() (GeneProteinAssociation method)@\spxentry{addAssociation()}\spxextra{GeneProteinAssociation method}}

\begin{fulllineitems}
\phantomsection\label{\detokenize{modules_doc:cbmpy.CBModel.GeneProteinAssociation.addAssociation}}
\pysigstartsignatures
\pysiglinewithargsret{\sphinxbfcode{\sphinxupquote{addAssociation}}}{\sphinxparam{\DUrole{n,n}{assoc}}}{}
\pysigstopsignatures
\sphinxAtStartPar
Add a gene/protein association expression

\end{fulllineitems}

\index{addGeneref() (GeneProteinAssociation method)@\spxentry{addGeneref()}\spxextra{GeneProteinAssociation method}}

\begin{fulllineitems}
\phantomsection\label{\detokenize{modules_doc:cbmpy.CBModel.GeneProteinAssociation.addGeneref}}
\pysigstartsignatures
\pysiglinewithargsret{\sphinxbfcode{\sphinxupquote{addGeneref}}}{\sphinxparam{\DUrole{n,n}{geneid}}}{}
\pysigstopsignatures
\sphinxAtStartPar
Add a gene reference to the list of gene references
\begin{itemize}
\item {} 
\sphinxAtStartPar
\sphinxstyleemphasis{geneid} a valid model Gene id

\end{itemize}

\end{fulllineitems}

\index{createAssociationAndGeneRefsFromString() (GeneProteinAssociation method)@\spxentry{createAssociationAndGeneRefsFromString()}\spxextra{GeneProteinAssociation method}}

\begin{fulllineitems}
\phantomsection\label{\detokenize{modules_doc:cbmpy.CBModel.GeneProteinAssociation.createAssociationAndGeneRefsFromString}}
\pysigstartsignatures
\pysiglinewithargsret{\sphinxbfcode{\sphinxupquote{createAssociationAndGeneRefsFromString}}}{\sphinxparam{\DUrole{n,n}{assoc}}\sphinxparamcomma \sphinxparam{\DUrole{n,n}{altlabels}\DUrole{o,o}{=}\DUrole{default_value}{None}}}{}
\pysigstopsignatures
\sphinxAtStartPar
Evaluate the gene/protein association and add the genes necessary to evaluate it
Note that this GPR should be added to a model with cmod.addGPRAssociation() before calling this method
\begin{itemize}
\item {} 
\sphinxAtStartPar
\sphinxstyleemphasis{assoc} the COBRA style gene protein association

\item {} 
\sphinxAtStartPar
\sphinxstyleemphasis{altlabels} {[}default=None{]} a dictionary containing a label\textless{}\textendash{}\textgreater{}id mapping

\end{itemize}

\end{fulllineitems}

\index{createAssociationAndGeneRefsFromTree() (GeneProteinAssociation method)@\spxentry{createAssociationAndGeneRefsFromTree()}\spxextra{GeneProteinAssociation method}}

\begin{fulllineitems}
\phantomsection\label{\detokenize{modules_doc:cbmpy.CBModel.GeneProteinAssociation.createAssociationAndGeneRefsFromTree}}
\pysigstartsignatures
\pysiglinewithargsret{\sphinxbfcode{\sphinxupquote{createAssociationAndGeneRefsFromTree}}}{\sphinxparam{\DUrole{n,n}{gprtree}}\sphinxparamcomma \sphinxparam{\DUrole{n,n}{altlabels}\DUrole{o,o}{=}\DUrole{default_value}{None}}}{}
\pysigstopsignatures
\sphinxAtStartPar
Evaluate the GPR tree and add the genes necessary to evaluate it
Note that this GPR should be added to a model with cmod.addGPRAssociation() before calling this method
\begin{itemize}
\item {} 
\sphinxAtStartPar
\sphinxstyleemphasis{gprtree} the CBMPy GPR tree data structure

\item {} 
\sphinxAtStartPar
\sphinxstyleemphasis{altlabels} {[}default=None{]} a dictionary containing a label\textless{}\textendash{}\textgreater{}id mapping

\end{itemize}

\end{fulllineitems}

\index{deleteGeneFromAssociation() (GeneProteinAssociation method)@\spxentry{deleteGeneFromAssociation()}\spxextra{GeneProteinAssociation method}}

\begin{fulllineitems}
\phantomsection\label{\detokenize{modules_doc:cbmpy.CBModel.GeneProteinAssociation.deleteGeneFromAssociation}}
\pysigstartsignatures
\pysiglinewithargsret{\sphinxbfcode{\sphinxupquote{deleteGeneFromAssociation}}}{\sphinxparam{\DUrole{n,n}{gid}}}{}
\pysigstopsignatures
\sphinxAtStartPar
Deletes a gene id from the gene association. \sphinxstyleemphasis{WARNING} this process is irreversible!!
\begin{itemize}
\item {} 
\sphinxAtStartPar
\sphinxstyleemphasis{gid} a valid gene identifier (not label)

\end{itemize}

\end{fulllineitems}

\index{deleteGeneref() (GeneProteinAssociation method)@\spxentry{deleteGeneref()}\spxextra{GeneProteinAssociation method}}

\begin{fulllineitems}
\phantomsection\label{\detokenize{modules_doc:cbmpy.CBModel.GeneProteinAssociation.deleteGeneref}}
\pysigstartsignatures
\pysiglinewithargsret{\sphinxbfcode{\sphinxupquote{deleteGeneref}}}{\sphinxparam{\DUrole{n,n}{gid}}}{}
\pysigstopsignatures
\sphinxAtStartPar
Deletes a gene reference
\begin{itemize}
\item {} 
\sphinxAtStartPar
\sphinxstyleemphasis{geneid} a valid model Gene id

\end{itemize}

\end{fulllineitems}

\index{evalAssociation() (GeneProteinAssociation method)@\spxentry{evalAssociation()}\spxextra{GeneProteinAssociation method}}

\begin{fulllineitems}
\phantomsection\label{\detokenize{modules_doc:cbmpy.CBModel.GeneProteinAssociation.evalAssociation}}
\pysigstartsignatures
\pysiglinewithargsret{\sphinxbfcode{\sphinxupquote{evalAssociation}}}{}{}
\pysigstopsignatures
\sphinxAtStartPar
Returns an integer value representing the logical associations or None.

\end{fulllineitems}

\index{getActiveGenes() (GeneProteinAssociation method)@\spxentry{getActiveGenes()}\spxextra{GeneProteinAssociation method}}

\begin{fulllineitems}
\phantomsection\label{\detokenize{modules_doc:cbmpy.CBModel.GeneProteinAssociation.getActiveGenes}}
\pysigstartsignatures
\pysiglinewithargsret{\sphinxbfcode{\sphinxupquote{getActiveGenes}}}{}{}
\pysigstopsignatures
\sphinxAtStartPar
Return a list of active gene objects

\end{fulllineitems}

\index{getAssociationStr() (GeneProteinAssociation method)@\spxentry{getAssociationStr()}\spxextra{GeneProteinAssociation method}}

\begin{fulllineitems}
\phantomsection\label{\detokenize{modules_doc:cbmpy.CBModel.GeneProteinAssociation.getAssociationStr}}
\pysigstartsignatures
\pysiglinewithargsret{\sphinxbfcode{\sphinxupquote{getAssociationStr}}}{\sphinxparam{\DUrole{n,n}{use\_labels}\DUrole{o,o}{=}\DUrole{default_value}{False}}}{}
\pysigstopsignatures
\sphinxAtStartPar
Return the gene association string, alternatively return string with labels
\begin{itemize}
\item {} 
\sphinxAtStartPar
\sphinxstyleemphasis{use\_lablels} {[}default=False{]} return the gene association string with labels rather than geneId’s (FBCv2 issue)

\end{itemize}

\end{fulllineitems}

\index{getGene() (GeneProteinAssociation method)@\spxentry{getGene()}\spxextra{GeneProteinAssociation method}}

\begin{fulllineitems}
\phantomsection\label{\detokenize{modules_doc:cbmpy.CBModel.GeneProteinAssociation.getGene}}
\pysigstartsignatures
\pysiglinewithargsret{\sphinxbfcode{\sphinxupquote{getGene}}}{\sphinxparam{\DUrole{n,n}{gid}}}{}
\pysigstopsignatures
\sphinxAtStartPar
Return a gene object with id

\end{fulllineitems}

\index{getGeneIds() (GeneProteinAssociation method)@\spxentry{getGeneIds()}\spxextra{GeneProteinAssociation method}}

\begin{fulllineitems}
\phantomsection\label{\detokenize{modules_doc:cbmpy.CBModel.GeneProteinAssociation.getGeneIds}}
\pysigstartsignatures
\pysiglinewithargsret{\sphinxbfcode{\sphinxupquote{getGeneIds}}}{}{}
\pysigstopsignatures
\sphinxAtStartPar
Return a list of gene id’s

\end{fulllineitems}

\index{getGeneLabels() (GeneProteinAssociation method)@\spxentry{getGeneLabels()}\spxextra{GeneProteinAssociation method}}

\begin{fulllineitems}
\phantomsection\label{\detokenize{modules_doc:cbmpy.CBModel.GeneProteinAssociation.getGeneLabels}}
\pysigstartsignatures
\pysiglinewithargsret{\sphinxbfcode{\sphinxupquote{getGeneLabels}}}{}{}
\pysigstopsignatures
\sphinxAtStartPar
Return a list of gene labels associated with this GPRass

\end{fulllineitems}

\index{getGenes() (GeneProteinAssociation method)@\spxentry{getGenes()}\spxextra{GeneProteinAssociation method}}

\begin{fulllineitems}
\phantomsection\label{\detokenize{modules_doc:cbmpy.CBModel.GeneProteinAssociation.getGenes}}
\pysigstartsignatures
\pysiglinewithargsret{\sphinxbfcode{\sphinxupquote{getGenes}}}{}{}
\pysigstopsignatures
\sphinxAtStartPar
Return a list of gene objects associated with this GPRass

\end{fulllineitems}

\index{getProtein() (GeneProteinAssociation method)@\spxentry{getProtein()}\spxextra{GeneProteinAssociation method}}

\begin{fulllineitems}
\phantomsection\label{\detokenize{modules_doc:cbmpy.CBModel.GeneProteinAssociation.getProtein}}
\pysigstartsignatures
\pysiglinewithargsret{\sphinxbfcode{\sphinxupquote{getProtein}}}{}{}
\pysigstopsignatures
\sphinxAtStartPar
Return the protein associated with this set of genes

\end{fulllineitems}

\index{getTree() (GeneProteinAssociation method)@\spxentry{getTree()}\spxextra{GeneProteinAssociation method}}

\begin{fulllineitems}
\phantomsection\label{\detokenize{modules_doc:cbmpy.CBModel.GeneProteinAssociation.getTree}}
\pysigstartsignatures
\pysiglinewithargsret{\sphinxbfcode{\sphinxupquote{getTree}}}{}{}
\pysigstopsignatures
\sphinxAtStartPar
Return the dictionary/tree representation of the GPR

\end{fulllineitems}

\index{getTreeCopy() (GeneProteinAssociation method)@\spxentry{getTreeCopy()}\spxextra{GeneProteinAssociation method}}

\begin{fulllineitems}
\phantomsection\label{\detokenize{modules_doc:cbmpy.CBModel.GeneProteinAssociation.getTreeCopy}}
\pysigstartsignatures
\pysiglinewithargsret{\sphinxbfcode{\sphinxupquote{getTreeCopy}}}{}{}
\pysigstopsignatures
\sphinxAtStartPar
Return a copy of the dictionary/tree representation of the GPR

\end{fulllineitems}

\index{isProteinActive() (GeneProteinAssociation method)@\spxentry{isProteinActive()}\spxextra{GeneProteinAssociation method}}

\begin{fulllineitems}
\phantomsection\label{\detokenize{modules_doc:cbmpy.CBModel.GeneProteinAssociation.isProteinActive}}
\pysigstartsignatures
\pysiglinewithargsret{\sphinxbfcode{\sphinxupquote{isProteinActive}}}{}{}
\pysigstopsignatures
\sphinxAtStartPar
This returns a boolean which indicates the result of evaluating the gene association. If the result is positive
then the protein is expressed and \sphinxstyleemphasis{True} is returned, otherwise if the expression evaluates to a value of 0 then
the protein is not expressed and  \sphinxstyleemphasis{False} is returned.

\end{fulllineitems}

\index{setAllGenesActive() (GeneProteinAssociation method)@\spxentry{setAllGenesActive()}\spxextra{GeneProteinAssociation method}}

\begin{fulllineitems}
\phantomsection\label{\detokenize{modules_doc:cbmpy.CBModel.GeneProteinAssociation.setAllGenesActive}}
\pysigstartsignatures
\pysiglinewithargsret{\sphinxbfcode{\sphinxupquote{setAllGenesActive}}}{}{}
\pysigstopsignatures
\sphinxAtStartPar
Activate all genes in association

\end{fulllineitems}

\index{setAllGenesInactive() (GeneProteinAssociation method)@\spxentry{setAllGenesInactive()}\spxextra{GeneProteinAssociation method}}

\begin{fulllineitems}
\phantomsection\label{\detokenize{modules_doc:cbmpy.CBModel.GeneProteinAssociation.setAllGenesInactive}}
\pysigstartsignatures
\pysiglinewithargsret{\sphinxbfcode{\sphinxupquote{setAllGenesInactive}}}{}{}
\pysigstopsignatures
\sphinxAtStartPar
Deactivates all genes in association

\end{fulllineitems}

\index{setGeneActive() (GeneProteinAssociation method)@\spxentry{setGeneActive()}\spxextra{GeneProteinAssociation method}}

\begin{fulllineitems}
\phantomsection\label{\detokenize{modules_doc:cbmpy.CBModel.GeneProteinAssociation.setGeneActive}}
\pysigstartsignatures
\pysiglinewithargsret{\sphinxbfcode{\sphinxupquote{setGeneActive}}}{\sphinxparam{\DUrole{n,n}{gid}}}{}
\pysigstopsignatures
\sphinxAtStartPar
Set a gene to be inactive

\end{fulllineitems}

\index{setGeneInactive() (GeneProteinAssociation method)@\spxentry{setGeneInactive()}\spxextra{GeneProteinAssociation method}}

\begin{fulllineitems}
\phantomsection\label{\detokenize{modules_doc:cbmpy.CBModel.GeneProteinAssociation.setGeneInactive}}
\pysigstartsignatures
\pysiglinewithargsret{\sphinxbfcode{\sphinxupquote{setGeneInactive}}}{\sphinxparam{\DUrole{n,n}{gid}}}{}
\pysigstopsignatures
\sphinxAtStartPar
Set a gene to be inactive

\end{fulllineitems}

\index{setProtein() (GeneProteinAssociation method)@\spxentry{setProtein()}\spxextra{GeneProteinAssociation method}}

\begin{fulllineitems}
\phantomsection\label{\detokenize{modules_doc:cbmpy.CBModel.GeneProteinAssociation.setProtein}}
\pysigstartsignatures
\pysiglinewithargsret{\sphinxbfcode{\sphinxupquote{setProtein}}}{\sphinxparam{\DUrole{n,n}{protein}}}{}
\pysigstopsignatures
\sphinxAtStartPar
Sets the protein associated with this set of genes

\end{fulllineitems}

\index{setTree() (GeneProteinAssociation method)@\spxentry{setTree()}\spxextra{GeneProteinAssociation method}}

\begin{fulllineitems}
\phantomsection\label{\detokenize{modules_doc:cbmpy.CBModel.GeneProteinAssociation.setTree}}
\pysigstartsignatures
\pysiglinewithargsret{\sphinxbfcode{\sphinxupquote{setTree}}}{\sphinxparam{\DUrole{n,n}{tree}}}{}
\pysigstopsignatures
\sphinxAtStartPar
Add a GPR dictionary/tree representation to the GPR.
\begin{itemize}
\item {} 
\sphinxAtStartPar
\sphinxstyleemphasis{tree} a dictionary representation of a GPR.

\end{itemize}

\end{fulllineitems}


\end{fulllineitems}

\index{Group (class in cbmpy.CBModel)@\spxentry{Group}\spxextra{class in cbmpy.CBModel}}

\begin{fulllineitems}
\phantomsection\label{\detokenize{modules_doc:cbmpy.CBModel.Group}}
\pysigstartsignatures
\pysiglinewithargsret{\sphinxbfcode{\sphinxupquote{class\DUrole{w,w}{  }}}\sphinxbfcode{\sphinxupquote{Group}}}{\sphinxparam{\DUrole{n,n}{pid}}}{}
\pysigstopsignatures
\sphinxAtStartPar
Container for SBML groups
\index{addMember() (Group method)@\spxentry{addMember()}\spxextra{Group method}}

\begin{fulllineitems}
\phantomsection\label{\detokenize{modules_doc:cbmpy.CBModel.Group.addMember}}
\pysigstartsignatures
\pysiglinewithargsret{\sphinxbfcode{\sphinxupquote{addMember}}}{\sphinxparam{\DUrole{n,n}{obj}}}{}
\pysigstopsignatures
\sphinxAtStartPar
Add member CBMPy object(s) to the group
\begin{itemize}
\item {} 
\sphinxAtStartPar
\sphinxstyleemphasis{obj} either a single, tuple or list of CBMPy objects

\end{itemize}

\end{fulllineitems}

\index{addSharedMIRIAMannotation() (Group method)@\spxentry{addSharedMIRIAMannotation()}\spxextra{Group method}}

\begin{fulllineitems}
\phantomsection\label{\detokenize{modules_doc:cbmpy.CBModel.Group.addSharedMIRIAMannotation}}
\pysigstartsignatures
\pysiglinewithargsret{\sphinxbfcode{\sphinxupquote{addSharedMIRIAMannotation}}}{\sphinxparam{\DUrole{n,n}{qual}}\sphinxparamcomma \sphinxparam{\DUrole{n,n}{entity}}\sphinxparamcomma \sphinxparam{\DUrole{n,n}{mid}}}{}
\pysigstopsignatures
\sphinxAtStartPar
Add a qualified MIRIAM annotation or entity to the list of members (all) rather than the group itself:
\begin{itemize}
\item {} 
\sphinxAtStartPar
\sphinxstyleemphasis{qual} a Biomodels biological qualifier e.g. “is” “isEncodedBy”

\item {} 
\sphinxAtStartPar
\sphinxstyleemphasis{entity} a MIRIAM resource entity e.g. “ChEBI”

\item {} 
\sphinxAtStartPar
\sphinxstyleemphasis{mid} the entity id e.g. CHEBI:17158 or fully qualifies url (if only\_qual\_uri)

\end{itemize}

\end{fulllineitems}

\index{assignAllSharedPropertiesToMembers() (Group method)@\spxentry{assignAllSharedPropertiesToMembers()}\spxextra{Group method}}

\begin{fulllineitems}
\phantomsection\label{\detokenize{modules_doc:cbmpy.CBModel.Group.assignAllSharedPropertiesToMembers}}
\pysigstartsignatures
\pysiglinewithargsret{\sphinxbfcode{\sphinxupquote{assignAllSharedPropertiesToMembers}}}{\sphinxparam{\DUrole{n,n}{overwrite}\DUrole{o,o}{=}\DUrole{default_value}{False}}}{}
\pysigstopsignatures
\sphinxAtStartPar
Assigns all group shared properties (notes, annotations, MIRIAM annotations, SBO) to the group members.
\begin{itemize}
\item {} 
\sphinxAtStartPar
\sphinxstyleemphasis{overwrite} {[}default=False{]} overwrite the target notes if they are defined

\end{itemize}

\end{fulllineitems}

\index{assignSharedAnnotationToMembers() (Group method)@\spxentry{assignSharedAnnotationToMembers()}\spxextra{Group method}}

\begin{fulllineitems}
\phantomsection\label{\detokenize{modules_doc:cbmpy.CBModel.Group.assignSharedAnnotationToMembers}}
\pysigstartsignatures
\pysiglinewithargsret{\sphinxbfcode{\sphinxupquote{assignSharedAnnotationToMembers}}}{}{}
\pysigstopsignatures
\sphinxAtStartPar
This function merges or updates the group member objects annotations with the group shared annotation.

\end{fulllineitems}

\index{assignSharedMIRIAMannotationToMembers() (Group method)@\spxentry{assignSharedMIRIAMannotationToMembers()}\spxextra{Group method}}

\begin{fulllineitems}
\phantomsection\label{\detokenize{modules_doc:cbmpy.CBModel.Group.assignSharedMIRIAMannotationToMembers}}
\pysigstartsignatures
\pysiglinewithargsret{\sphinxbfcode{\sphinxupquote{assignSharedMIRIAMannotationToMembers}}}{}{}
\pysigstopsignatures
\sphinxAtStartPar
This function merges or updates the group member objects MIRIAM annotations with the group shared MIRIAM annotation.

\end{fulllineitems}

\index{assignSharedNotesToMembers() (Group method)@\spxentry{assignSharedNotesToMembers()}\spxextra{Group method}}

\begin{fulllineitems}
\phantomsection\label{\detokenize{modules_doc:cbmpy.CBModel.Group.assignSharedNotesToMembers}}
\pysigstartsignatures
\pysiglinewithargsret{\sphinxbfcode{\sphinxupquote{assignSharedNotesToMembers}}}{\sphinxparam{\DUrole{n,n}{overwrite}\DUrole{o,o}{=}\DUrole{default_value}{False}}}{}
\pysigstopsignatures
\sphinxAtStartPar
Assigns the group shared notes to the group members.
\begin{itemize}
\item {} 
\sphinxAtStartPar
\sphinxstyleemphasis{overwrite} {[}default=False{]} overwrite the target notes if they are defined

\end{itemize}

\end{fulllineitems}

\index{assignSharedSBOtermsToMembers() (Group method)@\spxentry{assignSharedSBOtermsToMembers()}\spxextra{Group method}}

\begin{fulllineitems}
\phantomsection\label{\detokenize{modules_doc:cbmpy.CBModel.Group.assignSharedSBOtermsToMembers}}
\pysigstartsignatures
\pysiglinewithargsret{\sphinxbfcode{\sphinxupquote{assignSharedSBOtermsToMembers}}}{\sphinxparam{\DUrole{n,n}{overwrite}\DUrole{o,o}{=}\DUrole{default_value}{False}}}{}
\pysigstopsignatures
\sphinxAtStartPar
Assigns the group shared member SBO term to the group members.
\begin{itemize}
\item {} 
\sphinxAtStartPar
\sphinxstyleemphasis{overwrite} {[}default=False{]} overwrite the target SBO term if it is defined

\end{itemize}

\end{fulllineitems}

\index{clone() (Group method)@\spxentry{clone()}\spxextra{Group method}}

\begin{fulllineitems}
\phantomsection\label{\detokenize{modules_doc:cbmpy.CBModel.Group.clone}}
\pysigstartsignatures
\pysiglinewithargsret{\sphinxbfcode{\sphinxupquote{clone}}}{}{}
\pysigstopsignatures
\sphinxAtStartPar
Return a clone of this object. Note the for Groups this is a shallow copy, in that the reference
objects themselves are not cloned only the group (and attributes)

\end{fulllineitems}

\index{deleteMember() (Group method)@\spxentry{deleteMember()}\spxextra{Group method}}

\begin{fulllineitems}
\phantomsection\label{\detokenize{modules_doc:cbmpy.CBModel.Group.deleteMember}}
\pysigstartsignatures
\pysiglinewithargsret{\sphinxbfcode{\sphinxupquote{deleteMember}}}{\sphinxparam{\DUrole{n,n}{oid}}}{}
\pysigstopsignatures
\sphinxAtStartPar
Deletes a group member with group id.
\begin{itemize}
\item {} 
\sphinxAtStartPar
\sphinxstyleemphasis{oid} group member id

\end{itemize}

\end{fulllineitems}

\index{getKind() (Group method)@\spxentry{getKind()}\spxextra{Group method}}

\begin{fulllineitems}
\phantomsection\label{\detokenize{modules_doc:cbmpy.CBModel.Group.getKind}}
\pysigstartsignatures
\pysiglinewithargsret{\sphinxbfcode{\sphinxupquote{getKind}}}{}{}
\pysigstopsignatures
\sphinxAtStartPar
Return the group kind

\end{fulllineitems}

\index{getMember() (Group method)@\spxentry{getMember()}\spxextra{Group method}}

\begin{fulllineitems}
\phantomsection\label{\detokenize{modules_doc:cbmpy.CBModel.Group.getMember}}
\pysigstartsignatures
\pysiglinewithargsret{\sphinxbfcode{\sphinxupquote{getMember}}}{\sphinxparam{\DUrole{n,n}{mid}}}{}
\pysigstopsignatures
\sphinxAtStartPar
Returns the group member object that corresponds to mid
\begin{itemize}
\item {} 
\sphinxAtStartPar
\sphinxstyleemphasis{mid} the id of the requested object

\end{itemize}

\end{fulllineitems}

\index{getMemberIDs() (Group method)@\spxentry{getMemberIDs()}\spxextra{Group method}}

\begin{fulllineitems}
\phantomsection\label{\detokenize{modules_doc:cbmpy.CBModel.Group.getMemberIDs}}
\pysigstartsignatures
\pysiglinewithargsret{\sphinxbfcode{\sphinxupquote{getMemberIDs}}}{\sphinxparam{\DUrole{n,n}{as\_set}\DUrole{o,o}{=}\DUrole{default_value}{False}}}{}
\pysigstopsignatures
\sphinxAtStartPar
Return the ids of the member objects.
\begin{itemize}
\item {} 
\sphinxAtStartPar
\sphinxstyleemphasis{as\_set} return id’s as a set rather than a list

\end{itemize}

\end{fulllineitems}

\index{getMembers() (Group method)@\spxentry{getMembers()}\spxextra{Group method}}

\begin{fulllineitems}
\phantomsection\label{\detokenize{modules_doc:cbmpy.CBModel.Group.getMembers}}
\pysigstartsignatures
\pysiglinewithargsret{\sphinxbfcode{\sphinxupquote{getMembers}}}{\sphinxparam{\DUrole{n,n}{as\_set}\DUrole{o,o}{=}\DUrole{default_value}{False}}}{}
\pysigstopsignatures
\sphinxAtStartPar
Return the member objects of the group.
\begin{itemize}
\item {} 
\sphinxAtStartPar
\sphinxstyleemphasis{as\_set} return objects as a set rather than a list

\end{itemize}

\end{fulllineitems}

\index{getSharedAnnotations() (Group method)@\spxentry{getSharedAnnotations()}\spxextra{Group method}}

\begin{fulllineitems}
\phantomsection\label{\detokenize{modules_doc:cbmpy.CBModel.Group.getSharedAnnotations}}
\pysigstartsignatures
\pysiglinewithargsret{\sphinxbfcode{\sphinxupquote{getSharedAnnotations}}}{}{}
\pysigstopsignatures
\sphinxAtStartPar
Return a dictionary of the shared member annotations (rather than the group attribute).

\end{fulllineitems}

\index{getSharedMIRIAMannotations() (Group method)@\spxentry{getSharedMIRIAMannotations()}\spxextra{Group method}}

\begin{fulllineitems}
\phantomsection\label{\detokenize{modules_doc:cbmpy.CBModel.Group.getSharedMIRIAMannotations}}
\pysigstartsignatures
\pysiglinewithargsret{\sphinxbfcode{\sphinxupquote{getSharedMIRIAMannotations}}}{}{}
\pysigstopsignatures
\sphinxAtStartPar
Return a dictionary of the shared member MIRIAM annotations (rather than the group attribute).

\end{fulllineitems}

\index{getSharedNotes() (Group method)@\spxentry{getSharedNotes()}\spxextra{Group method}}

\begin{fulllineitems}
\phantomsection\label{\detokenize{modules_doc:cbmpy.CBModel.Group.getSharedNotes}}
\pysigstartsignatures
\pysiglinewithargsret{\sphinxbfcode{\sphinxupquote{getSharedNotes}}}{}{}
\pysigstopsignatures
\sphinxAtStartPar
Return the shared member notes (rather than the group attribute).

\end{fulllineitems}

\index{getSharedSBOterm() (Group method)@\spxentry{getSharedSBOterm()}\spxextra{Group method}}

\begin{fulllineitems}
\phantomsection\label{\detokenize{modules_doc:cbmpy.CBModel.Group.getSharedSBOterm}}
\pysigstartsignatures
\pysiglinewithargsret{\sphinxbfcode{\sphinxupquote{getSharedSBOterm}}}{}{}
\pysigstopsignatures
\sphinxAtStartPar
Return the shared member SBO term (rather than the group attribute).

\end{fulllineitems}

\index{hasMember() (Group method)@\spxentry{hasMember()}\spxextra{Group method}}

\begin{fulllineitems}
\phantomsection\label{\detokenize{modules_doc:cbmpy.CBModel.Group.hasMember}}
\pysigstartsignatures
\pysiglinewithargsret{\sphinxbfcode{\sphinxupquote{hasMember}}}{\sphinxparam{\DUrole{n,n}{mid}}}{}
\pysigstopsignatures
\sphinxAtStartPar
Returns a boolean indicating whether a member is in the group.
\begin{itemize}
\item {} 
\sphinxAtStartPar
\sphinxstyleemphasis{mid} the id to check

\end{itemize}

\end{fulllineitems}

\index{serialize() (Group method)@\spxentry{serialize()}\spxextra{Group method}}

\begin{fulllineitems}
\phantomsection\label{\detokenize{modules_doc:cbmpy.CBModel.Group.serialize}}
\pysigstartsignatures
\pysiglinewithargsret{\sphinxbfcode{\sphinxupquote{serialize}}}{\sphinxparam{\DUrole{n,n}{protocol}\DUrole{o,o}{=}\DUrole{default_value}{0}}}{}
\pysigstopsignatures
\sphinxAtStartPar
Serialize object, returns a string by default
\begin{itemize}
\item {} \begin{description}
\sphinxlineitem{\sphinxstyleemphasis{protocol} {[}default=0{]} serialize to a string or binary if required,}
\sphinxAtStartPar
see pickle module documentation for details

\end{description}

\end{itemize}

\sphinxAtStartPar
\# Reimplemented in Model

\end{fulllineitems}

\index{serializeToDisk() (Group method)@\spxentry{serializeToDisk()}\spxextra{Group method}}

\begin{fulllineitems}
\phantomsection\label{\detokenize{modules_doc:cbmpy.CBModel.Group.serializeToDisk}}
\pysigstartsignatures
\pysiglinewithargsret{\sphinxbfcode{\sphinxupquote{serializeToDisk}}}{\sphinxparam{\DUrole{n,n}{filename}}\sphinxparamcomma \sphinxparam{\DUrole{n,n}{protocol}\DUrole{o,o}{=}\DUrole{default_value}{2}}}{}
\pysigstopsignatures
\sphinxAtStartPar
Serialize to disk using pickle protocol:
\begin{itemize}
\item {} 
\sphinxAtStartPar
\sphinxstyleemphasis{filename} the name of the output file

\item {} \begin{description}
\sphinxlineitem{\sphinxstyleemphasis{protocol} {[}default=2{]} serialize to a string or binary if required,}
\sphinxAtStartPar
see pickle module documentation for details

\end{description}

\end{itemize}

\sphinxAtStartPar
\# Reimplemented in Model

\end{fulllineitems}

\index{setKind() (Group method)@\spxentry{setKind()}\spxextra{Group method}}

\begin{fulllineitems}
\phantomsection\label{\detokenize{modules_doc:cbmpy.CBModel.Group.setKind}}
\pysigstartsignatures
\pysiglinewithargsret{\sphinxbfcode{\sphinxupquote{setKind}}}{\sphinxparam{\DUrole{n,n}{kind}}}{}
\pysigstopsignatures
\sphinxAtStartPar
Sets the kind or type of the group, this must be one of: ‘collection’, ‘partonomy’, ‘classification’.
\begin{itemize}
\item {} 
\sphinxAtStartPar
\sphinxstyleemphasis{kind} the kind

\end{itemize}

\end{fulllineitems}

\index{setSharedAnnotation() (Group method)@\spxentry{setSharedAnnotation()}\spxextra{Group method}}

\begin{fulllineitems}
\phantomsection\label{\detokenize{modules_doc:cbmpy.CBModel.Group.setSharedAnnotation}}
\pysigstartsignatures
\pysiglinewithargsret{\sphinxbfcode{\sphinxupquote{setSharedAnnotation}}}{\sphinxparam{\DUrole{n,n}{key}}\sphinxparamcomma \sphinxparam{\DUrole{n,n}{value}}}{}
\pysigstopsignatures
\sphinxAtStartPar
Sets the list of members (all) annotation as a key : value pair.
\begin{itemize}
\item {} 
\sphinxAtStartPar
\sphinxstyleemphasis{key} the annotation key

\item {} 
\sphinxAtStartPar
\sphinxstyleemphasis{value} the annotation value

\end{itemize}

\end{fulllineitems}

\index{setSharedNotes() (Group method)@\spxentry{setSharedNotes()}\spxextra{Group method}}

\begin{fulllineitems}
\phantomsection\label{\detokenize{modules_doc:cbmpy.CBModel.Group.setSharedNotes}}
\pysigstartsignatures
\pysiglinewithargsret{\sphinxbfcode{\sphinxupquote{setSharedNotes}}}{\sphinxparam{\DUrole{n,n}{notes}}}{}
\pysigstopsignatures
\sphinxAtStartPar
Sets the group of objects notes attribute (all):
\begin{itemize}
\item {} 
\sphinxAtStartPar
\sphinxstyleemphasis{notes} the note string, should preferably be (X)HTML for SBML

\end{itemize}

\end{fulllineitems}

\index{setSharedSBOterm() (Group method)@\spxentry{setSharedSBOterm()}\spxextra{Group method}}

\begin{fulllineitems}
\phantomsection\label{\detokenize{modules_doc:cbmpy.CBModel.Group.setSharedSBOterm}}
\pysigstartsignatures
\pysiglinewithargsret{\sphinxbfcode{\sphinxupquote{setSharedSBOterm}}}{\sphinxparam{\DUrole{n,n}{sbo}}}{}
\pysigstopsignatures
\sphinxAtStartPar
Set the SBO term for the the members of the group (all).
\begin{itemize}
\item {} 
\sphinxAtStartPar
\sphinxstyleemphasis{sbo} the SBOterm with format: “SBO:\textless{}7 digit integer\textgreater{}”

\end{itemize}

\end{fulllineitems}


\end{fulllineitems}

\index{GroupMemberAttributes (class in cbmpy.CBModel)@\spxentry{GroupMemberAttributes}\spxextra{class in cbmpy.CBModel}}

\begin{fulllineitems}
\phantomsection\label{\detokenize{modules_doc:cbmpy.CBModel.GroupMemberAttributes}}
\pysigstartsignatures
\pysigline{\sphinxbfcode{\sphinxupquote{class\DUrole{w,w}{  }}}\sphinxbfcode{\sphinxupquote{GroupMemberAttributes}}}
\pysigstopsignatures
\sphinxAtStartPar
Contains the shared attributes of the group members (equivalent to SBML annotation on ListOfMembers)

\end{fulllineitems}

\index{Model (class in cbmpy.CBModel)@\spxentry{Model}\spxextra{class in cbmpy.CBModel}}

\begin{fulllineitems}
\phantomsection\label{\detokenize{modules_doc:cbmpy.CBModel.Model}}
\pysigstartsignatures
\pysiglinewithargsret{\sphinxbfcode{\sphinxupquote{class\DUrole{w,w}{  }}}\sphinxbfcode{\sphinxupquote{Model}}}{\sphinxparam{\DUrole{n,n}{pid}}}{}
\pysigstopsignatures
\sphinxAtStartPar
Container for constraint based model, adds methods for manipulating:
\begin{itemize}
\item {} 
\sphinxAtStartPar
objectives

\item {} 
\sphinxAtStartPar
constraints

\item {} 
\sphinxAtStartPar
reactions

\item {} 
\sphinxAtStartPar
species

\item {} 
\sphinxAtStartPar
compartments

\item {} 
\sphinxAtStartPar
groups

\item {} 
\sphinxAtStartPar
parameters

\item {} 
\sphinxAtStartPar
N a structmatrix object

\end{itemize}
\index{addCompartment() (Model method)@\spxentry{addCompartment()}\spxextra{Model method}}

\begin{fulllineitems}
\phantomsection\label{\detokenize{modules_doc:cbmpy.CBModel.Model.addCompartment}}
\pysigstartsignatures
\pysiglinewithargsret{\sphinxbfcode{\sphinxupquote{addCompartment}}}{\sphinxparam{\DUrole{n,n}{comp}}}{}
\pysigstopsignatures
\sphinxAtStartPar
Add an instantiated Compartment object to the CBM model
\begin{itemize}
\item {} 
\sphinxAtStartPar
\sphinxstyleemphasis{comp} an instance of the Compartment class

\end{itemize}

\end{fulllineitems}

\index{addFluxBound() (Model method)@\spxentry{addFluxBound()}\spxextra{Model method}}

\begin{fulllineitems}
\phantomsection\label{\detokenize{modules_doc:cbmpy.CBModel.Model.addFluxBound}}
\pysigstartsignatures
\pysiglinewithargsret{\sphinxbfcode{\sphinxupquote{addFluxBound}}}{\sphinxparam{\DUrole{n,n}{fluxbound}}\sphinxparamcomma \sphinxparam{\DUrole{n,n}{fbexists}\DUrole{o,o}{=}\DUrole{default_value}{None}}}{}
\pysigstopsignatures
\sphinxAtStartPar
Add an instantiated FluxBound object to the FBA model
\begin{itemize}
\item {} 
\sphinxAtStartPar
\sphinxstyleemphasis{fluxbound} an instance of the FluxBound class

\item {} 
\sphinxAtStartPar
\sphinxstyleemphasis{fbexists} {[}default=None{]} this is a list of strings which contains fluxbounds that have been added to the model, see sample code below.

\end{itemize}

\sphinxAtStartPar
The format of the string is ‘reactionid\_boundtype’

\sphinxAtStartPar
{\color{red}\bfseries{}\textasciigrave{}\textasciigrave{}}{\color{red}\bfseries{}\textasciigrave{}}python
fbexists = {[}{]}
for fluxbound in list\_of\_fluxbounds:
\begin{quote}

\sphinxAtStartPar
model.addFluxBound(fluxbound, fbexists=fbexists)
fbexists.append(“\{\}\_\{\}”.format(fluxbound.getReactionId(), fluxbound.getType()))
\end{quote}

\sphinxAtStartPar
{\color{red}\bfseries{}\textasciigrave{}\textasciigrave{}}{\color{red}\bfseries{}\textasciigrave{}}

\sphinxAtStartPar
Using the fbexists list drastically reduces the time it takes to add fluxbounds but circumvents any sort of existence checking and should \sphinxstyleemphasis{only} be
used as shown above when constructing a model from scratch. Alternativel, you need to prepopulate fbexists with existing fluxbound component information:

\sphinxAtStartPar
\sphinxcode{\sphinxupquote{\textasciigrave{}python
fbexists = {[}"\{\}\_\{\}".format(fluxbound.getReactionId(), fluxbound.getType()) for fluxbound in model.flux\_bounds{]}
\textasciigrave{}}}

\end{fulllineitems}

\index{addGPRAssociation() (Model method)@\spxentry{addGPRAssociation()}\spxextra{Model method}}

\begin{fulllineitems}
\phantomsection\label{\detokenize{modules_doc:cbmpy.CBModel.Model.addGPRAssociation}}
\pysigstartsignatures
\pysiglinewithargsret{\sphinxbfcode{\sphinxupquote{addGPRAssociation}}}{\sphinxparam{\DUrole{n,n}{gpr}}\sphinxparamcomma \sphinxparam{\DUrole{n,n}{update\_idx}\DUrole{o,o}{=}\DUrole{default_value}{True}}}{}
\pysigstopsignatures
\sphinxAtStartPar
Add a GeneProteinAssociation instance to the model
\begin{itemize}
\item {} 
\sphinxAtStartPar
\sphinxstyleemphasis{gpr} an instantiated GeneProteinAssociation object

\end{itemize}

\end{fulllineitems}

\index{addGene() (Model method)@\spxentry{addGene()}\spxextra{Model method}}

\begin{fulllineitems}
\phantomsection\label{\detokenize{modules_doc:cbmpy.CBModel.Model.addGene}}
\pysigstartsignatures
\pysiglinewithargsret{\sphinxbfcode{\sphinxupquote{addGene}}}{\sphinxparam{\DUrole{n,n}{gene}}}{}
\pysigstopsignatures
\sphinxAtStartPar
Add an instantiated Gene object to the FBA model
\begin{itemize}
\item {} 
\sphinxAtStartPar
\sphinxstyleemphasis{gene} an instance of the G class

\end{itemize}

\end{fulllineitems}

\index{addGroup() (Model method)@\spxentry{addGroup()}\spxextra{Model method}}

\begin{fulllineitems}
\phantomsection\label{\detokenize{modules_doc:cbmpy.CBModel.Model.addGroup}}
\pysigstartsignatures
\pysiglinewithargsret{\sphinxbfcode{\sphinxupquote{addGroup}}}{\sphinxparam{\DUrole{n,n}{obj}}}{}
\pysigstopsignatures
\sphinxAtStartPar
Add an instantiated group object to the model
\begin{itemize}
\item {} 
\sphinxAtStartPar
\sphinxstyleemphasis{obj} the Group instance

\end{itemize}

\end{fulllineitems}

\index{addMIRIAMannotation() (Model method)@\spxentry{addMIRIAMannotation()}\spxextra{Model method}}

\begin{fulllineitems}
\phantomsection\label{\detokenize{modules_doc:cbmpy.CBModel.Model.addMIRIAMannotation}}
\pysigstartsignatures
\pysiglinewithargsret{\sphinxbfcode{\sphinxupquote{addMIRIAMannotation}}}{\sphinxparam{\DUrole{n,n}{qual}}\sphinxparamcomma \sphinxparam{\DUrole{n,n}{entity}}\sphinxparamcomma \sphinxparam{\DUrole{n,n}{mid}}}{}
\pysigstopsignatures
\sphinxAtStartPar
Add a qualified MIRIAM annotation or entity:
\begin{itemize}
\item {} 
\sphinxAtStartPar
\sphinxstyleemphasis{qual} a Biomodels biological qualifier e.g. “is” “isEncodedBy”

\item {} 
\sphinxAtStartPar
\sphinxstyleemphasis{entity} a MIRIAM resource entity e.g. “ChEBI”

\item {} 
\sphinxAtStartPar
\sphinxstyleemphasis{mid} the entity id e.g. CHEBI:17158

\end{itemize}

\end{fulllineitems}

\index{addModelCreator() (Model method)@\spxentry{addModelCreator()}\spxextra{Model method}}

\begin{fulllineitems}
\phantomsection\label{\detokenize{modules_doc:cbmpy.CBModel.Model.addModelCreator}}
\pysigstartsignatures
\pysiglinewithargsret{\sphinxbfcode{\sphinxupquote{addModelCreator}}}{\sphinxparam{\DUrole{n,n}{firstname}}\sphinxparamcomma \sphinxparam{\DUrole{n,n}{lastname}}\sphinxparamcomma \sphinxparam{\DUrole{n,n}{organisation}\DUrole{o,o}{=}\DUrole{default_value}{None}}\sphinxparamcomma \sphinxparam{\DUrole{n,n}{email}\DUrole{o,o}{=}\DUrole{default_value}{None}}}{}
\pysigstopsignatures
\sphinxAtStartPar
Add a model creator to the list of model creators, only the first and fmaily names are mandatory:
\begin{itemize}
\item {} 
\sphinxAtStartPar
\sphinxstyleemphasis{firstname}

\item {} 
\sphinxAtStartPar
\sphinxstyleemphasis{lastname}

\item {} 
\sphinxAtStartPar
\sphinxstyleemphasis{organisation} {[}default=None{]}

\item {} 
\sphinxAtStartPar
\sphinxstyleemphasis{email}  {[}default=None{]}

\end{itemize}

\end{fulllineitems}

\index{addObjective() (Model method)@\spxentry{addObjective()}\spxextra{Model method}}

\begin{fulllineitems}
\phantomsection\label{\detokenize{modules_doc:cbmpy.CBModel.Model.addObjective}}
\pysigstartsignatures
\pysiglinewithargsret{\sphinxbfcode{\sphinxupquote{addObjective}}}{\sphinxparam{\DUrole{n,n}{obj}}\sphinxparamcomma \sphinxparam{\DUrole{n,n}{active}\DUrole{o,o}{=}\DUrole{default_value}{False}}}{}
\pysigstopsignatures
\sphinxAtStartPar
Add an instantiated Objective object to the FBA model
\begin{itemize}
\item {} 
\sphinxAtStartPar
\sphinxstyleemphasis{obj} an instance of the Objective class

\item {} 
\sphinxAtStartPar
\sphinxstyleemphasis{active} {[}default=False{]} flag this objective as the active objective (fba.activeObjIdx)

\end{itemize}

\end{fulllineitems}

\index{addParameter() (Model method)@\spxentry{addParameter()}\spxextra{Model method}}

\begin{fulllineitems}
\phantomsection\label{\detokenize{modules_doc:cbmpy.CBModel.Model.addParameter}}
\pysigstartsignatures
\pysiglinewithargsret{\sphinxbfcode{\sphinxupquote{addParameter}}}{\sphinxparam{\DUrole{n,n}{par}}}{}
\pysigstopsignatures
\sphinxAtStartPar
Add an instantiated Parameter object to the model
\begin{itemize}
\item {} 
\sphinxAtStartPar
\sphinxstyleemphasis{par} an instance of the Parameter class

\end{itemize}

\end{fulllineitems}

\index{addReaction() (Model method)@\spxentry{addReaction()}\spxextra{Model method}}

\begin{fulllineitems}
\phantomsection\label{\detokenize{modules_doc:cbmpy.CBModel.Model.addReaction}}
\pysigstartsignatures
\pysiglinewithargsret{\sphinxbfcode{\sphinxupquote{addReaction}}}{\sphinxparam{\DUrole{n,n}{reaction}}\sphinxparamcomma \sphinxparam{\DUrole{n,n}{create\_default\_bounds}\DUrole{o,o}{=}\DUrole{default_value}{False}}\sphinxparamcomma \sphinxparam{\DUrole{n,n}{silent}\DUrole{o,o}{=}\DUrole{default_value}{False}}}{}
\pysigstopsignatures
\sphinxAtStartPar
Adds a reaction object to the model
\begin{itemize}
\item {} 
\sphinxAtStartPar
\sphinxstyleemphasis{reaction} an instance of the Reaction class

\item {} 
\sphinxAtStartPar
\sphinxstyleemphasis{create\_default\_bounds} create default reaction bounds, irreversible 0 \textless{}= J \textless{}= INF, reversable \sphinxhyphen{}INF \textless{}= J \textless{}= INF

\item {} 
\sphinxAtStartPar
\sphinxstyleemphasis{silent} {[}default=False{]} if enabled this disables the printing of information messages

\end{itemize}

\end{fulllineitems}

\index{addSpecies() (Model method)@\spxentry{addSpecies()}\spxextra{Model method}}

\begin{fulllineitems}
\phantomsection\label{\detokenize{modules_doc:cbmpy.CBModel.Model.addSpecies}}
\pysigstartsignatures
\pysiglinewithargsret{\sphinxbfcode{\sphinxupquote{addSpecies}}}{\sphinxparam{\DUrole{n,n}{species}}}{}
\pysigstopsignatures
\sphinxAtStartPar
Add an instantiated Species object to the FBA model
\begin{itemize}
\item {} 
\sphinxAtStartPar
\sphinxstyleemphasis{species} an instance of the Species class

\end{itemize}

\end{fulllineitems}

\index{addUserConstraint() (Model method)@\spxentry{addUserConstraint()}\spxextra{Model method}}

\begin{fulllineitems}
\phantomsection\label{\detokenize{modules_doc:cbmpy.CBModel.Model.addUserConstraint}}
\pysigstartsignatures
\pysiglinewithargsret{\sphinxbfcode{\sphinxupquote{addUserConstraint}}}{\sphinxparam{\DUrole{n,n}{pid}}\sphinxparamcomma \sphinxparam{\DUrole{n,n}{fluxes}\DUrole{o,o}{=}\DUrole{default_value}{None}}\sphinxparamcomma \sphinxparam{\DUrole{n,n}{operator}\DUrole{o,o}{=}\DUrole{default_value}{\textquotesingle{}\textgreater{}=\textquotesingle{}}}\sphinxparamcomma \sphinxparam{\DUrole{n,n}{rhs}\DUrole{o,o}{=}\DUrole{default_value}{0.0}}}{}
\pysigstopsignatures
\sphinxAtStartPar
Add a user defined constraint to FBA model, this is additional to the automatically determined Stoichiometric constraints.
\begin{itemize}
\item {} 
\sphinxAtStartPar
\sphinxstyleemphasis{pid} user constraint name/id, use \sphinxtitleref{None} for auto\sphinxhyphen{}assign

\item {} 
\sphinxAtStartPar
\sphinxstyleemphasis{fluxes} a list of (coefficient, reaction id) pairs where coefficient is a float

\item {} 
\sphinxAtStartPar
\sphinxstyleemphasis{operator} is one of ‘=’, ‘\textgreater{}=’ or ‘\textless{}=’ (\textless{} and \textgreater{} will be interpreted as \textgreater{}= or \textless{}=)

\item {} 
\sphinxAtStartPar
\sphinxstyleemphasis{rhs} a float

\end{itemize}

\end{fulllineitems}

\index{addUserDefinedConstraint() (Model method)@\spxentry{addUserDefinedConstraint()}\spxextra{Model method}}

\begin{fulllineitems}
\phantomsection\label{\detokenize{modules_doc:cbmpy.CBModel.Model.addUserDefinedConstraint}}
\pysigstartsignatures
\pysiglinewithargsret{\sphinxbfcode{\sphinxupquote{addUserDefinedConstraint}}}{\sphinxparam{\DUrole{n,n}{udc}}}{}
\pysigstopsignatures
\sphinxAtStartPar
Add a  User Defined Constraint object to the FBA model
\begin{itemize}
\item {} 
\sphinxAtStartPar
\sphinxstyleemphasis{obj} an instance of the UserDefinedConstraint class

\end{itemize}

\end{fulllineitems}

\index{buildStoichMatrix() (Model method)@\spxentry{buildStoichMatrix()}\spxextra{Model method}}

\begin{fulllineitems}
\phantomsection\label{\detokenize{modules_doc:cbmpy.CBModel.Model.buildStoichMatrix}}
\pysigstartsignatures
\pysiglinewithargsret{\sphinxbfcode{\sphinxupquote{buildStoichMatrix}}}{\sphinxparam{\DUrole{n,n}{matrix\_type}\DUrole{o,o}{=}\DUrole{default_value}{\textquotesingle{}numpy\textquotesingle{}}}\sphinxparamcomma \sphinxparam{\DUrole{n,n}{only\_return}\DUrole{o,o}{=}\DUrole{default_value}{False}}}{}
\pysigstopsignatures
\sphinxAtStartPar
Build the stoichiometric matrix N and additional constraint matrix CN (if required)
\begin{itemize}
\item {} 
\sphinxAtStartPar
\sphinxstyleemphasis{matrix\_type} {[}default=’numpy’{]} the type of matrix to use to generate constraints
\begin{itemize}
\item {} 
\sphinxAtStartPar
\sphinxstyleemphasis{numpy} a NumPy matrix default

\item {} 
\sphinxAtStartPar
\sphinxstyleemphasis{sympy} a SymPy symbolic matrix, if available note the denominator limit can be set in \sphinxcode{\sphinxupquote{CBModel.\_\_CBCONFIG\_\_{[}\textquotesingle{}SYMPY\_DENOM\_LIMIT\textquotesingle{}{]} = 10**12}}

\item {} 
\sphinxAtStartPar
\sphinxstyleemphasis{scipy\_csr} create using NumPy but store as SciPy csr\_sparse

\end{itemize}

\end{itemize}
\begin{itemize}
\item {} 
\sphinxAtStartPar
\sphinxstyleemphasis{only\_return} {[}default=False{]} \sphinxstylestrong{IMPORTANT} only returns the stoichiometric matrix and constraint matrix (if required),
does not update the model

\end{itemize}

\end{fulllineitems}

\index{changeAllFluxBoundsWithValue() (Model method)@\spxentry{changeAllFluxBoundsWithValue()}\spxextra{Model method}}

\begin{fulllineitems}
\phantomsection\label{\detokenize{modules_doc:cbmpy.CBModel.Model.changeAllFluxBoundsWithValue}}
\pysigstartsignatures
\pysiglinewithargsret{\sphinxbfcode{\sphinxupquote{changeAllFluxBoundsWithValue}}}{\sphinxparam{\DUrole{n,n}{old}}\sphinxparamcomma \sphinxparam{\DUrole{n,n}{new}}}{}
\pysigstopsignatures
\sphinxAtStartPar
Replaces all flux bounds with value “old” with a new value “new”:
\begin{itemize}
\item {} 
\sphinxAtStartPar
\sphinxstyleemphasis{old} value

\item {} 
\sphinxAtStartPar
\sphinxstyleemphasis{new} value

\end{itemize}

\end{fulllineitems}

\index{clone() (Model method)@\spxentry{clone()}\spxextra{Model method}}

\begin{fulllineitems}
\phantomsection\label{\detokenize{modules_doc:cbmpy.CBModel.Model.clone}}
\pysigstartsignatures
\pysiglinewithargsret{\sphinxbfcode{\sphinxupquote{clone}}}{}{}
\pysigstopsignatures
\sphinxAtStartPar
Return a clone of this object.

\end{fulllineitems}

\index{convertUserConstraintsToUserDefinedConstraints() (Model method)@\spxentry{convertUserConstraintsToUserDefinedConstraints()}\spxextra{Model method}}

\begin{fulllineitems}
\phantomsection\label{\detokenize{modules_doc:cbmpy.CBModel.Model.convertUserConstraintsToUserDefinedConstraints}}
\pysigstartsignatures
\pysiglinewithargsret{\sphinxbfcode{\sphinxupquote{convertUserConstraintsToUserDefinedConstraints}}}{}{}
\pysigstopsignatures
\sphinxAtStartPar
If a model is loaded with the old CBMPy specific constraint data structures json files and dictionaries, this function will
upmark it to the new FBCv3 data structures

\end{fulllineitems}

\index{copyUserDefinedConstraintsToUserConstraints() (Model method)@\spxentry{copyUserDefinedConstraintsToUserConstraints()}\spxextra{Model method}}

\begin{fulllineitems}
\phantomsection\label{\detokenize{modules_doc:cbmpy.CBModel.Model.copyUserDefinedConstraintsToUserConstraints}}
\pysigstartsignatures
\pysiglinewithargsret{\sphinxbfcode{\sphinxupquote{copyUserDefinedConstraintsToUserConstraints}}}{}{}
\pysigstopsignatures
\sphinxAtStartPar
This is a workaround until I complete full UserDefinedConstraints support

\end{fulllineitems}

\index{createCompartment() (Model method)@\spxentry{createCompartment()}\spxextra{Model method}}

\begin{fulllineitems}
\phantomsection\label{\detokenize{modules_doc:cbmpy.CBModel.Model.createCompartment}}
\pysigstartsignatures
\pysiglinewithargsret{\sphinxbfcode{\sphinxupquote{createCompartment}}}{\sphinxparam{\DUrole{n,n}{cid}}\sphinxparamcomma \sphinxparam{\DUrole{n,n}{name}\DUrole{o,o}{=}\DUrole{default_value}{None}}\sphinxparamcomma \sphinxparam{\DUrole{n,n}{size}\DUrole{o,o}{=}\DUrole{default_value}{1}}\sphinxparamcomma \sphinxparam{\DUrole{n,n}{dimensions}\DUrole{o,o}{=}\DUrole{default_value}{3}}\sphinxparamcomma \sphinxparam{\DUrole{n,n}{volume}\DUrole{o,o}{=}\DUrole{default_value}{None}}}{}
\pysigstopsignatures
\sphinxAtStartPar
Create a new compartment and add it to the model if the id does not exist
\begin{itemize}
\item {} 
\sphinxAtStartPar
\sphinxstyleemphasis{cid} compartment id

\item {} 
\sphinxAtStartPar
\sphinxstyleemphasis{name} {[}None{]} compartment name

\item {} 
\sphinxAtStartPar
\sphinxstyleemphasis{size} {[}1{]} compartment size

\item {} 
\sphinxAtStartPar
\sphinxstyleemphasis{dimensions} {[}3{]} compartment size dimensions

\item {} 
\sphinxAtStartPar
\sphinxstyleemphasis{volume} {[}None{]} compartment volume

\end{itemize}

\end{fulllineitems}

\index{createGeneAssociationsFromAnnotations() (Model method)@\spxentry{createGeneAssociationsFromAnnotations()}\spxextra{Model method}}

\begin{fulllineitems}
\phantomsection\label{\detokenize{modules_doc:cbmpy.CBModel.Model.createGeneAssociationsFromAnnotations}}
\pysigstartsignatures
\pysiglinewithargsret{\sphinxbfcode{\sphinxupquote{createGeneAssociationsFromAnnotations}}}{\sphinxparam{\DUrole{n,n}{annotation\_key}\DUrole{o,o}{=}\DUrole{default_value}{\textquotesingle{}GENE ASSOCIATION\textquotesingle{}}}\sphinxparamcomma \sphinxparam{\DUrole{n,n}{replace\_existing}\DUrole{o,o}{=}\DUrole{default_value}{True}}}{}
\pysigstopsignatures
\sphinxAtStartPar
Add genes to the model using the definitions stored in the annotation key. If this fails it tries some standard annotation
keys: GENE ASSOCIATION, GENE\_ASSOCIATION, gene\_association, gene association.
\begin{itemize}
\item {} 
\sphinxAtStartPar
\sphinxstyleemphasis{annotation\_key} the annotation dictionary key that holds the gene association for the protein/enzyme

\item {} 
\sphinxAtStartPar
\sphinxstyleemphasis{replace\_existing} {[}default=True{]} replace existing annotations, otherwise only new ones are added

\end{itemize}

\end{fulllineitems}

\index{createGeneProteinAssociation() (Model method)@\spxentry{createGeneProteinAssociation()}\spxextra{Model method}}

\begin{fulllineitems}
\phantomsection\label{\detokenize{modules_doc:cbmpy.CBModel.Model.createGeneProteinAssociation}}
\pysigstartsignatures
\pysiglinewithargsret{\sphinxbfcode{\sphinxupquote{createGeneProteinAssociation}}}{\sphinxparam{\DUrole{n,n}{protein}}\sphinxparamcomma \sphinxparam{\DUrole{n,n}{assoc}}\sphinxparamcomma \sphinxparam{\DUrole{n,n}{gid}\DUrole{o,o}{=}\DUrole{default_value}{None}}\sphinxparamcomma \sphinxparam{\DUrole{n,n}{name}\DUrole{o,o}{=}\DUrole{default_value}{None}}\sphinxparamcomma \sphinxparam{\DUrole{n,n}{gene\_pattern}\DUrole{o,o}{=}\DUrole{default_value}{None}}\sphinxparamcomma \sphinxparam{\DUrole{n,n}{update\_idx}\DUrole{o,o}{=}\DUrole{default_value}{True}}\sphinxparamcomma \sphinxparam{\DUrole{n,n}{altlabels}\DUrole{o,o}{=}\DUrole{default_value}{None}}}{}
\pysigstopsignatures
\sphinxAtStartPar
Create and add a gene protein relationship to the model, note genes are mapped on protein objects which may or may not be reactions
\begin{itemize}
\item {} 
\sphinxAtStartPar
\sphinxstyleemphasis{protein} in this case the reaction

\item {} 
\sphinxAtStartPar
\sphinxstyleemphasis{assoc} the COBRA style gene protein association

\item {} 
\sphinxAtStartPar
\sphinxstyleemphasis{gid} the unique id

\item {} 
\sphinxAtStartPar
\sphinxstyleemphasis{name} the optional name

\item {} 
\sphinxAtStartPar
\sphinxstyleemphasis{gene\_pattern} deprecated, not needed anymore

\item {} 
\sphinxAtStartPar
\sphinxstyleemphasis{update\_idx} update the model gene index, not used

\item {} 
\sphinxAtStartPar
\sphinxstyleemphasis{altlabels} {[}default=None{]} alternative labels for genes, default uses geneIds

\end{itemize}

\end{fulllineitems}

\index{createGeneProteinAssociationFromTree() (Model method)@\spxentry{createGeneProteinAssociationFromTree()}\spxextra{Model method}}

\begin{fulllineitems}
\phantomsection\label{\detokenize{modules_doc:cbmpy.CBModel.Model.createGeneProteinAssociationFromTree}}
\pysigstartsignatures
\pysiglinewithargsret{\sphinxbfcode{\sphinxupquote{createGeneProteinAssociationFromTree}}}{\sphinxparam{\DUrole{n,n}{protein}}\sphinxparamcomma \sphinxparam{\DUrole{n,n}{gprtree}}\sphinxparamcomma \sphinxparam{\DUrole{n,n}{gid}\DUrole{o,o}{=}\DUrole{default_value}{None}}\sphinxparamcomma \sphinxparam{\DUrole{n,n}{name}\DUrole{o,o}{=}\DUrole{default_value}{None}}\sphinxparamcomma \sphinxparam{\DUrole{n,n}{gene\_pattern}\DUrole{o,o}{=}\DUrole{default_value}{None}}\sphinxparamcomma \sphinxparam{\DUrole{n,n}{update\_idx}\DUrole{o,o}{=}\DUrole{default_value}{True}}\sphinxparamcomma \sphinxparam{\DUrole{n,n}{altlabels}\DUrole{o,o}{=}\DUrole{default_value}{None}}}{}
\pysigstopsignatures
\sphinxAtStartPar
Create and add a gene protein relationship to the model, note genes are mapped on protein objects which may or may not be reactions
\begin{itemize}
\item {} 
\sphinxAtStartPar
\sphinxstyleemphasis{protein} in this case the reaction

\item {} 
\sphinxAtStartPar
\sphinxstyleemphasis{gprtree} the CBMPy GPR dictionary tree

\item {} 
\sphinxAtStartPar
\sphinxstyleemphasis{gid} the unique id

\item {} 
\sphinxAtStartPar
\sphinxstyleemphasis{name} the optional name

\item {} 
\sphinxAtStartPar
\sphinxstyleemphasis{gene\_pattern} deprecated, not needed anymore

\item {} 
\sphinxAtStartPar
\sphinxstyleemphasis{update\_idx} update the model gene index, not used

\item {} 
\sphinxAtStartPar
\sphinxstyleemphasis{altlabels} {[}default=None{]} alternative labels for genes, default uses geneIds

\end{itemize}

\end{fulllineitems}

\index{createGroup() (Model method)@\spxentry{createGroup()}\spxextra{Model method}}

\begin{fulllineitems}
\phantomsection\label{\detokenize{modules_doc:cbmpy.CBModel.Model.createGroup}}
\pysigstartsignatures
\pysiglinewithargsret{\sphinxbfcode{\sphinxupquote{createGroup}}}{\sphinxparam{\DUrole{n,n}{gid}}}{}
\pysigstopsignatures
\sphinxAtStartPar
Create an empty group with
\begin{itemize}
\item {} 
\sphinxAtStartPar
\sphinxstyleemphasis{gid} the unique group id

\end{itemize}

\end{fulllineitems}

\index{createObjectiveFunction() (Model method)@\spxentry{createObjectiveFunction()}\spxextra{Model method}}

\begin{fulllineitems}
\phantomsection\label{\detokenize{modules_doc:cbmpy.CBModel.Model.createObjectiveFunction}}
\pysigstartsignatures
\pysiglinewithargsret{\sphinxbfcode{\sphinxupquote{createObjectiveFunction}}}{\sphinxparam{\DUrole{n,n}{rid}}\sphinxparamcomma \sphinxparam{\DUrole{n,n}{coefficient}\DUrole{o,o}{=}\DUrole{default_value}{1}}\sphinxparamcomma \sphinxparam{\DUrole{n,n}{osense}\DUrole{o,o}{=}\DUrole{default_value}{\textquotesingle{}maximize\textquotesingle{}}}\sphinxparamcomma \sphinxparam{\DUrole{n,n}{active}\DUrole{o,o}{=}\DUrole{default_value}{True}}\sphinxparamcomma \sphinxparam{\DUrole{n,n}{delete\_current\_obj}\DUrole{o,o}{=}\DUrole{default_value}{True}}}{}
\pysigstopsignatures
\sphinxAtStartPar
Create a single variable objective function:
\begin{itemize}
\item {} 
\sphinxAtStartPar
\sphinxstylestrong{rid} The

\item {} 
\sphinxAtStartPar
\sphinxstylestrong{coefficient} {[}default=1{]}

\item {} 
\sphinxAtStartPar
\sphinxstylestrong{osense} {[}default=’maximize’{]}

\item {} 
\sphinxAtStartPar
\sphinxstylestrong{active} {[}default=True{]}

\item {} 
\sphinxAtStartPar
\sphinxstylestrong{delete\_current\_obj} {[}default=True{]}

\end{itemize}

\end{fulllineitems}

\index{createParameter() (Model method)@\spxentry{createParameter()}\spxextra{Model method}}

\begin{fulllineitems}
\phantomsection\label{\detokenize{modules_doc:cbmpy.CBModel.Model.createParameter}}
\pysigstartsignatures
\pysiglinewithargsret{\sphinxbfcode{\sphinxupquote{createParameter}}}{\sphinxparam{\DUrole{n,n}{pid}}\sphinxparamcomma \sphinxparam{\DUrole{n,n}{value}}\sphinxparamcomma \sphinxparam{\DUrole{n,n}{constant}\DUrole{o,o}{=}\DUrole{default_value}{True}}}{}
\pysigstopsignatures
\sphinxAtStartPar
Instantiates a parameter

\end{fulllineitems}

\index{createReaction() (Model method)@\spxentry{createReaction()}\spxextra{Model method}}

\begin{fulllineitems}
\phantomsection\label{\detokenize{modules_doc:cbmpy.CBModel.Model.createReaction}}
\pysigstartsignatures
\pysiglinewithargsret{\sphinxbfcode{\sphinxupquote{createReaction}}}{\sphinxparam{\DUrole{n,n}{rid}}\sphinxparamcomma \sphinxparam{\DUrole{n,n}{name}\DUrole{o,o}{=}\DUrole{default_value}{None}}\sphinxparamcomma \sphinxparam{\DUrole{n,n}{reversible}\DUrole{o,o}{=}\DUrole{default_value}{True}}\sphinxparamcomma \sphinxparam{\DUrole{n,n}{create\_default\_bounds}\DUrole{o,o}{=}\DUrole{default_value}{True}}\sphinxparamcomma \sphinxparam{\DUrole{n,n}{silent}\DUrole{o,o}{=}\DUrole{default_value}{False}}\sphinxparamcomma \sphinxparam{\DUrole{n,n}{new}\DUrole{o,o}{=}\DUrole{default_value}{False}}}{}
\pysigstopsignatures
\sphinxAtStartPar
Create a new blank reaction and add it to the model:
\begin{itemize}
\item {} 
\sphinxAtStartPar
\sphinxstyleemphasis{id} the unique reaction ID

\item {} 
\sphinxAtStartPar
\sphinxstyleemphasis{name} the reaction name

\item {} 
\sphinxAtStartPar
\sphinxstyleemphasis{reversible} {[}default=True{]} the reaction reversibility. True is reversible, False is irreversible

\item {} 
\sphinxAtStartPar
\sphinxstyleemphasis{create\_default\_bounds} create default reaction bounds, irreversible 0 \textless{}= J \textless{}= INF, reversable \sphinxhyphen{}INF \textless{}= J \textless{}= INF

\item {} 
\sphinxAtStartPar
\sphinxstyleemphasis{silent} {[}default=False{]} if enabled this disables the printing of information messages

\end{itemize}

\end{fulllineitems}

\index{createReactionBounds() (Model method)@\spxentry{createReactionBounds()}\spxextra{Model method}}

\begin{fulllineitems}
\phantomsection\label{\detokenize{modules_doc:cbmpy.CBModel.Model.createReactionBounds}}
\pysigstartsignatures
\pysiglinewithargsret{\sphinxbfcode{\sphinxupquote{createReactionBounds}}}{\sphinxparam{\DUrole{n,n}{reaction}}\sphinxparamcomma \sphinxparam{\DUrole{n,n}{lb\_value}}\sphinxparamcomma \sphinxparam{\DUrole{n,n}{ub\_value}}}{}
\pysigstopsignatures
\sphinxAtStartPar
Create a new lower bound for a reaction: value \textless{}= reaction
\begin{itemize}
\item {} 
\sphinxAtStartPar
\sphinxstylestrong{reaction} the reaction id

\item {} 
\sphinxAtStartPar
\sphinxstylestrong{lb\_value} the value of the lower bound

\item {} 
\sphinxAtStartPar
\sphinxstylestrong{ub\_value} the value of the upper bound

\end{itemize}

\end{fulllineitems}

\index{createReactionLowerBound() (Model method)@\spxentry{createReactionLowerBound()}\spxextra{Model method}}

\begin{fulllineitems}
\phantomsection\label{\detokenize{modules_doc:cbmpy.CBModel.Model.createReactionLowerBound}}
\pysigstartsignatures
\pysiglinewithargsret{\sphinxbfcode{\sphinxupquote{createReactionLowerBound}}}{\sphinxparam{\DUrole{n,n}{reaction}}\sphinxparamcomma \sphinxparam{\DUrole{n,n}{value}}}{}
\pysigstopsignatures
\sphinxAtStartPar
Create a new lower bound for a reaction: value \textless{}= reaction
\begin{itemize}
\item {} 
\sphinxAtStartPar
\sphinxstylestrong{reaction} the reaction id

\item {} 
\sphinxAtStartPar
\sphinxstylestrong{value} the value of the bound

\end{itemize}

\end{fulllineitems}

\index{createReactionNew() (Model method)@\spxentry{createReactionNew()}\spxextra{Model method}}

\begin{fulllineitems}
\phantomsection\label{\detokenize{modules_doc:cbmpy.CBModel.Model.createReactionNew}}
\pysigstartsignatures
\pysiglinewithargsret{\sphinxbfcode{\sphinxupquote{createReactionNew}}}{\sphinxparam{\DUrole{n,n}{rid}}\sphinxparamcomma \sphinxparam{\DUrole{n,n}{name}\DUrole{o,o}{=}\DUrole{default_value}{None}}\sphinxparamcomma \sphinxparam{\DUrole{n,n}{reversible}\DUrole{o,o}{=}\DUrole{default_value}{True}}\sphinxparamcomma \sphinxparam{\DUrole{n,n}{create\_default\_bounds}\DUrole{o,o}{=}\DUrole{default_value}{True}}\sphinxparamcomma \sphinxparam{\DUrole{n,n}{silent}\DUrole{o,o}{=}\DUrole{default_value}{False}}\sphinxparamcomma \sphinxparam{\DUrole{n,n}{new}\DUrole{o,o}{=}\DUrole{default_value}{False}}}{}
\pysigstopsignatures
\sphinxAtStartPar
Create a new blank reaction and add it to the model:
\begin{itemize}
\item {} 
\sphinxAtStartPar
\sphinxstyleemphasis{id} the unique reaction ID

\item {} 
\sphinxAtStartPar
\sphinxstyleemphasis{name} the reaction name

\item {} 
\sphinxAtStartPar
\sphinxstyleemphasis{reversible} {[}default=True{]} the reaction reversibility. True is reversible, False is irreversible

\item {} 
\sphinxAtStartPar
\sphinxstyleemphasis{create\_default\_bounds} create default reaction bounds, irreversible 0 \textless{}= J \textless{}= INF, reversable \sphinxhyphen{}INF \textless{}= J \textless{}= INF

\item {} 
\sphinxAtStartPar
\sphinxstyleemphasis{silent} {[}default=False{]} if enabled this disables the printing of information messages

\end{itemize}

\end{fulllineitems}

\index{createReactionReagent() (Model method)@\spxentry{createReactionReagent()}\spxextra{Model method}}

\begin{fulllineitems}
\phantomsection\label{\detokenize{modules_doc:cbmpy.CBModel.Model.createReactionReagent}}
\pysigstartsignatures
\pysiglinewithargsret{\sphinxbfcode{\sphinxupquote{createReactionReagent}}}{\sphinxparam{\DUrole{n,n}{reaction}}\sphinxparamcomma \sphinxparam{\DUrole{n,n}{metabolite}}\sphinxparamcomma \sphinxparam{\DUrole{n,n}{coefficient}}\sphinxparamcomma \sphinxparam{\DUrole{n,n}{silent}\DUrole{o,o}{=}\DUrole{default_value}{False}}}{}
\pysigstopsignatures
\sphinxAtStartPar
Add a reagent to an existing reaction, both reaction and metabolites must exist
\begin{itemize}
\item {} 
\sphinxAtStartPar
\sphinxstyleemphasis{reaction} a reaction id

\item {} 
\sphinxAtStartPar
\sphinxstyleemphasis{metabolite} a species/metabolite id

\item {} 
\sphinxAtStartPar
\sphinxstyleemphasis{coefficient} the reagent coefficient

\end{itemize}

\end{fulllineitems}

\index{createReactionUpperBound() (Model method)@\spxentry{createReactionUpperBound()}\spxextra{Model method}}

\begin{fulllineitems}
\phantomsection\label{\detokenize{modules_doc:cbmpy.CBModel.Model.createReactionUpperBound}}
\pysigstartsignatures
\pysiglinewithargsret{\sphinxbfcode{\sphinxupquote{createReactionUpperBound}}}{\sphinxparam{\DUrole{n,n}{reaction}}\sphinxparamcomma \sphinxparam{\DUrole{n,n}{value}}}{}
\pysigstopsignatures
\sphinxAtStartPar
Create a new upper bound for a reaction: reaction \textless{}= value
\begin{itemize}
\item {} 
\sphinxAtStartPar
\sphinxstylestrong{reaction} the reaction id

\item {} 
\sphinxAtStartPar
\sphinxstylestrong{value} the value of the bound

\end{itemize}

\end{fulllineitems}

\index{createSingleGeneEffectMap() (Model method)@\spxentry{createSingleGeneEffectMap()}\spxextra{Model method}}

\begin{fulllineitems}
\phantomsection\label{\detokenize{modules_doc:cbmpy.CBModel.Model.createSingleGeneEffectMap}}
\pysigstartsignatures
\pysiglinewithargsret{\sphinxbfcode{\sphinxupquote{createSingleGeneEffectMap}}}{}{}
\pysigstopsignatures
\sphinxAtStartPar
This takes a model and analyses the logical gene expression patterns. This only needs to be done once,
the result is a dictionary that has boolean effect patterns as keys and the (list of) genes that give rise to
those patterns as values. This map is used by the single gene deletion method for further analysis.

\sphinxAtStartPar
Note this dictionary can also be stored and retrieved separately as long as the model structure is not changed i.e.
the gene associations themselves or order of reactions (stored as the special entry ‘keyJ’).

\sphinxAtStartPar
Stored as self.\_\_single\_gene\_effect\_map\_\_

\end{fulllineitems}

\index{createSpecies() (Model method)@\spxentry{createSpecies()}\spxextra{Model method}}

\begin{fulllineitems}
\phantomsection\label{\detokenize{modules_doc:cbmpy.CBModel.Model.createSpecies}}
\pysigstartsignatures
\pysiglinewithargsret{\sphinxbfcode{\sphinxupquote{createSpecies}}}{\sphinxparam{\DUrole{n,n}{sid}}\sphinxparamcomma \sphinxparam{\DUrole{n,n}{boundary}\DUrole{o,o}{=}\DUrole{default_value}{False}}\sphinxparamcomma \sphinxparam{\DUrole{n,n}{name}\DUrole{o,o}{=}\DUrole{default_value}{\textquotesingle{}\textquotesingle{}}}\sphinxparamcomma \sphinxparam{\DUrole{n,n}{value}\DUrole{o,o}{=}\DUrole{default_value}{nan}}\sphinxparamcomma \sphinxparam{\DUrole{n,n}{compartment}\DUrole{o,o}{=}\DUrole{default_value}{None}}\sphinxparamcomma \sphinxparam{\DUrole{n,n}{charge}\DUrole{o,o}{=}\DUrole{default_value}{None}}\sphinxparamcomma \sphinxparam{\DUrole{n,n}{chemFormula}\DUrole{o,o}{=}\DUrole{default_value}{None}}}{}
\pysigstopsignatures
\sphinxAtStartPar
Create a new species and add it to the model:
\begin{itemize}
\item {} 
\sphinxAtStartPar
\sphinxstylestrong{id} the unique species id

\item {} 
\sphinxAtStartPar
\sphinxstylestrong{boundary} {[}default=False{]} whether the species is a variable (False) or is a boundary parameter (fixed)

\item {} 
\sphinxAtStartPar
\sphinxstylestrong{name} {[}default=’’{]} the species name

\item {} 
\sphinxAtStartPar
\sphinxstylestrong{value} {[}default=nan{]} the value \sphinxstyleemphasis{not currently used}

\item {} 
\sphinxAtStartPar
\sphinxstylestrong{compartment} {[}default=None{]} the compartment the species is located in

\item {} 
\sphinxAtStartPar
\sphinxstylestrong{charge} {[}default=None{]} the species charge

\item {} 
\sphinxAtStartPar
\sphinxstylestrong{chemFormula} {[}default=None{]} the chemical formula

\end{itemize}

\end{fulllineitems}

\index{createUserDefinedConstraint() (Model method)@\spxentry{createUserDefinedConstraint()}\spxextra{Model method}}

\begin{fulllineitems}
\phantomsection\label{\detokenize{modules_doc:cbmpy.CBModel.Model.createUserDefinedConstraint}}
\pysigstartsignatures
\pysiglinewithargsret{\sphinxbfcode{\sphinxupquote{createUserDefinedConstraint}}}{\sphinxparam{\DUrole{n,n}{pid}}\sphinxparamcomma \sphinxparam{\DUrole{n,n}{lb}}\sphinxparamcomma \sphinxparam{\DUrole{n,n}{ub}}\sphinxparamcomma \sphinxparam{\DUrole{n,n}{components}\DUrole{o,o}{=}\DUrole{default_value}{None}}}{}
\pysigstopsignatures
\sphinxAtStartPar
Create an FBCv3 UserDefinedConstraint
\begin{itemize}
\item {} 
\sphinxAtStartPar
\sphinxstyleemphasis{pid} unique id

\item {} 
\sphinxAtStartPar
\sphinxstyleemphasis{lb} lower bound float/parameter

\item {} 
\sphinxAtStartPar
\sphinxstyleemphasis{ub} upper bound float/parameter

\item {} \begin{description}
\sphinxlineitem{\sphinxstyleemphasis{componentents} optional, the user defined constraint componenents in the form of a list}
\sphinxAtStartPar
{[}(coefficient, variable, type, id), …{]} and coefficient and variable can be parameters
id is optional

\end{description}

\end{itemize}

\end{fulllineitems}

\index{deleteAllFluxBoundsWithValue() (Model method)@\spxentry{deleteAllFluxBoundsWithValue()}\spxextra{Model method}}

\begin{fulllineitems}
\phantomsection\label{\detokenize{modules_doc:cbmpy.CBModel.Model.deleteAllFluxBoundsWithValue}}
\pysigstartsignatures
\pysiglinewithargsret{\sphinxbfcode{\sphinxupquote{deleteAllFluxBoundsWithValue}}}{\sphinxparam{\DUrole{n,n}{value}}}{}
\pysigstopsignatures
\sphinxAtStartPar
Delete all flux bounds which have a specified value:
\begin{itemize}
\item {} 
\sphinxAtStartPar
\sphinxstyleemphasis{value} the value of the flux bound(s) to delete

\end{itemize}

\end{fulllineitems}

\index{deleteBoundsForReactionId() (Model method)@\spxentry{deleteBoundsForReactionId()}\spxextra{Model method}}

\begin{fulllineitems}
\phantomsection\label{\detokenize{modules_doc:cbmpy.CBModel.Model.deleteBoundsForReactionId}}
\pysigstartsignatures
\pysiglinewithargsret{\sphinxbfcode{\sphinxupquote{deleteBoundsForReactionId}}}{\sphinxparam{\DUrole{n,n}{rid}}\sphinxparamcomma \sphinxparam{\DUrole{n,n}{lower}\DUrole{o,o}{=}\DUrole{default_value}{True}}\sphinxparamcomma \sphinxparam{\DUrole{n,n}{upper}\DUrole{o,o}{=}\DUrole{default_value}{True}}}{}
\pysigstopsignatures
\sphinxAtStartPar
Delete bounds connected to reaction, rid
\begin{itemize}
\item {} 
\sphinxAtStartPar
\sphinxstyleemphasis{rid} a valid reaction id

\item {} 
\sphinxAtStartPar
\sphinxstyleemphasis{upper} {[}default=True{]} delete the upper bound

\item {} 
\sphinxAtStartPar
\sphinxstyleemphasis{lower} {[}default=True{]} delete the lower bound

\end{itemize}

\end{fulllineitems}

\index{deleteCompartment() (Model method)@\spxentry{deleteCompartment()}\spxextra{Model method}}

\begin{fulllineitems}
\phantomsection\label{\detokenize{modules_doc:cbmpy.CBModel.Model.deleteCompartment}}
\pysigstartsignatures
\pysiglinewithargsret{\sphinxbfcode{\sphinxupquote{deleteCompartment}}}{\sphinxparam{\DUrole{n,n}{sid}}\sphinxparamcomma \sphinxparam{\DUrole{n,n}{check\_components}\DUrole{o,o}{=}\DUrole{default_value}{True}}}{}
\pysigstopsignatures
\sphinxAtStartPar
Deletes a compartment object with id. Returns True if the compartment is deleted, False if not. In addition if components were checked
a list of id’s that reference the compartment are also returned.
\begin{itemize}
\item {} 
\sphinxAtStartPar
\sphinxstyleemphasis{sid} the compartment id

\item {} 
\sphinxAtStartPar
\sphinxstyleemphasis{check\_components} {[}default=True{]} if  enabled check that no species or reactions makes

\end{itemize}

\sphinxAtStartPar
use of the compartment, fail if it does.

\end{fulllineitems}

\index{deleteGPRAssociation() (Model method)@\spxentry{deleteGPRAssociation()}\spxextra{Model method}}

\begin{fulllineitems}
\phantomsection\label{\detokenize{modules_doc:cbmpy.CBModel.Model.deleteGPRAssociation}}
\pysigstartsignatures
\pysiglinewithargsret{\sphinxbfcode{\sphinxupquote{deleteGPRAssociation}}}{\sphinxparam{\DUrole{n,n}{gprid}}}{}
\pysigstopsignatures
\sphinxAtStartPar
Delete a GPR association with id
\begin{itemize}
\item {} 
\sphinxAtStartPar
\sphinxstyleemphasis{gprid} the GPR association id

\end{itemize}

\end{fulllineitems}

\index{deleteGene() (Model method)@\spxentry{deleteGene()}\spxextra{Model method}}

\begin{fulllineitems}
\phantomsection\label{\detokenize{modules_doc:cbmpy.CBModel.Model.deleteGene}}
\pysigstartsignatures
\pysiglinewithargsret{\sphinxbfcode{\sphinxupquote{deleteGene}}}{\sphinxparam{\DUrole{n,n}{gid}}\sphinxparamcomma \sphinxparam{\DUrole{n,n}{also\_delete\_gpr}\DUrole{o,o}{=}\DUrole{default_value}{True}}}{}
\pysigstopsignatures
\sphinxAtStartPar
Deletes the gene object with gid. Note if you want to delete a gene by label (locus tag etc)
use the deleteGeneByLabel() function.
\begin{itemize}
\item {} 
\sphinxAtStartPar
\sphinxstyleemphasis{gid} the gene Id

\item {} 
\sphinxAtStartPar
\sphinxstyleemphasis{also\_delete\_gpr} {[}default=True{]} automatically delete GPR’s that contain no gene references

\end{itemize}

\end{fulllineitems}

\index{deleteGeneByLabel() (Model method)@\spxentry{deleteGeneByLabel()}\spxextra{Model method}}

\begin{fulllineitems}
\phantomsection\label{\detokenize{modules_doc:cbmpy.CBModel.Model.deleteGeneByLabel}}
\pysigstartsignatures
\pysiglinewithargsret{\sphinxbfcode{\sphinxupquote{deleteGeneByLabel}}}{\sphinxparam{\DUrole{n,n}{label}}\sphinxparamcomma \sphinxparam{\DUrole{n,n}{also\_delete\_gpr}\DUrole{o,o}{=}\DUrole{default_value}{True}}}{}
\pysigstopsignatures
\sphinxAtStartPar
Deletes the gene object with label (b2003 etc).
\begin{itemize}
\item {} 
\sphinxAtStartPar
\sphinxstyleemphasis{label} the gene with label to be deleted

\item {} 
\sphinxAtStartPar
\sphinxstyleemphasis{also\_delete\_gpr} {[}default=True{]} automatically delete GPR’s that contain no gene references

\end{itemize}

\end{fulllineitems}

\index{deleteGroup() (Model method)@\spxentry{deleteGroup()}\spxextra{Model method}}

\begin{fulllineitems}
\phantomsection\label{\detokenize{modules_doc:cbmpy.CBModel.Model.deleteGroup}}
\pysigstartsignatures
\pysiglinewithargsret{\sphinxbfcode{\sphinxupquote{deleteGroup}}}{\sphinxparam{\DUrole{n,n}{gid}}}{}
\pysigstopsignatures
\sphinxAtStartPar
Delete a group with
\begin{itemize}
\item {} 
\sphinxAtStartPar
\sphinxstyleemphasis{gid} the unique group id

\end{itemize}

\end{fulllineitems}

\index{deleteNonReactingSpecies() (Model method)@\spxentry{deleteNonReactingSpecies()}\spxextra{Model method}}

\begin{fulllineitems}
\phantomsection\label{\detokenize{modules_doc:cbmpy.CBModel.Model.deleteNonReactingSpecies}}
\pysigstartsignatures
\pysiglinewithargsret{\sphinxbfcode{\sphinxupquote{deleteNonReactingSpecies}}}{\sphinxparam{\DUrole{n,n}{simulate}\DUrole{o,o}{=}\DUrole{default_value}{True}}}{}
\pysigstopsignatures
\sphinxAtStartPar
Deletes all species that are not reagents (do not to take part in a reaction).
\sphinxstyleemphasis{Warning} this deletion is permanent and greedy (not selective). Returns a list of (would be) deleted species
\begin{itemize}
\item {} 
\sphinxAtStartPar
\sphinxstyleemphasis{simulate} {[}default=True{]} only return a list of the speciesId’s that would have been deleted if False

\end{itemize}

\end{fulllineitems}

\index{deleteObjective() (Model method)@\spxentry{deleteObjective()}\spxextra{Model method}}

\begin{fulllineitems}
\phantomsection\label{\detokenize{modules_doc:cbmpy.CBModel.Model.deleteObjective}}
\pysigstartsignatures
\pysiglinewithargsret{\sphinxbfcode{\sphinxupquote{deleteObjective}}}{\sphinxparam{\DUrole{n,n}{objective\_id}}}{}
\pysigstopsignatures
\sphinxAtStartPar
Delete objective function:
\begin{quote}

\sphinxAtStartPar
\sphinxstyleemphasis{objective\_id} the id of the objective function. If objective\_id is given  as ‘active’ then the active objective is deleted.
\end{quote}

\end{fulllineitems}

\index{deleteReactionAndBounds() (Model method)@\spxentry{deleteReactionAndBounds()}\spxextra{Model method}}

\begin{fulllineitems}
\phantomsection\label{\detokenize{modules_doc:cbmpy.CBModel.Model.deleteReactionAndBounds}}
\pysigstartsignatures
\pysiglinewithargsret{\sphinxbfcode{\sphinxupquote{deleteReactionAndBounds}}}{\sphinxparam{\DUrole{n,n}{rid}}}{}
\pysigstopsignatures
\sphinxAtStartPar
Delete all reaction and bounds connected to reaction
\begin{itemize}
\item {} 
\sphinxAtStartPar
\sphinxstyleemphasis{rid} a valid reaction id

\end{itemize}

\end{fulllineitems}

\index{deleteSpecies() (Model method)@\spxentry{deleteSpecies()}\spxextra{Model method}}

\begin{fulllineitems}
\phantomsection\label{\detokenize{modules_doc:cbmpy.CBModel.Model.deleteSpecies}}
\pysigstartsignatures
\pysiglinewithargsret{\sphinxbfcode{\sphinxupquote{deleteSpecies}}}{\sphinxparam{\DUrole{n,n}{sid}}\sphinxparamcomma \sphinxparam{\DUrole{n,n}{also\_delete}\DUrole{o,o}{=}\DUrole{default_value}{None}}}{}
\pysigstopsignatures
\sphinxAtStartPar
Deletes a species object with id
\begin{itemize}
\item {} 
\sphinxAtStartPar
\sphinxstyleemphasis{sid} the species id

\item {} 
\sphinxAtStartPar
\sphinxstyleemphasis{also\_delete} {[}default=None{]} only delete the species

\end{itemize}

\sphinxAtStartPar
\textendash{} ‘reagents’ delete the species from the reactions it participates in as a \sphinxstylestrong{reagent}
\textendash{} ‘reactions’ deletes the \sphinxstylestrong{reactions} that the species participates in

\end{fulllineitems}

\index{emptyUndelete() (Model method)@\spxentry{emptyUndelete()}\spxextra{Model method}}

\begin{fulllineitems}
\phantomsection\label{\detokenize{modules_doc:cbmpy.CBModel.Model.emptyUndelete}}
\pysigstartsignatures
\pysiglinewithargsret{\sphinxbfcode{\sphinxupquote{emptyUndelete}}}{}{}
\pysigstopsignatures
\sphinxAtStartPar
Empties the undelete cache

\end{fulllineitems}

\index{exportFVAdata() (Model method)@\spxentry{exportFVAdata()}\spxextra{Model method}}

\begin{fulllineitems}
\phantomsection\label{\detokenize{modules_doc:cbmpy.CBModel.Model.exportFVAdata}}
\pysigstartsignatures
\pysiglinewithargsret{\sphinxbfcode{\sphinxupquote{exportFVAdata}}}{}{}
\pysigstopsignatures
\sphinxAtStartPar
Export the fva data as an array and list of reaction id’s

\end{fulllineitems}

\index{exportUserConstraints() (Model method)@\spxentry{exportUserConstraints()}\spxextra{Model method}}

\begin{fulllineitems}
\phantomsection\label{\detokenize{modules_doc:cbmpy.CBModel.Model.exportUserConstraints}}
\pysigstartsignatures
\pysiglinewithargsret{\sphinxbfcode{\sphinxupquote{exportUserConstraints}}}{\sphinxparam{\DUrole{n,n}{filename}}}{}
\pysigstopsignatures
\sphinxAtStartPar
Exports user constraints in json

\end{fulllineitems}

\index{findFluxesForConnectedSpecies() (Model method)@\spxentry{findFluxesForConnectedSpecies()}\spxextra{Model method}}

\begin{fulllineitems}
\phantomsection\label{\detokenize{modules_doc:cbmpy.CBModel.Model.findFluxesForConnectedSpecies}}
\pysigstartsignatures
\pysiglinewithargsret{\sphinxbfcode{\sphinxupquote{findFluxesForConnectedSpecies}}}{\sphinxparam{\DUrole{n,n}{metab}}}{}
\pysigstopsignatures
\sphinxAtStartPar
Returns a list of (reaction, flux value) pairs that this metabolite appears as a reagent of
\begin{itemize}
\item {} 
\sphinxAtStartPar
\sphinxstyleemphasis{metab} the metabolite name

\end{itemize}

\end{fulllineitems}

\index{getActiveObjective() (Model method)@\spxentry{getActiveObjective()}\spxextra{Model method}}

\begin{fulllineitems}
\phantomsection\label{\detokenize{modules_doc:cbmpy.CBModel.Model.getActiveObjective}}
\pysigstartsignatures
\pysiglinewithargsret{\sphinxbfcode{\sphinxupquote{getActiveObjective}}}{}{}
\pysigstopsignatures
\sphinxAtStartPar
Returns the active objective object.

\end{fulllineitems}

\index{getActiveObjectiveReactionIds() (Model method)@\spxentry{getActiveObjectiveReactionIds()}\spxextra{Model method}}

\begin{fulllineitems}
\phantomsection\label{\detokenize{modules_doc:cbmpy.CBModel.Model.getActiveObjectiveReactionIds}}
\pysigstartsignatures
\pysiglinewithargsret{\sphinxbfcode{\sphinxupquote{getActiveObjectiveReactionIds}}}{}{}
\pysigstopsignatures
\sphinxAtStartPar
Returns the active objective flux objective reaction id’s

\end{fulllineitems}

\index{getActiveObjectiveStoichiometry() (Model method)@\spxentry{getActiveObjectiveStoichiometry()}\spxextra{Model method}}

\begin{fulllineitems}
\phantomsection\label{\detokenize{modules_doc:cbmpy.CBModel.Model.getActiveObjectiveStoichiometry}}
\pysigstartsignatures
\pysiglinewithargsret{\sphinxbfcode{\sphinxupquote{getActiveObjectiveStoichiometry}}}{}{}
\pysigstopsignatures
\sphinxAtStartPar
Returns a list of (coefficient, flux\_objective) tuples

\end{fulllineitems}

\index{getAllFluxBounds() (Model method)@\spxentry{getAllFluxBounds()}\spxextra{Model method}}

\begin{fulllineitems}
\phantomsection\label{\detokenize{modules_doc:cbmpy.CBModel.Model.getAllFluxBounds}}
\pysigstartsignatures
\pysiglinewithargsret{\sphinxbfcode{\sphinxupquote{getAllFluxBounds}}}{}{}
\pysigstopsignatures
\sphinxAtStartPar
DEPRECATED
Returns a dictionary of all flux bounds {[}id:value{]}

\end{fulllineitems}

\index{getAllGeneActivities() (Model method)@\spxentry{getAllGeneActivities()}\spxextra{Model method}}

\begin{fulllineitems}
\phantomsection\label{\detokenize{modules_doc:cbmpy.CBModel.Model.getAllGeneActivities}}
\pysigstartsignatures
\pysiglinewithargsret{\sphinxbfcode{\sphinxupquote{getAllGeneActivities}}}{}{}
\pysigstopsignatures
\sphinxAtStartPar
Returns a dictionary of genes (if defined) and whether they are active or not

\end{fulllineitems}

\index{getAllGeneProteinAssociations() (Model method)@\spxentry{getAllGeneProteinAssociations()}\spxextra{Model method}}

\begin{fulllineitems}
\phantomsection\label{\detokenize{modules_doc:cbmpy.CBModel.Model.getAllGeneProteinAssociations}}
\pysigstartsignatures
\pysiglinewithargsret{\sphinxbfcode{\sphinxupquote{getAllGeneProteinAssociations}}}{\sphinxparam{\DUrole{n,n}{use\_labels}\DUrole{o,o}{=}\DUrole{default_value}{False}}}{}
\pysigstopsignatures
\sphinxAtStartPar
Returns a dictionary of genes associated with each protein
\begin{itemize}
\item {} 
\sphinxAtStartPar
\sphinxstyleemphasis{use\_labels} use V2 gene labels rather than ID’s

\end{itemize}

\end{fulllineitems}

\index{getAllProteinActivities() (Model method)@\spxentry{getAllProteinActivities()}\spxextra{Model method}}

\begin{fulllineitems}
\phantomsection\label{\detokenize{modules_doc:cbmpy.CBModel.Model.getAllProteinActivities}}
\pysigstartsignatures
\pysiglinewithargsret{\sphinxbfcode{\sphinxupquote{getAllProteinActivities}}}{}{}
\pysigstopsignatures
\sphinxAtStartPar
Returns a dictionary of reactions (if genes and GPR’s are defined) and whether they are active or not

\end{fulllineitems}

\index{getAllProteinGeneAssociations() (Model method)@\spxentry{getAllProteinGeneAssociations()}\spxextra{Model method}}

\begin{fulllineitems}
\phantomsection\label{\detokenize{modules_doc:cbmpy.CBModel.Model.getAllProteinGeneAssociations}}
\pysigstartsignatures
\pysiglinewithargsret{\sphinxbfcode{\sphinxupquote{getAllProteinGeneAssociations}}}{\sphinxparam{\DUrole{n,n}{use\_labels}\DUrole{o,o}{=}\DUrole{default_value}{False}}}{}
\pysigstopsignatures
\sphinxAtStartPar
Returns a dictionary of the proteins associated with each gene
\begin{itemize}
\item {} 
\sphinxAtStartPar
\sphinxstyleemphasis{use\_labels} use V2 gene labels rather than ID’s

\end{itemize}

\end{fulllineitems}

\index{getBoundarySpeciesIds() (Model method)@\spxentry{getBoundarySpeciesIds()}\spxextra{Model method}}

\begin{fulllineitems}
\phantomsection\label{\detokenize{modules_doc:cbmpy.CBModel.Model.getBoundarySpeciesIds}}
\pysigstartsignatures
\pysiglinewithargsret{\sphinxbfcode{\sphinxupquote{getBoundarySpeciesIds}}}{\sphinxparam{\DUrole{n,n}{rid}\DUrole{o,o}{=}\DUrole{default_value}{None}}}{}
\pysigstopsignatures
\sphinxAtStartPar
Return all boundary species associated with reaction
\begin{itemize}
\item {} 
\sphinxAtStartPar
rid {[}default=None{]} by default return all boundary species in a model, alternatively a string containing a reaction id or list of reaction id’s

\end{itemize}

\end{fulllineitems}

\index{getCompartment() (Model method)@\spxentry{getCompartment()}\spxextra{Model method}}

\begin{fulllineitems}
\phantomsection\label{\detokenize{modules_doc:cbmpy.CBModel.Model.getCompartment}}
\pysigstartsignatures
\pysiglinewithargsret{\sphinxbfcode{\sphinxupquote{getCompartment}}}{\sphinxparam{\DUrole{n,n}{cid}}}{}
\pysigstopsignatures
\sphinxAtStartPar
Returns a compartment object with \sphinxstyleemphasis{cid}
\begin{itemize}
\item {} 
\sphinxAtStartPar
\sphinxstyleemphasis{cid} compartment ID

\end{itemize}

\end{fulllineitems}

\index{getCompartmentIds() (Model method)@\spxentry{getCompartmentIds()}\spxextra{Model method}}

\begin{fulllineitems}
\phantomsection\label{\detokenize{modules_doc:cbmpy.CBModel.Model.getCompartmentIds}}
\pysigstartsignatures
\pysiglinewithargsret{\sphinxbfcode{\sphinxupquote{getCompartmentIds}}}{\sphinxparam{\DUrole{n,n}{substring}\DUrole{o,o}{=}\DUrole{default_value}{None}}}{}
\pysigstopsignatures
\sphinxAtStartPar
Returns a list of compartment Ids, applies a substring search if substring is defined
\begin{itemize}
\item {} 
\sphinxAtStartPar
\sphinxstyleemphasis{substring} search for this pattern anywhere in the id

\end{itemize}

\end{fulllineitems}

\index{getCompartmentObjects() (Model method)@\spxentry{getCompartmentObjects()}\spxextra{Model method}}

\begin{fulllineitems}
\phantomsection\label{\detokenize{modules_doc:cbmpy.CBModel.Model.getCompartmentObjects}}
\pysigstartsignatures
\pysiglinewithargsret{\sphinxbfcode{\sphinxupquote{getCompartmentObjects}}}{\sphinxparam{\DUrole{n,n}{substring}\DUrole{o,o}{=}\DUrole{default_value}{None}}}{}
\pysigstopsignatures
\sphinxAtStartPar
Returns a list of compartment objects, applies a substring search if substring is defined
\begin{itemize}
\item {} 
\sphinxAtStartPar
\sphinxstyleemphasis{substring} search for this pattern anywhere in the id

\end{itemize}

\end{fulllineitems}

\index{getDescription() (Model method)@\spxentry{getDescription()}\spxextra{Model method}}

\begin{fulllineitems}
\phantomsection\label{\detokenize{modules_doc:cbmpy.CBModel.Model.getDescription}}
\pysigstartsignatures
\pysiglinewithargsret{\sphinxbfcode{\sphinxupquote{getDescription}}}{}{}
\pysigstopsignatures
\sphinxAtStartPar
Returns the model description which was stored in the SBML \textless{}notes\textgreater{} field

\end{fulllineitems}

\index{getExchangeReactionIds() (Model method)@\spxentry{getExchangeReactionIds()}\spxextra{Model method}}

\begin{fulllineitems}
\phantomsection\label{\detokenize{modules_doc:cbmpy.CBModel.Model.getExchangeReactionIds}}
\pysigstartsignatures
\pysiglinewithargsret{\sphinxbfcode{\sphinxupquote{getExchangeReactionIds}}}{}{}
\pysigstopsignatures
\sphinxAtStartPar
Returns id’s of reactions where the ‘is\_exchange’ attribute set to True. This is by default
reactions that contain a boundary species.

\end{fulllineitems}

\index{getExchangeReactions() (Model method)@\spxentry{getExchangeReactions()}\spxextra{Model method}}

\begin{fulllineitems}
\phantomsection\label{\detokenize{modules_doc:cbmpy.CBModel.Model.getExchangeReactions}}
\pysigstartsignatures
\pysiglinewithargsret{\sphinxbfcode{\sphinxupquote{getExchangeReactions}}}{}{}
\pysigstopsignatures
\sphinxAtStartPar
Returns reaction instances where the ‘is\_exchange’ attribute set to True. This is by default
reactions that contain a boundary species.

\end{fulllineitems}

\index{getFluxBoundByID() (Model method)@\spxentry{getFluxBoundByID()}\spxextra{Model method}}

\begin{fulllineitems}
\phantomsection\label{\detokenize{modules_doc:cbmpy.CBModel.Model.getFluxBoundByID}}
\pysigstartsignatures
\pysiglinewithargsret{\sphinxbfcode{\sphinxupquote{getFluxBoundByID}}}{\sphinxparam{\DUrole{n,n}{fid}}}{}
\pysigstopsignatures
\sphinxAtStartPar
Returns a FluxBound/Parameter with id
\begin{itemize}
\item {} 
\sphinxAtStartPar
\sphinxstyleemphasis{fid} the fluxBound ID

\end{itemize}

\end{fulllineitems}

\index{getFluxBoundByReactionID() (Model method)@\spxentry{getFluxBoundByReactionID()}\spxextra{Model method}}

\begin{fulllineitems}
\phantomsection\label{\detokenize{modules_doc:cbmpy.CBModel.Model.getFluxBoundByReactionID}}
\pysigstartsignatures
\pysiglinewithargsret{\sphinxbfcode{\sphinxupquote{getFluxBoundByReactionID}}}{\sphinxparam{\DUrole{n,n}{rid}}\sphinxparamcomma \sphinxparam{\DUrole{n,n}{bound}}}{}
\pysigstopsignatures
\sphinxAtStartPar
Returns a FluxBound/Parameter instance
\begin{itemize}
\item {} 
\sphinxAtStartPar
\sphinxstyleemphasis{rid} the reaction ID

\item {} 
\sphinxAtStartPar
\sphinxstyleemphasis{bound} the bound: ‘upper’, ‘lower’, ‘equality’

\end{itemize}

\end{fulllineitems}

\index{getFluxBoundIds() (Model method)@\spxentry{getFluxBoundIds()}\spxextra{Model method}}

\begin{fulllineitems}
\phantomsection\label{\detokenize{modules_doc:cbmpy.CBModel.Model.getFluxBoundIds}}
\pysigstartsignatures
\pysiglinewithargsret{\sphinxbfcode{\sphinxupquote{getFluxBoundIds}}}{\sphinxparam{\DUrole{n,n}{substring}\DUrole{o,o}{=}\DUrole{default_value}{None}}}{}
\pysigstopsignatures
\sphinxAtStartPar
Returns a list of fluxbound Ids, applies a substring search if substring is defined
\begin{itemize}
\item {} 
\sphinxAtStartPar
\sphinxstyleemphasis{substring} search for this pattern anywhere in the id

\end{itemize}

\end{fulllineitems}

\index{getFluxBoundsByReactionID() (Model method)@\spxentry{getFluxBoundsByReactionID()}\spxextra{Model method}}

\begin{fulllineitems}
\phantomsection\label{\detokenize{modules_doc:cbmpy.CBModel.Model.getFluxBoundsByReactionID}}
\pysigstartsignatures
\pysiglinewithargsret{\sphinxbfcode{\sphinxupquote{getFluxBoundsByReactionID}}}{\sphinxparam{\DUrole{n,n}{rid}}}{}
\pysigstopsignatures
\sphinxAtStartPar
Returns all FluxBound instances connected to a reactionId as a tuple of valid
(lower, upper, None) or (None, None, equality) or alternatively invalid (lower, upper, equality).
\begin{quote}
\begin{itemize}
\item {} 
\sphinxAtStartPar
\sphinxstyleemphasis{rid} the reaction ID

\end{itemize}

\sphinxAtStartPar
\sphinxstyleemphasis{under evaluation}
\end{quote}

\end{fulllineitems}

\index{getFluxesAssociatedWithCompartments() (Model method)@\spxentry{getFluxesAssociatedWithCompartments()}\spxextra{Model method}}

\begin{fulllineitems}
\phantomsection\label{\detokenize{modules_doc:cbmpy.CBModel.Model.getFluxesAssociatedWithCompartments}}
\pysigstartsignatures
\pysiglinewithargsret{\sphinxbfcode{\sphinxupquote{getFluxesAssociatedWithCompartments}}}{\sphinxparam{\DUrole{n,n}{compartments}}}{}
\pysigstopsignatures
\sphinxAtStartPar
Determines all reactions and flux values associated with a list of
compartments. This function can be used to find all transport reactions
between compartments, e.g. the cytosol and mitochondria. If the
compartment IDs are ‘cyt’ and ‘mit’, respectively, you can call
“your\_model.getFluxesAssociatedWithCompartments({[}‘cyt’, ‘mit’{]})”
to get all fluxes between these compartments.
\begin{description}
\sphinxlineitem{\sphinxstyleemphasis{compartments}: a list or set of compartment IDs.}
\sphinxAtStartPar
To check the existing compartment IDs in your model
call “your\_model.getCompartmentIds()”

\end{description}

\sphinxAtStartPar
:returns a dictionary with reaction IDs as keys and corresponding
flux values as values

\end{fulllineitems}

\index{getFluxesAssociatedWithSpecies() (Model method)@\spxentry{getFluxesAssociatedWithSpecies()}\spxextra{Model method}}

\begin{fulllineitems}
\phantomsection\label{\detokenize{modules_doc:cbmpy.CBModel.Model.getFluxesAssociatedWithSpecies}}
\pysigstartsignatures
\pysiglinewithargsret{\sphinxbfcode{\sphinxupquote{getFluxesAssociatedWithSpecies}}}{\sphinxparam{\DUrole{n,n}{metab}}}{}
\pysigstopsignatures
\sphinxAtStartPar
Returns a list of (reaction, flux value) pairs that this metabolite appears as a reagent in
\begin{itemize}
\item {} 
\sphinxAtStartPar
\sphinxstyleemphasis{metab} the metabolite name

\end{itemize}

\end{fulllineitems}

\index{getGPRIdAssociatedWithGeneId() (Model method)@\spxentry{getGPRIdAssociatedWithGeneId()}\spxextra{Model method}}

\begin{fulllineitems}
\phantomsection\label{\detokenize{modules_doc:cbmpy.CBModel.Model.getGPRIdAssociatedWithGeneId}}
\pysigstartsignatures
\pysiglinewithargsret{\sphinxbfcode{\sphinxupquote{getGPRIdAssociatedWithGeneId}}}{\sphinxparam{\DUrole{n,n}{gid}}}{}
\pysigstopsignatures
\sphinxAtStartPar
Return the GPR(s) associated with the gene id:
\begin{itemize}
\item {} 
\sphinxAtStartPar
\sphinxstyleemphasis{gid} a gene id

\end{itemize}

\end{fulllineitems}

\index{getGPRIdAssociatedWithGeneLabel() (Model method)@\spxentry{getGPRIdAssociatedWithGeneLabel()}\spxextra{Model method}}

\begin{fulllineitems}
\phantomsection\label{\detokenize{modules_doc:cbmpy.CBModel.Model.getGPRIdAssociatedWithGeneLabel}}
\pysigstartsignatures
\pysiglinewithargsret{\sphinxbfcode{\sphinxupquote{getGPRIdAssociatedWithGeneLabel}}}{\sphinxparam{\DUrole{n,n}{label}}}{}
\pysigstopsignatures
\sphinxAtStartPar
Return the GPR Id’s associated with the gene label:
\begin{itemize}
\item {} 
\sphinxAtStartPar
\sphinxstyleemphasis{label} a gene label

\end{itemize}

\end{fulllineitems}

\index{getGPRIds() (Model method)@\spxentry{getGPRIds()}\spxextra{Model method}}

\begin{fulllineitems}
\phantomsection\label{\detokenize{modules_doc:cbmpy.CBModel.Model.getGPRIds}}
\pysigstartsignatures
\pysiglinewithargsret{\sphinxbfcode{\sphinxupquote{getGPRIds}}}{\sphinxparam{\DUrole{n,n}{substring}\DUrole{o,o}{=}\DUrole{default_value}{None}}}{}
\pysigstopsignatures
\sphinxAtStartPar
Returns a list of GPR Id’s, applies a substring search if substring is defined
\begin{itemize}
\item {} 
\sphinxAtStartPar
\sphinxstyleemphasis{substring} search for this pattern anywhere in the id

\end{itemize}

\end{fulllineitems}

\index{getGPRObjects() (Model method)@\spxentry{getGPRObjects()}\spxextra{Model method}}

\begin{fulllineitems}
\phantomsection\label{\detokenize{modules_doc:cbmpy.CBModel.Model.getGPRObjects}}
\pysigstartsignatures
\pysiglinewithargsret{\sphinxbfcode{\sphinxupquote{getGPRObjects}}}{\sphinxparam{\DUrole{n,n}{substring}\DUrole{o,o}{=}\DUrole{default_value}{None}}}{}
\pysigstopsignatures
\sphinxAtStartPar
Returns a list of GPR objects, applies a substring search if substring is defined
\begin{itemize}
\item {} 
\sphinxAtStartPar
\sphinxstyleemphasis{substring} search for this pattern anywhere in the id

\end{itemize}

\end{fulllineitems}

\index{getGPRassociation() (Model method)@\spxentry{getGPRassociation()}\spxextra{Model method}}

\begin{fulllineitems}
\phantomsection\label{\detokenize{modules_doc:cbmpy.CBModel.Model.getGPRassociation}}
\pysigstartsignatures
\pysiglinewithargsret{\sphinxbfcode{\sphinxupquote{getGPRassociation}}}{\sphinxparam{\DUrole{n,n}{gprid}}}{}
\pysigstopsignatures
\sphinxAtStartPar
Returns a gene protein association object that has the identifier:
\begin{itemize}
\item {} 
\sphinxAtStartPar
\sphinxstyleemphasis{gprid} the gene protein identifier

\end{itemize}

\end{fulllineitems}

\index{getGPRforReaction() (Model method)@\spxentry{getGPRforReaction()}\spxextra{Model method}}

\begin{fulllineitems}
\phantomsection\label{\detokenize{modules_doc:cbmpy.CBModel.Model.getGPRforReaction}}
\pysigstartsignatures
\pysiglinewithargsret{\sphinxbfcode{\sphinxupquote{getGPRforReaction}}}{\sphinxparam{\DUrole{n,n}{rid}}}{}
\pysigstopsignatures
\sphinxAtStartPar
Return the GPR associated with the reaction id:
\begin{itemize}
\item {} 
\sphinxAtStartPar
\sphinxstyleemphasis{rid} a reaction id

\end{itemize}

\end{fulllineitems}

\index{getGPRforReactionAsDict() (Model method)@\spxentry{getGPRforReactionAsDict()}\spxextra{Model method}}

\begin{fulllineitems}
\phantomsection\label{\detokenize{modules_doc:cbmpy.CBModel.Model.getGPRforReactionAsDict}}
\pysigstartsignatures
\pysiglinewithargsret{\sphinxbfcode{\sphinxupquote{getGPRforReactionAsDict}}}{\sphinxparam{\DUrole{n,n}{rid}}\sphinxparamcomma \sphinxparam{\DUrole{n,n}{useweakref}\DUrole{o,o}{=}\DUrole{default_value}{True}}}{}
\pysigstopsignatures
\sphinxAtStartPar
Return the GPR associated with the reaction id as a nested dictionary structure:
\begin{itemize}
\item {} 
\sphinxAtStartPar
\sphinxstyleemphasis{rid} a reaction id

\end{itemize}

\end{fulllineitems}

\index{getGene() (Model method)@\spxentry{getGene()}\spxextra{Model method}}

\begin{fulllineitems}
\phantomsection\label{\detokenize{modules_doc:cbmpy.CBModel.Model.getGene}}
\pysigstartsignatures
\pysiglinewithargsret{\sphinxbfcode{\sphinxupquote{getGene}}}{\sphinxparam{\DUrole{n,n}{gid}}}{}
\pysigstopsignatures
\sphinxAtStartPar
Returns a gene object that has the identifier:
\begin{itemize}
\item {} 
\sphinxAtStartPar
\sphinxstyleemphasis{gid} the gene identifier

\end{itemize}

\end{fulllineitems}

\index{getGeneByLabel() (Model method)@\spxentry{getGeneByLabel()}\spxextra{Model method}}

\begin{fulllineitems}
\phantomsection\label{\detokenize{modules_doc:cbmpy.CBModel.Model.getGeneByLabel}}
\pysigstartsignatures
\pysiglinewithargsret{\sphinxbfcode{\sphinxupquote{getGeneByLabel}}}{\sphinxparam{\DUrole{n,n}{label}}}{}
\pysigstopsignatures
\sphinxAtStartPar
Given a gene label return the corresponding Gene object
\begin{itemize}
\item {} 
\sphinxAtStartPar
\sphinxstyleemphasis{label}

\end{itemize}

\end{fulllineitems}

\index{getGeneIdFromLabel() (Model method)@\spxentry{getGeneIdFromLabel()}\spxextra{Model method}}

\begin{fulllineitems}
\phantomsection\label{\detokenize{modules_doc:cbmpy.CBModel.Model.getGeneIdFromLabel}}
\pysigstartsignatures
\pysiglinewithargsret{\sphinxbfcode{\sphinxupquote{getGeneIdFromLabel}}}{\sphinxparam{\DUrole{n,n}{label}}}{}
\pysigstopsignatures
\sphinxAtStartPar
Given a gene label it returns the corresponding Gene id or None
\begin{itemize}
\item {} 
\sphinxAtStartPar
\sphinxstyleemphasis{label}

\end{itemize}

\end{fulllineitems}

\index{getGeneIds() (Model method)@\spxentry{getGeneIds()}\spxextra{Model method}}

\begin{fulllineitems}
\phantomsection\label{\detokenize{modules_doc:cbmpy.CBModel.Model.getGeneIds}}
\pysigstartsignatures
\pysiglinewithargsret{\sphinxbfcode{\sphinxupquote{getGeneIds}}}{\sphinxparam{\DUrole{n,n}{substring}\DUrole{o,o}{=}\DUrole{default_value}{None}}}{}
\pysigstopsignatures
\sphinxAtStartPar
Returns a list of gene Ids, applies a substring search if substring is defined
\begin{itemize}
\item {} 
\sphinxAtStartPar
\sphinxstyleemphasis{substring} search for this pattern anywhere in the id

\end{itemize}

\end{fulllineitems}

\index{getGeneLabels() (Model method)@\spxentry{getGeneLabels()}\spxextra{Model method}}

\begin{fulllineitems}
\phantomsection\label{\detokenize{modules_doc:cbmpy.CBModel.Model.getGeneLabels}}
\pysigstartsignatures
\pysiglinewithargsret{\sphinxbfcode{\sphinxupquote{getGeneLabels}}}{\sphinxparam{\DUrole{n,n}{substring}\DUrole{o,o}{=}\DUrole{default_value}{None}}}{}
\pysigstopsignatures
\sphinxAtStartPar
Returns a list of gene labels (locus tags), applies a substring search if substring is defined
\begin{itemize}
\item {} 
\sphinxAtStartPar
\sphinxstyleemphasis{substring} search for this pattern anywhere in the label

\end{itemize}

\end{fulllineitems}

\index{getGeneObjects() (Model method)@\spxentry{getGeneObjects()}\spxextra{Model method}}

\begin{fulllineitems}
\phantomsection\label{\detokenize{modules_doc:cbmpy.CBModel.Model.getGeneObjects}}
\pysigstartsignatures
\pysiglinewithargsret{\sphinxbfcode{\sphinxupquote{getGeneObjects}}}{\sphinxparam{\DUrole{n,n}{substring}\DUrole{o,o}{=}\DUrole{default_value}{None}}}{}
\pysigstopsignatures
\sphinxAtStartPar
Returns a list of gene objects, applies a substring search if substring is defined
\begin{itemize}
\item {} 
\sphinxAtStartPar
\sphinxstyleemphasis{substring} search for this pattern anywhere in the id

\end{itemize}

\end{fulllineitems}

\index{getGeneObjectsByLabel() (Model method)@\spxentry{getGeneObjectsByLabel()}\spxextra{Model method}}

\begin{fulllineitems}
\phantomsection\label{\detokenize{modules_doc:cbmpy.CBModel.Model.getGeneObjectsByLabel}}
\pysigstartsignatures
\pysiglinewithargsret{\sphinxbfcode{\sphinxupquote{getGeneObjectsByLabel}}}{\sphinxparam{\DUrole{n,n}{substring}\DUrole{o,o}{=}\DUrole{default_value}{None}}}{}
\pysigstopsignatures
\sphinxAtStartPar
Returns a list of gene objects, applies a substring search if substring is defined
\begin{itemize}
\item {} 
\sphinxAtStartPar
\sphinxstyleemphasis{substring} search for this pattern anywhere in the label

\end{itemize}

\end{fulllineitems}

\index{getGroup() (Model method)@\spxentry{getGroup()}\spxextra{Model method}}

\begin{fulllineitems}
\phantomsection\label{\detokenize{modules_doc:cbmpy.CBModel.Model.getGroup}}
\pysigstartsignatures
\pysiglinewithargsret{\sphinxbfcode{\sphinxupquote{getGroup}}}{\sphinxparam{\DUrole{n,n}{gid}}}{}
\pysigstopsignatures
\sphinxAtStartPar
Return a group with
\begin{itemize}
\item {} 
\sphinxAtStartPar
\sphinxstyleemphasis{gid} the unique group id

\end{itemize}

\end{fulllineitems}

\index{getGroupIds() (Model method)@\spxentry{getGroupIds()}\spxextra{Model method}}

\begin{fulllineitems}
\phantomsection\label{\detokenize{modules_doc:cbmpy.CBModel.Model.getGroupIds}}
\pysigstartsignatures
\pysiglinewithargsret{\sphinxbfcode{\sphinxupquote{getGroupIds}}}{}{}
\pysigstopsignatures
\sphinxAtStartPar
Get all group ids

\end{fulllineitems}

\index{getGroupMembership() (Model method)@\spxentry{getGroupMembership()}\spxextra{Model method}}

\begin{fulllineitems}
\phantomsection\label{\detokenize{modules_doc:cbmpy.CBModel.Model.getGroupMembership}}
\pysigstartsignatures
\pysiglinewithargsret{\sphinxbfcode{\sphinxupquote{getGroupMembership}}}{}{}
\pysigstopsignatures
\sphinxAtStartPar
Returns group membership of items in groups. Returns \{object\_id: {[}‘group\_id1’, ‘group\_id2’{]}\}

\end{fulllineitems}

\index{getGroupNames() (Model method)@\spxentry{getGroupNames()}\spxextra{Model method}}

\begin{fulllineitems}
\phantomsection\label{\detokenize{modules_doc:cbmpy.CBModel.Model.getGroupNames}}
\pysigstartsignatures
\pysiglinewithargsret{\sphinxbfcode{\sphinxupquote{getGroupNames}}}{}{}
\pysigstopsignatures
\sphinxAtStartPar
Get all group names

\end{fulllineitems}

\index{getIrreversibleReactionIds() (Model method)@\spxentry{getIrreversibleReactionIds()}\spxextra{Model method}}

\begin{fulllineitems}
\phantomsection\label{\detokenize{modules_doc:cbmpy.CBModel.Model.getIrreversibleReactionIds}}
\pysigstartsignatures
\pysiglinewithargsret{\sphinxbfcode{\sphinxupquote{getIrreversibleReactionIds}}}{}{}
\pysigstopsignatures
\sphinxAtStartPar
Return a list of irreversible reaction Id’s

\end{fulllineitems}

\index{getModel() (Model method)@\spxentry{getModel()}\spxextra{Model method}}

\begin{fulllineitems}
\phantomsection\label{\detokenize{modules_doc:cbmpy.CBModel.Model.getModel}}
\pysigstartsignatures
\pysiglinewithargsret{\sphinxbfcode{\sphinxupquote{getModel}}}{}{}
\pysigstopsignatures
\sphinxAtStartPar
Overrides the FBase inherited method, returns own instance.

\end{fulllineitems}

\index{getModelCreators() (Model method)@\spxentry{getModelCreators()}\spxextra{Model method}}

\begin{fulllineitems}
\phantomsection\label{\detokenize{modules_doc:cbmpy.CBModel.Model.getModelCreators}}
\pysigstartsignatures
\pysiglinewithargsret{\sphinxbfcode{\sphinxupquote{getModelCreators}}}{}{}
\pysigstopsignatures
\sphinxAtStartPar
Return model creator information

\end{fulllineitems}

\index{getObjFuncValue() (Model method)@\spxentry{getObjFuncValue()}\spxextra{Model method}}

\begin{fulllineitems}
\phantomsection\label{\detokenize{modules_doc:cbmpy.CBModel.Model.getObjFuncValue}}
\pysigstartsignatures
\pysiglinewithargsret{\sphinxbfcode{\sphinxupquote{getObjFuncValue}}}{}{}
\pysigstopsignatures
\sphinxAtStartPar
Returns the objective function value

\end{fulllineitems}

\index{getObject() (Model method)@\spxentry{getObject()}\spxextra{Model method}}

\begin{fulllineitems}
\phantomsection\label{\detokenize{modules_doc:cbmpy.CBModel.Model.getObject}}
\pysigstartsignatures
\pysiglinewithargsret{\sphinxbfcode{\sphinxupquote{getObject}}}{\sphinxparam{\DUrole{n,n}{pid}}}{}
\pysigstopsignatures\begin{itemize}
\item {} 
\sphinxAtStartPar
\sphinxstyleemphasis{pid} returns a model object with pid or None

\end{itemize}

\end{fulllineitems}

\index{getObjectiveIds() (Model method)@\spxentry{getObjectiveIds()}\spxextra{Model method}}

\begin{fulllineitems}
\phantomsection\label{\detokenize{modules_doc:cbmpy.CBModel.Model.getObjectiveIds}}
\pysigstartsignatures
\pysiglinewithargsret{\sphinxbfcode{\sphinxupquote{getObjectiveIds}}}{\sphinxparam{\DUrole{n,n}{substring}\DUrole{o,o}{=}\DUrole{default_value}{None}}}{}
\pysigstopsignatures
\sphinxAtStartPar
Returns a list of objective function Ids, applies a substring search if substring is defined
\begin{itemize}
\item {} 
\sphinxAtStartPar
\sphinxstyleemphasis{substring} search for this pattern anywhere in the id

\end{itemize}

\end{fulllineitems}

\index{getOptimalValue() (Model method)@\spxentry{getOptimalValue()}\spxextra{Model method}}

\begin{fulllineitems}
\phantomsection\label{\detokenize{modules_doc:cbmpy.CBModel.Model.getOptimalValue}}
\pysigstartsignatures
\pysiglinewithargsret{\sphinxbfcode{\sphinxupquote{getOptimalValue}}}{}{}
\pysigstopsignatures
\sphinxAtStartPar
Returns the optimal value of the objective function

\end{fulllineitems}

\index{getParameter() (Model method)@\spxentry{getParameter()}\spxextra{Model method}}

\begin{fulllineitems}
\phantomsection\label{\detokenize{modules_doc:cbmpy.CBModel.Model.getParameter}}
\pysigstartsignatures
\pysiglinewithargsret{\sphinxbfcode{\sphinxupquote{getParameter}}}{\sphinxparam{\DUrole{n,n}{pid}}}{}
\pysigstopsignatures
\sphinxAtStartPar
Returns a parameter object with pid

\end{fulllineitems}

\index{getReaction() (Model method)@\spxentry{getReaction()}\spxextra{Model method}}

\begin{fulllineitems}
\phantomsection\label{\detokenize{modules_doc:cbmpy.CBModel.Model.getReaction}}
\pysigstartsignatures
\pysiglinewithargsret{\sphinxbfcode{\sphinxupquote{getReaction}}}{\sphinxparam{\DUrole{n,n}{rid}}}{}
\pysigstopsignatures
\sphinxAtStartPar
Returns a reaction object with \sphinxstyleemphasis{id}
\begin{itemize}
\item {} 
\sphinxAtStartPar
\sphinxstyleemphasis{rid} reaction ID

\end{itemize}

\end{fulllineitems}

\index{getReactionActivity() (Model method)@\spxentry{getReactionActivity()}\spxextra{Model method}}

\begin{fulllineitems}
\phantomsection\label{\detokenize{modules_doc:cbmpy.CBModel.Model.getReactionActivity}}
\pysigstartsignatures
\pysiglinewithargsret{\sphinxbfcode{\sphinxupquote{getReactionActivity}}}{\sphinxparam{\DUrole{n,n}{rid}}}{}
\pysigstopsignatures
\sphinxAtStartPar
If there is a GPR and genes associated with the reaction ID then return either active=True or inactive=False
Note if there is no gene associated information then this will return active.
\begin{itemize}
\item {} 
\sphinxAtStartPar
\sphinxstyleemphasis{rid} a reaction id

\end{itemize}

\end{fulllineitems}

\index{getReactionBounds() (Model method)@\spxentry{getReactionBounds()}\spxextra{Model method}}

\begin{fulllineitems}
\phantomsection\label{\detokenize{modules_doc:cbmpy.CBModel.Model.getReactionBounds}}
\pysigstartsignatures
\pysiglinewithargsret{\sphinxbfcode{\sphinxupquote{getReactionBounds}}}{\sphinxparam{\DUrole{n,n}{rid}}}{}
\pysigstopsignatures
\sphinxAtStartPar
Get the bounds of a reaction, returns a tuple of rid, lowerbound value, upperbound value and equality value (None means bound does not exist).
\begin{itemize}
\item {} 
\sphinxAtStartPar
\sphinxstyleemphasis{rid} the reaction ID

\end{itemize}

\end{fulllineitems}

\index{getReactionIds() (Model method)@\spxentry{getReactionIds()}\spxextra{Model method}}

\begin{fulllineitems}
\phantomsection\label{\detokenize{modules_doc:cbmpy.CBModel.Model.getReactionIds}}
\pysigstartsignatures
\pysiglinewithargsret{\sphinxbfcode{\sphinxupquote{getReactionIds}}}{\sphinxparam{\DUrole{n,n}{substring}\DUrole{o,o}{=}\DUrole{default_value}{None}}}{}
\pysigstopsignatures
\sphinxAtStartPar
Returns a list of reaction Ids, applies a substring search if substring is defined
\begin{itemize}
\item {} 
\sphinxAtStartPar
\sphinxstyleemphasis{substring} search for this pattern anywhere in the id

\end{itemize}

\end{fulllineitems}

\index{getReactionIdsAssociatedWithSpecies() (Model method)@\spxentry{getReactionIdsAssociatedWithSpecies()}\spxextra{Model method}}

\begin{fulllineitems}
\phantomsection\label{\detokenize{modules_doc:cbmpy.CBModel.Model.getReactionIdsAssociatedWithSpecies}}
\pysigstartsignatures
\pysiglinewithargsret{\sphinxbfcode{\sphinxupquote{getReactionIdsAssociatedWithSpecies}}}{\sphinxparam{\DUrole{n,n}{metab}}}{}
\pysigstopsignatures
\sphinxAtStartPar
Returns a list of (reaction, flux value) pairs that this metabolite appears as a reagent in
\begin{itemize}
\item {} 
\sphinxAtStartPar
\sphinxstyleemphasis{metab} the metabolite name

\end{itemize}

\end{fulllineitems}

\index{getReactionLowerBound() (Model method)@\spxentry{getReactionLowerBound()}\spxextra{Model method}}

\begin{fulllineitems}
\phantomsection\label{\detokenize{modules_doc:cbmpy.CBModel.Model.getReactionLowerBound}}
\pysigstartsignatures
\pysiglinewithargsret{\sphinxbfcode{\sphinxupquote{getReactionLowerBound}}}{\sphinxparam{\DUrole{n,n}{rid}}}{}
\pysigstopsignatures
\sphinxAtStartPar
Returns the lower bound of a reaction (it it exists) or None
\begin{itemize}
\item {} 
\sphinxAtStartPar
\sphinxstyleemphasis{rid} the reaction ID

\end{itemize}

\end{fulllineitems}

\index{getReactionNames() (Model method)@\spxentry{getReactionNames()}\spxextra{Model method}}

\begin{fulllineitems}
\phantomsection\label{\detokenize{modules_doc:cbmpy.CBModel.Model.getReactionNames}}
\pysigstartsignatures
\pysiglinewithargsret{\sphinxbfcode{\sphinxupquote{getReactionNames}}}{\sphinxparam{\DUrole{n,n}{substring}\DUrole{o,o}{=}\DUrole{default_value}{None}}}{}
\pysigstopsignatures
\sphinxAtStartPar
Returns a list of reaction names, applies a substring search if substring is defined
\begin{itemize}
\item {} 
\sphinxAtStartPar
\sphinxstyleemphasis{substring} search for this pattern anywhere in the name

\end{itemize}

\end{fulllineitems}

\index{getReactionObjects() (Model method)@\spxentry{getReactionObjects()}\spxextra{Model method}}

\begin{fulllineitems}
\phantomsection\label{\detokenize{modules_doc:cbmpy.CBModel.Model.getReactionObjects}}
\pysigstartsignatures
\pysiglinewithargsret{\sphinxbfcode{\sphinxupquote{getReactionObjects}}}{\sphinxparam{\DUrole{n,n}{substring}\DUrole{o,o}{=}\DUrole{default_value}{None}}}{}
\pysigstopsignatures
\sphinxAtStartPar
Returns a list of reaction objects, applies a substring search if substring is defined
\begin{itemize}
\item {} 
\sphinxAtStartPar
\sphinxstyleemphasis{substring} search for this pattern anywhere in the id

\end{itemize}

\end{fulllineitems}

\index{getReactionUpperBound() (Model method)@\spxentry{getReactionUpperBound()}\spxextra{Model method}}

\begin{fulllineitems}
\phantomsection\label{\detokenize{modules_doc:cbmpy.CBModel.Model.getReactionUpperBound}}
\pysigstartsignatures
\pysiglinewithargsret{\sphinxbfcode{\sphinxupquote{getReactionUpperBound}}}{\sphinxparam{\DUrole{n,n}{rid}}}{}
\pysigstopsignatures
\sphinxAtStartPar
Returns the upper bound of a reaction (it it exists) or None
\begin{itemize}
\item {} 
\sphinxAtStartPar
\sphinxstyleemphasis{rid} the reaction ID

\end{itemize}

\end{fulllineitems}

\index{getReactionValues() (Model method)@\spxentry{getReactionValues()}\spxextra{Model method}}

\begin{fulllineitems}
\phantomsection\label{\detokenize{modules_doc:cbmpy.CBModel.Model.getReactionValues}}
\pysigstartsignatures
\pysiglinewithargsret{\sphinxbfcode{\sphinxupquote{getReactionValues}}}{\sphinxparam{\DUrole{n,n}{only\_exchange}\DUrole{o,o}{=}\DUrole{default_value}{False}}}{}
\pysigstopsignatures
\sphinxAtStartPar
Returns a dictionary of ReactionID : ReactionValue pairs:
\begin{itemize}
\item {} 
\sphinxAtStartPar
\sphinxstyleemphasis{only\_exchange} {[}default=False{]} only return the reactions labelled as exchange

\end{itemize}

\end{fulllineitems}

\index{getReversibleReactionIds() (Model method)@\spxentry{getReversibleReactionIds()}\spxextra{Model method}}

\begin{fulllineitems}
\phantomsection\label{\detokenize{modules_doc:cbmpy.CBModel.Model.getReversibleReactionIds}}
\pysigstartsignatures
\pysiglinewithargsret{\sphinxbfcode{\sphinxupquote{getReversibleReactionIds}}}{}{}
\pysigstopsignatures
\sphinxAtStartPar
Return a list of reversible reaction Id’s

\end{fulllineitems}

\index{getSolutionVector() (Model method)@\spxentry{getSolutionVector()}\spxextra{Model method}}

\begin{fulllineitems}
\phantomsection\label{\detokenize{modules_doc:cbmpy.CBModel.Model.getSolutionVector}}
\pysigstartsignatures
\pysiglinewithargsret{\sphinxbfcode{\sphinxupquote{getSolutionVector}}}{\sphinxparam{\DUrole{n,n}{names}\DUrole{o,o}{=}\DUrole{default_value}{False}}}{}
\pysigstopsignatures
\sphinxAtStartPar
Return a vector of solution values
\begin{itemize}
\item {} 
\sphinxAtStartPar
\sphinxstyleemphasis{names} {[}default=False{]} if True return a solution vector and list of names

\end{itemize}

\end{fulllineitems}

\index{getSpecies() (Model method)@\spxentry{getSpecies()}\spxextra{Model method}}

\begin{fulllineitems}
\phantomsection\label{\detokenize{modules_doc:cbmpy.CBModel.Model.getSpecies}}
\pysigstartsignatures
\pysiglinewithargsret{\sphinxbfcode{\sphinxupquote{getSpecies}}}{\sphinxparam{\DUrole{n,n}{sid}}}{}
\pysigstopsignatures
\sphinxAtStartPar
Returns a species object with \sphinxstyleemphasis{sid}
\begin{itemize}
\item {} 
\sphinxAtStartPar
\sphinxstyleemphasis{sid} a specied ID

\end{itemize}

\end{fulllineitems}

\index{getSpeciesIds() (Model method)@\spxentry{getSpeciesIds()}\spxextra{Model method}}

\begin{fulllineitems}
\phantomsection\label{\detokenize{modules_doc:cbmpy.CBModel.Model.getSpeciesIds}}
\pysigstartsignatures
\pysiglinewithargsret{\sphinxbfcode{\sphinxupquote{getSpeciesIds}}}{\sphinxparam{\DUrole{n,n}{substring}\DUrole{o,o}{=}\DUrole{default_value}{None}}}{}
\pysigstopsignatures
\sphinxAtStartPar
Returns a list of species Ids, applies a substring search if substring is defined
\begin{itemize}
\item {} 
\sphinxAtStartPar
\sphinxstyleemphasis{substring} search for this pattern anywhere in the id

\end{itemize}

\end{fulllineitems}

\index{getSpeciesObjects() (Model method)@\spxentry{getSpeciesObjects()}\spxextra{Model method}}

\begin{fulllineitems}
\phantomsection\label{\detokenize{modules_doc:cbmpy.CBModel.Model.getSpeciesObjects}}
\pysigstartsignatures
\pysiglinewithargsret{\sphinxbfcode{\sphinxupquote{getSpeciesObjects}}}{\sphinxparam{\DUrole{n,n}{substring}\DUrole{o,o}{=}\DUrole{default_value}{None}}}{}
\pysigstopsignatures
\sphinxAtStartPar
Returns a list of species objects, applies a substring search if substring is defined
\begin{itemize}
\item {} 
\sphinxAtStartPar
\sphinxstyleemphasis{substring} search for this pattern anywhere in the id

\end{itemize}

\end{fulllineitems}

\index{hasObject() (Model method)@\spxentry{hasObject()}\spxextra{Model method}}

\begin{fulllineitems}
\phantomsection\label{\detokenize{modules_doc:cbmpy.CBModel.Model.hasObject}}
\pysigstartsignatures
\pysiglinewithargsret{\sphinxbfcode{\sphinxupquote{hasObject}}}{\sphinxparam{\DUrole{n,n}{pid}}}{}
\pysigstopsignatures\begin{itemize}
\item {} 
\sphinxAtStartPar
\sphinxstyleemphasis{pid} returns a boolean if there is a registered object with pid\sphinxhyphen{}

\end{itemize}

\end{fulllineitems}

\index{importUserConstraints() (Model method)@\spxentry{importUserConstraints()}\spxextra{Model method}}

\begin{fulllineitems}
\phantomsection\label{\detokenize{modules_doc:cbmpy.CBModel.Model.importUserConstraints}}
\pysigstartsignatures
\pysiglinewithargsret{\sphinxbfcode{\sphinxupquote{importUserConstraints}}}{\sphinxparam{\DUrole{n,n}{filename}}}{}
\pysigstopsignatures
\sphinxAtStartPar
Exports user constraints in json

\end{fulllineitems}

\index{registerObjectInGlobalStore() (Model method)@\spxentry{registerObjectInGlobalStore()}\spxextra{Model method}}

\begin{fulllineitems}
\phantomsection\label{\detokenize{modules_doc:cbmpy.CBModel.Model.registerObjectInGlobalStore}}
\pysigstartsignatures
\pysiglinewithargsret{\sphinxbfcode{\sphinxupquote{registerObjectInGlobalStore}}}{\sphinxparam{\DUrole{n,n}{obj}}}{}
\pysigstopsignatures\begin{itemize}
\item {} 
\sphinxAtStartPar
\sphinxstyleemphasis{object}

\end{itemize}

\end{fulllineitems}

\index{renameObjectIds() (Model method)@\spxentry{renameObjectIds()}\spxextra{Model method}}

\begin{fulllineitems}
\phantomsection\label{\detokenize{modules_doc:cbmpy.CBModel.Model.renameObjectIds}}
\pysigstartsignatures
\pysiglinewithargsret{\sphinxbfcode{\sphinxupquote{renameObjectIds}}}{\sphinxparam{\DUrole{n,n}{prefix}\DUrole{o,o}{=}\DUrole{default_value}{None}}\sphinxparamcomma \sphinxparam{\DUrole{n,n}{suffix}\DUrole{o,o}{=}\DUrole{default_value}{None}}\sphinxparamcomma \sphinxparam{\DUrole{n,n}{target}\DUrole{o,o}{=}\DUrole{default_value}{\textquotesingle{}all\textquotesingle{}}}\sphinxparamcomma \sphinxparam{\DUrole{n,n}{ignore}\DUrole{o,o}{=}\DUrole{default_value}{None}}}{}
\pysigstopsignatures
\sphinxAtStartPar
This method is designed for target=”all” other use may result in inconsistent models. Update: “species” and “reactions”
should also work as advertised, please check results.
\begin{quote}
\begin{itemize}
\item {} 
\sphinxAtStartPar
\sphinxstyleemphasis{prefix} {[}None{]} if supplied add as a prefix

\item {} 
\sphinxAtStartPar
\sphinxstyleemphasis{suffix} {[}None{]} if supplied add as a suffix

\item {} 
\sphinxAtStartPar
\sphinxstyleemphasis{target} {[}‘all’{]} specify what class of objects to rename

\end{itemize}
\begin{itemize}
\item {} 
\sphinxAtStartPar
‘species’

\item {} 
\sphinxAtStartPar
‘reactions’

\item {} 
\sphinxAtStartPar
‘bounds’

\item {} 
\sphinxAtStartPar
‘objectives’

\item {} 
\sphinxAtStartPar
‘all’

\end{itemize}
\begin{itemize}
\item {} 
\sphinxAtStartPar
\sphinxstyleemphasis{ignore} {[}default=None{]} a list of id’s to ignore

\end{itemize}
\end{quote}

\end{fulllineitems}

\index{resetAllGenes() (Model method)@\spxentry{resetAllGenes()}\spxextra{Model method}}

\begin{fulllineitems}
\phantomsection\label{\detokenize{modules_doc:cbmpy.CBModel.Model.resetAllGenes}}
\pysigstartsignatures
\pysiglinewithargsret{\sphinxbfcode{\sphinxupquote{resetAllGenes}}}{\sphinxparam{\DUrole{n,n}{update\_reactions}\DUrole{o,o}{=}\DUrole{default_value}{False}}}{}
\pysigstopsignatures
\sphinxAtStartPar
Resets all genes to their default activity state (normally on)
\begin{itemize}
\item {} 
\sphinxAtStartPar
\sphinxstyleemphasis{update\_reactions} {[}default=False{]} update the associated reactions fluxbounds from the gene deletion bounds if they exist

\end{itemize}

\end{fulllineitems}

\index{resetAllInactiveGPRBounds() (Model method)@\spxentry{resetAllInactiveGPRBounds()}\spxextra{Model method}}

\begin{fulllineitems}
\phantomsection\label{\detokenize{modules_doc:cbmpy.CBModel.Model.resetAllInactiveGPRBounds}}
\pysigstartsignatures
\pysiglinewithargsret{\sphinxbfcode{\sphinxupquote{resetAllInactiveGPRBounds}}}{}{}
\pysigstopsignatures
\sphinxAtStartPar
Resets all reaction bounds modified by the \sphinxcode{\sphinxupquote{cmod.setAllInactiveGeneReactionBounds()}} method to their previous values

\end{fulllineitems}

\index{serialize() (Model method)@\spxentry{serialize()}\spxextra{Model method}}

\begin{fulllineitems}
\phantomsection\label{\detokenize{modules_doc:cbmpy.CBModel.Model.serialize}}
\pysigstartsignatures
\pysiglinewithargsret{\sphinxbfcode{\sphinxupquote{serialize}}}{\sphinxparam{\DUrole{n,n}{protocol}\DUrole{o,o}{=}\DUrole{default_value}{0}}}{}
\pysigstopsignatures
\sphinxAtStartPar
Serialize object, returns a string by default
\begin{itemize}
\item {} \begin{description}
\sphinxlineitem{\sphinxstyleemphasis{protocol} {[}default=0{]} serialize to a string or binary if required,}
\sphinxAtStartPar
see pickle module documentation for details

\end{description}

\end{itemize}

\sphinxAtStartPar
\# overloaded in CBModel

\end{fulllineitems}

\index{serializeToDisk() (Model method)@\spxentry{serializeToDisk()}\spxextra{Model method}}

\begin{fulllineitems}
\phantomsection\label{\detokenize{modules_doc:cbmpy.CBModel.Model.serializeToDisk}}
\pysigstartsignatures
\pysiglinewithargsret{\sphinxbfcode{\sphinxupquote{serializeToDisk}}}{\sphinxparam{\DUrole{n,n}{filename}}\sphinxparamcomma \sphinxparam{\DUrole{n,n}{protocol}\DUrole{o,o}{=}\DUrole{default_value}{2}}}{}
\pysigstopsignatures
\sphinxAtStartPar
Serialize to disk using pickle protocol:
\begin{itemize}
\item {} 
\sphinxAtStartPar
\sphinxstyleemphasis{filename} the name of the output file

\item {} \begin{description}
\sphinxlineitem{\sphinxstyleemphasis{protocol} {[}default=2{]} serialize to a string or binary if required,}
\sphinxAtStartPar
see pickle module documentation for details

\end{description}

\end{itemize}

\sphinxAtStartPar
\# overloaded in CBModel

\end{fulllineitems}

\index{setAllFluxBounds() (Model method)@\spxentry{setAllFluxBounds()}\spxextra{Model method}}

\begin{fulllineitems}
\phantomsection\label{\detokenize{modules_doc:cbmpy.CBModel.Model.setAllFluxBounds}}
\pysigstartsignatures
\pysiglinewithargsret{\sphinxbfcode{\sphinxupquote{setAllFluxBounds}}}{\sphinxparam{\DUrole{n,n}{bounds}}}{}
\pysigstopsignatures
\sphinxAtStartPar
DEPRECATED! use setFluxBoundsFromDict()

\sphinxAtStartPar
Sets all the fluxbounds present in bounds
\begin{itemize}
\item {} 
\sphinxAtStartPar
\sphinxstyleemphasis{bounds} a dictionary of {[}fluxbound\_id : value{]} pairs (not per reaction!!!)

\end{itemize}

\end{fulllineitems}

\index{setAllInactiveGPRBounds() (Model method)@\spxentry{setAllInactiveGPRBounds()}\spxextra{Model method}}

\begin{fulllineitems}
\phantomsection\label{\detokenize{modules_doc:cbmpy.CBModel.Model.setAllInactiveGPRBounds}}
\pysigstartsignatures
\pysiglinewithargsret{\sphinxbfcode{\sphinxupquote{setAllInactiveGPRBounds}}}{\sphinxparam{\DUrole{n,n}{lower}\DUrole{o,o}{=}\DUrole{default_value}{0.0}}\sphinxparamcomma \sphinxparam{\DUrole{n,n}{upper}\DUrole{o,o}{=}\DUrole{default_value}{0.0}}}{}
\pysigstopsignatures
\sphinxAtStartPar
Set all reactions that are inactive (as determined by gene and gpr evaluation) to bounds:
\begin{itemize}
\item {} 
\sphinxAtStartPar
\sphinxstyleemphasis{lower} {[}default=0.0{]} the new lower bound

\item {} 
\sphinxAtStartPar
\sphinxstyleemphasis{upper} {[}default=0.0{]} the new upper bound

\end{itemize}

\end{fulllineitems}

\index{setAllProteinActivities() (Model method)@\spxentry{setAllProteinActivities()}\spxextra{Model method}}

\begin{fulllineitems}
\phantomsection\label{\detokenize{modules_doc:cbmpy.CBModel.Model.setAllProteinActivities}}
\pysigstartsignatures
\pysiglinewithargsret{\sphinxbfcode{\sphinxupquote{setAllProteinActivities}}}{\sphinxparam{\DUrole{n,n}{activites}}\sphinxparamcomma \sphinxparam{\DUrole{n,n}{lower}\DUrole{o,o}{=}\DUrole{default_value}{0.0}}\sphinxparamcomma \sphinxparam{\DUrole{n,n}{upper}\DUrole{o,o}{=}\DUrole{default_value}{0.0}}}{}
\pysigstopsignatures
\sphinxAtStartPar
Given a dictionary of activities {[}rid : boolean{]} pairs set all the corresponding reactions:
\begin{itemize}
\item {} 
\sphinxAtStartPar
\sphinxstyleemphasis{activities} a dictionary of {[}rid : boolean{]} pairs

\item {} 
\sphinxAtStartPar
\sphinxstyleemphasis{lower} {[}default=0.0{]} the lower bound of the deactivated flux

\item {} 
\sphinxAtStartPar
\sphinxstyleemphasis{upper} {[}default=0.0{]} the upper bound of the deactivated flux

\end{itemize}

\end{fulllineitems}

\index{setBoundValueByName() (Model method)@\spxentry{setBoundValueByName()}\spxextra{Model method}}

\begin{fulllineitems}
\phantomsection\label{\detokenize{modules_doc:cbmpy.CBModel.Model.setBoundValueByName}}
\pysigstartsignatures
\pysiglinewithargsret{\sphinxbfcode{\sphinxupquote{setBoundValueByName}}}{\sphinxparam{\DUrole{n,n}{rid}}\sphinxparamcomma \sphinxparam{\DUrole{n,n}{value}}\sphinxparamcomma \sphinxparam{\DUrole{n,n}{bound}}}{}
\pysigstopsignatures
\sphinxAtStartPar
Deprecated use setReactionBound
\begin{description}
\sphinxlineitem{Set a reaction bound}\begin{itemize}
\item {} 
\sphinxAtStartPar
\sphinxstyleemphasis{rid} the reactions id

\item {} 
\sphinxAtStartPar
\sphinxstyleemphasis{value} the new value

\item {} 
\sphinxAtStartPar
\sphinxstyleemphasis{bound} this is either ‘lower’ or ‘upper’

\end{itemize}

\end{description}

\end{fulllineitems}

\index{setCreatedDate() (Model method)@\spxentry{setCreatedDate()}\spxextra{Model method}}

\begin{fulllineitems}
\phantomsection\label{\detokenize{modules_doc:cbmpy.CBModel.Model.setCreatedDate}}
\pysigstartsignatures
\pysiglinewithargsret{\sphinxbfcode{\sphinxupquote{setCreatedDate}}}{\sphinxparam{\DUrole{n,n}{date}\DUrole{o,o}{=}\DUrole{default_value}{None}}}{}
\pysigstopsignatures
\sphinxAtStartPar
Set the model created date tuple(year, month, day, hour, minute, second)
\begin{itemize}
\item {} 
\sphinxAtStartPar
\sphinxstyleemphasis{date} {[}default=None{]} default is now (automatic) otherwise (year, month, day, hour, minute, second) e.g. (2012, 09, 24, 13, 34, 00)

\end{itemize}

\end{fulllineitems}

\index{setDescription() (Model method)@\spxentry{setDescription()}\spxextra{Model method}}

\begin{fulllineitems}
\phantomsection\label{\detokenize{modules_doc:cbmpy.CBModel.Model.setDescription}}
\pysigstartsignatures
\pysiglinewithargsret{\sphinxbfcode{\sphinxupquote{setDescription}}}{\sphinxparam{\DUrole{n,n}{html}}}{}
\pysigstopsignatures
\sphinxAtStartPar
Sets the model description which translates into the SBML \textless{}notes\textgreater{} field.
\begin{itemize}
\item {} 
\sphinxAtStartPar
\sphinxstyleemphasis{html} any valid html or the empty string to clear ‘’

\end{itemize}

\end{fulllineitems}

\index{setFluxBoundsFromDict() (Model method)@\spxentry{setFluxBoundsFromDict()}\spxextra{Model method}}

\begin{fulllineitems}
\phantomsection\label{\detokenize{modules_doc:cbmpy.CBModel.Model.setFluxBoundsFromDict}}
\pysigstartsignatures
\pysiglinewithargsret{\sphinxbfcode{\sphinxupquote{setFluxBoundsFromDict}}}{\sphinxparam{\DUrole{n,n}{bounds}}}{}
\pysigstopsignatures
\sphinxAtStartPar
DEPRECATED! This method will be modified to use reaction Idin CBMPy 0.9.0
Sets all the fluxbounds present in bounds
\begin{itemize}
\item {} 
\sphinxAtStartPar
\sphinxstyleemphasis{bounds} a dictionary of {[}fluxbound\_id : value{]} pairs (not per reaction!!!)

\end{itemize}

\end{fulllineitems}

\index{setGeneActive() (Model method)@\spxentry{setGeneActive()}\spxextra{Model method}}

\begin{fulllineitems}
\phantomsection\label{\detokenize{modules_doc:cbmpy.CBModel.Model.setGeneActive}}
\pysigstartsignatures
\pysiglinewithargsret{\sphinxbfcode{\sphinxupquote{setGeneActive}}}{\sphinxparam{\DUrole{n,n}{g\_id}}\sphinxparamcomma \sphinxparam{\DUrole{n,n}{update\_reactions}\DUrole{o,o}{=}\DUrole{default_value}{False}}}{}
\pysigstopsignatures
\sphinxAtStartPar
Effectively restores a gene by setting it’s active flag
\begin{itemize}
\item {} 
\sphinxAtStartPar
\sphinxstyleemphasis{g\_id} a gene ID

\item {} 
\sphinxAtStartPar
\sphinxstyleemphasis{update\_reactions} {[}default=False{]} update the associated reactions fluxbounds from the gene deletion bounds if they exist

\end{itemize}

\end{fulllineitems}

\index{setGeneInactive() (Model method)@\spxentry{setGeneInactive()}\spxextra{Model method}}

\begin{fulllineitems}
\phantomsection\label{\detokenize{modules_doc:cbmpy.CBModel.Model.setGeneInactive}}
\pysigstartsignatures
\pysiglinewithargsret{\sphinxbfcode{\sphinxupquote{setGeneInactive}}}{\sphinxparam{\DUrole{n,n}{g\_id}}\sphinxparamcomma \sphinxparam{\DUrole{n,n}{update\_reactions}\DUrole{o,o}{=}\DUrole{default_value}{False}}\sphinxparamcomma \sphinxparam{\DUrole{n,n}{lower}\DUrole{o,o}{=}\DUrole{default_value}{0.0}}\sphinxparamcomma \sphinxparam{\DUrole{n,n}{upper}\DUrole{o,o}{=}\DUrole{default_value}{0.0}}}{}
\pysigstopsignatures
\sphinxAtStartPar
Effectively deletes a gene by setting it’s inactive flag while optionally updating the GPR associated reactions
\begin{itemize}
\item {} 
\sphinxAtStartPar
\sphinxstyleemphasis{g\_id} a gene ID

\item {} 
\sphinxAtStartPar
\sphinxstyleemphasis{update\_reactions} {[}default=False{]} update the associated reactions fluxbounds

\item {} 
\sphinxAtStartPar
\sphinxstyleemphasis{lower} {[}default=0.0{]} the deactivated reaction lower bound

\item {} 
\sphinxAtStartPar
\sphinxstyleemphasis{upper} {[}default=0.0{]} the deactivated reaction upper bound

\end{itemize}

\end{fulllineitems}

\index{setModifiedDate() (Model method)@\spxentry{setModifiedDate()}\spxextra{Model method}}

\begin{fulllineitems}
\phantomsection\label{\detokenize{modules_doc:cbmpy.CBModel.Model.setModifiedDate}}
\pysigstartsignatures
\pysiglinewithargsret{\sphinxbfcode{\sphinxupquote{setModifiedDate}}}{\sphinxparam{\DUrole{n,n}{date}\DUrole{o,o}{=}\DUrole{default_value}{None}}}{}
\pysigstopsignatures
\sphinxAtStartPar
Set the model modification date: tuple(year, month, day, hour, minute, second)
\begin{itemize}
\item {} 
\sphinxAtStartPar
\sphinxstyleemphasis{date} {[}default=None{]} default is now (automatic) otherwise (year, month, day, hour, minute, second) e.g. (2012, 09, 24, 13, 34, 00)

\end{itemize}

\end{fulllineitems}

\index{setObjectiveFlux() (Model method)@\spxentry{setObjectiveFlux()}\spxextra{Model method}}

\begin{fulllineitems}
\phantomsection\label{\detokenize{modules_doc:cbmpy.CBModel.Model.setObjectiveFlux}}
\pysigstartsignatures
\pysiglinewithargsret{\sphinxbfcode{\sphinxupquote{setObjectiveFlux}}}{\sphinxparam{\DUrole{n,n}{rid}}\sphinxparamcomma \sphinxparam{\DUrole{n,n}{coefficient}\DUrole{o,o}{=}\DUrole{default_value}{1}}\sphinxparamcomma \sphinxparam{\DUrole{n,n}{osense}\DUrole{o,o}{=}\DUrole{default_value}{\textquotesingle{}maximize\textquotesingle{}}}\sphinxparamcomma \sphinxparam{\DUrole{n,n}{delete\_objflx}\DUrole{o,o}{=}\DUrole{default_value}{True}}}{}
\pysigstopsignatures
\sphinxAtStartPar
Set single target reaction flux for the current active objective function.
\begin{itemize}
\item {} 
\sphinxAtStartPar
\sphinxstyleemphasis{rid} a string containing a reaction id

\item {} 
\sphinxAtStartPar
\sphinxstyleemphasis{coefficient} {[}default=1{]} an objective flux coefficient

\item {} 
\sphinxAtStartPar
\sphinxstyleemphasis{osense} the optimization sense must be \sphinxstylestrong{maximize} or \sphinxstylestrong{minimize}

\item {} 
\sphinxAtStartPar
\sphinxstyleemphasis{delete\_objflx} {[}default=True{]} delete all existing fluxObjectives in the active objective function

\end{itemize}

\end{fulllineitems}

\index{setPrefix() (Model method)@\spxentry{setPrefix()}\spxextra{Model method}}

\begin{fulllineitems}
\phantomsection\label{\detokenize{modules_doc:cbmpy.CBModel.Model.setPrefix}}
\pysigstartsignatures
\pysiglinewithargsret{\sphinxbfcode{\sphinxupquote{setPrefix}}}{\sphinxparam{\DUrole{n,n}{prefix}}\sphinxparamcomma \sphinxparam{\DUrole{n,n}{target}}}{}
\pysigstopsignatures
\sphinxAtStartPar
This is alpha stuff, target can be:
\begin{itemize}
\item {} 
\sphinxAtStartPar
‘species’

\item {} 
\sphinxAtStartPar
‘reactions’

\item {} 
\sphinxAtStartPar
‘constraints’

\item {} 
\sphinxAtStartPar
‘objectives’

\item {} 
\sphinxAtStartPar
‘all’

\end{itemize}

\end{fulllineitems}

\index{setReactionBound() (Model method)@\spxentry{setReactionBound()}\spxextra{Model method}}

\begin{fulllineitems}
\phantomsection\label{\detokenize{modules_doc:cbmpy.CBModel.Model.setReactionBound}}
\pysigstartsignatures
\pysiglinewithargsret{\sphinxbfcode{\sphinxupquote{setReactionBound}}}{\sphinxparam{\DUrole{n,n}{rid}}\sphinxparamcomma \sphinxparam{\DUrole{n,n}{value}}\sphinxparamcomma \sphinxparam{\DUrole{n,n}{bound}}}{}
\pysigstopsignatures
\sphinxAtStartPar
Set a reaction bound
\begin{itemize}
\item {} 
\sphinxAtStartPar
\sphinxstyleemphasis{rid} the reactions id

\item {} 
\sphinxAtStartPar
\sphinxstyleemphasis{value} the new value

\item {} 
\sphinxAtStartPar
\sphinxstyleemphasis{bound} this is either ‘lower’ or ‘upper’, or ‘equal’

\end{itemize}

\end{fulllineitems}

\index{setReactionBounds() (Model method)@\spxentry{setReactionBounds()}\spxextra{Model method}}

\begin{fulllineitems}
\phantomsection\label{\detokenize{modules_doc:cbmpy.CBModel.Model.setReactionBounds}}
\pysigstartsignatures
\pysiglinewithargsret{\sphinxbfcode{\sphinxupquote{setReactionBounds}}}{\sphinxparam{\DUrole{n,n}{rid}}\sphinxparamcomma \sphinxparam{\DUrole{n,n}{lower}}\sphinxparamcomma \sphinxparam{\DUrole{n,n}{upper}}}{}
\pysigstopsignatures
\sphinxAtStartPar
Set both the upper and lower bound of a reaction:
\begin{itemize}
\item {} 
\sphinxAtStartPar
\sphinxstyleemphasis{rid} the good old reaction id

\item {} 
\sphinxAtStartPar
\sphinxstyleemphasis{lower} the lower bound value

\item {} 
\sphinxAtStartPar
\sphinxstyleemphasis{upper} the upper bound value

\end{itemize}

\end{fulllineitems}

\index{setReactionLowerBound() (Model method)@\spxentry{setReactionLowerBound()}\spxextra{Model method}}

\begin{fulllineitems}
\phantomsection\label{\detokenize{modules_doc:cbmpy.CBModel.Model.setReactionLowerBound}}
\pysigstartsignatures
\pysiglinewithargsret{\sphinxbfcode{\sphinxupquote{setReactionLowerBound}}}{\sphinxparam{\DUrole{n,n}{rid}}\sphinxparamcomma \sphinxparam{\DUrole{n,n}{value}}}{}
\pysigstopsignatures
\sphinxAtStartPar
Set a reactions lower bound (if it exists)
\begin{itemize}
\item {} 
\sphinxAtStartPar
\sphinxstyleemphasis{rid} the reactions id

\item {} 
\sphinxAtStartPar
\sphinxstyleemphasis{value} the new value

\end{itemize}

\end{fulllineitems}

\index{setReactionUpperBound() (Model method)@\spxentry{setReactionUpperBound()}\spxextra{Model method}}

\begin{fulllineitems}
\phantomsection\label{\detokenize{modules_doc:cbmpy.CBModel.Model.setReactionUpperBound}}
\pysigstartsignatures
\pysiglinewithargsret{\sphinxbfcode{\sphinxupquote{setReactionUpperBound}}}{\sphinxparam{\DUrole{n,n}{rid}}\sphinxparamcomma \sphinxparam{\DUrole{n,n}{value}}}{}
\pysigstopsignatures
\sphinxAtStartPar
Set a reactions upper bound (if it exists)
\begin{itemize}
\item {} 
\sphinxAtStartPar
\sphinxstyleemphasis{rid} the reaction id

\item {} 
\sphinxAtStartPar
\sphinxstyleemphasis{value} the new value

\end{itemize}

\end{fulllineitems}

\index{setSuffix() (Model method)@\spxentry{setSuffix()}\spxextra{Model method}}

\begin{fulllineitems}
\phantomsection\label{\detokenize{modules_doc:cbmpy.CBModel.Model.setSuffix}}
\pysigstartsignatures
\pysiglinewithargsret{\sphinxbfcode{\sphinxupquote{setSuffix}}}{\sphinxparam{\DUrole{n,n}{suffix}}\sphinxparamcomma \sphinxparam{\DUrole{n,n}{target}}}{}
\pysigstopsignatures
\sphinxAtStartPar
This is alpha stuff, target can be:
\begin{itemize}
\item {} 
\sphinxAtStartPar
‘species’

\item {} 
\sphinxAtStartPar
‘reactions’

\item {} 
\sphinxAtStartPar
‘constraints’

\item {} 
\sphinxAtStartPar
‘objectives’

\item {} 
\sphinxAtStartPar
‘all’

\end{itemize}

\end{fulllineitems}

\index{sortReactionsById() (Model method)@\spxentry{sortReactionsById()}\spxextra{Model method}}

\begin{fulllineitems}
\phantomsection\label{\detokenize{modules_doc:cbmpy.CBModel.Model.sortReactionsById}}
\pysigstartsignatures
\pysiglinewithargsret{\sphinxbfcode{\sphinxupquote{sortReactionsById}}}{}{}
\pysigstopsignatures
\sphinxAtStartPar
Sorts the reactions by Reaction.id uses the python string sort

\end{fulllineitems}

\index{sortSpeciesById() (Model method)@\spxentry{sortSpeciesById()}\spxextra{Model method}}

\begin{fulllineitems}
\phantomsection\label{\detokenize{modules_doc:cbmpy.CBModel.Model.sortSpeciesById}}
\pysigstartsignatures
\pysiglinewithargsret{\sphinxbfcode{\sphinxupquote{sortSpeciesById}}}{}{}
\pysigstopsignatures
\sphinxAtStartPar
Sorts the reaction list by Reaction.id uses the python string sort

\end{fulllineitems}

\index{splitEqualityFluxBounds() (Model method)@\spxentry{splitEqualityFluxBounds()}\spxextra{Model method}}

\begin{fulllineitems}
\phantomsection\label{\detokenize{modules_doc:cbmpy.CBModel.Model.splitEqualityFluxBounds}}
\pysigstartsignatures
\pysiglinewithargsret{\sphinxbfcode{\sphinxupquote{splitEqualityFluxBounds}}}{}{}
\pysigstopsignatures
\sphinxAtStartPar
Splits any equalit flux bounds into lower and upper bounds.

\end{fulllineitems}

\index{testGeneProteinAssociations() (Model method)@\spxentry{testGeneProteinAssociations()}\spxextra{Model method}}

\begin{fulllineitems}
\phantomsection\label{\detokenize{modules_doc:cbmpy.CBModel.Model.testGeneProteinAssociations}}
\pysigstartsignatures
\pysiglinewithargsret{\sphinxbfcode{\sphinxupquote{testGeneProteinAssociations}}}{}{}
\pysigstopsignatures
\sphinxAtStartPar
This method will test the GeneProtein associations and return a list of protein, association pairs

\end{fulllineitems}

\index{unRegisterObjectInGlobalStore() (Model method)@\spxentry{unRegisterObjectInGlobalStore()}\spxextra{Model method}}

\begin{fulllineitems}
\phantomsection\label{\detokenize{modules_doc:cbmpy.CBModel.Model.unRegisterObjectInGlobalStore}}
\pysigstartsignatures
\pysiglinewithargsret{\sphinxbfcode{\sphinxupquote{unRegisterObjectInGlobalStore}}}{\sphinxparam{\DUrole{n,n}{sid}}}{}
\pysigstopsignatures\begin{itemize}
\item {} 
\sphinxAtStartPar
\sphinxstyleemphasis{sid}

\end{itemize}

\end{fulllineitems}

\index{updateNetwork() (Model method)@\spxentry{updateNetwork()}\spxextra{Model method}}

\begin{fulllineitems}
\phantomsection\label{\detokenize{modules_doc:cbmpy.CBModel.Model.updateNetwork}}
\pysigstartsignatures
\pysiglinewithargsret{\sphinxbfcode{\sphinxupquote{updateNetwork}}}{\sphinxparam{\DUrole{n,n}{lower}\DUrole{o,o}{=}\DUrole{default_value}{0.0}}\sphinxparamcomma \sphinxparam{\DUrole{n,n}{upper}\DUrole{o,o}{=}\DUrole{default_value}{0.0}}\sphinxparamcomma \sphinxparam{\DUrole{n,n}{silent}\DUrole{o,o}{=}\DUrole{default_value}{False}}}{}
\pysigstopsignatures
\sphinxAtStartPar
Update the reaction network based on gene activity. If reaction is deactivated then lower and upper bounds are used
\begin{itemize}
\item {} 
\sphinxAtStartPar
\sphinxstyleemphasis{lower} {[}default=0.0{]} deactivated lower bound

\item {} 
\sphinxAtStartPar
\sphinxstyleemphasis{upper} {[}default=0.0{]} deactivated upper bound

\end{itemize}

\end{fulllineitems}


\end{fulllineitems}

\index{Objective (class in cbmpy.CBModel)@\spxentry{Objective}\spxextra{class in cbmpy.CBModel}}

\begin{fulllineitems}
\phantomsection\label{\detokenize{modules_doc:cbmpy.CBModel.Objective}}
\pysigstartsignatures
\pysiglinewithargsret{\sphinxbfcode{\sphinxupquote{class\DUrole{w,w}{  }}}\sphinxbfcode{\sphinxupquote{Objective}}}{\sphinxparam{\DUrole{n,n}{pid}}\sphinxparamcomma \sphinxparam{\DUrole{n,n}{operation}}}{}
\pysigstopsignatures
\sphinxAtStartPar
An objective function
\index{addFluxObjective() (Objective method)@\spxentry{addFluxObjective()}\spxextra{Objective method}}

\begin{fulllineitems}
\phantomsection\label{\detokenize{modules_doc:cbmpy.CBModel.Objective.addFluxObjective}}
\pysigstartsignatures
\pysiglinewithargsret{\sphinxbfcode{\sphinxupquote{addFluxObjective}}}{\sphinxparam{\DUrole{n,n}{fobj}}\sphinxparamcomma \sphinxparam{\DUrole{n,n}{override}\DUrole{o,o}{=}\DUrole{default_value}{False}}}{}
\pysigstopsignatures
\sphinxAtStartPar
Adds a FluxObjective instance to the Objective
\begin{itemize}
\item {} 
\sphinxAtStartPar
\sphinxstyleemphasis{fobj} the FluxObjective object

\item {} 
\sphinxAtStartPar
\sphinxstyleemphasis{override} {[}default=False{]} override pushing the global id map, this should never be used

\end{itemize}

\end{fulllineitems}

\index{createFluxObjectives() (Objective method)@\spxentry{createFluxObjectives()}\spxextra{Objective method}}

\begin{fulllineitems}
\phantomsection\label{\detokenize{modules_doc:cbmpy.CBModel.Objective.createFluxObjectives}}
\pysigstartsignatures
\pysiglinewithargsret{\sphinxbfcode{\sphinxupquote{createFluxObjectives}}}{\sphinxparam{\DUrole{n,n}{fluxlist}}}{}
\pysigstopsignatures
\sphinxAtStartPar
Create and add flux objective objects to this objective function.
\begin{itemize}
\item {} 
\sphinxAtStartPar
\sphinxstyleemphasis{fluxlist} a list of one or more (‘coefficient’, ‘rid’, ‘type’) triples

\end{itemize}

\end{fulllineitems}

\index{createQuadraticFluxObjectives() (Objective method)@\spxentry{createQuadraticFluxObjectives()}\spxextra{Objective method}}

\begin{fulllineitems}
\phantomsection\label{\detokenize{modules_doc:cbmpy.CBModel.Objective.createQuadraticFluxObjectives}}
\pysigstartsignatures
\pysiglinewithargsret{\sphinxbfcode{\sphinxupquote{createQuadraticFluxObjectives}}}{\sphinxparam{\DUrole{n,n}{fluxlist}}}{}
\pysigstopsignatures
\sphinxAtStartPar
Create and add quadratic flux objective objects to this objective function.
\begin{itemize}
\item {} 
\sphinxAtStartPar
\sphinxstyleemphasis{fluxlist} a list of one or more (‘coefficient’, ‘rid’, ‘rid2’, ‘type’) triples

\end{itemize}

\end{fulllineitems}

\index{deleteAllFluxObjectives() (Objective method)@\spxentry{deleteAllFluxObjectives()}\spxextra{Objective method}}

\begin{fulllineitems}
\phantomsection\label{\detokenize{modules_doc:cbmpy.CBModel.Objective.deleteAllFluxObjectives}}
\pysigstartsignatures
\pysiglinewithargsret{\sphinxbfcode{\sphinxupquote{deleteAllFluxObjectives}}}{}{}
\pysigstopsignatures
\sphinxAtStartPar
Delete all flux objectives

\end{fulllineitems}

\index{getFluxObjective() (Objective method)@\spxentry{getFluxObjective()}\spxextra{Objective method}}

\begin{fulllineitems}
\phantomsection\label{\detokenize{modules_doc:cbmpy.CBModel.Objective.getFluxObjective}}
\pysigstartsignatures
\pysiglinewithargsret{\sphinxbfcode{\sphinxupquote{getFluxObjective}}}{\sphinxparam{\DUrole{n,n}{foid}}}{}
\pysigstopsignatures
\sphinxAtStartPar
Return the flux objective with id.
\begin{itemize}
\item {} 
\sphinxAtStartPar
\sphinxstyleemphasis{foid} the flux objective id returns either an object or a list if there are multiply defined flux objectives

\end{itemize}

\end{fulllineitems}

\index{getFluxObjectiveData() (Objective method)@\spxentry{getFluxObjectiveData()}\spxextra{Objective method}}

\begin{fulllineitems}
\phantomsection\label{\detokenize{modules_doc:cbmpy.CBModel.Objective.getFluxObjectiveData}}
\pysigstartsignatures
\pysiglinewithargsret{\sphinxbfcode{\sphinxupquote{getFluxObjectiveData}}}{}{}
\pysigstopsignatures
\sphinxAtStartPar
Returns a list of ObjectiveFunction components as (coefficient, flux, type) pairs

\end{fulllineitems}

\index{getFluxObjectiveForReaction() (Objective method)@\spxentry{getFluxObjectiveForReaction()}\spxextra{Objective method}}

\begin{fulllineitems}
\phantomsection\label{\detokenize{modules_doc:cbmpy.CBModel.Objective.getFluxObjectiveForReaction}}
\pysigstartsignatures
\pysiglinewithargsret{\sphinxbfcode{\sphinxupquote{getFluxObjectiveForReaction}}}{\sphinxparam{\DUrole{n,n}{rid}}}{}
\pysigstopsignatures
\sphinxAtStartPar
Returns the FluxObjective associated with the suplied rid. If there is more than fluxObjective associated with a reaction (illegal)
then a list of fluxObjectives is returned.
\begin{quote}

\sphinxAtStartPar
\sphinxstyleemphasis{rid} a reaction id
\end{quote}

\end{fulllineitems}

\index{getFluxObjectiveIDs() (Objective method)@\spxentry{getFluxObjectiveIDs()}\spxextra{Objective method}}

\begin{fulllineitems}
\phantomsection\label{\detokenize{modules_doc:cbmpy.CBModel.Objective.getFluxObjectiveIDs}}
\pysigstartsignatures
\pysiglinewithargsret{\sphinxbfcode{\sphinxupquote{getFluxObjectiveIDs}}}{}{}
\pysigstopsignatures
\sphinxAtStartPar
Returns a list of ObjectiveFlux ids, for the reaction id’s use \sphinxstyleemphasis{getFluxObjectiveReactions()}
or for coefficient, fluxobjective pairs use \sphinxstyleemphasis{getFluxObjectiveData()}

\end{fulllineitems}

\index{getFluxObjectiveReactions() (Objective method)@\spxentry{getFluxObjectiveReactions()}\spxextra{Objective method}}

\begin{fulllineitems}
\phantomsection\label{\detokenize{modules_doc:cbmpy.CBModel.Objective.getFluxObjectiveReactions}}
\pysigstartsignatures
\pysiglinewithargsret{\sphinxbfcode{\sphinxupquote{getFluxObjectiveReactions}}}{}{}
\pysigstopsignatures
\sphinxAtStartPar
Returns a list of reactions that are used as flux\_objectives

\end{fulllineitems}

\index{getFluxObjectives() (Objective method)@\spxentry{getFluxObjectives()}\spxextra{Objective method}}

\begin{fulllineitems}
\phantomsection\label{\detokenize{modules_doc:cbmpy.CBModel.Objective.getFluxObjectives}}
\pysigstartsignatures
\pysiglinewithargsret{\sphinxbfcode{\sphinxupquote{getFluxObjectives}}}{}{}
\pysigstopsignatures
\sphinxAtStartPar
Returns the list of FluxObjective objects.

\end{fulllineitems}

\index{getLinearFluxObjectives() (Objective method)@\spxentry{getLinearFluxObjectives()}\spxextra{Objective method}}

\begin{fulllineitems}
\phantomsection\label{\detokenize{modules_doc:cbmpy.CBModel.Objective.getLinearFluxObjectives}}
\pysigstartsignatures
\pysiglinewithargsret{\sphinxbfcode{\sphinxupquote{getLinearFluxObjectives}}}{}{}
\pysigstopsignatures
\sphinxAtStartPar
Returns a list of linear variable flux objective objects

\end{fulllineitems}

\index{getOperation() (Objective method)@\spxentry{getOperation()}\spxextra{Objective method}}

\begin{fulllineitems}
\phantomsection\label{\detokenize{modules_doc:cbmpy.CBModel.Objective.getOperation}}
\pysigstartsignatures
\pysiglinewithargsret{\sphinxbfcode{\sphinxupquote{getOperation}}}{}{}
\pysigstopsignatures
\sphinxAtStartPar
Returns the operation or sense of the objective

\end{fulllineitems}

\index{getQuadraticBivariateFluxObjectives() (Objective method)@\spxentry{getQuadraticBivariateFluxObjectives()}\spxextra{Objective method}}

\begin{fulllineitems}
\phantomsection\label{\detokenize{modules_doc:cbmpy.CBModel.Objective.getQuadraticBivariateFluxObjectives}}
\pysigstartsignatures
\pysiglinewithargsret{\sphinxbfcode{\sphinxupquote{getQuadraticBivariateFluxObjectives}}}{}{}
\pysigstopsignatures
\sphinxAtStartPar
Returns a list of bivariate quadratic variable flux objective objects

\end{fulllineitems}

\index{getQuadraticFluxObjectives() (Objective method)@\spxentry{getQuadraticFluxObjectives()}\spxextra{Objective method}}

\begin{fulllineitems}
\phantomsection\label{\detokenize{modules_doc:cbmpy.CBModel.Objective.getQuadraticFluxObjectives}}
\pysigstartsignatures
\pysiglinewithargsret{\sphinxbfcode{\sphinxupquote{getQuadraticFluxObjectives}}}{}{}
\pysigstopsignatures
\sphinxAtStartPar
Returns a list of quadratic variable flux objective objects

\end{fulllineitems}

\index{getValue() (Objective method)@\spxentry{getValue()}\spxextra{Objective method}}

\begin{fulllineitems}
\phantomsection\label{\detokenize{modules_doc:cbmpy.CBModel.Objective.getValue}}
\pysigstartsignatures
\pysiglinewithargsret{\sphinxbfcode{\sphinxupquote{getValue}}}{}{}
\pysigstopsignatures
\sphinxAtStartPar
Returns the current value of the attribute (input/solution)

\end{fulllineitems}

\index{setOperation() (Objective method)@\spxentry{setOperation()}\spxextra{Objective method}}

\begin{fulllineitems}
\phantomsection\label{\detokenize{modules_doc:cbmpy.CBModel.Objective.setOperation}}
\pysigstartsignatures
\pysiglinewithargsret{\sphinxbfcode{\sphinxupquote{setOperation}}}{\sphinxparam{\DUrole{n,n}{operation}}}{}
\pysigstopsignatures
\sphinxAtStartPar
Sets the objective operation (sense)
\begin{itemize}
\item {} 
\sphinxAtStartPar
\sphinxstyleemphasis{operation} {[}default=’maximize’{]} one of ‘maximize’, ‘maximise’, ‘max’, ‘minimize’, ‘minimise’, ‘min’

\end{itemize}

\end{fulllineitems}

\index{setValue() (Objective method)@\spxentry{setValue()}\spxextra{Objective method}}

\begin{fulllineitems}
\phantomsection\label{\detokenize{modules_doc:cbmpy.CBModel.Objective.setValue}}
\pysigstartsignatures
\pysiglinewithargsret{\sphinxbfcode{\sphinxupquote{setValue}}}{\sphinxparam{\DUrole{n,n}{value}}}{}
\pysigstopsignatures
\sphinxAtStartPar
Sets the attribute ‘’value’’

\end{fulllineitems}


\end{fulllineitems}

\index{Parameter (class in cbmpy.CBModel)@\spxentry{Parameter}\spxextra{class in cbmpy.CBModel}}

\begin{fulllineitems}
\phantomsection\label{\detokenize{modules_doc:cbmpy.CBModel.Parameter}}
\pysigstartsignatures
\pysiglinewithargsret{\sphinxbfcode{\sphinxupquote{class\DUrole{w,w}{  }}}\sphinxbfcode{\sphinxupquote{Parameter}}}{\sphinxparam{\DUrole{n,n}{pid}}\sphinxparamcomma \sphinxparam{\DUrole{n,n}{value}}\sphinxparamcomma \sphinxparam{\DUrole{n,n}{name}\DUrole{o,o}{=}\DUrole{default_value}{None}}\sphinxparamcomma \sphinxparam{\DUrole{n,n}{constant}\DUrole{o,o}{=}\DUrole{default_value}{True}}}{}
\pysigstopsignatures
\sphinxAtStartPar
Holds parameter information
\index{addAssociation() (Parameter method)@\spxentry{addAssociation()}\spxextra{Parameter method}}

\begin{fulllineitems}
\phantomsection\label{\detokenize{modules_doc:cbmpy.CBModel.Parameter.addAssociation}}
\pysigstartsignatures
\pysiglinewithargsret{\sphinxbfcode{\sphinxupquote{addAssociation}}}{\sphinxparam{\DUrole{n,n}{assoc}}}{}
\pysigstopsignatures
\sphinxAtStartPar
Add an object ID to associate with this object

\end{fulllineitems}

\index{deleteAssociation() (Parameter method)@\spxentry{deleteAssociation()}\spxextra{Parameter method}}

\begin{fulllineitems}
\phantomsection\label{\detokenize{modules_doc:cbmpy.CBModel.Parameter.deleteAssociation}}
\pysigstartsignatures
\pysiglinewithargsret{\sphinxbfcode{\sphinxupquote{deleteAssociation}}}{\sphinxparam{\DUrole{n,n}{assoc}}}{}
\pysigstopsignatures
\sphinxAtStartPar
Delete the object id associated with this object

\end{fulllineitems}

\index{getAssociations() (Parameter method)@\spxentry{getAssociations()}\spxextra{Parameter method}}

\begin{fulllineitems}
\phantomsection\label{\detokenize{modules_doc:cbmpy.CBModel.Parameter.getAssociations}}
\pysigstartsignatures
\pysiglinewithargsret{\sphinxbfcode{\sphinxupquote{getAssociations}}}{}{}
\pysigstopsignatures
\sphinxAtStartPar
Return the Object ID’s associated with this parameter

\end{fulllineitems}

\index{getValue() (Parameter method)@\spxentry{getValue()}\spxextra{Parameter method}}

\begin{fulllineitems}
\phantomsection\label{\detokenize{modules_doc:cbmpy.CBModel.Parameter.getValue}}
\pysigstartsignatures
\pysiglinewithargsret{\sphinxbfcode{\sphinxupquote{getValue}}}{}{}
\pysigstopsignatures
\sphinxAtStartPar
Returns the current value of the attribute (input/solution)

\end{fulllineitems}

\index{setValue() (Parameter method)@\spxentry{setValue()}\spxextra{Parameter method}}

\begin{fulllineitems}
\phantomsection\label{\detokenize{modules_doc:cbmpy.CBModel.Parameter.setValue}}
\pysigstartsignatures
\pysiglinewithargsret{\sphinxbfcode{\sphinxupquote{setValue}}}{\sphinxparam{\DUrole{n,n}{value}}}{}
\pysigstopsignatures
\sphinxAtStartPar
Sets the value attribute:
\begin{itemize}
\item {} 
\sphinxAtStartPar
\sphinxstyleemphasis{value} a float

\end{itemize}

\end{fulllineitems}

\index{value (Parameter property)@\spxentry{value}\spxextra{Parameter property}}

\begin{fulllineitems}
\phantomsection\label{\detokenize{modules_doc:cbmpy.CBModel.Parameter.value}}
\pysigstartsignatures
\pysigline{\sphinxbfcode{\sphinxupquote{property\DUrole{w,w}{  }}}\sphinxbfcode{\sphinxupquote{value}}}
\pysigstopsignatures
\sphinxAtStartPar
Returns the current value of the attribute (input/solution)

\end{fulllineitems}


\end{fulllineitems}

\index{Reaction (class in cbmpy.CBModel)@\spxentry{Reaction}\spxextra{class in cbmpy.CBModel}}

\begin{fulllineitems}
\phantomsection\label{\detokenize{modules_doc:cbmpy.CBModel.Reaction}}
\pysigstartsignatures
\pysiglinewithargsret{\sphinxbfcode{\sphinxupquote{class\DUrole{w,w}{  }}}\sphinxbfcode{\sphinxupquote{Reaction}}}{\sphinxparam{\DUrole{n,n}{pid}}\sphinxparamcomma \sphinxparam{\DUrole{n,n}{name}\DUrole{o,o}{=}\DUrole{default_value}{None}}\sphinxparamcomma \sphinxparam{\DUrole{n,n}{reversible}\DUrole{o,o}{=}\DUrole{default_value}{True}}}{}
\pysigstopsignatures
\sphinxAtStartPar
Holds reaction information
\index{addReagent() (Reaction method)@\spxentry{addReagent()}\spxextra{Reaction method}}

\begin{fulllineitems}
\phantomsection\label{\detokenize{modules_doc:cbmpy.CBModel.Reaction.addReagent}}
\pysigstartsignatures
\pysiglinewithargsret{\sphinxbfcode{\sphinxupquote{addReagent}}}{\sphinxparam{\DUrole{n,n}{reag}}}{}
\pysigstopsignatures
\sphinxAtStartPar
Adds an instantiated Reagent object to the reaction

\end{fulllineitems}

\index{changeId() (Reaction method)@\spxentry{changeId()}\spxextra{Reaction method}}

\begin{fulllineitems}
\phantomsection\label{\detokenize{modules_doc:cbmpy.CBModel.Reaction.changeId}}
\pysigstartsignatures
\pysiglinewithargsret{\sphinxbfcode{\sphinxupquote{changeId}}}{\sphinxparam{\DUrole{n,n}{pid}}}{}
\pysigstopsignatures
\sphinxAtStartPar
Changes the Id of the reaction and updates associated FluxBounds

\end{fulllineitems}

\index{createReagent() (Reaction method)@\spxentry{createReagent()}\spxextra{Reaction method}}

\begin{fulllineitems}
\phantomsection\label{\detokenize{modules_doc:cbmpy.CBModel.Reaction.createReagent}}
\pysigstartsignatures
\pysiglinewithargsret{\sphinxbfcode{\sphinxupquote{createReagent}}}{\sphinxparam{\DUrole{n,n}{metabolite}}\sphinxparamcomma \sphinxparam{\DUrole{n,n}{coefficient}}}{}
\pysigstopsignatures
\sphinxAtStartPar
Create a new reagent and add it to the reaction:
\begin{quote}
\begin{itemize}
\item {} 
\sphinxAtStartPar
\sphinxstylestrong{metabolite} the metabolite name

\item {} 
\sphinxAtStartPar
\sphinxstylestrong{coefficient} the

\end{itemize}

\begin{flushright}
---negative coefficient is a substrate
\textendash{} positive coefficient is a product
\end{flushright}
\end{quote}

\sphinxAtStartPar
Will fail if a species reference already exists

\end{fulllineitems}

\index{deactivateReaction() (Reaction method)@\spxentry{deactivateReaction()}\spxextra{Reaction method}}

\begin{fulllineitems}
\phantomsection\label{\detokenize{modules_doc:cbmpy.CBModel.Reaction.deactivateReaction}}
\pysigstartsignatures
\pysiglinewithargsret{\sphinxbfcode{\sphinxupquote{deactivateReaction}}}{\sphinxparam{\DUrole{n,n}{lower}\DUrole{o,o}{=}\DUrole{default_value}{0.0}}\sphinxparamcomma \sphinxparam{\DUrole{n,n}{upper}\DUrole{o,o}{=}\DUrole{default_value}{0.0}}\sphinxparamcomma \sphinxparam{\DUrole{n,n}{silent}\DUrole{o,o}{=}\DUrole{default_value}{True}}}{}
\pysigstopsignatures
\sphinxAtStartPar
Deactivates a reaction by setting its bounds to lower and upper. Restore with reactivateReaction()
\begin{itemize}
\item {} 
\sphinxAtStartPar
\sphinxstyleemphasis{lower} {[}default=0.0{]} bound

\item {} 
\sphinxAtStartPar
\sphinxstyleemphasis{upper} {[}default=0.0{]} bound

\end{itemize}

\end{fulllineitems}

\index{deleteReagentWithSpeciesRef() (Reaction method)@\spxentry{deleteReagentWithSpeciesRef()}\spxextra{Reaction method}}

\begin{fulllineitems}
\phantomsection\label{\detokenize{modules_doc:cbmpy.CBModel.Reaction.deleteReagentWithSpeciesRef}}
\pysigstartsignatures
\pysiglinewithargsret{\sphinxbfcode{\sphinxupquote{deleteReagentWithSpeciesRef}}}{\sphinxparam{\DUrole{n,n}{sid}}}{}
\pysigstopsignatures
\sphinxAtStartPar
Delete a reagent (or reagents) that refers to the species id:
\begin{itemize}
\item {} 
\sphinxAtStartPar
\sphinxstyleemphasis{sid} a species/metabolite id

\end{itemize}

\end{fulllineitems}

\index{getEquation() (Reaction method)@\spxentry{getEquation()}\spxextra{Reaction method}}

\begin{fulllineitems}
\phantomsection\label{\detokenize{modules_doc:cbmpy.CBModel.Reaction.getEquation}}
\pysigstartsignatures
\pysiglinewithargsret{\sphinxbfcode{\sphinxupquote{getEquation}}}{\sphinxparam{\DUrole{n,n}{reverse\_symb}\DUrole{o,o}{=}\DUrole{default_value}{\textquotesingle{}=\textquotesingle{}}}\sphinxparamcomma \sphinxparam{\DUrole{n,n}{irreverse\_symb}\DUrole{o,o}{=}\DUrole{default_value}{\textquotesingle{}\textgreater{}\textquotesingle{}}}\sphinxparamcomma \sphinxparam{\DUrole{n,n}{use\_names}\DUrole{o,o}{=}\DUrole{default_value}{False}}}{}
\pysigstopsignatures
\sphinxAtStartPar
Return a pretty printed string containing the reaction equation
\begin{itemize}
\item {} 
\sphinxAtStartPar
\sphinxstyleemphasis{reverse\_symb} {[}default = ‘=’{]} the symbol to use for reversible reactions

\item {} 
\sphinxAtStartPar
\sphinxstyleemphasis{irreverse\_symb} {[}default = ‘\textgreater{}’{]} the symbol to use for irreversible reactions

\item {} 
\sphinxAtStartPar
\sphinxstyleemphasis{use\_names} {[}defualt = False{]} use species names rather than id’s

\end{itemize}

\end{fulllineitems}

\index{getFVAdata() (Reaction method)@\spxentry{getFVAdata()}\spxextra{Reaction method}}

\begin{fulllineitems}
\phantomsection\label{\detokenize{modules_doc:cbmpy.CBModel.Reaction.getFVAdata}}
\pysigstartsignatures
\pysiglinewithargsret{\sphinxbfcode{\sphinxupquote{getFVAdata}}}{\sphinxparam{\DUrole{n,n}{roundnum}\DUrole{o,o}{=}\DUrole{default_value}{None}}\sphinxparamcomma \sphinxparam{\DUrole{n,n}{silent}\DUrole{o,o}{=}\DUrole{default_value}{True}}}{}
\pysigstopsignatures
\sphinxAtStartPar
Returns the data generated by CBSolver.FluxVariabilityAnalysis() for this reaction as a tuple of
(Flux, FVAmin, FVAmax, span) where span is abs(FVAmax \sphinxhyphen{} FVAmin). FVAmin or FVAmax is None this indicates no solution
to that particular optimization (infeasible).
\begin{itemize}
\item {} 
\sphinxAtStartPar
\sphinxstyleemphasis{roundnum} {[}default=None{]} the integer number of roundoff decimals the default is no rounding

\item {} 
\sphinxAtStartPar
\sphinxstyleemphasis{silent} {[}default=True{]} supress output to stdout

\end{itemize}

\end{fulllineitems}

\index{getGPRassociationString() (Reaction method)@\spxentry{getGPRassociationString()}\spxextra{Reaction method}}

\begin{fulllineitems}
\phantomsection\label{\detokenize{modules_doc:cbmpy.CBModel.Reaction.getGPRassociationString}}
\pysigstartsignatures
\pysiglinewithargsret{\sphinxbfcode{\sphinxupquote{getGPRassociationString}}}{\sphinxparam{\DUrole{n,n}{use\_labels}\DUrole{o,o}{=}\DUrole{default_value}{True}}}{}
\pysigstopsignatures
\sphinxAtStartPar
Return the GPR string associated with this reaction (assuming it exists) or None.
\begin{itemize}
\item {} 
\sphinxAtStartPar
\sphinxstyleemphasis{use\_labels} {[}default=True{]} return string with lab

\end{itemize}

\end{fulllineitems}

\index{getLowerBound() (Reaction method)@\spxentry{getLowerBound()}\spxextra{Reaction method}}

\begin{fulllineitems}
\phantomsection\label{\detokenize{modules_doc:cbmpy.CBModel.Reaction.getLowerBound}}
\pysigstartsignatures
\pysiglinewithargsret{\sphinxbfcode{\sphinxupquote{getLowerBound}}}{}{}
\pysigstopsignatures
\sphinxAtStartPar
Get the value of the reactions lower bound

\end{fulllineitems}

\index{getProductIds() (Reaction method)@\spxentry{getProductIds()}\spxextra{Reaction method}}

\begin{fulllineitems}
\phantomsection\label{\detokenize{modules_doc:cbmpy.CBModel.Reaction.getProductIds}}
\pysigstartsignatures
\pysiglinewithargsret{\sphinxbfcode{\sphinxupquote{getProductIds}}}{\sphinxparam{\DUrole{n,n}{use\_names}\DUrole{o,o}{=}\DUrole{default_value}{False}}}{}
\pysigstopsignatures
\sphinxAtStartPar
Returns a list of the reaction products, species identifiers
\begin{itemize}
\item {} 
\sphinxAtStartPar
\sphinxstyleemphasis{use\_names} {[}default = False{]} use species names rather than id’s

\end{itemize}

\end{fulllineitems}

\index{getReagent() (Reaction method)@\spxentry{getReagent()}\spxextra{Reaction method}}

\begin{fulllineitems}
\phantomsection\label{\detokenize{modules_doc:cbmpy.CBModel.Reaction.getReagent}}
\pysigstartsignatures
\pysiglinewithargsret{\sphinxbfcode{\sphinxupquote{getReagent}}}{\sphinxparam{\DUrole{n,n}{rid}}}{}
\pysigstopsignatures
\sphinxAtStartPar
Return the one or more reagent objects which have \sphinxstyleemphasis{rid}:
\begin{itemize}
\item {} 
\sphinxAtStartPar
\sphinxstyleemphasis{rid} a reagent \sphinxstyleemphasis{rid}

\end{itemize}

\end{fulllineitems}

\index{getReagentObjIds() (Reaction method)@\spxentry{getReagentObjIds()}\spxextra{Reaction method}}

\begin{fulllineitems}
\phantomsection\label{\detokenize{modules_doc:cbmpy.CBModel.Reaction.getReagentObjIds}}
\pysigstartsignatures
\pysiglinewithargsret{\sphinxbfcode{\sphinxupquote{getReagentObjIds}}}{}{}
\pysigstopsignatures
\sphinxAtStartPar
Returns a list of the reagent id’s. For the name of the reagents/metabolites use \sphinxstyleemphasis{\textless{}reaction\textgreater{}.getSpeciesIds()}

\end{fulllineitems}

\index{getReagentRefs() (Reaction method)@\spxentry{getReagentRefs()}\spxextra{Reaction method}}

\begin{fulllineitems}
\phantomsection\label{\detokenize{modules_doc:cbmpy.CBModel.Reaction.getReagentRefs}}
\pysigstartsignatures
\pysiglinewithargsret{\sphinxbfcode{\sphinxupquote{getReagentRefs}}}{}{}
\pysigstopsignatures
\sphinxAtStartPar
Returns a list of the reagents/metabolites

\end{fulllineitems}

\index{getReagentWithSpeciesRef() (Reaction method)@\spxentry{getReagentWithSpeciesRef()}\spxextra{Reaction method}}

\begin{fulllineitems}
\phantomsection\label{\detokenize{modules_doc:cbmpy.CBModel.Reaction.getReagentWithSpeciesRef}}
\pysigstartsignatures
\pysiglinewithargsret{\sphinxbfcode{\sphinxupquote{getReagentWithSpeciesRef}}}{\sphinxparam{\DUrole{n,n}{sid}}}{}
\pysigstopsignatures
\sphinxAtStartPar
Return the reagent object which refers to the \sphinxstyleemphasis{species} id. If there are multiple reagents that
refer to the same species a list is returned.
\begin{itemize}
\item {} 
\sphinxAtStartPar
\sphinxstyleemphasis{sid} the species/metabolite id

\end{itemize}

\end{fulllineitems}

\index{getSpeciesIds() (Reaction method)@\spxentry{getSpeciesIds()}\spxextra{Reaction method}}

\begin{fulllineitems}
\phantomsection\label{\detokenize{modules_doc:cbmpy.CBModel.Reaction.getSpeciesIds}}
\pysigstartsignatures
\pysiglinewithargsret{\sphinxbfcode{\sphinxupquote{getSpeciesIds}}}{}{}
\pysigstopsignatures
\sphinxAtStartPar
Returns a list of the reagents/metabolites

\end{fulllineitems}

\index{getSpeciesObj() (Reaction method)@\spxentry{getSpeciesObj()}\spxextra{Reaction method}}

\begin{fulllineitems}
\phantomsection\label{\detokenize{modules_doc:cbmpy.CBModel.Reaction.getSpeciesObj}}
\pysigstartsignatures
\pysiglinewithargsret{\sphinxbfcode{\sphinxupquote{getSpeciesObj}}}{}{}
\pysigstopsignatures
\sphinxAtStartPar
Returns a list of the species objects that are reagents

\end{fulllineitems}

\index{getStoichiometry() (Reaction method)@\spxentry{getStoichiometry()}\spxextra{Reaction method}}

\begin{fulllineitems}
\phantomsection\label{\detokenize{modules_doc:cbmpy.CBModel.Reaction.getStoichiometry}}
\pysigstartsignatures
\pysiglinewithargsret{\sphinxbfcode{\sphinxupquote{getStoichiometry}}}{\sphinxparam{\DUrole{n,n}{use\_names}\DUrole{o,o}{=}\DUrole{default_value}{False}}\sphinxparamcomma \sphinxparam{\DUrole{n,n}{altout}\DUrole{o,o}{=}\DUrole{default_value}{False}}}{}
\pysigstopsignatures
\sphinxAtStartPar
Returns a list of (coefficient, species) pairs for this reaction
\begin{itemize}
\item {} 
\sphinxAtStartPar
\sphinxstyleemphasis{use\_names} {[}default = False{]} use species names rather than id’s

\item {} 
\sphinxAtStartPar
\sphinxstyleemphasis{altout} {[}default = False{]} returns a dictionary {[}DEPRECATED{]}

\end{itemize}

\end{fulllineitems}

\index{getSubstrateIds() (Reaction method)@\spxentry{getSubstrateIds()}\spxextra{Reaction method}}

\begin{fulllineitems}
\phantomsection\label{\detokenize{modules_doc:cbmpy.CBModel.Reaction.getSubstrateIds}}
\pysigstartsignatures
\pysiglinewithargsret{\sphinxbfcode{\sphinxupquote{getSubstrateIds}}}{\sphinxparam{\DUrole{n,n}{use\_names}\DUrole{o,o}{=}\DUrole{default_value}{False}}}{}
\pysigstopsignatures
\sphinxAtStartPar
Returns a list of the reaction substrates, species identifiers
\begin{itemize}
\item {} 
\sphinxAtStartPar
\sphinxstyleemphasis{use\_names} {[}defualt = False{]} use species names rather than id’s

\end{itemize}

\end{fulllineitems}

\index{getUpperBound() (Reaction method)@\spxentry{getUpperBound()}\spxextra{Reaction method}}

\begin{fulllineitems}
\phantomsection\label{\detokenize{modules_doc:cbmpy.CBModel.Reaction.getUpperBound}}
\pysigstartsignatures
\pysiglinewithargsret{\sphinxbfcode{\sphinxupquote{getUpperBound}}}{}{}
\pysigstopsignatures
\sphinxAtStartPar
Get the value of the reactions upper bound

\end{fulllineitems}

\index{getValue() (Reaction method)@\spxentry{getValue()}\spxextra{Reaction method}}

\begin{fulllineitems}
\phantomsection\label{\detokenize{modules_doc:cbmpy.CBModel.Reaction.getValue}}
\pysigstartsignatures
\pysiglinewithargsret{\sphinxbfcode{\sphinxupquote{getValue}}}{}{}
\pysigstopsignatures
\sphinxAtStartPar
Returns the current value of the flux.

\end{fulllineitems}

\index{reactivateReaction() (Reaction method)@\spxentry{reactivateReaction()}\spxextra{Reaction method}}

\begin{fulllineitems}
\phantomsection\label{\detokenize{modules_doc:cbmpy.CBModel.Reaction.reactivateReaction}}
\pysigstartsignatures
\pysiglinewithargsret{\sphinxbfcode{\sphinxupquote{reactivateReaction}}}{\sphinxparam{\DUrole{n,n}{silent}\DUrole{o,o}{=}\DUrole{default_value}{True}}}{}
\pysigstopsignatures
\sphinxAtStartPar
Activates a reaction deactivated with deactivateReaction

\end{fulllineitems}

\index{setId() (Reaction method)@\spxentry{setId()}\spxextra{Reaction method}}

\begin{fulllineitems}
\phantomsection\label{\detokenize{modules_doc:cbmpy.CBModel.Reaction.setId}}
\pysigstartsignatures
\pysiglinewithargsret{\sphinxbfcode{\sphinxupquote{setId}}}{\sphinxparam{\DUrole{n,n}{fid}}}{}
\pysigstopsignatures
\sphinxAtStartPar
Sets the object Id
\begin{quote}
\begin{itemize}
\item {} 
\sphinxAtStartPar
\sphinxstyleemphasis{fid} a valid c variable style id string

\end{itemize}

\sphinxAtStartPar
Reimplements @FBase.setId()
\end{quote}

\end{fulllineitems}

\index{setLowerBound() (Reaction method)@\spxentry{setLowerBound()}\spxextra{Reaction method}}

\begin{fulllineitems}
\phantomsection\label{\detokenize{modules_doc:cbmpy.CBModel.Reaction.setLowerBound}}
\pysigstartsignatures
\pysiglinewithargsret{\sphinxbfcode{\sphinxupquote{setLowerBound}}}{\sphinxparam{\DUrole{n,n}{value}}}{}
\pysigstopsignatures
\sphinxAtStartPar
Set the value of the reactions lower bound
\begin{itemize}
\item {} 
\sphinxAtStartPar
\sphinxstyleemphasis{value} a floating point value

\end{itemize}

\end{fulllineitems}

\index{setStoichCoefficient() (Reaction method)@\spxentry{setStoichCoefficient()}\spxextra{Reaction method}}

\begin{fulllineitems}
\phantomsection\label{\detokenize{modules_doc:cbmpy.CBModel.Reaction.setStoichCoefficient}}
\pysigstartsignatures
\pysiglinewithargsret{\sphinxbfcode{\sphinxupquote{setStoichCoefficient}}}{\sphinxparam{\DUrole{n,n}{sid}}\sphinxparamcomma \sphinxparam{\DUrole{n,n}{value}}}{}
\pysigstopsignatures
\sphinxAtStartPar
Sets the stoichiometric coefficient of a reagent that refers to a metabolite. Note \sphinxstyleemphasis{negative coefficients} are \sphinxstyleemphasis{substrates}
while \sphinxstyleemphasis{positive} ones are \sphinxstyleemphasis{products}. At this point zero coefficients are not allowed
\begin{itemize}
\item {} 
\sphinxAtStartPar
\sphinxstyleemphasis{sid} the species/metabolite id

\item {} 
\sphinxAtStartPar
\sphinxstyleemphasis{value} a floating point value != 0

\end{itemize}

\end{fulllineitems}

\index{setUpperBound() (Reaction method)@\spxentry{setUpperBound()}\spxextra{Reaction method}}

\begin{fulllineitems}
\phantomsection\label{\detokenize{modules_doc:cbmpy.CBModel.Reaction.setUpperBound}}
\pysigstartsignatures
\pysiglinewithargsret{\sphinxbfcode{\sphinxupquote{setUpperBound}}}{\sphinxparam{\DUrole{n,n}{value}}}{}
\pysigstopsignatures
\sphinxAtStartPar
Set the value of the reactions upper bound
\begin{itemize}
\item {} 
\sphinxAtStartPar
\sphinxstyleemphasis{value} a floating point value

\end{itemize}

\end{fulllineitems}

\index{setValue() (Reaction method)@\spxentry{setValue()}\spxextra{Reaction method}}

\begin{fulllineitems}
\phantomsection\label{\detokenize{modules_doc:cbmpy.CBModel.Reaction.setValue}}
\pysigstartsignatures
\pysiglinewithargsret{\sphinxbfcode{\sphinxupquote{setValue}}}{\sphinxparam{\DUrole{n,n}{value}}}{}
\pysigstopsignatures
\sphinxAtStartPar
Sets the attribute \sphinxstyleemphasis{value} in this case the flux.

\end{fulllineitems}


\end{fulllineitems}

\index{ReactionNew (class in cbmpy.CBModel)@\spxentry{ReactionNew}\spxextra{class in cbmpy.CBModel}}

\begin{fulllineitems}
\phantomsection\label{\detokenize{modules_doc:cbmpy.CBModel.ReactionNew}}
\pysigstartsignatures
\pysiglinewithargsret{\sphinxbfcode{\sphinxupquote{class\DUrole{w,w}{  }}}\sphinxbfcode{\sphinxupquote{ReactionNew}}}{\sphinxparam{\DUrole{n,n}{pid}}\sphinxparamcomma \sphinxparam{\DUrole{n,n}{name}\DUrole{o,o}{=}\DUrole{default_value}{None}}\sphinxparamcomma \sphinxparam{\DUrole{n,n}{lb}\DUrole{o,o}{=}\DUrole{default_value}{\sphinxhyphen{}inf}}\sphinxparamcomma \sphinxparam{\DUrole{n,n}{ub}\DUrole{o,o}{=}\DUrole{default_value}{inf}}\sphinxparamcomma \sphinxparam{\DUrole{n,n}{reversible}\DUrole{o,o}{=}\DUrole{default_value}{True}}}{}
\pysigstopsignatures
\sphinxAtStartPar
Extended reaction class with new upper/lower bound structure
\index{deactivateReaction() (ReactionNew method)@\spxentry{deactivateReaction()}\spxextra{ReactionNew method}}

\begin{fulllineitems}
\phantomsection\label{\detokenize{modules_doc:cbmpy.CBModel.ReactionNew.deactivateReaction}}
\pysigstartsignatures
\pysiglinewithargsret{\sphinxbfcode{\sphinxupquote{deactivateReaction}}}{\sphinxparam{\DUrole{n,n}{lower}\DUrole{o,o}{=}\DUrole{default_value}{0.0}}\sphinxparamcomma \sphinxparam{\DUrole{n,n}{upper}\DUrole{o,o}{=}\DUrole{default_value}{0.0}}\sphinxparamcomma \sphinxparam{\DUrole{n,n}{silent}\DUrole{o,o}{=}\DUrole{default_value}{True}}}{}
\pysigstopsignatures
\sphinxAtStartPar
Deactivates a reaction by setting its bounds to lower and upper. Restore with reactivateReaction()
\begin{itemize}
\item {} 
\sphinxAtStartPar
\sphinxstyleemphasis{lower} {[}default=0.0{]} bound

\item {} 
\sphinxAtStartPar
\sphinxstyleemphasis{upper} {[}default=0.0{]} bound

\end{itemize}

\end{fulllineitems}

\index{getLowerBound() (ReactionNew method)@\spxentry{getLowerBound()}\spxextra{ReactionNew method}}

\begin{fulllineitems}
\phantomsection\label{\detokenize{modules_doc:cbmpy.CBModel.ReactionNew.getLowerBound}}
\pysigstartsignatures
\pysiglinewithargsret{\sphinxbfcode{\sphinxupquote{getLowerBound}}}{}{}
\pysigstopsignatures
\sphinxAtStartPar
Get the value of the reactions lower bound

\end{fulllineitems}

\index{getUpperBound() (ReactionNew method)@\spxentry{getUpperBound()}\spxextra{ReactionNew method}}

\begin{fulllineitems}
\phantomsection\label{\detokenize{modules_doc:cbmpy.CBModel.ReactionNew.getUpperBound}}
\pysigstartsignatures
\pysiglinewithargsret{\sphinxbfcode{\sphinxupquote{getUpperBound}}}{}{}
\pysigstopsignatures
\sphinxAtStartPar
Get the value of the reactions upper bound

\end{fulllineitems}

\index{reactivateReaction() (ReactionNew method)@\spxentry{reactivateReaction()}\spxextra{ReactionNew method}}

\begin{fulllineitems}
\phantomsection\label{\detokenize{modules_doc:cbmpy.CBModel.ReactionNew.reactivateReaction}}
\pysigstartsignatures
\pysiglinewithargsret{\sphinxbfcode{\sphinxupquote{reactivateReaction}}}{\sphinxparam{\DUrole{n,n}{silent}\DUrole{o,o}{=}\DUrole{default_value}{True}}}{}
\pysigstopsignatures
\sphinxAtStartPar
Activates a reaction deactivated with deactivateReaction

\end{fulllineitems}

\index{setId() (ReactionNew method)@\spxentry{setId()}\spxextra{ReactionNew method}}

\begin{fulllineitems}
\phantomsection\label{\detokenize{modules_doc:cbmpy.CBModel.ReactionNew.setId}}
\pysigstartsignatures
\pysiglinewithargsret{\sphinxbfcode{\sphinxupquote{setId}}}{\sphinxparam{\DUrole{n,n}{fid}}}{}
\pysigstopsignatures
\sphinxAtStartPar
Sets the object Id
\begin{quote}
\begin{itemize}
\item {} 
\sphinxAtStartPar
\sphinxstyleemphasis{fid} a valid c variable style id string

\end{itemize}

\sphinxAtStartPar
Reimplements @FBase.setId()
\end{quote}

\end{fulllineitems}

\index{setLowerBound() (ReactionNew method)@\spxentry{setLowerBound()}\spxextra{ReactionNew method}}

\begin{fulllineitems}
\phantomsection\label{\detokenize{modules_doc:cbmpy.CBModel.ReactionNew.setLowerBound}}
\pysigstartsignatures
\pysiglinewithargsret{\sphinxbfcode{\sphinxupquote{setLowerBound}}}{\sphinxparam{\DUrole{n,n}{value}}}{}
\pysigstopsignatures
\sphinxAtStartPar
Set the value of the reactions lower bound
\begin{itemize}
\item {} 
\sphinxAtStartPar
\sphinxstyleemphasis{value} a floating point value

\end{itemize}

\end{fulllineitems}

\index{setUpperBound() (ReactionNew method)@\spxentry{setUpperBound()}\spxextra{ReactionNew method}}

\begin{fulllineitems}
\phantomsection\label{\detokenize{modules_doc:cbmpy.CBModel.ReactionNew.setUpperBound}}
\pysigstartsignatures
\pysiglinewithargsret{\sphinxbfcode{\sphinxupquote{setUpperBound}}}{\sphinxparam{\DUrole{n,n}{value}}}{}
\pysigstopsignatures
\sphinxAtStartPar
Set the value of the reactions upper bound
\begin{itemize}
\item {} 
\sphinxAtStartPar
\sphinxstyleemphasis{value} a floating point value

\end{itemize}

\end{fulllineitems}


\end{fulllineitems}

\index{Reagent (class in cbmpy.CBModel)@\spxentry{Reagent}\spxextra{class in cbmpy.CBModel}}

\begin{fulllineitems}
\phantomsection\label{\detokenize{modules_doc:cbmpy.CBModel.Reagent}}
\pysigstartsignatures
\pysiglinewithargsret{\sphinxbfcode{\sphinxupquote{class\DUrole{w,w}{  }}}\sphinxbfcode{\sphinxupquote{Reagent}}}{\sphinxparam{\DUrole{n,n}{pid}}\sphinxparamcomma \sphinxparam{\DUrole{n,n}{species\_ref}}\sphinxparamcomma \sphinxparam{\DUrole{n,n}{coef}}}{}
\pysigstopsignatures\begin{description}
\sphinxlineitem{Has a reactive species id and stoichiometric coefficient:}\begin{itemize}
\item {} 
\sphinxAtStartPar
negative = substrate

\item {} 
\sphinxAtStartPar
positive = product

\item {} 
\sphinxAtStartPar
species\_ref a reference to a species obj

\end{itemize}

\end{description}
\index{getCoefficient() (Reagent method)@\spxentry{getCoefficient()}\spxextra{Reagent method}}

\begin{fulllineitems}
\phantomsection\label{\detokenize{modules_doc:cbmpy.CBModel.Reagent.getCoefficient}}
\pysigstartsignatures
\pysiglinewithargsret{\sphinxbfcode{\sphinxupquote{getCoefficient}}}{}{}
\pysigstopsignatures
\sphinxAtStartPar
Returns the reagent coefficient

\end{fulllineitems}

\index{getRole() (Reagent method)@\spxentry{getRole()}\spxextra{Reagent method}}

\begin{fulllineitems}
\phantomsection\label{\detokenize{modules_doc:cbmpy.CBModel.Reagent.getRole}}
\pysigstartsignatures
\pysiglinewithargsret{\sphinxbfcode{\sphinxupquote{getRole}}}{}{}
\pysigstopsignatures
\sphinxAtStartPar
Returns the reagents role, “substrate”, “product” or None

\end{fulllineitems}

\index{getSpecies() (Reagent method)@\spxentry{getSpecies()}\spxextra{Reagent method}}

\begin{fulllineitems}
\phantomsection\label{\detokenize{modules_doc:cbmpy.CBModel.Reagent.getSpecies}}
\pysigstartsignatures
\pysiglinewithargsret{\sphinxbfcode{\sphinxupquote{getSpecies}}}{}{}
\pysigstopsignatures
\sphinxAtStartPar
Returns the metabolite/species that the reagent reference refers to

\end{fulllineitems}

\index{setCoefficient() (Reagent method)@\spxentry{setCoefficient()}\spxextra{Reagent method}}

\begin{fulllineitems}
\phantomsection\label{\detokenize{modules_doc:cbmpy.CBModel.Reagent.setCoefficient}}
\pysigstartsignatures
\pysiglinewithargsret{\sphinxbfcode{\sphinxupquote{setCoefficient}}}{\sphinxparam{\DUrole{n,n}{coef}}}{}
\pysigstopsignatures
\sphinxAtStartPar
Sets the reagent coefficient and role, negative coefficients are substrates and positive ones are products
\begin{itemize}
\item {} 
\sphinxAtStartPar
\sphinxstyleemphasis{coeff} the new coefficient

\end{itemize}

\end{fulllineitems}

\index{setSpecies() (Reagent method)@\spxentry{setSpecies()}\spxextra{Reagent method}}

\begin{fulllineitems}
\phantomsection\label{\detokenize{modules_doc:cbmpy.CBModel.Reagent.setSpecies}}
\pysigstartsignatures
\pysiglinewithargsret{\sphinxbfcode{\sphinxupquote{setSpecies}}}{\sphinxparam{\DUrole{n,n}{spe}}}{}
\pysigstopsignatures
\sphinxAtStartPar
Sets the metabolite/species that the reagent reference refers to

\end{fulllineitems}


\end{fulllineitems}

\index{Species (class in cbmpy.CBModel)@\spxentry{Species}\spxextra{class in cbmpy.CBModel}}

\begin{fulllineitems}
\phantomsection\label{\detokenize{modules_doc:cbmpy.CBModel.Species}}
\pysigstartsignatures
\pysiglinewithargsret{\sphinxbfcode{\sphinxupquote{class\DUrole{w,w}{  }}}\sphinxbfcode{\sphinxupquote{Species}}}{\sphinxparam{\DUrole{n,n}{pid}}\sphinxparamcomma \sphinxparam{\DUrole{n,n}{boundary}\DUrole{o,o}{=}\DUrole{default_value}{False}}\sphinxparamcomma \sphinxparam{\DUrole{n,n}{name}\DUrole{o,o}{=}\DUrole{default_value}{None}}\sphinxparamcomma \sphinxparam{\DUrole{n,n}{value}\DUrole{o,o}{=}\DUrole{default_value}{nan}}\sphinxparamcomma \sphinxparam{\DUrole{n,n}{compartment}\DUrole{o,o}{=}\DUrole{default_value}{None}}\sphinxparamcomma \sphinxparam{\DUrole{n,n}{charge}\DUrole{o,o}{=}\DUrole{default_value}{None}}\sphinxparamcomma \sphinxparam{\DUrole{n,n}{chemFormula}\DUrole{o,o}{=}\DUrole{default_value}{None}}}{}
\pysigstopsignatures
\sphinxAtStartPar
Holds species/metabolite information
\index{getCharge() (Species method)@\spxentry{getCharge()}\spxextra{Species method}}

\begin{fulllineitems}
\phantomsection\label{\detokenize{modules_doc:cbmpy.CBModel.Species.getCharge}}
\pysigstartsignatures
\pysiglinewithargsret{\sphinxbfcode{\sphinxupquote{getCharge}}}{}{}
\pysigstopsignatures
\sphinxAtStartPar
Returns the species charge

\end{fulllineitems}

\index{getChemFormula() (Species method)@\spxentry{getChemFormula()}\spxextra{Species method}}

\begin{fulllineitems}
\phantomsection\label{\detokenize{modules_doc:cbmpy.CBModel.Species.getChemFormula}}
\pysigstartsignatures
\pysiglinewithargsret{\sphinxbfcode{\sphinxupquote{getChemFormula}}}{}{}
\pysigstopsignatures
\sphinxAtStartPar
Returns the species chemical formula

\end{fulllineitems}

\index{getReagentOf() (Species method)@\spxentry{getReagentOf()}\spxextra{Species method}}

\begin{fulllineitems}
\phantomsection\label{\detokenize{modules_doc:cbmpy.CBModel.Species.getReagentOf}}
\pysigstartsignatures
\pysiglinewithargsret{\sphinxbfcode{\sphinxupquote{getReagentOf}}}{}{}
\pysigstopsignatures
\sphinxAtStartPar
Returns a list of reaction id’s that this metabolite occurs in

\end{fulllineitems}

\index{getValue() (Species method)@\spxentry{getValue()}\spxextra{Species method}}

\begin{fulllineitems}
\phantomsection\label{\detokenize{modules_doc:cbmpy.CBModel.Species.getValue}}
\pysigstartsignatures
\pysiglinewithargsret{\sphinxbfcode{\sphinxupquote{getValue}}}{}{}
\pysigstopsignatures
\sphinxAtStartPar
Returns the current value of the attribute (input/solution)

\end{fulllineitems}

\index{isReagentOf() (Species method)@\spxentry{isReagentOf()}\spxextra{Species method}}

\begin{fulllineitems}
\phantomsection\label{\detokenize{modules_doc:cbmpy.CBModel.Species.isReagentOf}}
\pysigstartsignatures
\pysiglinewithargsret{\sphinxbfcode{\sphinxupquote{isReagentOf}}}{}{}
\pysigstopsignatures
\sphinxAtStartPar
Returns a dynamically generated list of reactions that this species occurs as a reagent

\end{fulllineitems}

\index{setBoundary() (Species method)@\spxentry{setBoundary()}\spxextra{Species method}}

\begin{fulllineitems}
\phantomsection\label{\detokenize{modules_doc:cbmpy.CBModel.Species.setBoundary}}
\pysigstartsignatures
\pysiglinewithargsret{\sphinxbfcode{\sphinxupquote{setBoundary}}}{}{}
\pysigstopsignatures
\sphinxAtStartPar
Sets the species so it is a boundary metabolite or fixed which does not occur in the stoichiometric matrix N

\end{fulllineitems}

\index{setCharge() (Species method)@\spxentry{setCharge()}\spxextra{Species method}}

\begin{fulllineitems}
\phantomsection\label{\detokenize{modules_doc:cbmpy.CBModel.Species.setCharge}}
\pysigstartsignatures
\pysiglinewithargsret{\sphinxbfcode{\sphinxupquote{setCharge}}}{\sphinxparam{\DUrole{n,n}{charge}}}{}
\pysigstopsignatures
\sphinxAtStartPar
Sets the species charge:
\begin{itemize}
\item {} 
\sphinxAtStartPar
\sphinxstyleemphasis{charge} a signed double but generally a signed int is used

\end{itemize}

\end{fulllineitems}

\index{setChemFormula() (Species method)@\spxentry{setChemFormula()}\spxextra{Species method}}

\begin{fulllineitems}
\phantomsection\label{\detokenize{modules_doc:cbmpy.CBModel.Species.setChemFormula}}
\pysigstartsignatures
\pysiglinewithargsret{\sphinxbfcode{\sphinxupquote{setChemFormula}}}{\sphinxparam{\DUrole{n,n}{cf}}}{}
\pysigstopsignatures
\sphinxAtStartPar
Sets the species chemical formula
\begin{itemize}
\item {} 
\sphinxAtStartPar
\sphinxstyleemphasis{cf} a chemical formula e.g. CH3NO2

\end{itemize}

\end{fulllineitems}

\index{setId() (Species method)@\spxentry{setId()}\spxextra{Species method}}

\begin{fulllineitems}
\phantomsection\label{\detokenize{modules_doc:cbmpy.CBModel.Species.setId}}
\pysigstartsignatures
\pysiglinewithargsret{\sphinxbfcode{\sphinxupquote{setId}}}{\sphinxparam{\DUrole{n,n}{fid}}\sphinxparamcomma \sphinxparam{\DUrole{n,n}{allow\_rename}\DUrole{o,o}{=}\DUrole{default_value}{False}}}{}
\pysigstopsignatures
\sphinxAtStartPar
Sets the object Id
\begin{quote}
\begin{itemize}
\item {} 
\sphinxAtStartPar
\sphinxstyleemphasis{fid} a valid c variable style id string

\end{itemize}

\sphinxAtStartPar
Reimplements @FBase.setId()
\end{quote}

\end{fulllineitems}

\index{setReagentOf() (Species method)@\spxentry{setReagentOf()}\spxextra{Species method}}

\begin{fulllineitems}
\phantomsection\label{\detokenize{modules_doc:cbmpy.CBModel.Species.setReagentOf}}
\pysigstartsignatures
\pysiglinewithargsret{\sphinxbfcode{\sphinxupquote{setReagentOf}}}{\sphinxparam{\DUrole{n,n}{rid}}}{}
\pysigstopsignatures
\sphinxAtStartPar
Adds the supplied reaction id to the reagent\_of list (if it isn’t one already)
\begin{itemize}
\item {} 
\sphinxAtStartPar
\sphinxstyleemphasis{rid} a valid reaction id

\end{itemize}

\end{fulllineitems}

\index{setValue() (Species method)@\spxentry{setValue()}\spxextra{Species method}}

\begin{fulllineitems}
\phantomsection\label{\detokenize{modules_doc:cbmpy.CBModel.Species.setValue}}
\pysigstartsignatures
\pysiglinewithargsret{\sphinxbfcode{\sphinxupquote{setValue}}}{\sphinxparam{\DUrole{n,n}{value}}}{}
\pysigstopsignatures
\sphinxAtStartPar
Sets the attribute ‘’value’’

\end{fulllineitems}

\index{unsetBoundary() (Species method)@\spxentry{unsetBoundary()}\spxextra{Species method}}

\begin{fulllineitems}
\phantomsection\label{\detokenize{modules_doc:cbmpy.CBModel.Species.unsetBoundary}}
\pysigstartsignatures
\pysiglinewithargsret{\sphinxbfcode{\sphinxupquote{unsetBoundary}}}{}{}
\pysigstopsignatures
\sphinxAtStartPar
Unsets the species boundary attribute so that the metabolite is free and therefore occurs in the stoichiometric matrix N

\end{fulllineitems}


\end{fulllineitems}

\index{UserDefinedConstraint (class in cbmpy.CBModel)@\spxentry{UserDefinedConstraint}\spxextra{class in cbmpy.CBModel}}

\begin{fulllineitems}
\phantomsection\label{\detokenize{modules_doc:cbmpy.CBModel.UserDefinedConstraint}}
\pysigstartsignatures
\pysiglinewithargsret{\sphinxbfcode{\sphinxupquote{class\DUrole{w,w}{  }}}\sphinxbfcode{\sphinxupquote{UserDefinedConstraint}}}{\sphinxparam{\DUrole{n,n}{pid}}\sphinxparamcomma \sphinxparam{\DUrole{n,n}{lb}}\sphinxparamcomma \sphinxparam{\DUrole{n,n}{ub}}}{}
\pysigstopsignatures
\sphinxAtStartPar
This is an FBCv3 class that defines a set of user defined constraints, it is similar to an objective constraint except allows parameters as
coefficients and values in the constraint components
\index{addConstraintComponent() (UserDefinedConstraint method)@\spxentry{addConstraintComponent()}\spxextra{UserDefinedConstraint method}}

\begin{fulllineitems}
\phantomsection\label{\detokenize{modules_doc:cbmpy.CBModel.UserDefinedConstraint.addConstraintComponent}}
\pysigstartsignatures
\pysiglinewithargsret{\sphinxbfcode{\sphinxupquote{addConstraintComponent}}}{\sphinxparam{\DUrole{n,n}{cc}}}{}
\pysigstopsignatures\begin{itemize}
\item {} 
\sphinxAtStartPar
\sphinxstyleemphasis{cc} UserConstraintComponent

\end{itemize}

\end{fulllineitems}

\index{createConstraintComponent() (UserDefinedConstraint method)@\spxentry{createConstraintComponent()}\spxextra{UserDefinedConstraint method}}

\begin{fulllineitems}
\phantomsection\label{\detokenize{modules_doc:cbmpy.CBModel.UserDefinedConstraint.createConstraintComponent}}
\pysigstartsignatures
\pysiglinewithargsret{\sphinxbfcode{\sphinxupquote{createConstraintComponent}}}{\sphinxparam{\DUrole{n,n}{pid}}\sphinxparamcomma \sphinxparam{\DUrole{n,n}{coefficient}}\sphinxparamcomma \sphinxparam{\DUrole{n,n}{variable}}\sphinxparamcomma \sphinxparam{\DUrole{n,n}{ctype}}}{}
\pysigstopsignatures
\end{fulllineitems}

\index{getConstraintComponent() (UserDefinedConstraint method)@\spxentry{getConstraintComponent()}\spxextra{UserDefinedConstraint method}}

\begin{fulllineitems}
\phantomsection\label{\detokenize{modules_doc:cbmpy.CBModel.UserDefinedConstraint.getConstraintComponent}}
\pysigstartsignatures
\pysiglinewithargsret{\sphinxbfcode{\sphinxupquote{getConstraintComponent}}}{\sphinxparam{\DUrole{n,n}{cid}}}{}
\pysigstopsignatures\begin{itemize}
\item {} 
\sphinxAtStartPar
\sphinxstyleemphasis{cid} get a constraint component that matches cid

\end{itemize}

\end{fulllineitems}

\index{getConstraintComponentData() (UserDefinedConstraint method)@\spxentry{getConstraintComponentData()}\spxextra{UserDefinedConstraint method}}

\begin{fulllineitems}
\phantomsection\label{\detokenize{modules_doc:cbmpy.CBModel.UserDefinedConstraint.getConstraintComponentData}}
\pysigstartsignatures
\pysiglinewithargsret{\sphinxbfcode{\sphinxupquote{getConstraintComponentData}}}{}{}
\pysigstopsignatures
\end{fulllineitems}

\index{getConstraintComponentForVariable() (UserDefinedConstraint method)@\spxentry{getConstraintComponentForVariable()}\spxextra{UserDefinedConstraint method}}

\begin{fulllineitems}
\phantomsection\label{\detokenize{modules_doc:cbmpy.CBModel.UserDefinedConstraint.getConstraintComponentForVariable}}
\pysigstartsignatures
\pysiglinewithargsret{\sphinxbfcode{\sphinxupquote{getConstraintComponentForVariable}}}{\sphinxparam{\DUrole{n,n}{rid}}}{}
\pysigstopsignatures
\sphinxAtStartPar
\sphinxstyleemphasis{rid} a component id

\end{fulllineitems}

\index{getConstraintComponentIDs() (UserDefinedConstraint method)@\spxentry{getConstraintComponentIDs()}\spxextra{UserDefinedConstraint method}}

\begin{fulllineitems}
\phantomsection\label{\detokenize{modules_doc:cbmpy.CBModel.UserDefinedConstraint.getConstraintComponentIDs}}
\pysigstartsignatures
\pysiglinewithargsret{\sphinxbfcode{\sphinxupquote{getConstraintComponentIDs}}}{}{}
\pysigstopsignatures
\end{fulllineitems}

\index{getConstraintComponentVariableTypes() (UserDefinedConstraint method)@\spxentry{getConstraintComponentVariableTypes()}\spxextra{UserDefinedConstraint method}}

\begin{fulllineitems}
\phantomsection\label{\detokenize{modules_doc:cbmpy.CBModel.UserDefinedConstraint.getConstraintComponentVariableTypes}}
\pysigstartsignatures
\pysiglinewithargsret{\sphinxbfcode{\sphinxupquote{getConstraintComponentVariableTypes}}}{}{}
\pysigstopsignatures
\end{fulllineitems}

\index{getConstraintComponentVariables() (UserDefinedConstraint method)@\spxentry{getConstraintComponentVariables()}\spxextra{UserDefinedConstraint method}}

\begin{fulllineitems}
\phantomsection\label{\detokenize{modules_doc:cbmpy.CBModel.UserDefinedConstraint.getConstraintComponentVariables}}
\pysigstartsignatures
\pysiglinewithargsret{\sphinxbfcode{\sphinxupquote{getConstraintComponentVariables}}}{}{}
\pysigstopsignatures
\end{fulllineitems}

\index{getConstraintComponents() (UserDefinedConstraint method)@\spxentry{getConstraintComponents()}\spxextra{UserDefinedConstraint method}}

\begin{fulllineitems}
\phantomsection\label{\detokenize{modules_doc:cbmpy.CBModel.UserDefinedConstraint.getConstraintComponents}}
\pysigstartsignatures
\pysiglinewithargsret{\sphinxbfcode{\sphinxupquote{getConstraintComponents}}}{}{}
\pysigstopsignatures
\end{fulllineitems}


\end{fulllineitems}

\phantomsection\label{\detokenize{modules_doc:module-cbmpy.CBModelTools}}\index{module@\spxentry{module}!cbmpy.CBModelTools@\spxentry{cbmpy.CBModelTools}}\index{cbmpy.CBModelTools@\spxentry{cbmpy.CBModelTools}!module@\spxentry{module}}

\section{CBMPy: CBModelTools module}
\label{\detokenize{modules_doc:cbmpy-cbmodeltools-module}}
\sphinxAtStartPar
PySCeS Constraint Based Modelling (\sphinxurl{http://cbmpy.sourceforge.net})
Copyright (C) 2009\sphinxhyphen{}2024 Brett G. Olivier, VU University Amsterdam, Amsterdam, The Netherlands

\sphinxAtStartPar
This program is free software: you can redistribute it and/or modify
it under the terms of the GNU General Public License as published by
the Free Software Foundation, either version 3 of the License, or
(at your option) any later version.

\sphinxAtStartPar
This program is distributed in the hope that it will be useful,
but WITHOUT ANY WARRANTY; without even the implied warranty of
MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
GNU General Public License for more details.

\sphinxAtStartPar
You should have received a copy of the GNU General Public License
along with this program.  If not, see \textless{}\sphinxurl{http://www.gnu.org/licenses/}\textgreater{}

\sphinxAtStartPar
Author: Brett G. Olivier PhD
Contact developers: \sphinxurl{https://github.com/SystemsBioinformatics/cbmpy/issues}
Last edit: \$Author: bgoli \$ (\$Id: CBModelTools.py 710 2020\sphinxhyphen{}04\sphinxhyphen{}27 14:22:34Z bgoli \$)
\phantomsection\label{\detokenize{modules_doc:module-cbmpy.CBMultiCore}}\index{module@\spxentry{module}!cbmpy.CBMultiCore@\spxentry{cbmpy.CBMultiCore}}\index{cbmpy.CBMultiCore@\spxentry{cbmpy.CBMultiCore}!module@\spxentry{module}}

\section{CBMPy: CBMultiCore module}
\label{\detokenize{modules_doc:cbmpy-cbmulticore-module}}
\sphinxAtStartPar
PySCeS Constraint Based Modelling (\sphinxurl{http://cbmpy.sourceforge.net})
Copyright (C) 2009\sphinxhyphen{}2024 Brett G. Olivier, VU University Amsterdam, Amsterdam, The Netherlands

\sphinxAtStartPar
This program is free software: you can redistribute it and/or modify
it under the terms of the GNU General Public License as published by
the Free Software Foundation, either version 3 of the License, or
(at your option) any later version.

\sphinxAtStartPar
This program is distributed in the hope that it will be useful,
but WITHOUT ANY WARRANTY; without even the implied warranty of
MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
GNU General Public License for more details.

\sphinxAtStartPar
You should have received a copy of the GNU General Public License
along with this program.  If not, see \textless{}\sphinxurl{http://www.gnu.org/licenses/}\textgreater{}

\sphinxAtStartPar
Author: Brett G. Olivier PhD
Contact developers: \sphinxurl{https://github.com/SystemsBioinformatics/cbmpy/issues}
Last edit: \$Author: bgoli \$ (\$Id: CBMultiCore.py 710 2020\sphinxhyphen{}04\sphinxhyphen{}27 14:22:34Z bgoli \$)
\index{grouper() (in module cbmpy.CBMultiCore)@\spxentry{grouper()}\spxextra{in module cbmpy.CBMultiCore}}

\begin{fulllineitems}
\phantomsection\label{\detokenize{modules_doc:cbmpy.CBMultiCore.grouper}}
\pysigstartsignatures
\pysiglinewithargsret{\sphinxbfcode{\sphinxupquote{grouper}}}{\sphinxparam{\DUrole{n,n}{3}}\sphinxparamcomma \sphinxparam{\DUrole{n,n}{\textquotesingle{}abcdefg\textquotesingle{}}}\sphinxparamcomma \sphinxparam{\DUrole{n,n}{\textquotesingle{}x\textquotesingle{}) \sphinxhyphen{}\sphinxhyphen{}\textgreater{} (\textquotesingle{}a\textquotesingle{}}}\sphinxparamcomma \sphinxparam{\DUrole{n,n}{\textquotesingle{}b\textquotesingle{}}}\sphinxparamcomma \sphinxparam{\DUrole{n,n}{\textquotesingle{}c\textquotesingle{})}}\sphinxparamcomma \sphinxparam{\DUrole{n,n}{(\textquotesingle{}d\textquotesingle{}}}\sphinxparamcomma \sphinxparam{\DUrole{n,n}{\textquotesingle{}e\textquotesingle{}}}\sphinxparamcomma \sphinxparam{\DUrole{n,n}{\textquotesingle{}f\textquotesingle{})}}\sphinxparamcomma \sphinxparam{\DUrole{n,n}{(\textquotesingle{}g\textquotesingle{}}}\sphinxparamcomma \sphinxparam{\DUrole{n,n}{\textquotesingle{}x\textquotesingle{}}}\sphinxparamcomma \sphinxparam{\DUrole{n,n}{\textquotesingle{}x\textquotesingle{}}}}{}
\pysigstopsignatures
\end{fulllineitems}

\index{runMultiCoreFVA() (in module cbmpy.CBMultiCore)@\spxentry{runMultiCoreFVA()}\spxextra{in module cbmpy.CBMultiCore}}

\begin{fulllineitems}
\phantomsection\label{\detokenize{modules_doc:cbmpy.CBMultiCore.runMultiCoreFVA}}
\pysigstartsignatures
\pysiglinewithargsret{\sphinxbfcode{\sphinxupquote{runMultiCoreFVA}}}{\sphinxparam{\DUrole{n,n}{fba}}\sphinxparamcomma \sphinxparam{\DUrole{n,n}{selected\_reactions}\DUrole{o,o}{=}\DUrole{default_value}{None}}\sphinxparamcomma \sphinxparam{\DUrole{n,n}{pre\_opt}\DUrole{o,o}{=}\DUrole{default_value}{True}}\sphinxparamcomma \sphinxparam{\DUrole{n,n}{tol}\DUrole{o,o}{=}\DUrole{default_value}{None}}\sphinxparamcomma \sphinxparam{\DUrole{n,n}{objF2constr}\DUrole{o,o}{=}\DUrole{default_value}{True}}\sphinxparamcomma \sphinxparam{\DUrole{n,n}{rhs\_sense}\DUrole{o,o}{=}\DUrole{default_value}{\textquotesingle{}lower\textquotesingle{}}}\sphinxparamcomma \sphinxparam{\DUrole{n,n}{optPercentage}\DUrole{o,o}{=}\DUrole{default_value}{100.0}}\sphinxparamcomma \sphinxparam{\DUrole{n,n}{work\_dir}\DUrole{o,o}{=}\DUrole{default_value}{None}}\sphinxparamcomma \sphinxparam{\DUrole{n,n}{quiet}\DUrole{o,o}{=}\DUrole{default_value}{True}}\sphinxparamcomma \sphinxparam{\DUrole{n,n}{debug}\DUrole{o,o}{=}\DUrole{default_value}{False}}\sphinxparamcomma \sphinxparam{\DUrole{n,n}{oldlpgen}\DUrole{o,o}{=}\DUrole{default_value}{False}}\sphinxparamcomma \sphinxparam{\DUrole{n,n}{markupmodel}\DUrole{o,o}{=}\DUrole{default_value}{True}}\sphinxparamcomma \sphinxparam{\DUrole{n,n}{procs}\DUrole{o,o}{=}\DUrole{default_value}{2}}\sphinxparamcomma \sphinxparam{\DUrole{n,n}{override\_bin}\DUrole{o,o}{=}\DUrole{default_value}{None}}}{}
\pysigstopsignatures
\sphinxAtStartPar
Run a multicore FVA where:
\begin{itemize}
\item {} 
\sphinxAtStartPar
\sphinxstyleemphasis{fba} is an fba model instance

\item {} 
\sphinxAtStartPar
\sphinxstyleemphasis{procs} {[}default=2{]} number of processing threads (optimum seems to be about the number of physical cores)

\item {} 
\sphinxAtStartPar
\sphinxstyleemphasis{python\_override\_bin} allows customization of the Python bin used for the multicore process

\end{itemize}

\end{fulllineitems}

\phantomsection\label{\detokenize{modules_doc:module-cbmpy.CBMultiEnv}}\index{module@\spxentry{module}!cbmpy.CBMultiEnv@\spxentry{cbmpy.CBMultiEnv}}\index{cbmpy.CBMultiEnv@\spxentry{cbmpy.CBMultiEnv}!module@\spxentry{module}}

\section{CBMPy: CBMultiEnv module}
\label{\detokenize{modules_doc:cbmpy-cbmultienv-module}}
\sphinxAtStartPar
PySCeS Constraint Based Modelling (\sphinxurl{http://cbmpy.sourceforge.net})
Copyright (C) 2009\sphinxhyphen{}2024 Brett G. Olivier, VU University Amsterdam, Amsterdam, The Netherlands

\sphinxAtStartPar
This program is free software: you can redistribute it and/or modify
it under the terms of the GNU General Public License as published by
the Free Software Foundation, either version 3 of the License, or
(at your option) any later version.

\sphinxAtStartPar
This program is distributed in the hope that it will be useful,
but WITHOUT ANY WARRANTY; without even the implied warranty of
MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
GNU General Public License for more details.

\sphinxAtStartPar
You should have received a copy of the GNU General Public License
along with this program.  If not, see \textless{}\sphinxurl{http://www.gnu.org/licenses/}\textgreater{}

\sphinxAtStartPar
Author: Brett G. Olivier PhD
Contact developers: \sphinxurl{https://github.com/SystemsBioinformatics/cbmpy/issues}
Last edit: \$Author: bgoli \$ (\$Id: CBMultiEnv.py 710 2020\sphinxhyphen{}04\sphinxhyphen{}27 14:22:34Z bgoli \$)
\phantomsection\label{\detokenize{modules_doc:module-cbmpy.CBNetDB}}\index{module@\spxentry{module}!cbmpy.CBNetDB@\spxentry{cbmpy.CBNetDB}}\index{cbmpy.CBNetDB@\spxentry{cbmpy.CBNetDB}!module@\spxentry{module}}

\section{CBMPy: CBNetDB module}
\label{\detokenize{modules_doc:cbmpy-cbnetdb-module}}
\sphinxAtStartPar
PySCeS Constraint Based Modelling (\sphinxurl{http://cbmpy.sourceforge.net})
Copyright (C) 2009\sphinxhyphen{}2024 Brett G. Olivier, VU University Amsterdam, Amsterdam, The Netherlands

\sphinxAtStartPar
This program is free software: you can redistribute it and/or modify
it under the terms of the GNU General Public License as published by
the Free Software Foundation, either version 3 of the License, or
(at your option) any later version.

\sphinxAtStartPar
This program is distributed in the hope that it will be useful,
but WITHOUT ANY WARRANTY; without even the implied warranty of
MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
GNU General Public License for more details.

\sphinxAtStartPar
You should have received a copy of the GNU General Public License
along with this program.  If not, see \textless{}\sphinxurl{http://www.gnu.org/licenses/}\textgreater{}

\sphinxAtStartPar
Author: Brett G. Olivier PhD
Contact developers: \sphinxurl{https://github.com/SystemsBioinformatics/cbmpy/issues}
Last edit: \$Author: bgoli \$ (\$Id: CBNetDB.py 710 2020\sphinxhyphen{}04\sphinxhyphen{}27 14:22:34Z bgoli \$)
\index{DBTools (class in cbmpy.CBNetDB)@\spxentry{DBTools}\spxextra{class in cbmpy.CBNetDB}}

\begin{fulllineitems}
\phantomsection\label{\detokenize{modules_doc:cbmpy.CBNetDB.DBTools}}
\pysigstartsignatures
\pysigline{\sphinxbfcode{\sphinxupquote{class\DUrole{w,w}{  }}}\sphinxbfcode{\sphinxupquote{DBTools}}}
\pysigstopsignatures
\sphinxAtStartPar
Tools to work with SQLite DB’s (optimized, no SQL required).
\index{checkEntryInColumn() (DBTools method)@\spxentry{checkEntryInColumn()}\spxextra{DBTools method}}

\begin{fulllineitems}
\phantomsection\label{\detokenize{modules_doc:cbmpy.CBNetDB.DBTools.checkEntryInColumn}}
\pysigstartsignatures
\pysiglinewithargsret{\sphinxbfcode{\sphinxupquote{checkEntryInColumn}}}{\sphinxparam{\DUrole{n,n}{table}}\sphinxparamcomma \sphinxparam{\DUrole{n,n}{col}}\sphinxparamcomma \sphinxparam{\DUrole{n,n}{rid}}}{}
\pysigstopsignatures
\sphinxAtStartPar
Check if an entry exists in a table
\begin{itemize}
\item {} 
\sphinxAtStartPar
\sphinxstyleemphasis{table} the table name

\item {} 
\sphinxAtStartPar
\sphinxstyleemphasis{col} the column name

\item {} 
\sphinxAtStartPar
\sphinxstyleemphasis{rid} the row to search for

\end{itemize}

\end{fulllineitems}

\index{closeDB() (DBTools method)@\spxentry{closeDB()}\spxextra{DBTools method}}

\begin{fulllineitems}
\phantomsection\label{\detokenize{modules_doc:cbmpy.CBNetDB.DBTools.closeDB}}
\pysigstartsignatures
\pysiglinewithargsret{\sphinxbfcode{\sphinxupquote{closeDB}}}{}{}
\pysigstopsignatures
\sphinxAtStartPar
Close the DB connection and reset the DBTools instance (can be reconnected)

\end{fulllineitems}

\index{commitDB() (DBTools method)@\spxentry{commitDB()}\spxextra{DBTools method}}

\begin{fulllineitems}
\phantomsection\label{\detokenize{modules_doc:cbmpy.CBNetDB.DBTools.commitDB}}
\pysigstartsignatures
\pysiglinewithargsret{\sphinxbfcode{\sphinxupquote{commitDB}}}{}{}
\pysigstopsignatures
\sphinxAtStartPar
Commits all curent changes to DB, returns a boolean.

\end{fulllineitems}

\index{connectSQLiteDB() (DBTools method)@\spxentry{connectSQLiteDB()}\spxextra{DBTools method}}

\begin{fulllineitems}
\phantomsection\label{\detokenize{modules_doc:cbmpy.CBNetDB.DBTools.connectSQLiteDB}}
\pysigstartsignatures
\pysiglinewithargsret{\sphinxbfcode{\sphinxupquote{connectSQLiteDB}}}{\sphinxparam{\DUrole{n,n}{db\_name}}\sphinxparamcomma \sphinxparam{\DUrole{n,n}{work\_dir}\DUrole{o,o}{=}\DUrole{default_value}{None}}}{}
\pysigstopsignatures
\sphinxAtStartPar
Connect to a sqlite database.
\begin{itemize}
\item {} 
\sphinxAtStartPar
\sphinxstyleemphasis{db\_name} the name of the sqlite database

\item {} 
\sphinxAtStartPar
\sphinxstyleemphasis{work\_dir} the optional database path

\end{itemize}

\end{fulllineitems}

\index{createDBTable() (DBTools method)@\spxentry{createDBTable()}\spxextra{DBTools method}}

\begin{fulllineitems}
\phantomsection\label{\detokenize{modules_doc:cbmpy.CBNetDB.DBTools.createDBTable}}
\pysigstartsignatures
\pysiglinewithargsret{\sphinxbfcode{\sphinxupquote{createDBTable}}}{\sphinxparam{\DUrole{n,n}{table}}\sphinxparamcomma \sphinxparam{\DUrole{n,n}{sqlcols}}}{}
\pysigstopsignatures
\sphinxAtStartPar
Create a database table if it does not exist:
\begin{itemize}
\item {} 
\sphinxAtStartPar
\sphinxstyleemphasis{table} the table name

\item {} 
\sphinxAtStartPar
\sphinxstyleemphasis{sqlcols} a list containing the SQL definitions of the table columns: \textless{}id\textgreater{} \textless{}type\textgreater{} for example \sphinxtitleref{{[}‘gene TEXT PRIMARY KEY’, ‘aa\_seq TEXT’, ‘nuc\_seq TEXT’, ‘aa\_len INT’, ‘nuc\_len INT’{]}}

\end{itemize}

\sphinxAtStartPar
Effectively writes CREATE TABLE “table” (\textless{}id\textgreater{} \textless{}type\textgreater{}, gene TEXT PRIMARY KEY, aa\_seq TEXT, nuc\_seq TEXT, aa\_len INT, nuc\_len INT) \% table

\end{fulllineitems}

\index{dumpTableToCSV() (DBTools method)@\spxentry{dumpTableToCSV()}\spxextra{DBTools method}}

\begin{fulllineitems}
\phantomsection\label{\detokenize{modules_doc:cbmpy.CBNetDB.DBTools.dumpTableToCSV}}
\pysigstartsignatures
\pysiglinewithargsret{\sphinxbfcode{\sphinxupquote{dumpTableToCSV}}}{\sphinxparam{\DUrole{n,n}{table}}\sphinxparamcomma \sphinxparam{\DUrole{n,n}{filename}}}{}
\pysigstopsignatures
\sphinxAtStartPar
Save a table as tab separated txt file
\begin{itemize}
\item {} 
\sphinxAtStartPar
\sphinxstyleemphasis{table} the table to export

\item {} 
\sphinxAtStartPar
\sphinxstyleemphasis{filename} the filename of the table dump

\end{itemize}

\end{fulllineitems}

\index{dumpTableToTxt() (DBTools method)@\spxentry{dumpTableToTxt()}\spxextra{DBTools method}}

\begin{fulllineitems}
\phantomsection\label{\detokenize{modules_doc:cbmpy.CBNetDB.DBTools.dumpTableToTxt}}
\pysigstartsignatures
\pysiglinewithargsret{\sphinxbfcode{\sphinxupquote{dumpTableToTxt}}}{\sphinxparam{\DUrole{n,n}{table}}\sphinxparamcomma \sphinxparam{\DUrole{n,n}{filename}}}{}
\pysigstopsignatures
\sphinxAtStartPar
Save a table as tab separated txt file
\begin{itemize}
\item {} 
\sphinxAtStartPar
\sphinxstyleemphasis{table} the table to export

\item {} 
\sphinxAtStartPar
\sphinxstyleemphasis{filename} the filename of the table dump

\end{itemize}

\end{fulllineitems}

\index{executeSQL() (DBTools method)@\spxentry{executeSQL()}\spxextra{DBTools method}}

\begin{fulllineitems}
\phantomsection\label{\detokenize{modules_doc:cbmpy.CBNetDB.DBTools.executeSQL}}
\pysigstartsignatures
\pysiglinewithargsret{\sphinxbfcode{\sphinxupquote{executeSQL}}}{\sphinxparam{\DUrole{n,n}{sql}}}{}
\pysigstopsignatures
\sphinxAtStartPar
Execute a SQL command:
\begin{itemize}
\item {} 
\sphinxAtStartPar
\sphinxstyleemphasis{sql} a string containing a SQL command

\end{itemize}

\end{fulllineitems}

\index{fetchAll() (DBTools method)@\spxentry{fetchAll()}\spxextra{DBTools method}}

\begin{fulllineitems}
\phantomsection\label{\detokenize{modules_doc:cbmpy.CBNetDB.DBTools.fetchAll}}
\pysigstartsignatures
\pysiglinewithargsret{\sphinxbfcode{\sphinxupquote{fetchAll}}}{\sphinxparam{\DUrole{n,n}{sql}}}{}
\pysigstopsignatures
\sphinxAtStartPar
Raw SQL query e.g. ‘SELECT id FROM gene WHERE gene=”G”’

\end{fulllineitems}

\index{getCell() (DBTools method)@\spxentry{getCell()}\spxextra{DBTools method}}

\begin{fulllineitems}
\phantomsection\label{\detokenize{modules_doc:cbmpy.CBNetDB.DBTools.getCell}}
\pysigstartsignatures
\pysiglinewithargsret{\sphinxbfcode{\sphinxupquote{getCell}}}{\sphinxparam{\DUrole{n,n}{table}}\sphinxparamcomma \sphinxparam{\DUrole{n,n}{col}}\sphinxparamcomma \sphinxparam{\DUrole{n,n}{rid}}\sphinxparamcomma \sphinxparam{\DUrole{n,n}{cell}}}{}
\pysigstopsignatures
\sphinxAtStartPar
Get the table cell which correspond to rid in column. Returns the value or None
\begin{itemize}
\item {} 
\sphinxAtStartPar
\sphinxstyleemphasis{table} the database table

\item {} 
\sphinxAtStartPar
\sphinxstyleemphasis{col} the column id

\item {} 
\sphinxAtStartPar
\sphinxstyleemphasis{rid} the row index id

\item {} 
\sphinxAtStartPar
\sphinxstyleemphasis{cell} the column of the cell you want tp extract

\end{itemize}

\end{fulllineitems}

\index{getColumns() (DBTools method)@\spxentry{getColumns()}\spxextra{DBTools method}}

\begin{fulllineitems}
\phantomsection\label{\detokenize{modules_doc:cbmpy.CBNetDB.DBTools.getColumns}}
\pysigstartsignatures
\pysiglinewithargsret{\sphinxbfcode{\sphinxupquote{getColumns}}}{\sphinxparam{\DUrole{n,n}{table}}\sphinxparamcomma \sphinxparam{\DUrole{n,n}{cols}}}{}
\pysigstopsignatures
\sphinxAtStartPar
Fetch the contents of one or more columns of data in a table
\begin{itemize}
\item {} 
\sphinxAtStartPar
\sphinxstyleemphasis{table} the database table

\item {} 
\sphinxAtStartPar
\sphinxstyleemphasis{cols} a list of one or more column id’s

\end{itemize}

\end{fulllineitems}

\index{getRow() (DBTools method)@\spxentry{getRow()}\spxextra{DBTools method}}

\begin{fulllineitems}
\phantomsection\label{\detokenize{modules_doc:cbmpy.CBNetDB.DBTools.getRow}}
\pysigstartsignatures
\pysiglinewithargsret{\sphinxbfcode{\sphinxupquote{getRow}}}{\sphinxparam{\DUrole{n,n}{table}}\sphinxparamcomma \sphinxparam{\DUrole{n,n}{col}}\sphinxparamcomma \sphinxparam{\DUrole{n,n}{rid}}}{}
\pysigstopsignatures
\sphinxAtStartPar
Get the table row(s) which correspond to rid in column. Returns the row(s) as a list, if the column is the primary key
this is always a single entry.
\begin{itemize}
\item {} 
\sphinxAtStartPar
\sphinxstyleemphasis{table} the database table

\item {} 
\sphinxAtStartPar
\sphinxstyleemphasis{col} the column id

\item {} 
\sphinxAtStartPar
\sphinxstyleemphasis{rid} the row index id

\end{itemize}

\end{fulllineitems}

\index{getTable() (DBTools method)@\spxentry{getTable()}\spxextra{DBTools method}}

\begin{fulllineitems}
\phantomsection\label{\detokenize{modules_doc:cbmpy.CBNetDB.DBTools.getTable}}
\pysigstartsignatures
\pysiglinewithargsret{\sphinxbfcode{\sphinxupquote{getTable}}}{\sphinxparam{\DUrole{n,n}{table}}\sphinxparamcomma \sphinxparam{\DUrole{n,n}{colOut}\DUrole{o,o}{=}\DUrole{default_value}{False}}}{}
\pysigstopsignatures
\sphinxAtStartPar
Returns an entire database table
\begin{itemize}
\item {} 
\sphinxAtStartPar
\sphinxstyleemphasis{table} the table name

\item {} 
\sphinxAtStartPar
\sphinxstyleemphasis{colOut} optionally return a tuple of (data,ColNames)

\end{itemize}

\end{fulllineitems}

\index{insertData() (DBTools method)@\spxentry{insertData()}\spxextra{DBTools method}}

\begin{fulllineitems}
\phantomsection\label{\detokenize{modules_doc:cbmpy.CBNetDB.DBTools.insertData}}
\pysigstartsignatures
\pysiglinewithargsret{\sphinxbfcode{\sphinxupquote{insertData}}}{\sphinxparam{\DUrole{n,n}{table}}\sphinxparamcomma \sphinxparam{\DUrole{n,n}{data}}\sphinxparamcomma \sphinxparam{\DUrole{n,n}{commit}\DUrole{o,o}{=}\DUrole{default_value}{True}}}{}
\pysigstopsignatures\begin{description}
\sphinxlineitem{Insert data into a table: “INSERT INTO \%s (?, ?, ?, ?, ?) VALUES (?, ?, ?, ?, ?)” \% table,}\begin{quote}

\sphinxAtStartPar
(?, ?, ?, ?, ?)) )
\end{quote}
\begin{itemize}
\item {} 
\sphinxAtStartPar
\sphinxstyleemphasis{table} the DB table name

\item {} 
\sphinxAtStartPar
\sphinxstyleemphasis{data} a dictionary of \{id:value\} pairs

\item {} 
\sphinxAtStartPar
\sphinxstyleemphasis{commit} whether to commit the data insertions

\end{itemize}

\end{description}

\end{fulllineitems}

\index{updateData() (DBTools method)@\spxentry{updateData()}\spxextra{DBTools method}}

\begin{fulllineitems}
\phantomsection\label{\detokenize{modules_doc:cbmpy.CBNetDB.DBTools.updateData}}
\pysigstartsignatures
\pysiglinewithargsret{\sphinxbfcode{\sphinxupquote{updateData}}}{\sphinxparam{\DUrole{n,n}{table}}\sphinxparamcomma \sphinxparam{\DUrole{n,n}{col}}\sphinxparamcomma \sphinxparam{\DUrole{n,n}{rid}}\sphinxparamcomma \sphinxparam{\DUrole{n,n}{data}}\sphinxparamcomma \sphinxparam{\DUrole{n,n}{commit}\DUrole{o,o}{=}\DUrole{default_value}{True}}}{}
\pysigstopsignatures
\sphinxAtStartPar
Update already defined data
\begin{quote}
\begin{itemize}
\item {} 
\sphinxAtStartPar
\sphinxstyleemphasis{table} the table name

\item {} 
\sphinxAtStartPar
\sphinxstyleemphasis{col} the column name

\item {} 
\sphinxAtStartPar
\sphinxstyleemphasis{rid} the row id to update

\item {} 
\sphinxAtStartPar
\sphinxstyleemphasis{data} a dictionary of \{id:value\} pairs

\item {} 
\sphinxAtStartPar
\sphinxstyleemphasis{commit} whether to commit the data updates

\end{itemize}

\sphinxAtStartPar
UPDATE COMPANY SET ADDRESS = ‘Texas’ WHERE ID = 6;
\end{quote}

\end{fulllineitems}


\end{fulllineitems}

\index{KeGGSequenceTools (class in cbmpy.CBNetDB)@\spxentry{KeGGSequenceTools}\spxextra{class in cbmpy.CBNetDB}}

\begin{fulllineitems}
\phantomsection\label{\detokenize{modules_doc:cbmpy.CBNetDB.KeGGSequenceTools}}
\pysigstartsignatures
\pysiglinewithargsret{\sphinxbfcode{\sphinxupquote{class\DUrole{w,w}{  }}}\sphinxbfcode{\sphinxupquote{KeGGSequenceTools}}}{\sphinxparam{\DUrole{n,n}{url}}\sphinxparamcomma \sphinxparam{\DUrole{n,n}{db\_name}}\sphinxparamcomma \sphinxparam{\DUrole{n,n}{work\_dir}}}{}
\pysigstopsignatures
\sphinxAtStartPar
Using the KeGG connector this class provides tools to construct an organims specific sequence database

\end{fulllineitems}

\index{KeGGTools (class in cbmpy.CBNetDB)@\spxentry{KeGGTools}\spxextra{class in cbmpy.CBNetDB}}

\begin{fulllineitems}
\phantomsection\label{\detokenize{modules_doc:cbmpy.CBNetDB.KeGGTools}}
\pysigstartsignatures
\pysiglinewithargsret{\sphinxbfcode{\sphinxupquote{class\DUrole{w,w}{  }}}\sphinxbfcode{\sphinxupquote{KeGGTools}}}{\sphinxparam{\DUrole{n,n}{url}}}{}
\pysigstopsignatures
\sphinxAtStartPar
Class that holds useful methods for querying KeGG via a SUDS provided soap client
\index{fetchSeqfromKeGG() (KeGGTools method)@\spxentry{fetchSeqfromKeGG()}\spxextra{KeGGTools method}}

\begin{fulllineitems}
\phantomsection\label{\detokenize{modules_doc:cbmpy.CBNetDB.KeGGTools.fetchSeqfromKeGG}}
\pysigstartsignatures
\pysiglinewithargsret{\sphinxbfcode{\sphinxupquote{fetchSeqfromKeGG}}}{\sphinxparam{\DUrole{n,n}{k\_gene}}}{}
\pysigstopsignatures
\sphinxAtStartPar
Given a gene name try and retrieve the gene and amino acid sequence

\end{fulllineitems}


\end{fulllineitems}

\index{MIRIAMTools (class in cbmpy.CBNetDB)@\spxentry{MIRIAMTools}\spxextra{class in cbmpy.CBNetDB}}

\begin{fulllineitems}
\phantomsection\label{\detokenize{modules_doc:cbmpy.CBNetDB.MIRIAMTools}}
\pysigstartsignatures
\pysigline{\sphinxbfcode{\sphinxupquote{class\DUrole{w,w}{  }}}\sphinxbfcode{\sphinxupquote{MIRIAMTools}}}
\pysigstopsignatures
\sphinxAtStartPar
Tools dealing with MIRIAM annotations

\end{fulllineitems}

\index{RESTClient (class in cbmpy.CBNetDB)@\spxentry{RESTClient}\spxextra{class in cbmpy.CBNetDB}}

\begin{fulllineitems}
\phantomsection\label{\detokenize{modules_doc:cbmpy.CBNetDB.RESTClient}}
\pysigstartsignatures
\pysigline{\sphinxbfcode{\sphinxupquote{class\DUrole{w,w}{  }}}\sphinxbfcode{\sphinxupquote{RESTClient}}}
\pysigstopsignatures
\sphinxAtStartPar
Class that provides the basis for application specific connectors to REST web services
\index{Close() (RESTClient method)@\spxentry{Close()}\spxextra{RESTClient method}}

\begin{fulllineitems}
\phantomsection\label{\detokenize{modules_doc:cbmpy.CBNetDB.RESTClient.Close}}
\pysigstartsignatures
\pysiglinewithargsret{\sphinxbfcode{\sphinxupquote{Close}}}{}{}
\pysigstopsignatures
\sphinxAtStartPar
Close the currently active connection

\end{fulllineitems}

\index{Connect() (RESTClient method)@\spxentry{Connect()}\spxextra{RESTClient method}}

\begin{fulllineitems}
\phantomsection\label{\detokenize{modules_doc:cbmpy.CBNetDB.RESTClient.Connect}}
\pysigstartsignatures
\pysiglinewithargsret{\sphinxbfcode{\sphinxupquote{Connect}}}{\sphinxparam{\DUrole{n,n}{root}}}{}
\pysigstopsignatures
\sphinxAtStartPar
Establish HTTP connection to
\begin{itemize}
\item {} 
\sphinxAtStartPar
\sphinxstyleemphasis{root} the site root “www.google.com”

\end{itemize}

\end{fulllineitems}

\index{Get() (RESTClient method)@\spxentry{Get()}\spxextra{RESTClient method}}

\begin{fulllineitems}
\phantomsection\label{\detokenize{modules_doc:cbmpy.CBNetDB.RESTClient.Get}}
\pysigstartsignatures
\pysiglinewithargsret{\sphinxbfcode{\sphinxupquote{Get}}}{\sphinxparam{\DUrole{n,n}{query}}}{}
\pysigstopsignatures
\sphinxAtStartPar
Perform an http GET using:
\begin{itemize}
\item {} 
\sphinxAtStartPar
\sphinxstyleemphasis{query} e.g.

\item {} 
\sphinxAtStartPar
\sphinxstyleemphasis{reply\_mode} {[}default=’’{]} this is the reply mode

\end{itemize}

\sphinxAtStartPar
For example “/semanticSBML/annotate/search.xml?q=ATP”

\end{fulllineitems}

\index{GetLog() (RESTClient method)@\spxentry{GetLog()}\spxextra{RESTClient method}}

\begin{fulllineitems}
\phantomsection\label{\detokenize{modules_doc:cbmpy.CBNetDB.RESTClient.GetLog}}
\pysigstartsignatures
\pysiglinewithargsret{\sphinxbfcode{\sphinxupquote{GetLog}}}{}{}
\pysigstopsignatures
\sphinxAtStartPar
Return the logged history

\end{fulllineitems}

\index{Log() (RESTClient method)@\spxentry{Log()}\spxextra{RESTClient method}}

\begin{fulllineitems}
\phantomsection\label{\detokenize{modules_doc:cbmpy.CBNetDB.RESTClient.Log}}
\pysigstartsignatures
\pysiglinewithargsret{\sphinxbfcode{\sphinxupquote{Log}}}{\sphinxparam{\DUrole{n,n}{txt}}}{}
\pysigstopsignatures
\sphinxAtStartPar
Add txt to logfile history
\begin{itemize}
\item {} 
\sphinxAtStartPar
\sphinxstyleemphasis{txt} a string

\end{itemize}

\end{fulllineitems}


\end{fulllineitems}

\index{SemanticSBML (class in cbmpy.CBNetDB)@\spxentry{SemanticSBML}\spxextra{class in cbmpy.CBNetDB}}

\begin{fulllineitems}
\phantomsection\label{\detokenize{modules_doc:cbmpy.CBNetDB.SemanticSBML}}
\pysigstartsignatures
\pysigline{\sphinxbfcode{\sphinxupquote{class\DUrole{w,w}{  }}}\sphinxbfcode{\sphinxupquote{SemanticSBML}}}
\pysigstopsignatures
\sphinxAtStartPar
REST client for connecting to SemanticSBML services
\index{parseXMLtoText() (SemanticSBML method)@\spxentry{parseXMLtoText()}\spxextra{SemanticSBML method}}

\begin{fulllineitems}
\phantomsection\label{\detokenize{modules_doc:cbmpy.CBNetDB.SemanticSBML.parseXMLtoText}}
\pysigstartsignatures
\pysiglinewithargsret{\sphinxbfcode{\sphinxupquote{parseXMLtoText}}}{\sphinxparam{\DUrole{n,n}{xml}}}{}
\pysigstopsignatures
\sphinxAtStartPar
Parse the xml output by quickLookup() into a list of URL
\begin{itemize}
\item {} 
\sphinxAtStartPar
\sphinxstyleemphasis{xml} XML returns from SemanticSBML

\end{itemize}

\end{fulllineitems}

\index{quickLookup() (SemanticSBML method)@\spxentry{quickLookup()}\spxextra{SemanticSBML method}}

\begin{fulllineitems}
\phantomsection\label{\detokenize{modules_doc:cbmpy.CBNetDB.SemanticSBML.quickLookup}}
\pysigstartsignatures
\pysiglinewithargsret{\sphinxbfcode{\sphinxupquote{quickLookup}}}{\sphinxparam{\DUrole{n,n}{txt}}}{}
\pysigstopsignatures
\sphinxAtStartPar
Do a quick lookpup for txt using SemanticSBML (connectic if required) and return results. Returns
a list of identifiers.org id’s in descending priority (as return)
\begin{itemize}
\item {} 
\sphinxAtStartPar
\sphinxstyleemphasis{txt} the string to lookup

\end{itemize}

\end{fulllineitems}

\index{viewDataInWebrowser() (SemanticSBML method)@\spxentry{viewDataInWebrowser()}\spxextra{SemanticSBML method}}

\begin{fulllineitems}
\phantomsection\label{\detokenize{modules_doc:cbmpy.CBNetDB.SemanticSBML.viewDataInWebrowser}}
\pysigstartsignatures
\pysiglinewithargsret{\sphinxbfcode{\sphinxupquote{viewDataInWebrowser}}}{\sphinxparam{\DUrole{n,n}{maxres}\DUrole{o,o}{=}\DUrole{default_value}{10}}}{}
\pysigstopsignatures
\sphinxAtStartPar
Attempt to view \#maxres results returned by SemanticSBML in the default browser
\begin{itemize}
\item {} 
\sphinxAtStartPar
\sphinxstyleemphasis{maxres} default maximum number of results to display.

\end{itemize}

\end{fulllineitems}


\end{fulllineitems}

\phantomsection\label{\detokenize{modules_doc:module-cbmpy.CBPlot}}\index{module@\spxentry{module}!cbmpy.CBPlot@\spxentry{cbmpy.CBPlot}}\index{cbmpy.CBPlot@\spxentry{cbmpy.CBPlot}!module@\spxentry{module}}

\section{CBMPy: CBPlot module}
\label{\detokenize{modules_doc:cbmpy-cbplot-module}}
\sphinxAtStartPar
PySCeS Constraint Based Modelling (\sphinxurl{http://cbmpy.sourceforge.net})
Copyright (C) 2009\sphinxhyphen{}2024 Brett G. Olivier, VU University Amsterdam, Amsterdam, The Netherlands

\sphinxAtStartPar
This program is free software: you can redistribute it and/or modify
it under the terms of the GNU General Public License as published by
the Free Software Foundation, either version 3 of the License, or
(at your option) any later version.

\sphinxAtStartPar
This program is distributed in the hope that it will be useful,
but WITHOUT ANY WARRANTY; without even the implied warranty of
MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
GNU General Public License for more details.

\sphinxAtStartPar
You should have received a copy of the GNU General Public License
along with this program.  If not, see \textless{}\sphinxurl{http://www.gnu.org/licenses/}\textgreater{}

\sphinxAtStartPar
Author: Brett G. Olivier PhD
Contact developers: \sphinxurl{https://github.com/SystemsBioinformatics/cbmpy/issues}
Last edit: \$Author: bgoli \$ (\$Id: CBPlot.py 710 2020\sphinxhyphen{}04\sphinxhyphen{}27 14:22:34Z bgoli \$)
\index{plotFluxVariability() (in module cbmpy.CBPlot)@\spxentry{plotFluxVariability()}\spxextra{in module cbmpy.CBPlot}}

\begin{fulllineitems}
\phantomsection\label{\detokenize{modules_doc:cbmpy.CBPlot.plotFluxVariability}}
\pysigstartsignatures
\pysiglinewithargsret{\sphinxbfcode{\sphinxupquote{plotFluxVariability}}}{\sphinxparam{\DUrole{n,n}{fva\_data}}\sphinxparamcomma \sphinxparam{\DUrole{n,n}{fva\_names}}\sphinxparamcomma \sphinxparam{\DUrole{n,n}{fname}}\sphinxparamcomma \sphinxparam{\DUrole{n,n}{work\_dir}\DUrole{o,o}{=}\DUrole{default_value}{None}}\sphinxparamcomma \sphinxparam{\DUrole{n,n}{title}\DUrole{o,o}{=}\DUrole{default_value}{None}}\sphinxparamcomma \sphinxparam{\DUrole{n,n}{ySlice}\DUrole{o,o}{=}\DUrole{default_value}{None}}\sphinxparamcomma \sphinxparam{\DUrole{n,n}{minHeight}\DUrole{o,o}{=}\DUrole{default_value}{None}}\sphinxparamcomma \sphinxparam{\DUrole{n,n}{maxHeight}\DUrole{o,o}{=}\DUrole{default_value}{None}}\sphinxparamcomma \sphinxparam{\DUrole{n,n}{roundec}\DUrole{o,o}{=}\DUrole{default_value}{None}}\sphinxparamcomma \sphinxparam{\DUrole{n,n}{autoclose}\DUrole{o,o}{=}\DUrole{default_value}{True}}\sphinxparamcomma \sphinxparam{\DUrole{n,n}{fluxval}\DUrole{o,o}{=}\DUrole{default_value}{True}}\sphinxparamcomma \sphinxparam{\DUrole{n,n}{type}\DUrole{o,o}{=}\DUrole{default_value}{\textquotesingle{}png\textquotesingle{}}}}{}
\pysigstopsignatures
\sphinxAtStartPar
Plots and saves as an image the flux variability results as generated by CBSolver.FluxVariabilityAnalysis.
\begin{itemize}
\item {} 
\sphinxAtStartPar
\sphinxstyleemphasis{fva\_data} FluxVariabilityAnalysis() FVA OUTPUT\_ARRAY

\item {} 
\sphinxAtStartPar
\sphinxstyleemphasis{fva\_names} FluxVariabilityAnalysis() FVA OUTPUT\_NAMES

\item {} 
\sphinxAtStartPar
\sphinxstyleemphasis{fname} filename\_base for the CSV output

\item {} 
\sphinxAtStartPar
\sphinxstyleemphasis{work\_dir} {[}default=None{]} if set the output directory for the csv files

\item {} 
\sphinxAtStartPar
\sphinxstyleemphasis{title} {[}default=None{]} the user defined title for the graph

\item {} 
\sphinxAtStartPar
\sphinxstyleemphasis{ySlice} {[}default=None{]} this sets an absolute (fixed) limit on the Y\sphinxhyphen{}axis (+\sphinxhyphen{} ySlice)

\item {} 
\sphinxAtStartPar
\sphinxstyleemphasis{minHeight} {[}default=None{]} the minimum length that defined a span

\item {} 
\sphinxAtStartPar
\sphinxstyleemphasis{maxHeight} {[}default=None{]} the maximum length a span can obtain, bar will be limited to maxHeight and coloured yellow

\item {} 
\sphinxAtStartPar
\sphinxstyleemphasis{roundec} {[}default=None{]} an integer indicating at which decimal to round off output. Default is no rounding.

\item {} 
\sphinxAtStartPar
\sphinxstyleemphasis{autoclose} {[}default=True{]} autoclose plot after save

\item {} 
\sphinxAtStartPar
\sphinxstyleemphasis{fluxval} {[}default=True{]} plot the flux value

\item {} 
\sphinxAtStartPar
\sphinxstyleemphasis{type} {[}default=’png’{]} the output format, depends on matplotlib backend e.g. ‘png’, ‘pdf’, ‘eps’

\end{itemize}

\end{fulllineitems}

\phantomsection\label{\detokenize{modules_doc:module-cbmpy.CBRead}}\index{module@\spxentry{module}!cbmpy.CBRead@\spxentry{cbmpy.CBRead}}\index{cbmpy.CBRead@\spxentry{cbmpy.CBRead}!module@\spxentry{module}}

\section{CBMPy: CBRead module}
\label{\detokenize{modules_doc:cbmpy-cbread-module}}
\sphinxAtStartPar
PySCeS Constraint Based Modelling (\sphinxurl{http://cbmpy.sourceforge.net})
Copyright (C) 2009\sphinxhyphen{}2024 Brett G. Olivier, VU University Amsterdam, Amsterdam, The Netherlands

\sphinxAtStartPar
This program is free software: you can redistribute it and/or modify
it under the terms of the GNU General Public License as published by
the Free Software Foundation, either version 3 of the License, or
(at your option) any later version.

\sphinxAtStartPar
This program is distributed in the hope that it will be useful,
but WITHOUT ANY WARRANTY; without even the implied warranty of
MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
GNU General Public License for more details.

\sphinxAtStartPar
You should have received a copy of the GNU General Public License
along with this program.  If not, see \textless{}\sphinxurl{http://www.gnu.org/licenses/}\textgreater{}

\sphinxAtStartPar
Author: Brett G. Olivier PhD
Contact developers: \sphinxurl{https://github.com/SystemsBioinformatics/cbmpy/issues}
Last edit: \$Author: bgoli \$ (\$Id: CBRead.py 669 2019\sphinxhyphen{}02\sphinxhyphen{}18 22:58:19Z bgoli \$)
\index{loadModel() (in module cbmpy.CBRead)@\spxentry{loadModel()}\spxextra{in module cbmpy.CBRead}}

\begin{fulllineitems}
\phantomsection\label{\detokenize{modules_doc:cbmpy.CBRead.loadModel}}
\pysigstartsignatures
\pysiglinewithargsret{\sphinxbfcode{\sphinxupquote{loadModel}}}{\sphinxparam{\DUrole{n,n}{sbmlfile}}}{}
\pysigstopsignatures
\sphinxAtStartPar
Loads any SBML model in COBRA, FAME (SBML2FBA), SBML3FBCv1, SBML3FBCv2 format.
\begin{itemize}
\item {} 
\sphinxAtStartPar
\sphinxstyleemphasis{sbmlfile} an SBML model file

\end{itemize}

\end{fulllineitems}

\index{readCOBRASBML() (in module cbmpy.CBRead)@\spxentry{readCOBRASBML()}\spxextra{in module cbmpy.CBRead}}

\begin{fulllineitems}
\phantomsection\label{\detokenize{modules_doc:cbmpy.CBRead.readCOBRASBML}}
\pysigstartsignatures
\pysiglinewithargsret{\sphinxbfcode{\sphinxupquote{readCOBRASBML}}}{\sphinxparam{\DUrole{n,n}{fname}}\sphinxparamcomma \sphinxparam{\DUrole{n,n}{work\_dir}\DUrole{o,o}{=}\DUrole{default_value}{None}}\sphinxparamcomma \sphinxparam{\DUrole{n,n}{return\_sbml\_model}\DUrole{o,o}{=}\DUrole{default_value}{False}}\sphinxparamcomma \sphinxparam{\DUrole{n,n}{delete\_intermediate}\DUrole{o,o}{=}\DUrole{default_value}{False}}\sphinxparamcomma \sphinxparam{\DUrole{n,n}{fake\_boundary\_species\_search}\DUrole{o,o}{=}\DUrole{default_value}{False}}\sphinxparamcomma \sphinxparam{\DUrole{n,n}{output\_dir}\DUrole{o,o}{=}\DUrole{default_value}{None}}\sphinxparamcomma \sphinxparam{\DUrole{n,n}{skip\_genes}\DUrole{o,o}{=}\DUrole{default_value}{False}}\sphinxparamcomma \sphinxparam{\DUrole{n,n}{scan\_notes\_gpr}\DUrole{o,o}{=}\DUrole{default_value}{True}}}{}
\pysigstopsignatures
\sphinxAtStartPar
Read in a COBRA format SBML Level 2 file with FBA annotation where and return either a CBM model object
or a (cbm\_mod, sbml\_mod) pair if return\_sbml\_model=True
\begin{itemize}
\item {} 
\sphinxAtStartPar
\sphinxstyleemphasis{fname} is the filename

\item {} 
\sphinxAtStartPar
\sphinxstyleemphasis{work\_dir} is the working directory

\item {} 
\sphinxAtStartPar
\sphinxstyleemphasis{delete\_intermediate} {[}default=False{]} delete the intermediate SBML Level 3 FBC file

\item {} 
\sphinxAtStartPar
\sphinxstyleemphasis{fake\_boundary\_species\_search} {[}default=False{]} after looking for the boundary\_condition of a species search for overloaded id’s \textless{}id\textgreater{}\_b

\item {} 
\sphinxAtStartPar
\sphinxstyleemphasis{output\_dir} {[}default=None{]} the directory to output the intermediate SBML L3 files (if generated) default to input directory

\item {} 
\sphinxAtStartPar
\sphinxstyleemphasis{skip\_genes} {[}default=False{]} do not load GPR data

\item {} 
\sphinxAtStartPar
\sphinxstyleemphasis{scan\_notes\_gpr} {[}default=True{]} if the model is loaded and no genes are detected the scan the \textless{}notes\textgreater{} field for GPR associationa

\end{itemize}

\end{fulllineitems}

\index{readExcel97Model() (in module cbmpy.CBRead)@\spxentry{readExcel97Model()}\spxextra{in module cbmpy.CBRead}}

\begin{fulllineitems}
\phantomsection\label{\detokenize{modules_doc:cbmpy.CBRead.readExcel97Model}}
\pysigstartsignatures
\pysiglinewithargsret{\sphinxbfcode{\sphinxupquote{readExcel97Model}}}{\sphinxparam{\DUrole{n,n}{xlname}}\sphinxparamcomma \sphinxparam{\DUrole{n,n}{write\_sbml}\DUrole{o,o}{=}\DUrole{default_value}{True}}\sphinxparamcomma \sphinxparam{\DUrole{n,n}{sbml\_level}\DUrole{o,o}{=}\DUrole{default_value}{3}}\sphinxparamcomma \sphinxparam{\DUrole{n,n}{return\_dictionaries}\DUrole{o,o}{=}\DUrole{default_value}{False}}}{}
\pysigstopsignatures
\sphinxAtStartPar
Reads a model encoded as an Excel97 workbook and returns it as a CBMPy model object and SBML file. Note the workbook must be formatted
exactly like those produced by cbm.writeModelToExcel97(). Note that reactions have to be defined in \sphinxstylestrong{both} the \sphinxstyleemphasis{reaction}
and \sphinxstyleemphasis{network\_react} sheets to be included in the model.
\begin{itemize}
\item {} 
\sphinxAtStartPar
\sphinxstyleemphasis{xlpath} the filename of the Excel workbook

\item {} 
\sphinxAtStartPar
\sphinxstyleemphasis{return\_model} {[}default=True{]} construct and return the CBMPy model

\item {} 
\sphinxAtStartPar
\sphinxstyleemphasis{write\_sbml} {[}default=True{]} write the SBML file to fname

\item {} 
\sphinxAtStartPar
\sphinxstyleemphasis{return\_dictionaries} {[}default=False{]} return the dictionaries constructed when reading the Excel file (in place of the model)

\item {} 
\sphinxAtStartPar
\sphinxstyleemphasis{sbml\_level} {[}default=3{]} write the SBML file as either SBML L2 FBA or SBML L3 FBC file.

\end{itemize}

\end{fulllineitems}

\index{readSBML2FBA() (in module cbmpy.CBRead)@\spxentry{readSBML2FBA()}\spxextra{in module cbmpy.CBRead}}

\begin{fulllineitems}
\phantomsection\label{\detokenize{modules_doc:cbmpy.CBRead.readSBML2FBA}}
\pysigstartsignatures
\pysiglinewithargsret{\sphinxbfcode{\sphinxupquote{readSBML2FBA}}}{\sphinxparam{\DUrole{n,n}{fname}}\sphinxparamcomma \sphinxparam{\DUrole{n,n}{work\_dir}\DUrole{o,o}{=}\DUrole{default_value}{None}}\sphinxparamcomma \sphinxparam{\DUrole{n,n}{return\_sbml\_model}\DUrole{o,o}{=}\DUrole{default_value}{False}}\sphinxparamcomma \sphinxparam{\DUrole{n,n}{fake\_boundary\_species\_search}\DUrole{o,o}{=}\DUrole{default_value}{False}}\sphinxparamcomma \sphinxparam{\DUrole{n,n}{scan\_notes\_gpr}\DUrole{o,o}{=}\DUrole{default_value}{True}}}{}
\pysigstopsignatures
\sphinxAtStartPar
Read in an SBML Level 2 file with FBA annotation where:
\begin{itemize}
\item {} 
\sphinxAtStartPar
\sphinxstyleemphasis{fname} is the filename

\item {} 
\sphinxAtStartPar
\sphinxstyleemphasis{work\_dir} is the working directory if None then only fname is used

\item {} 
\sphinxAtStartPar
\sphinxstyleemphasis{return\_sbml\_model} {[}default=False{]} return a a (cbm\_mod, sbml\_mod) pair

\item {} 
\sphinxAtStartPar
\sphinxstyleemphasis{fake\_boundary\_species\_search} {[}default=False{]} after looking for the boundary\_condition of a species search for overloaded id’s \textless{}id\textgreater{}\_b

\item {} 
\sphinxAtStartPar
\sphinxstyleemphasis{scan\_notes\_gpr} {[}default=True{]} if the model is loaded and no genes are detected the scan the \textless{}notes\textgreater{} field for GPR associationa

\end{itemize}

\end{fulllineitems}

\index{readSBML3FBC() (in module cbmpy.CBRead)@\spxentry{readSBML3FBC()}\spxextra{in module cbmpy.CBRead}}

\begin{fulllineitems}
\phantomsection\label{\detokenize{modules_doc:cbmpy.CBRead.readSBML3FBC}}
\pysigstartsignatures
\pysiglinewithargsret{\sphinxbfcode{\sphinxupquote{readSBML3FBC}}}{\sphinxparam{\DUrole{n,n}{fname}}\sphinxparamcomma \sphinxparam{\DUrole{n,n}{work\_dir}\DUrole{o,o}{=}\DUrole{default_value}{None}}\sphinxparamcomma \sphinxparam{\DUrole{n,n}{return\_sbml\_model}\DUrole{o,o}{=}\DUrole{default_value}{False}}\sphinxparamcomma \sphinxparam{\DUrole{n,n}{xoptions}\DUrole{o,o}{=}\DUrole{default_value}{\{\textquotesingle{}validate\textquotesingle{}: False\}}}\sphinxparamcomma \sphinxparam{\DUrole{n,n}{scan\_notes\_gpr}\DUrole{o,o}{=}\DUrole{default_value}{True}}}{}
\pysigstopsignatures
\sphinxAtStartPar
Read in an SBML Level 3 file with FBC annotation where and return a CBM model object
\begin{itemize}
\item {} 
\sphinxAtStartPar
\sphinxstyleemphasis{fname} is the filename

\item {} 
\sphinxAtStartPar
\sphinxstyleemphasis{work\_dir} is the working directory

\item {} 
\sphinxAtStartPar
\sphinxstyleemphasis{return\_sbml\_model} deprecated and ignored please update code

\item {} 
\sphinxAtStartPar
\sphinxstyleemphasis{xoptions} special load options, enable with option=True except for nmatrix\_type which has a type.
\begin{itemize}
\item {} 
\sphinxAtStartPar
\sphinxstyleemphasis{nogenes} do not load/process genes

\item {} 
\sphinxAtStartPar
\sphinxstyleemphasis{noannot} do not load/process any annotations

\item {} 
\sphinxAtStartPar
\sphinxstyleemphasis{validate} validate model and display errors and warnings before loading

\item {} 
\sphinxAtStartPar
\sphinxstyleemphasis{readcobra} read the cobra annotation

\item {} 
\sphinxAtStartPar
\sphinxstyleemphasis{read\_model\_string} {[}default=False{]} read the model from a string (instead of a filename) containing an SBML document

\item {} 
\sphinxAtStartPar
\sphinxstyleemphasis{nmatrix\_type} {[}default=’normal’{]} define the type of stoichiometrich matrix to be built
\begin{itemize}
\item {} 
\sphinxAtStartPar
‘numpy’ dense numpy array (best performance)

\item {} 
\sphinxAtStartPar
‘scipy\_csr’ scipy sparse matrix (lower performance, low memory)

\item {} 
\sphinxAtStartPar
‘sympy’ a sympy rational matrix (low performance, high memory, cast to dense to analyse)

\item {} 
\sphinxAtStartPar
None do not build matrix

\end{itemize}

\end{itemize}

\end{itemize}
\begin{itemize}
\item {} 
\sphinxAtStartPar
\sphinxstyleemphasis{scan\_notes\_gpr} {[}default=True{]} if the model is loaded and no genes are detected scan the \textless{}notes\textgreater{} field for GPR associationa

\end{itemize}

\end{fulllineitems}

\index{readSK\_FVA() (in module cbmpy.CBRead)@\spxentry{readSK\_FVA()}\spxextra{in module cbmpy.CBRead}}

\begin{fulllineitems}
\phantomsection\label{\detokenize{modules_doc:cbmpy.CBRead.readSK_FVA}}
\pysigstartsignatures
\pysiglinewithargsret{\sphinxbfcode{\sphinxupquote{readSK\_FVA}}}{\sphinxparam{\DUrole{n,n}{filename}}}{}
\pysigstopsignatures
\sphinxAtStartPar
Read Stevens FVA results (opt.fva) file and return a list of dictionaries

\end{fulllineitems}

\index{readSK\_vertex() (in module cbmpy.CBRead)@\spxentry{readSK\_vertex()}\spxextra{in module cbmpy.CBRead}}

\begin{fulllineitems}
\phantomsection\label{\detokenize{modules_doc:cbmpy.CBRead.readSK_vertex}}
\pysigstartsignatures
\pysiglinewithargsret{\sphinxbfcode{\sphinxupquote{readSK\_vertex}}}{\sphinxparam{\DUrole{n,n}{fname}}\sphinxparamcomma \sphinxparam{\DUrole{n,n}{bigfile}\DUrole{o,o}{=}\DUrole{default_value}{True}}\sphinxparamcomma \sphinxparam{\DUrole{n,n}{fast\_rational}\DUrole{o,o}{=}\DUrole{default_value}{False}}\sphinxparamcomma \sphinxparam{\DUrole{n,n}{nformat}\DUrole{o,o}{=}\DUrole{default_value}{\textquotesingle{}\%.14f\textquotesingle{}}}\sphinxparamcomma \sphinxparam{\DUrole{n,n}{compression}\DUrole{o,o}{=}\DUrole{default_value}{None}}\sphinxparamcomma \sphinxparam{\DUrole{n,n}{hdf5file}\DUrole{o,o}{=}\DUrole{default_value}{None}}}{}
\pysigstopsignatures
\sphinxAtStartPar
Reads in Stevens vertex analysis file:
\begin{itemize}
\item {} 
\sphinxAtStartPar
\sphinxstyleemphasis{fname} the input filename (.all file that results from Stevens pipeline)

\item {} 
\sphinxAtStartPar
\sphinxstyleemphasis{bigfile} {[}default=True{]} this option is now always true and is left in for backwards compatability

\item {} 
\sphinxAtStartPar
\sphinxstyleemphasis{fast\_rational} {[}default=False{]} by default off and uses SymPy for rational\textendash{}\textgreater{}float conversion, when on uses float decomposition with a slight (2th decimal) decrease in accuracy

\item {} 
\sphinxAtStartPar
\sphinxstyleemphasis{nformat} {[}default=’\%.14f’{]} the number format used in output files

\item {} 
\sphinxAtStartPar
\sphinxstyleemphasis{compression} {[}default=None{]} compression to be used in hdf5 files can be one of {[}None, ‘lzf’, ‘gz?’, ‘szip’{]}

\item {} 
\sphinxAtStartPar
\sphinxstyleemphasis{hdf5file} {[}default=None{]} if None then generic filename ‘\_vtx\_.tmp.hdf5’ is uses otherwise \textless{}hdf5file\textgreater{}.hdf5

\end{itemize}

\sphinxAtStartPar
and returns an hdf5 \sphinxstyleemphasis{filename} of the results with a single group named \sphinxstylestrong{data} which countains datasets
\begin{itemize}
\item {} 
\sphinxAtStartPar
vertices

\item {} 
\sphinxAtStartPar
rays

\item {} 
\sphinxAtStartPar
lin

\end{itemize}

\sphinxAtStartPar
where all vectors are in terms of the column space of N.

\end{fulllineitems}

\index{readSK\_vertexOld() (in module cbmpy.CBRead)@\spxentry{readSK\_vertexOld()}\spxextra{in module cbmpy.CBRead}}

\begin{fulllineitems}
\phantomsection\label{\detokenize{modules_doc:cbmpy.CBRead.readSK_vertexOld}}
\pysigstartsignatures
\pysiglinewithargsret{\sphinxbfcode{\sphinxupquote{readSK\_vertexOld}}}{\sphinxparam{\DUrole{n,n}{fname}}\sphinxparamcomma \sphinxparam{\DUrole{n,n}{bigfile}\DUrole{o,o}{=}\DUrole{default_value}{False}}\sphinxparamcomma \sphinxparam{\DUrole{n,n}{fast\_rational}\DUrole{o,o}{=}\DUrole{default_value}{False}}\sphinxparamcomma \sphinxparam{\DUrole{n,n}{nformat}\DUrole{o,o}{=}\DUrole{default_value}{\textquotesingle{}\%.14f\textquotesingle{}}}\sphinxparamcomma \sphinxparam{\DUrole{n,n}{compresslevel}\DUrole{o,o}{=}\DUrole{default_value}{3}}}{}
\pysigstopsignatures
\sphinxAtStartPar
Reads in Stevens vertex analysis file and returns, even more optimized for large datasets than the original.
\begin{itemize}
\item {} 
\sphinxAtStartPar
a list of vertex vectors

\item {} 
\sphinxAtStartPar
a list of ray vectors

\item {} 
\sphinxAtStartPar
the basis of the lineality space as a list of vectors

\end{itemize}

\sphinxAtStartPar
all vectors in terms of the column space of N

\end{fulllineitems}

\phantomsection\label{\detokenize{modules_doc:module-cbmpy.CBReadtxt}}\index{module@\spxentry{module}!cbmpy.CBReadtxt@\spxentry{cbmpy.CBReadtxt}}\index{cbmpy.CBReadtxt@\spxentry{cbmpy.CBReadtxt}!module@\spxentry{module}}

\section{CBMPy: CBReadtxt module}
\label{\detokenize{modules_doc:cbmpy-cbreadtxt-module}}
\sphinxAtStartPar
PySCeS Constraint Based Modelling (\sphinxurl{http://cbmpy.sourceforge.net})
Copyright (C) 2009\sphinxhyphen{}2024 Brett G. Olivier, VU University Amsterdam, Amsterdam, The Netherlands

\sphinxAtStartPar
This program is free software: you can redistribute it and/or modify
it under the terms of the GNU General Public License as published by
the Free Software Foundation, either version 3 of the License, or
(at your option) any later version.

\sphinxAtStartPar
This program is distributed in the hope that it will be useful,
but WITHOUT ANY WARRANTY; without even the implied warranty of
MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
GNU General Public License for more details.

\sphinxAtStartPar
You should have received a copy of the GNU General Public License
along with this program.  If not, see \textless{}\sphinxurl{http://www.gnu.org/licenses/}\textgreater{}

\sphinxAtStartPar
Author: Brett G. Olivier PhD
Contact developers: \sphinxurl{https://github.com/SystemsBioinformatics/cbmpy/issues}
Last edit: \$Author: bgoli \$ (\$Id: CBReadtxt.py 710 2020\sphinxhyphen{}04\sphinxhyphen{}27 14:22:34Z bgoli \$)
\index{readCSV() (in module cbmpy.CBReadtxt)@\spxentry{readCSV()}\spxextra{in module cbmpy.CBReadtxt}}

\begin{fulllineitems}
\phantomsection\label{\detokenize{modules_doc:cbmpy.CBReadtxt.readCSV}}
\pysigstartsignatures
\pysiglinewithargsret{\sphinxbfcode{\sphinxupquote{readCSV}}}{\sphinxparam{\DUrole{n,n}{model\_file}}\sphinxparamcomma \sphinxparam{\DUrole{n,n}{bounds\_file}\DUrole{o,o}{=}\DUrole{default_value}{None}}\sphinxparamcomma \sphinxparam{\DUrole{n,n}{biomass\_flux}\DUrole{o,o}{=}\DUrole{default_value}{None}}\sphinxparamcomma \sphinxparam{\DUrole{n,n}{model\_id}\DUrole{o,o}{=}\DUrole{default_value}{\textquotesingle{}FBAModel\textquotesingle{}}}\sphinxparamcomma \sphinxparam{\DUrole{n,n}{reaction\_prefix}\DUrole{o,o}{=}\DUrole{default_value}{\textquotesingle{}R\_\textquotesingle{}}}\sphinxparamcomma \sphinxparam{\DUrole{n,n}{has\_header}\DUrole{o,o}{=}\DUrole{default_value}{False}}}{}
\pysigstopsignatures
\sphinxAtStartPar
This function loads a CSV file and translates it into a Python object:

\begin{sphinxVerbatim}[commandchars=\\\{\}]
\PYG{o}{\PYGZhy{}} \PYG{o}{*}\PYG{n}{model\PYGZus{}file}\PYG{o}{*} \PYG{n}{the} \PYG{n}{name} \PYG{n}{of} \PYG{n}{the} \PYG{n}{CSV} \PYG{n}{file} \PYG{n}{that} \PYG{n}{contains} \PYG{n}{the} \PYG{n}{model}
\PYG{o}{\PYGZhy{}} \PYG{o}{*}\PYG{n}{bounds\PYGZus{}file}\PYG{o}{*} \PYG{n}{the} \PYG{n}{name} \PYG{n}{of} \PYG{n}{the} \PYG{n}{CSV} \PYG{n}{file} \PYG{n}{that} \PYG{n}{contains} \PYG{n}{the} \PYG{n}{flux} \PYG{n}{bounds}
\PYG{o}{\PYGZhy{}} \PYG{o}{*}\PYG{n}{biomass\PYGZus{}flux}\PYG{o}{*} \PYG{n}{the} \PYG{n}{name} \PYG{n}{of} \PYG{n}{the} \PYG{n}{reaction} \PYG{n}{that} \PYG{o+ow}{is} \PYG{n}{the} \PYG{n}{objective} \PYG{n}{function}
\PYG{o}{\PYGZhy{}} \PYG{o}{*}\PYG{n}{reaction\PYGZus{}prefix}\PYG{o}{*} \PYG{p}{[}\PYG{n}{default}\PYG{o}{=}\PYG{l+s+s1}{\PYGZsq{}}\PYG{l+s+s1}{R \PYGZus{}}\PYG{l+s+s1}{\PYGZsq{}}\PYG{p}{]} \PYG{n}{the} \PYG{n}{prefix} \PYG{n}{to} \PYG{n}{add} \PYG{n}{to} \PYG{n+nb}{input} \PYG{n}{reaction} \PYG{n}{ID}\PYG{l+s+s1}{\PYGZsq{}}\PYG{l+s+s1}{s}
\PYG{o}{\PYGZhy{}} \PYG{o}{*}\PYG{n}{has\PYGZus{}header}\PYG{o}{*} \PYG{p}{[}\PYG{n}{default}\PYG{o}{=}\PYG{k+kc}{False}\PYG{p}{]} \PYG{k}{if} \PYG{n}{there} \PYG{o+ow}{is} \PYG{n}{a} \PYG{n}{header} \PYG{n}{row} \PYG{o+ow}{in} \PYG{n}{the} \PYG{n}{csv} \PYG{n}{file}
\end{sphinxVerbatim}

\end{fulllineitems}

\phantomsection\label{\detokenize{modules_doc:module-cbmpy.CBSolver}}\index{module@\spxentry{module}!cbmpy.CBSolver@\spxentry{cbmpy.CBSolver}}\index{cbmpy.CBSolver@\spxentry{cbmpy.CBSolver}!module@\spxentry{module}}

\section{CBMPy: CBSolver module}
\label{\detokenize{modules_doc:cbmpy-cbsolver-module}}
\sphinxAtStartPar
PySCeS Constraint Based Modelling (\sphinxurl{http://cbmpy.sourceforge.net})
Copyright (C) 2009\sphinxhyphen{}2024 Brett G. Olivier, VU University Amsterdam, Amsterdam, The Netherlands

\sphinxAtStartPar
This program is free software: you can redistribute it and/or modify
it under the terms of the GNU General Public License as published by
the Free Software Foundation, either version 3 of the License, or
(at your option) any later version.

\sphinxAtStartPar
This program is distributed in the hope that it will be useful,
but WITHOUT ANY WARRANTY; without even the implied warranty of
MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
GNU General Public License for more details.

\sphinxAtStartPar
You should have received a copy of the GNU General Public License
along with this program.  If not, see \textless{}\sphinxurl{http://www.gnu.org/licenses/}\textgreater{}

\sphinxAtStartPar
Author: Brett G. Olivier PhD
Contact developers: \sphinxurl{https://github.com/SystemsBioinformatics/cbmpy/issues}
Last edit: \$Author: bgoli \$ (\$Id: CBSolver.py 710 2020\sphinxhyphen{}04\sphinxhyphen{}27 14:22:34Z bgoli \$)
\phantomsection\label{\detokenize{modules_doc:module-cbmpy.CBTools}}\index{module@\spxentry{module}!cbmpy.CBTools@\spxentry{cbmpy.CBTools}}\index{cbmpy.CBTools@\spxentry{cbmpy.CBTools}!module@\spxentry{module}}

\section{CBMPy: CBTools module}
\label{\detokenize{modules_doc:cbmpy-cbtools-module}}
\sphinxAtStartPar
PySCeS Constraint Based Modelling (\sphinxurl{http://cbmpy.sourceforge.net})
Copyright (C) 2009\sphinxhyphen{}2024 Brett G. Olivier, VU University Amsterdam, Amsterdam, The Netherlands

\sphinxAtStartPar
This program is free software: you can redistribute it and/or modify
it under the terms of the GNU General Public License as published by
the Free Software Foundation, either version 3 of the License, or
(at your option) any later version.

\sphinxAtStartPar
This program is distributed in the hope that it will be useful,
but WITHOUT ANY WARRANTY; without even the implied warranty of
MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
GNU General Public License for more details.

\sphinxAtStartPar
You should have received a copy of the GNU General Public License
along with this program.  If not, see \textless{}\sphinxurl{http://www.gnu.org/licenses/}\textgreater{}

\sphinxAtStartPar
Author: Brett G. Olivier PhD
Contact developers: \sphinxurl{https://github.com/SystemsBioinformatics/cbmpy/issues}
Last edit: \$Author: bgoli \$ (\$Id: CBTools.py 710 2020\sphinxhyphen{}04\sphinxhyphen{}27 14:22:34Z bgoli \$)
\index{addFluxAsActiveObjective() (in module cbmpy.CBTools)@\spxentry{addFluxAsActiveObjective()}\spxextra{in module cbmpy.CBTools}}

\begin{fulllineitems}
\phantomsection\label{\detokenize{modules_doc:cbmpy.CBTools.addFluxAsActiveObjective}}
\pysigstartsignatures
\pysiglinewithargsret{\sphinxbfcode{\sphinxupquote{addFluxAsActiveObjective}}}{\sphinxparam{\DUrole{n,n}{f}}\sphinxparamcomma \sphinxparam{\DUrole{n,n}{reaction\_id}}\sphinxparamcomma \sphinxparam{\DUrole{n,n}{osense}}\sphinxparamcomma \sphinxparam{\DUrole{n,n}{coefficient}\DUrole{o,o}{=}\DUrole{default_value}{1}}}{}
\pysigstopsignatures
\sphinxAtStartPar
Adds a flux as an active objective function
\begin{itemize}
\item {} 
\sphinxAtStartPar
\sphinxstyleemphasis{reaction\_id} a string containing a reaction id

\item {} 
\sphinxAtStartPar
\sphinxstyleemphasis{osense} objective sense must be \sphinxstylestrong{maximize} or \sphinxstylestrong{minimize}

\item {} 
\sphinxAtStartPar
\sphinxstyleemphasis{coefficient} the objective funtion coefficient {[}default=1{]}

\end{itemize}

\end{fulllineitems}

\index{addGenesFromAnnotations() (in module cbmpy.CBTools)@\spxentry{addGenesFromAnnotations()}\spxextra{in module cbmpy.CBTools}}

\begin{fulllineitems}
\phantomsection\label{\detokenize{modules_doc:cbmpy.CBTools.addGenesFromAnnotations}}
\pysigstartsignatures
\pysiglinewithargsret{\sphinxbfcode{\sphinxupquote{addGenesFromAnnotations}}}{\sphinxparam{\DUrole{n,n}{fba}}\sphinxparamcomma \sphinxparam{\DUrole{n,n}{annotation\_key}\DUrole{o,o}{=}\DUrole{default_value}{\textquotesingle{}GENE ASSOCIATION\textquotesingle{}}}\sphinxparamcomma \sphinxparam{\DUrole{n,n}{gene\_pattern}\DUrole{o,o}{=}\DUrole{default_value}{None}}}{}
\pysigstopsignatures
\sphinxAtStartPar
THIS METHOD IS DERPRECATED PLEASE USE cmod.createGeneAssociationsFromAnnotations()

\sphinxAtStartPar
Add genes to the model using the definitions stored in the annotation key
\begin{itemize}
\item {} 
\sphinxAtStartPar
\sphinxstyleemphasis{fba} and fba object

\item {} 
\sphinxAtStartPar
\sphinxstyleemphasis{annotation\_key} the annotation dictionary key that holds the gene association for the protein/enzyme

\item {} 
\sphinxAtStartPar
\sphinxstyleemphasis{gene\_pattern} deprecated, not needed anymore

\end{itemize}

\end{fulllineitems}

\index{addSinkReaction() (in module cbmpy.CBTools)@\spxentry{addSinkReaction()}\spxextra{in module cbmpy.CBTools}}

\begin{fulllineitems}
\phantomsection\label{\detokenize{modules_doc:cbmpy.CBTools.addSinkReaction}}
\pysigstartsignatures
\pysiglinewithargsret{\sphinxbfcode{\sphinxupquote{addSinkReaction}}}{\sphinxparam{\DUrole{n,n}{fbam}}\sphinxparamcomma \sphinxparam{\DUrole{n,n}{species}}\sphinxparamcomma \sphinxparam{\DUrole{n,n}{lb}\DUrole{o,o}{=}\DUrole{default_value}{0.0}}\sphinxparamcomma \sphinxparam{\DUrole{n,n}{ub}\DUrole{o,o}{=}\DUrole{default_value}{1000.0}}}{}
\pysigstopsignatures
\sphinxAtStartPar
Adds a sink reactions that consumes a model \sphinxstyleemphasis{species} so that X \textendash{}\textgreater{}
\begin{itemize}
\item {} 
\sphinxAtStartPar
\sphinxstyleemphasis{fbam} an fba model object

\item {} 
\sphinxAtStartPar
\sphinxstyleemphasis{species} a valid species name

\item {} 
\sphinxAtStartPar
\sphinxstyleemphasis{lb} lower flux bound {[}default = 0.0{]}

\item {} 
\sphinxAtStartPar
\sphinxstyleemphasis{ub} upper flux bound {[}default = 1000.0{]}

\end{itemize}

\end{fulllineitems}

\index{addSourceReaction() (in module cbmpy.CBTools)@\spxentry{addSourceReaction()}\spxextra{in module cbmpy.CBTools}}

\begin{fulllineitems}
\phantomsection\label{\detokenize{modules_doc:cbmpy.CBTools.addSourceReaction}}
\pysigstartsignatures
\pysiglinewithargsret{\sphinxbfcode{\sphinxupquote{addSourceReaction}}}{\sphinxparam{\DUrole{n,n}{fbam}}\sphinxparamcomma \sphinxparam{\DUrole{n,n}{species}}\sphinxparamcomma \sphinxparam{\DUrole{n,n}{lb}\DUrole{o,o}{=}\DUrole{default_value}{0.0}}\sphinxparamcomma \sphinxparam{\DUrole{n,n}{ub}\DUrole{o,o}{=}\DUrole{default_value}{1000.0}}}{}
\pysigstopsignatures
\sphinxAtStartPar
Adds a source reactions that produces a model \sphinxstyleemphasis{species} so that \textendash{}\textgreater{} X
\begin{itemize}
\item {} 
\sphinxAtStartPar
\sphinxstyleemphasis{fbam} an fba model object

\item {} 
\sphinxAtStartPar
\sphinxstyleemphasis{species} a valid species name

\item {} 
\sphinxAtStartPar
\sphinxstyleemphasis{lb} lower flux bound {[}default = 0.0{]}

\item {} 
\sphinxAtStartPar
\sphinxstyleemphasis{ub} upper flux bound {[}default = 1000.0{]}

\end{itemize}

\sphinxAtStartPar
Note reversiblity is determined by the lower bound, default 0 = irreversible. If
negative then reversible.

\end{fulllineitems}

\index{addStoichToFBAModel() (in module cbmpy.CBTools)@\spxentry{addStoichToFBAModel()}\spxextra{in module cbmpy.CBTools}}

\begin{fulllineitems}
\phantomsection\label{\detokenize{modules_doc:cbmpy.CBTools.addStoichToFBAModel}}
\pysigstartsignatures
\pysiglinewithargsret{\sphinxbfcode{\sphinxupquote{addStoichToFBAModel}}}{\sphinxparam{\DUrole{n,n}{fm}}}{}
\pysigstopsignatures
\sphinxAtStartPar
Build stoichiometry: this method has been refactored into the model class \sphinxhyphen{} cmod.buildStoichMatrix()

\end{fulllineitems}

\index{checkExchangeReactions() (in module cbmpy.CBTools)@\spxentry{checkExchangeReactions()}\spxextra{in module cbmpy.CBTools}}

\begin{fulllineitems}
\phantomsection\label{\detokenize{modules_doc:cbmpy.CBTools.checkExchangeReactions}}
\pysigstartsignatures
\pysiglinewithargsret{\sphinxbfcode{\sphinxupquote{checkExchangeReactions}}}{\sphinxparam{\DUrole{n,n}{fba}}\sphinxparamcomma \sphinxparam{\DUrole{n,n}{autocorrect}\DUrole{o,o}{=}\DUrole{default_value}{True}}}{}
\pysigstopsignatures
\sphinxAtStartPar
Scan all reactions for exchange reactions (reactions containing a boundary species), return a list of
inconsistent reactions or correct automatically.
\begin{itemize}
\item {} 
\sphinxAtStartPar
\sphinxstyleemphasis{fba} a CBMPy model

\item {} 
\sphinxAtStartPar
\sphinxstyleemphasis{autocorrect} {[}default=True{]} correctly set the “is\_exchange” attribute on a reaction

\end{itemize}

\end{fulllineitems}

\index{checkFluxBoundConsistency() (in module cbmpy.CBTools)@\spxentry{checkFluxBoundConsistency()}\spxextra{in module cbmpy.CBTools}}

\begin{fulllineitems}
\phantomsection\label{\detokenize{modules_doc:cbmpy.CBTools.checkFluxBoundConsistency}}
\pysigstartsignatures
\pysiglinewithargsret{\sphinxbfcode{\sphinxupquote{checkFluxBoundConsistency}}}{\sphinxparam{\DUrole{n,n}{fba}}}{}
\pysigstopsignatures
\sphinxAtStartPar
Check flux bound consistency checks for multiply defined bounds, bounds without a reaction, inconsistent bounds with respect to each other
and reaction reversbility. Returns a dictionary of bounds/reactions where errors occur.

\end{fulllineitems}

\index{checkIds() (in module cbmpy.CBTools)@\spxentry{checkIds()}\spxextra{in module cbmpy.CBTools}}

\begin{fulllineitems}
\phantomsection\label{\detokenize{modules_doc:cbmpy.CBTools.checkIds}}
\pysigstartsignatures
\pysiglinewithargsret{\sphinxbfcode{\sphinxupquote{checkIds}}}{\sphinxparam{\DUrole{n,n}{fba}}\sphinxparamcomma \sphinxparam{\DUrole{n,n}{items}\DUrole{o,o}{=}\DUrole{default_value}{\textquotesingle{}all\textquotesingle{}}}}{}
\pysigstopsignatures
\sphinxAtStartPar
Checks the id’s of the specified model attributes to see if the name is legal and if there are duplicates.
Returns a list of items with errors.
\begin{itemize}
\item {} 
\sphinxAtStartPar
\sphinxstyleemphasis{fba} a CBMPy model instance

\item {} 
\sphinxAtStartPar
\sphinxstyleemphasis{items} {[}default=’all’{]} ‘all’ means ‘species,reactions,flux\_bounds,objectives’ of which one or more can be specified

\end{itemize}

\end{fulllineitems}

\index{checkReactionBalanceElemental() (in module cbmpy.CBTools)@\spxentry{checkReactionBalanceElemental()}\spxextra{in module cbmpy.CBTools}}

\begin{fulllineitems}
\phantomsection\label{\detokenize{modules_doc:cbmpy.CBTools.checkReactionBalanceElemental}}
\pysigstartsignatures
\pysiglinewithargsret{\sphinxbfcode{\sphinxupquote{checkReactionBalanceElemental}}}{\sphinxparam{\DUrole{n,n}{f}}\sphinxparamcomma \sphinxparam{\DUrole{n,n}{Rid}\DUrole{o,o}{=}\DUrole{default_value}{None}}\sphinxparamcomma \sphinxparam{\DUrole{n,n}{zero\_tol}\DUrole{o,o}{=}\DUrole{default_value}{1e\sphinxhyphen{}12}}}{}
\pysigstopsignatures
\sphinxAtStartPar
Check if the reaction is balanced using the chemical formula
\begin{itemize}
\item {} 
\sphinxAtStartPar
\sphinxstyleemphasis{f} the FBA object

\item {} 
\sphinxAtStartPar
\sphinxstyleemphasis{Rid} {[}default = None{]} the reaction to check, defaults to all

\item {} 
\sphinxAtStartPar
\sphinxstyleemphasis{zero\_tol} {[}default=1.0e\sphinxhyphen{}12{]} the floating point zero used for elemental balancing

\end{itemize}

\sphinxAtStartPar
This function is derived from the code found here: \sphinxurl{http://pyparsing.wikispaces.com/file/view/chemicalFormulas.py}

\end{fulllineitems}

\index{createTempFileName() (in module cbmpy.CBTools)@\spxentry{createTempFileName()}\spxextra{in module cbmpy.CBTools}}

\begin{fulllineitems}
\phantomsection\label{\detokenize{modules_doc:cbmpy.CBTools.createTempFileName}}
\pysigstartsignatures
\pysiglinewithargsret{\sphinxbfcode{\sphinxupquote{createTempFileName}}}{}{}
\pysigstopsignatures
\sphinxAtStartPar
Return a temporary filename

\end{fulllineitems}

\index{createZipArchive() (in module cbmpy.CBTools)@\spxentry{createZipArchive()}\spxextra{in module cbmpy.CBTools}}

\begin{fulllineitems}
\phantomsection\label{\detokenize{modules_doc:cbmpy.CBTools.createZipArchive}}
\pysigstartsignatures
\pysiglinewithargsret{\sphinxbfcode{\sphinxupquote{createZipArchive}}}{\sphinxparam{\DUrole{n,n}{zipname}}\sphinxparamcomma \sphinxparam{\DUrole{n,n}{files}}\sphinxparamcomma \sphinxparam{\DUrole{n,n}{move}\DUrole{o,o}{=}\DUrole{default_value}{False}}\sphinxparamcomma \sphinxparam{\DUrole{n,n}{compression}\DUrole{o,o}{=}\DUrole{default_value}{\textquotesingle{}normal\textquotesingle{}}}}{}
\pysigstopsignatures
\sphinxAtStartPar
Create a zip archive which contains one or more files
\begin{itemize}
\item {} 
\sphinxAtStartPar
\sphinxstyleemphasis{zipname} the name of the zip archive to create (fully qualified)

\item {} 
\sphinxAtStartPar
\sphinxstyleemphasis{files} either a valid filename or a list of filenames (fully qualified)

\item {} 
\sphinxAtStartPar
\sphinxstyleemphasis{move} {[}default=False{]} attempt to delete input files after zip\sphinxhyphen{}archive creation

\item {} 
\sphinxAtStartPar
\sphinxstyleemphasis{compression} {[}default=’normal’{]} normal zip compression, set as None for no compression only store files (zlib not required)

\end{itemize}

\end{fulllineitems}

\index{deSerialize() (in module cbmpy.CBTools)@\spxentry{deSerialize()}\spxextra{in module cbmpy.CBTools}}

\begin{fulllineitems}
\phantomsection\label{\detokenize{modules_doc:cbmpy.CBTools.deSerialize}}
\pysigstartsignatures
\pysiglinewithargsret{\sphinxbfcode{\sphinxupquote{deSerialize}}}{\sphinxparam{\DUrole{n,n}{s}}}{}
\pysigstopsignatures
\sphinxAtStartPar
Deserializes a serialised object contained in a string

\end{fulllineitems}

\index{deSerializeFromDisk() (in module cbmpy.CBTools)@\spxentry{deSerializeFromDisk()}\spxextra{in module cbmpy.CBTools}}

\begin{fulllineitems}
\phantomsection\label{\detokenize{modules_doc:cbmpy.CBTools.deSerializeFromDisk}}
\pysigstartsignatures
\pysiglinewithargsret{\sphinxbfcode{\sphinxupquote{deSerializeFromDisk}}}{\sphinxparam{\DUrole{n,n}{filename}}}{}
\pysigstopsignatures
\sphinxAtStartPar
Loads a serialised Python pickle from \sphinxstyleemphasis{filename} returns the Python object(s)

\end{fulllineitems}

\index{exportArray2CSV() (in module cbmpy.CBTools)@\spxentry{exportArray2CSV()}\spxextra{in module cbmpy.CBTools}}

\begin{fulllineitems}
\phantomsection\label{\detokenize{modules_doc:cbmpy.CBTools.exportArray2CSV}}
\pysigstartsignatures
\pysiglinewithargsret{\sphinxbfcode{\sphinxupquote{exportArray2CSV}}}{\sphinxparam{\DUrole{n,n}{arr}}\sphinxparamcomma \sphinxparam{\DUrole{n,n}{fname}}}{}
\pysigstopsignatures
\sphinxAtStartPar
Export an array to fname.csv
\begin{itemize}
\item {} 
\sphinxAtStartPar
\sphinxstyleemphasis{arr} the an array like object

\item {} 
\sphinxAtStartPar
\sphinxstyleemphasis{fname} the output filename

\item {} 
\sphinxAtStartPar
\sphinxstyleemphasis{sep} {[}default=’,’{]} the column separator

\end{itemize}

\end{fulllineitems}

\index{exportArray2TXT() (in module cbmpy.CBTools)@\spxentry{exportArray2TXT()}\spxextra{in module cbmpy.CBTools}}

\begin{fulllineitems}
\phantomsection\label{\detokenize{modules_doc:cbmpy.CBTools.exportArray2TXT}}
\pysigstartsignatures
\pysiglinewithargsret{\sphinxbfcode{\sphinxupquote{exportArray2TXT}}}{\sphinxparam{\DUrole{n,n}{arr}}\sphinxparamcomma \sphinxparam{\DUrole{n,n}{fname}}}{}
\pysigstopsignatures
\sphinxAtStartPar
Export an array to fname.txt
\begin{itemize}
\item {} 
\sphinxAtStartPar
\sphinxstyleemphasis{arr} the an array like object

\item {} 
\sphinxAtStartPar
\sphinxstyleemphasis{fname} the output filename

\item {} 
\sphinxAtStartPar
\sphinxstyleemphasis{sep} {[}default=’,’{]} the column separator

\end{itemize}

\end{fulllineitems}

\index{exportLabelledArray() (in module cbmpy.CBTools)@\spxentry{exportLabelledArray()}\spxextra{in module cbmpy.CBTools}}

\begin{fulllineitems}
\phantomsection\label{\detokenize{modules_doc:cbmpy.CBTools.exportLabelledArray}}
\pysigstartsignatures
\pysiglinewithargsret{\sphinxbfcode{\sphinxupquote{exportLabelledArray}}}{\sphinxparam{\DUrole{n,n}{arr}}\sphinxparamcomma \sphinxparam{\DUrole{n,n}{fname}}\sphinxparamcomma \sphinxparam{\DUrole{n,n}{names}\DUrole{o,o}{=}\DUrole{default_value}{None}}\sphinxparamcomma \sphinxparam{\DUrole{n,n}{sep}\DUrole{o,o}{=}\DUrole{default_value}{\textquotesingle{},\textquotesingle{}}}\sphinxparamcomma \sphinxparam{\DUrole{n,n}{fmt}\DUrole{o,o}{=}\DUrole{default_value}{\textquotesingle{}\%f\textquotesingle{}}}}{}
\pysigstopsignatures
\sphinxAtStartPar
Write a 2D array type object to file
\begin{itemize}
\item {} 
\sphinxAtStartPar
\sphinxstyleemphasis{arr} the an array like object

\item {} 
\sphinxAtStartPar
\sphinxstyleemphasis{names} {[}default=None{]} the list of row names

\item {} 
\sphinxAtStartPar
\sphinxstyleemphasis{fname} the output filename

\item {} 
\sphinxAtStartPar
\sphinxstyleemphasis{sep} {[}default=’,’{]} the column separator

\item {} 
\sphinxAtStartPar
\sphinxstyleemphasis{fmt} {[}default=’\%s’{]} the output number format

\end{itemize}

\end{fulllineitems}

\index{exportLabelledArray2CSV() (in module cbmpy.CBTools)@\spxentry{exportLabelledArray2CSV()}\spxextra{in module cbmpy.CBTools}}

\begin{fulllineitems}
\phantomsection\label{\detokenize{modules_doc:cbmpy.CBTools.exportLabelledArray2CSV}}
\pysigstartsignatures
\pysiglinewithargsret{\sphinxbfcode{\sphinxupquote{exportLabelledArray2CSV}}}{\sphinxparam{\DUrole{n,n}{arr}}\sphinxparamcomma \sphinxparam{\DUrole{n,n}{fname}}\sphinxparamcomma \sphinxparam{\DUrole{n,n}{names}\DUrole{o,o}{=}\DUrole{default_value}{None}}}{}
\pysigstopsignatures
\sphinxAtStartPar
Export an array with row names to fname.csv
\begin{itemize}
\item {} 
\sphinxAtStartPar
\sphinxstyleemphasis{arr} the an array like object

\item {} 
\sphinxAtStartPar
\sphinxstyleemphasis{fname} the output filename

\item {} 
\sphinxAtStartPar
\sphinxstyleemphasis{names} {[}default=None{]} the list of row names

\end{itemize}

\end{fulllineitems}

\index{exportLabelledArray2TXT() (in module cbmpy.CBTools)@\spxentry{exportLabelledArray2TXT()}\spxextra{in module cbmpy.CBTools}}

\begin{fulllineitems}
\phantomsection\label{\detokenize{modules_doc:cbmpy.CBTools.exportLabelledArray2TXT}}
\pysigstartsignatures
\pysiglinewithargsret{\sphinxbfcode{\sphinxupquote{exportLabelledArray2TXT}}}{\sphinxparam{\DUrole{n,n}{arr}}\sphinxparamcomma \sphinxparam{\DUrole{n,n}{fname}}\sphinxparamcomma \sphinxparam{\DUrole{n,n}{names}\DUrole{o,o}{=}\DUrole{default_value}{None}}}{}
\pysigstopsignatures
\sphinxAtStartPar
Export an array with row names to fname.txt
\begin{itemize}
\item {} 
\sphinxAtStartPar
\sphinxstyleemphasis{arr} the an array like object

\item {} 
\sphinxAtStartPar
\sphinxstyleemphasis{names} {[}default=None{]} the list of row names

\item {} 
\sphinxAtStartPar
\sphinxstyleemphasis{fname} the output filename

\end{itemize}

\end{fulllineitems}

\index{exportLabelledArrayWithHeader() (in module cbmpy.CBTools)@\spxentry{exportLabelledArrayWithHeader()}\spxextra{in module cbmpy.CBTools}}

\begin{fulllineitems}
\phantomsection\label{\detokenize{modules_doc:cbmpy.CBTools.exportLabelledArrayWithHeader}}
\pysigstartsignatures
\pysiglinewithargsret{\sphinxbfcode{\sphinxupquote{exportLabelledArrayWithHeader}}}{\sphinxparam{\DUrole{n,n}{arr}}\sphinxparamcomma \sphinxparam{\DUrole{n,n}{fname}}\sphinxparamcomma \sphinxparam{\DUrole{n,n}{names}\DUrole{o,o}{=}\DUrole{default_value}{None}}\sphinxparamcomma \sphinxparam{\DUrole{n,n}{header}\DUrole{o,o}{=}\DUrole{default_value}{None}}\sphinxparamcomma \sphinxparam{\DUrole{n,n}{sep}\DUrole{o,o}{=}\DUrole{default_value}{\textquotesingle{},\textquotesingle{}}}\sphinxparamcomma \sphinxparam{\DUrole{n,n}{fmt}\DUrole{o,o}{=}\DUrole{default_value}{\textquotesingle{}\%f\textquotesingle{}}}}{}
\pysigstopsignatures
\sphinxAtStartPar
Export an array with row names and header
\begin{itemize}
\item {} 
\sphinxAtStartPar
\sphinxstyleemphasis{arr} the an array like object

\item {} 
\sphinxAtStartPar
\sphinxstyleemphasis{names} {[}default=None{]} the list of row names

\item {} 
\sphinxAtStartPar
\sphinxstyleemphasis{header} {[}default=None{]} the list of column names

\item {} 
\sphinxAtStartPar
\sphinxstyleemphasis{fname} the output filename

\item {} 
\sphinxAtStartPar
\sphinxstyleemphasis{sep} {[}default=’,’{]} the column separator

\item {} 
\sphinxAtStartPar
\sphinxstyleemphasis{fmt} {[}default=’\%s’{]} the output number format

\item {} 
\sphinxAtStartPar
\sphinxstyleemphasis{appendlist} {[}default=False{]} if True append the array to \sphinxstyleemphasis{fname} otherwise create a new file

\end{itemize}

\end{fulllineitems}

\index{exportLabelledArrayWithHeader2CSV() (in module cbmpy.CBTools)@\spxentry{exportLabelledArrayWithHeader2CSV()}\spxextra{in module cbmpy.CBTools}}

\begin{fulllineitems}
\phantomsection\label{\detokenize{modules_doc:cbmpy.CBTools.exportLabelledArrayWithHeader2CSV}}
\pysigstartsignatures
\pysiglinewithargsret{\sphinxbfcode{\sphinxupquote{exportLabelledArrayWithHeader2CSV}}}{\sphinxparam{\DUrole{n,n}{arr}}\sphinxparamcomma \sphinxparam{\DUrole{n,n}{fname}}\sphinxparamcomma \sphinxparam{\DUrole{n,n}{names}\DUrole{o,o}{=}\DUrole{default_value}{None}}\sphinxparamcomma \sphinxparam{\DUrole{n,n}{header}\DUrole{o,o}{=}\DUrole{default_value}{None}}}{}
\pysigstopsignatures
\sphinxAtStartPar
Export an array with row names and header to fname.csv
\begin{itemize}
\item {} 
\sphinxAtStartPar
\sphinxstyleemphasis{arr} the an array like object

\item {} 
\sphinxAtStartPar
\sphinxstyleemphasis{fname} the output filename

\item {} 
\sphinxAtStartPar
\sphinxstyleemphasis{names} {[}default=None{]} the list of row names

\item {} 
\sphinxAtStartPar
\sphinxstyleemphasis{header} {[}default=None{]} the list of column names

\end{itemize}

\end{fulllineitems}

\index{exportLabelledArrayWithHeader2TXT() (in module cbmpy.CBTools)@\spxentry{exportLabelledArrayWithHeader2TXT()}\spxextra{in module cbmpy.CBTools}}

\begin{fulllineitems}
\phantomsection\label{\detokenize{modules_doc:cbmpy.CBTools.exportLabelledArrayWithHeader2TXT}}
\pysigstartsignatures
\pysiglinewithargsret{\sphinxbfcode{\sphinxupquote{exportLabelledArrayWithHeader2TXT}}}{\sphinxparam{\DUrole{n,n}{arr}}\sphinxparamcomma \sphinxparam{\DUrole{n,n}{fname}}\sphinxparamcomma \sphinxparam{\DUrole{n,n}{names}\DUrole{o,o}{=}\DUrole{default_value}{None}}\sphinxparamcomma \sphinxparam{\DUrole{n,n}{header}\DUrole{o,o}{=}\DUrole{default_value}{None}}}{}
\pysigstopsignatures
\sphinxAtStartPar
Export an array with row names and header to fname.txt
\begin{itemize}
\item {} 
\sphinxAtStartPar
\sphinxstyleemphasis{arr} the an array like object

\item {} 
\sphinxAtStartPar
\sphinxstyleemphasis{names} the list of row names

\item {} 
\sphinxAtStartPar
\sphinxstyleemphasis{header} the list of column names

\item {} 
\sphinxAtStartPar
\sphinxstyleemphasis{fname} the output filename

\end{itemize}

\end{fulllineitems}

\index{exportLabelledLinkedList() (in module cbmpy.CBTools)@\spxentry{exportLabelledLinkedList()}\spxextra{in module cbmpy.CBTools}}

\begin{fulllineitems}
\phantomsection\label{\detokenize{modules_doc:cbmpy.CBTools.exportLabelledLinkedList}}
\pysigstartsignatures
\pysiglinewithargsret{\sphinxbfcode{\sphinxupquote{exportLabelledLinkedList}}}{\sphinxparam{\DUrole{n,n}{arr}}\sphinxparamcomma \sphinxparam{\DUrole{n,n}{fname}}\sphinxparamcomma \sphinxparam{\DUrole{n,n}{names}\DUrole{o,o}{=}\DUrole{default_value}{None}}\sphinxparamcomma \sphinxparam{\DUrole{n,n}{sep}\DUrole{o,o}{=}\DUrole{default_value}{\textquotesingle{},\textquotesingle{}}}\sphinxparamcomma \sphinxparam{\DUrole{n,n}{fmt}\DUrole{o,o}{=}\DUrole{default_value}{\textquotesingle{}\%s\textquotesingle{}}}\sphinxparamcomma \sphinxparam{\DUrole{n,n}{appendlist}\DUrole{o,o}{=}\DUrole{default_value}{False}}}{}
\pysigstopsignatures
\sphinxAtStartPar
Write a 2D linked list {[}{[}…{]},{[}…{]},{[}…{]},{[}…{]}{]} and optionally a list of row labels to file:
\begin{itemize}
\item {} 
\sphinxAtStartPar
\sphinxstyleemphasis{arr} the linked list

\item {} 
\sphinxAtStartPar
\sphinxstyleemphasis{fname} the output filename

\item {} 
\sphinxAtStartPar
\sphinxstyleemphasis{names} {[}default=None{]} the list of row names

\item {} 
\sphinxAtStartPar
\sphinxstyleemphasis{sep} {[}default=’,’{]} the column separator

\item {} 
\sphinxAtStartPar
\sphinxstyleemphasis{fmt} {[}default=’\%s’{]} the output number format

\item {} 
\sphinxAtStartPar
\sphinxstyleemphasis{appendlist} {[}default=False{]} if True append the array to \sphinxstyleemphasis{fname} otherwise create a new file

\end{itemize}

\end{fulllineitems}

\index{findDeadEndMetabolites() (in module cbmpy.CBTools)@\spxentry{findDeadEndMetabolites()}\spxextra{in module cbmpy.CBTools}}

\begin{fulllineitems}
\phantomsection\label{\detokenize{modules_doc:cbmpy.CBTools.findDeadEndMetabolites}}
\pysigstartsignatures
\pysiglinewithargsret{\sphinxbfcode{\sphinxupquote{findDeadEndMetabolites}}}{\sphinxparam{\DUrole{n,n}{fbam}}}{}
\pysigstopsignatures
\sphinxAtStartPar
Finds dead\sphinxhyphen{}end (single reaction) metabolites rows in N with a single entry), returns a list of (metabolite, reaction) ids

\end{fulllineitems}

\index{findDeadEndReactions() (in module cbmpy.CBTools)@\spxentry{findDeadEndReactions()}\spxextra{in module cbmpy.CBTools}}

\begin{fulllineitems}
\phantomsection\label{\detokenize{modules_doc:cbmpy.CBTools.findDeadEndReactions}}
\pysigstartsignatures
\pysiglinewithargsret{\sphinxbfcode{\sphinxupquote{findDeadEndReactions}}}{\sphinxparam{\DUrole{n,n}{fbam}}}{}
\pysigstopsignatures
\sphinxAtStartPar
Finds dead\sphinxhyphen{}end (single substrate/product) reactions (cols in N with a single entry), returns a list of (metabolite, reaction) ids

\end{fulllineitems}

\index{fixReversibility() (in module cbmpy.CBTools)@\spxentry{fixReversibility()}\spxextra{in module cbmpy.CBTools}}

\begin{fulllineitems}
\phantomsection\label{\detokenize{modules_doc:cbmpy.CBTools.fixReversibility}}
\pysigstartsignatures
\pysiglinewithargsret{\sphinxbfcode{\sphinxupquote{fixReversibility}}}{\sphinxparam{\DUrole{n,n}{fbam}}\sphinxparamcomma \sphinxparam{\DUrole{n,n}{auto\_correct}\DUrole{o,o}{=}\DUrole{default_value}{False}}}{}
\pysigstopsignatures
\sphinxAtStartPar
Set fluxbound lower bound from reactions reversibility information.
\begin{itemize}
\item {} 
\sphinxAtStartPar
\sphinxstyleemphasis{fbam} and FBAModel instance

\item {} 
\sphinxAtStartPar
\sphinxstyleemphasis{auto\_correct} (default=False) if True automatically sets lower bound to zero if required, otherwise prints a warning if false.

\end{itemize}

\end{fulllineitems}

\index{getBoundsDict() (in module cbmpy.CBTools)@\spxentry{getBoundsDict()}\spxextra{in module cbmpy.CBTools}}

\begin{fulllineitems}
\phantomsection\label{\detokenize{modules_doc:cbmpy.CBTools.getBoundsDict}}
\pysigstartsignatures
\pysiglinewithargsret{\sphinxbfcode{\sphinxupquote{getBoundsDict}}}{\sphinxparam{\DUrole{n,n}{fbamod}}\sphinxparamcomma \sphinxparam{\DUrole{n,n}{substring}\DUrole{o,o}{=}\DUrole{default_value}{None}}}{}
\pysigstopsignatures
\sphinxAtStartPar
Return a dictionary of reactions\&bounds

\end{fulllineitems}

\index{getExchBoundsDict() (in module cbmpy.CBTools)@\spxentry{getExchBoundsDict()}\spxextra{in module cbmpy.CBTools}}

\begin{fulllineitems}
\phantomsection\label{\detokenize{modules_doc:cbmpy.CBTools.getExchBoundsDict}}
\pysigstartsignatures
\pysiglinewithargsret{\sphinxbfcode{\sphinxupquote{getExchBoundsDict}}}{\sphinxparam{\DUrole{n,n}{fbamod}}}{}
\pysigstopsignatures
\sphinxAtStartPar
Return a dictionary of all exchange reactions (as determined by the is\_exchange attribute of Reaction)
\begin{itemize}
\item {} 
\sphinxAtStartPar
\sphinxstyleemphasis{fbamod} a CBMPy model

\end{itemize}

\end{fulllineitems}

\index{getModelGenesPerReaction() (in module cbmpy.CBTools)@\spxentry{getModelGenesPerReaction()}\spxextra{in module cbmpy.CBTools}}

\begin{fulllineitems}
\phantomsection\label{\detokenize{modules_doc:cbmpy.CBTools.getModelGenesPerReaction}}
\pysigstartsignatures
\pysiglinewithargsret{\sphinxbfcode{\sphinxupquote{getModelGenesPerReaction}}}{\sphinxparam{\DUrole{n,n}{fba}}\sphinxparamcomma \sphinxparam{\DUrole{n,n}{gene\_pattern}\DUrole{o,o}{=}\DUrole{default_value}{None}}\sphinxparamcomma \sphinxparam{\DUrole{n,n}{gene\_annotation\_key}\DUrole{o,o}{=}\DUrole{default_value}{\textquotesingle{}GENE ASSOCIATION\textquotesingle{}}}}{}
\pysigstopsignatures
\sphinxAtStartPar
Parse a BiGG style gene annotation string using default gene\_pattern=’((W*w*W*))’ or
(\textless{}any non\sphinxhyphen{}alphanum\textgreater{}\textless{}any alphanum\textgreater{}\textless{}any non\sphinxhyphen{}alphanum\textgreater{})

\sphinxAtStartPar
Old eColi specific pattern ‘(bw*W)’

\sphinxAtStartPar
It is advisable to use the model methods directly rather than this function

\end{fulllineitems}

\index{loadObj() (in module cbmpy.CBTools)@\spxentry{loadObj()}\spxextra{in module cbmpy.CBTools}}

\begin{fulllineitems}
\phantomsection\label{\detokenize{modules_doc:cbmpy.CBTools.loadObj}}
\pysigstartsignatures
\pysiglinewithargsret{\sphinxbfcode{\sphinxupquote{loadObj}}}{\sphinxparam{\DUrole{n,n}{filename}}}{}
\pysigstopsignatures
\sphinxAtStartPar
Loads a serialised Python pickle from \sphinxstyleemphasis{filename}.dat returns the Python object(s)

\end{fulllineitems}

\index{merge2Models() (in module cbmpy.CBTools)@\spxentry{merge2Models()}\spxextra{in module cbmpy.CBTools}}

\begin{fulllineitems}
\phantomsection\label{\detokenize{modules_doc:cbmpy.CBTools.merge2Models}}
\pysigstartsignatures
\pysiglinewithargsret{\sphinxbfcode{\sphinxupquote{merge2Models}}}{\sphinxparam{\DUrole{n,n}{m1}}\sphinxparamcomma \sphinxparam{\DUrole{n,n}{m2}}\sphinxparamcomma \sphinxparam{\DUrole{n,n}{ignore}\DUrole{o,o}{=}\DUrole{default_value}{None}}\sphinxparamcomma \sphinxparam{\DUrole{n,n}{ignore\_duplicate\_ids}\DUrole{o,o}{=}\DUrole{default_value}{False}}}{}
\pysigstopsignatures
\sphinxAtStartPar
Merge 2 models, this method does a raw merge of model 2 into model 1 without any model checking.
Component id’s in ignore are ignored in both models and the first objective of model 1 is arbitrarily
set as active. Compartments are also merged and a new “OuterMerge” compartment is also created.

\sphinxAtStartPar
In all cases duplicate id’s are tracked and ignored, essentially using the object id encountered first \sphinxhyphen{}
usually that of model 1. Duplicate checking can be disabled by setting the \sphinxstyleemphasis{ignore\_duplicate\_ids} flag.
\begin{itemize}
\item {} 
\sphinxAtStartPar
\sphinxstyleemphasis{m1} model 1

\item {} 
\sphinxAtStartPar
\sphinxstyleemphasis{m2} model 2

\item {} 
\sphinxAtStartPar
\sphinxstyleemphasis{ignore} {[}{[}{]}{]} do not merge these id’s

\item {} 
\sphinxAtStartPar
\sphinxstyleemphasis{ignore\_duplicate\_ids} {[}False{]} default behaviour that can be enabled

\end{itemize}

\sphinxAtStartPar
In development: merging genes and gpr’s.

\end{fulllineitems}

\index{mergeGroups() (in module cbmpy.CBTools)@\spxentry{mergeGroups()}\spxextra{in module cbmpy.CBTools}}

\begin{fulllineitems}
\phantomsection\label{\detokenize{modules_doc:cbmpy.CBTools.mergeGroups}}
\pysigstartsignatures
\pysiglinewithargsret{\sphinxbfcode{\sphinxupquote{mergeGroups}}}{\sphinxparam{\DUrole{n,n}{m}}\sphinxparamcomma \sphinxparam{\DUrole{n,n}{groups}}\sphinxparamcomma \sphinxparam{\DUrole{n,n}{new\_id}}\sphinxparamcomma \sphinxparam{\DUrole{n,n}{new\_name}\DUrole{o,o}{=}\DUrole{default_value}{\textquotesingle{}\textquotesingle{}}}\sphinxparamcomma \sphinxparam{\DUrole{n,n}{auto\_delete}\DUrole{o,o}{=}\DUrole{default_value}{False}}}{}
\pysigstopsignatures
\sphinxAtStartPar
Merge a list of groups into a new group. Note, annotations are not merged!
\begin{itemize}
\item {} 
\sphinxAtStartPar
\sphinxstyleemphasis{m} the model containing the source groups

\item {} 
\sphinxAtStartPar
\sphinxstyleemphasis{groups} a list of groups

\item {} 
\sphinxAtStartPar
\sphinxstyleemphasis{new\_id} the new, merged, group id

\item {} 
\sphinxAtStartPar
\sphinxstyleemphasis{new\_name} {[}default=’’{]} the new group name, the default behaviour is to merge the old names

\item {} 
\sphinxAtStartPar
\sphinxstyleemphasis{auto\_delete} {[}default=False{]} delete the source groups

\end{itemize}

\end{fulllineitems}

\index{processBiGGannotationNote() (in module cbmpy.CBTools)@\spxentry{processBiGGannotationNote()}\spxextra{in module cbmpy.CBTools}}

\begin{fulllineitems}
\phantomsection\label{\detokenize{modules_doc:cbmpy.CBTools.processBiGGannotationNote}}
\pysigstartsignatures
\pysiglinewithargsret{\sphinxbfcode{\sphinxupquote{processBiGGannotationNote}}}{\sphinxparam{\DUrole{n,n}{fba}}\sphinxparamcomma \sphinxparam{\DUrole{n,n}{annotation\_key}\DUrole{o,o}{=}\DUrole{default_value}{\textquotesingle{}note\textquotesingle{}}}}{}
\pysigstopsignatures
\sphinxAtStartPar
Parse the HTML formatted reaction information stored in the BiGG notes field.
This function is being deprecated and replaced by \sphinxtitleref{CBTools.processSBMLAnnotationNotes()}
\begin{itemize}
\item {} 
\sphinxAtStartPar
requires an \sphinxstyleemphasis{annotation\_key} which contains a BiGG HTML fragment

\end{itemize}

\end{fulllineitems}

\index{processBiGGchemFormula() (in module cbmpy.CBTools)@\spxentry{processBiGGchemFormula()}\spxextra{in module cbmpy.CBTools}}

\begin{fulllineitems}
\phantomsection\label{\detokenize{modules_doc:cbmpy.CBTools.processBiGGchemFormula}}
\pysigstartsignatures
\pysiglinewithargsret{\sphinxbfcode{\sphinxupquote{processBiGGchemFormula}}}{\sphinxparam{\DUrole{n,n}{fba}}}{}
\pysigstopsignatures
\sphinxAtStartPar
Disambiguates the overloaded BiGG name NAME\_CHEMFORMULA into
\begin{itemize}
\item {} 
\sphinxAtStartPar
\sphinxstyleemphasis{species.name} NAME

\item {} 
\sphinxAtStartPar
\sphinxstyleemphasis{species.chemFormula} CHEMFORMULA

\end{itemize}

\end{fulllineitems}

\index{processExchangeReactions() (in module cbmpy.CBTools)@\spxentry{processExchangeReactions()}\spxextra{in module cbmpy.CBTools}}

\begin{fulllineitems}
\phantomsection\label{\detokenize{modules_doc:cbmpy.CBTools.processExchangeReactions}}
\pysigstartsignatures
\pysiglinewithargsret{\sphinxbfcode{\sphinxupquote{processExchangeReactions}}}{\sphinxparam{\DUrole{n,n}{fba}}\sphinxparamcomma \sphinxparam{\DUrole{n,n}{key}}}{}
\pysigstopsignatures
\sphinxAtStartPar
Extract exchange reactions from model using \sphinxstyleemphasis{key} and return:
\begin{itemize}
\item {} 
\sphinxAtStartPar
a dictionary of all exchange reactions without \sphinxstyleemphasis{medium} reactions

\item {} 
\sphinxAtStartPar
a dictionary of \sphinxstyleemphasis{medium} exchange reactions (negative lower bound)

\end{itemize}

\end{fulllineitems}

\index{processSBMLAnnotationNotes() (in module cbmpy.CBTools)@\spxentry{processSBMLAnnotationNotes()}\spxextra{in module cbmpy.CBTools}}

\begin{fulllineitems}
\phantomsection\label{\detokenize{modules_doc:cbmpy.CBTools.processSBMLAnnotationNotes}}
\pysigstartsignatures
\pysiglinewithargsret{\sphinxbfcode{\sphinxupquote{processSBMLAnnotationNotes}}}{\sphinxparam{\DUrole{n,n}{fba}}\sphinxparamcomma \sphinxparam{\DUrole{n,n}{annotation\_key}\DUrole{o,o}{=}\DUrole{default_value}{\textquotesingle{}note\textquotesingle{}}}\sphinxparamcomma \sphinxparam{\DUrole{n,n}{level}\DUrole{o,o}{=}\DUrole{default_value}{3}}}{}
\pysigstopsignatures
\sphinxAtStartPar
Parse the HTML formatted reaction information stored in the SBML notes field currently
processes BiGG and PySCeSCBM style annotations it looks for the the annotation indexed
with the \sphinxstyleemphasis{annotation\_key}
\begin{itemize}
\item {} 
\sphinxAtStartPar
\sphinxstyleemphasis{annotation\_key} {[}default=’note’{]} which contains a HTML/XHTML fragment in BiGG/PySCeSCBM format (ignored in L3)

\end{itemize}

\end{fulllineitems}

\index{removeFixedSpeciesReactions() (in module cbmpy.CBTools)@\spxentry{removeFixedSpeciesReactions()}\spxextra{in module cbmpy.CBTools}}

\begin{fulllineitems}
\phantomsection\label{\detokenize{modules_doc:cbmpy.CBTools.removeFixedSpeciesReactions}}
\pysigstartsignatures
\pysiglinewithargsret{\sphinxbfcode{\sphinxupquote{removeFixedSpeciesReactions}}}{\sphinxparam{\DUrole{n,n}{f}}}{}
\pysigstopsignatures
\sphinxAtStartPar
This function is a hack that removes reactions which only have boundary species as reactants
and products. These are typically gene associations encoded in the Manchester style and there
is probably a better way of working around this problem …
\begin{itemize}
\item {} 
\sphinxAtStartPar
\sphinxstyleemphasis{f} an instantiated fba model object

\end{itemize}

\end{fulllineitems}

\index{roundOffWithSense() (in module cbmpy.CBTools)@\spxentry{roundOffWithSense()}\spxextra{in module cbmpy.CBTools}}

\begin{fulllineitems}
\phantomsection\label{\detokenize{modules_doc:cbmpy.CBTools.roundOffWithSense}}
\pysigstartsignatures
\pysiglinewithargsret{\sphinxbfcode{\sphinxupquote{roundOffWithSense}}}{\sphinxparam{\DUrole{n,n}{val}}\sphinxparamcomma \sphinxparam{\DUrole{n,n}{osense}\DUrole{o,o}{=}\DUrole{default_value}{\textquotesingle{}max\textquotesingle{}}}\sphinxparamcomma \sphinxparam{\DUrole{n,n}{tol}\DUrole{o,o}{=}\DUrole{default_value}{1e\sphinxhyphen{}08}}}{}
\pysigstopsignatures
\sphinxAtStartPar
Round of a value in a way that takes into consideration the sense of the operation that generated it
\begin{itemize}
\item {} 
\sphinxAtStartPar
\sphinxstyleemphasis{val} the value

\item {} 
\sphinxAtStartPar
\sphinxstyleemphasis{osense} {[}default=’max’{]} the sense

\item {} 
\sphinxAtStartPar
\sphinxstyleemphasis{tol} {[}default=1e\sphinxhyphen{}8{]} the tolerance of the roundoff factor

\end{itemize}

\end{fulllineitems}

\index{scanForReactionDuplicates() (in module cbmpy.CBTools)@\spxentry{scanForReactionDuplicates()}\spxextra{in module cbmpy.CBTools}}

\begin{fulllineitems}
\phantomsection\label{\detokenize{modules_doc:cbmpy.CBTools.scanForReactionDuplicates}}
\pysigstartsignatures
\pysiglinewithargsret{\sphinxbfcode{\sphinxupquote{scanForReactionDuplicates}}}{\sphinxparam{\DUrole{n,n}{f}}\sphinxparamcomma \sphinxparam{\DUrole{n,n}{ignore\_coefficients}\DUrole{o,o}{=}\DUrole{default_value}{False}}}{}
\pysigstopsignatures
\sphinxAtStartPar
This method uses uses a brute force apprach to finding reactions with matching
stoichiometry

\end{fulllineitems}

\index{scanForUnbalancedReactions() (in module cbmpy.CBTools)@\spxentry{scanForUnbalancedReactions()}\spxextra{in module cbmpy.CBTools}}

\begin{fulllineitems}
\phantomsection\label{\detokenize{modules_doc:cbmpy.CBTools.scanForUnbalancedReactions}}
\pysigstartsignatures
\pysiglinewithargsret{\sphinxbfcode{\sphinxupquote{scanForUnbalancedReactions}}}{\sphinxparam{\DUrole{n,n}{f}}\sphinxparamcomma \sphinxparam{\DUrole{n,n}{output}\DUrole{o,o}{=}\DUrole{default_value}{\textquotesingle{}all\textquotesingle{}}}}{}
\pysigstopsignatures
\sphinxAtStartPar
Scan a model for unbalanced reactions, returns a tuple of dictionaries balanced and unbalanced:
\begin{itemize}
\item {} 
\sphinxAtStartPar
\sphinxstyleemphasis{f} an FBA model instance

\item {} 
\sphinxAtStartPar
\sphinxstyleemphasis{output} {[}default=’all’{]} can be one of {[}‘all’,’charge’,’element’{]}

\item {} 
\sphinxAtStartPar
\sphinxstyleemphasis{charge} return all charge \sphinxstylestrong{un} balanced reactions

\item {} 
\sphinxAtStartPar
\sphinxstyleemphasis{element} return all element \sphinxstylestrong{un} balanced reactions

\end{itemize}

\end{fulllineitems}

\index{setSpeciesPropertiesFromAnnotations() (in module cbmpy.CBTools)@\spxentry{setSpeciesPropertiesFromAnnotations()}\spxextra{in module cbmpy.CBTools}}

\begin{fulllineitems}
\phantomsection\label{\detokenize{modules_doc:cbmpy.CBTools.setSpeciesPropertiesFromAnnotations}}
\pysigstartsignatures
\pysiglinewithargsret{\sphinxbfcode{\sphinxupquote{setSpeciesPropertiesFromAnnotations}}}{\sphinxparam{\DUrole{n,n}{fbam}}\sphinxparamcomma \sphinxparam{\DUrole{n,n}{overwriteCharge}\DUrole{o,o}{=}\DUrole{default_value}{False}}\sphinxparamcomma \sphinxparam{\DUrole{n,n}{overwriteChemFormula}\DUrole{o,o}{=}\DUrole{default_value}{False}}}{}
\pysigstopsignatures
\sphinxAtStartPar
This will attempt to set the model Species properties from the annotation. With the default options
it will only replace missing data. With ChemicalFormula this is easy to detect however charge may
have an “unknown value” of 0. Setting the optional values to true will replace any existing value
with any valid annotation.
\begin{itemize}
\item {} 
\sphinxAtStartPar
\sphinxstyleemphasis{overwriteChemFormula} {[}default=False{]}

\item {} 
\sphinxAtStartPar
\sphinxstyleemphasis{overwriteCharge} {[}default=False{]}

\end{itemize}

\end{fulllineitems}

\index{splitReversibleReactions() (in module cbmpy.CBTools)@\spxentry{splitReversibleReactions()}\spxextra{in module cbmpy.CBTools}}

\begin{fulllineitems}
\phantomsection\label{\detokenize{modules_doc:cbmpy.CBTools.splitReversibleReactions}}
\pysigstartsignatures
\pysiglinewithargsret{\sphinxbfcode{\sphinxupquote{splitReversibleReactions}}}{\sphinxparam{\DUrole{n,n}{fba}}\sphinxparamcomma \sphinxparam{\DUrole{n,n}{selected\_reactions}\DUrole{o,o}{=}\DUrole{default_value}{None}}}{}
\pysigstopsignatures
\sphinxAtStartPar
Split a (set of) reactions into reversible reactions returns a copy of the original model

\sphinxAtStartPar
R1: A = B
R1f: A \sphinxhyphen{}\textgreater{} B
R1r: B \sphinxhyphen{}\textgreater{} A
\begin{itemize}
\item {} 
\sphinxAtStartPar
\sphinxstyleemphasis{fba} an instantiated CBMPy model object

\item {} 
\sphinxAtStartPar
\sphinxstyleemphasis{selected\_reactions} if a reversible reaction id is in here split it

\end{itemize}

\end{fulllineitems}

\index{splitSingleReversibleReaction() (in module cbmpy.CBTools)@\spxentry{splitSingleReversibleReaction()}\spxextra{in module cbmpy.CBTools}}

\begin{fulllineitems}
\phantomsection\label{\detokenize{modules_doc:cbmpy.CBTools.splitSingleReversibleReaction}}
\pysigstartsignatures
\pysiglinewithargsret{\sphinxbfcode{\sphinxupquote{splitSingleReversibleReaction}}}{\sphinxparam{\DUrole{n,n}{fba}}\sphinxparamcomma \sphinxparam{\DUrole{n,n}{rid}}\sphinxparamcomma \sphinxparam{\DUrole{n,n}{fwd\_id}\DUrole{o,o}{=}\DUrole{default_value}{None}}\sphinxparamcomma \sphinxparam{\DUrole{n,n}{rev\_id}\DUrole{o,o}{=}\DUrole{default_value}{None}}}{}
\pysigstopsignatures
\sphinxAtStartPar
Split a single reversible reaction into two irreversible reactions, returns the original reversible reaction and bounds
while deleting them from model.

\sphinxAtStartPar
R1: A = B
R1\_fwd: A \sphinxhyphen{}\textgreater{} B
R1\_rev: B \sphinxhyphen{}\textgreater{} A
\begin{itemize}
\item {} 
\sphinxAtStartPar
\sphinxstyleemphasis{fba} an instantiated CBMPy model object

\item {} 
\sphinxAtStartPar
\sphinxstyleemphasis{rid} a valid reaction id

\item {} 
\sphinxAtStartPar
\sphinxstyleemphasis{fwd\_id} {[}default=None{]} the new forward reaction id, defaults to rid\_fwd

\item {} 
\sphinxAtStartPar
\sphinxstyleemphasis{rev\_id} {[}default=None{]} the new forward reaction id, defaults to rid\_rev

\end{itemize}

\end{fulllineitems}

\index{storeObj() (in module cbmpy.CBTools)@\spxentry{storeObj()}\spxextra{in module cbmpy.CBTools}}

\begin{fulllineitems}
\phantomsection\label{\detokenize{modules_doc:cbmpy.CBTools.storeObj}}
\pysigstartsignatures
\pysiglinewithargsret{\sphinxbfcode{\sphinxupquote{storeObj}}}{\sphinxparam{\DUrole{n,n}{obj}}\sphinxparamcomma \sphinxparam{\DUrole{n,n}{filename}}\sphinxparamcomma \sphinxparam{\DUrole{n,n}{compress}\DUrole{o,o}{=}\DUrole{default_value}{False}}}{}
\pysigstopsignatures
\sphinxAtStartPar
Stores a Python \sphinxstyleemphasis{obj} as a serialised binary object in \sphinxstyleemphasis{filename}.dat
\begin{itemize}
\item {} 
\sphinxAtStartPar
\sphinxstyleemphasis{obj} a python object

\item {} 
\sphinxAtStartPar
\sphinxstyleemphasis{filename} the base filename

\item {} 
\sphinxAtStartPar
\sphinxstyleemphasis{compress} {[}False{]} use gzip compression not \sphinxstyleemphasis{implemented}

\end{itemize}

\end{fulllineitems}

\index{stringReplace() (in module cbmpy.CBTools)@\spxentry{stringReplace()}\spxextra{in module cbmpy.CBTools}}

\begin{fulllineitems}
\phantomsection\label{\detokenize{modules_doc:cbmpy.CBTools.stringReplace}}
\pysigstartsignatures
\pysiglinewithargsret{\sphinxbfcode{\sphinxupquote{stringReplace}}}{\sphinxparam{\DUrole{n,n}{fbamod}}\sphinxparamcomma \sphinxparam{\DUrole{n,n}{old}}\sphinxparamcomma \sphinxparam{\DUrole{n,n}{new}}\sphinxparamcomma \sphinxparam{\DUrole{n,n}{target}}}{}
\pysigstopsignatures
\sphinxAtStartPar
This is alpha stuff, target can be:
\begin{itemize}
\item {} 
\sphinxAtStartPar
‘species’

\item {} 
\sphinxAtStartPar
‘reactions’

\item {} 
\sphinxAtStartPar
‘constraints’

\item {} 
\sphinxAtStartPar
‘objectives’

\item {} 
\sphinxAtStartPar
‘all’

\end{itemize}

\end{fulllineitems}

\phantomsection\label{\detokenize{modules_doc:module-cbmpy.CBWrite}}\index{module@\spxentry{module}!cbmpy.CBWrite@\spxentry{cbmpy.CBWrite}}\index{cbmpy.CBWrite@\spxentry{cbmpy.CBWrite}!module@\spxentry{module}}

\section{CBMPy: CBWrite module}
\label{\detokenize{modules_doc:cbmpy-cbwrite-module}}
\sphinxAtStartPar
PySCeS Constraint Based Modelling (\sphinxurl{http://cbmpy.sourceforge.net})
Copyright (C) 2009\sphinxhyphen{}2024 Brett G. Olivier, VU University Amsterdam, Amsterdam, The Netherlands

\sphinxAtStartPar
This program is free software: you can redistribute it and/or modify
it under the terms of the GNU General Public License as published by
the Free Software Foundation, either version 3 of the License, or
(at your option) any later version.

\sphinxAtStartPar
This program is distributed in the hope that it will be useful,
but WITHOUT ANY WARRANTY; without even the implied warranty of
MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
GNU General Public License for more details.

\sphinxAtStartPar
You should have received a copy of the GNU General Public License
along with this program.  If not, see \textless{}\sphinxurl{http://www.gnu.org/licenses/}\textgreater{}

\sphinxAtStartPar
Author: Brett G. Olivier PhD
Contact developers: \sphinxurl{https://github.com/SystemsBioinformatics/cbmpy/issues}
Last edit: \$Author: bgoli \$ (\$Id: CBWrite.py 710 2020\sphinxhyphen{}04\sphinxhyphen{}27 14:22:34Z bgoli \$)
\index{BuildHformatFluxBounds() (in module cbmpy.CBWrite)@\spxentry{BuildHformatFluxBounds()}\spxextra{in module cbmpy.CBWrite}}

\begin{fulllineitems}
\phantomsection\label{\detokenize{modules_doc:cbmpy.CBWrite.BuildHformatFluxBounds}}
\pysigstartsignatures
\pysiglinewithargsret{\sphinxbfcode{\sphinxupquote{BuildHformatFluxBounds}}}{\sphinxparam{\DUrole{n,n}{fba}}\sphinxparamcomma \sphinxparam{\DUrole{n,n}{infinity\_replace}\DUrole{o,o}{=}\DUrole{default_value}{None}}\sphinxparamcomma \sphinxparam{\DUrole{n,n}{use\_rational}\DUrole{o,o}{=}\DUrole{default_value}{False}}}{}
\pysigstopsignatures
\sphinxAtStartPar
Build and return a csio that contains the flux bounds in H format
\begin{itemize}
\item {} 
\sphinxAtStartPar
\sphinxstyleemphasis{fba} a PySCeS\sphinxhyphen{}CBM FBA object

\item {} 
\sphinxAtStartPar
\sphinxstyleemphasis{infinity\_replace} {[}default=None{]} if defined this is the abs(value) of +\sphinxhyphen{}\textless{}infinity\textgreater{}

\end{itemize}

\end{fulllineitems}

\index{BuildLPConstraints() (in module cbmpy.CBWrite)@\spxentry{BuildLPConstraints()}\spxextra{in module cbmpy.CBWrite}}

\begin{fulllineitems}
\phantomsection\label{\detokenize{modules_doc:cbmpy.CBWrite.BuildLPConstraints}}
\pysigstartsignatures
\pysiglinewithargsret{\sphinxbfcode{\sphinxupquote{BuildLPConstraints}}}{\sphinxparam{\DUrole{n,n}{fba}}\sphinxparamcomma \sphinxparam{\DUrole{n,n}{use\_rational}\DUrole{o,o}{=}\DUrole{default_value}{False}}}{}
\pysigstopsignatures
\sphinxAtStartPar
Build and return a csio that contains constraint constructed from
the StoichiometeryLP object
\begin{itemize}
\item {} 
\sphinxAtStartPar
\sphinxstyleemphasis{fba} an fba model object which has a stoichiometry

\item {} 
\sphinxAtStartPar
\sphinxstyleemphasis{use\_rational} write rational number output {[}default=False{]}

\end{itemize}

\end{fulllineitems}

\index{BuildLPConstraintsMath() (in module cbmpy.CBWrite)@\spxentry{BuildLPConstraintsMath()}\spxextra{in module cbmpy.CBWrite}}

\begin{fulllineitems}
\phantomsection\label{\detokenize{modules_doc:cbmpy.CBWrite.BuildLPConstraintsMath}}
\pysigstartsignatures
\pysiglinewithargsret{\sphinxbfcode{\sphinxupquote{BuildLPConstraintsMath}}}{\sphinxparam{\DUrole{n,n}{fba}}\sphinxparamcomma \sphinxparam{\DUrole{n,n}{use\_rational}\DUrole{o,o}{=}\DUrole{default_value}{False}}}{}
\pysigstopsignatures
\sphinxAtStartPar
Build and return a csio that contains the constaints in LP format
Strict refers to dS/dt =\textgreater{} 0 and dS/dt \textless{}= 0

\end{fulllineitems}

\index{BuildLPConstraintsRelaxed() (in module cbmpy.CBWrite)@\spxentry{BuildLPConstraintsRelaxed()}\spxextra{in module cbmpy.CBWrite}}

\begin{fulllineitems}
\phantomsection\label{\detokenize{modules_doc:cbmpy.CBWrite.BuildLPConstraintsRelaxed}}
\pysigstartsignatures
\pysiglinewithargsret{\sphinxbfcode{\sphinxupquote{BuildLPConstraintsRelaxed}}}{\sphinxparam{\DUrole{n,n}{fba}}}{}
\pysigstopsignatures
\sphinxAtStartPar
Build and return a csio that contains the constaints in LP format
Relaxed refers to dS/dt \textgreater{}= 0

\end{fulllineitems}

\index{BuildLPConstraintsStrict() (in module cbmpy.CBWrite)@\spxentry{BuildLPConstraintsStrict()}\spxextra{in module cbmpy.CBWrite}}

\begin{fulllineitems}
\phantomsection\label{\detokenize{modules_doc:cbmpy.CBWrite.BuildLPConstraintsStrict}}
\pysigstartsignatures
\pysiglinewithargsret{\sphinxbfcode{\sphinxupquote{BuildLPConstraintsStrict}}}{\sphinxparam{\DUrole{n,n}{fba}}\sphinxparamcomma \sphinxparam{\DUrole{n,n}{use\_rational}\DUrole{o,o}{=}\DUrole{default_value}{False}}}{}
\pysigstopsignatures
\sphinxAtStartPar
Build and return a csio that contains the constaints in LP format
Strict refers to dS/dt = 0

\end{fulllineitems}

\index{BuildLPFluxBounds() (in module cbmpy.CBWrite)@\spxentry{BuildLPFluxBounds()}\spxextra{in module cbmpy.CBWrite}}

\begin{fulllineitems}
\phantomsection\label{\detokenize{modules_doc:cbmpy.CBWrite.BuildLPFluxBounds}}
\pysigstartsignatures
\pysiglinewithargsret{\sphinxbfcode{\sphinxupquote{BuildLPFluxBounds}}}{\sphinxparam{\DUrole{n,n}{fba}}\sphinxparamcomma \sphinxparam{\DUrole{n,n}{use\_rational}\DUrole{o,o}{=}\DUrole{default_value}{False}}}{}
\pysigstopsignatures
\sphinxAtStartPar
Build and return a csio that contains the flux bounds in LP format

\end{fulllineitems}

\index{BuildLPUserConstraints() (in module cbmpy.CBWrite)@\spxentry{BuildLPUserConstraints()}\spxextra{in module cbmpy.CBWrite}}

\begin{fulllineitems}
\phantomsection\label{\detokenize{modules_doc:cbmpy.CBWrite.BuildLPUserConstraints}}
\pysigstartsignatures
\pysiglinewithargsret{\sphinxbfcode{\sphinxupquote{BuildLPUserConstraints}}}{\sphinxparam{\DUrole{n,n}{fba}}\sphinxparamcomma \sphinxparam{\DUrole{n,n}{use\_rational}\DUrole{o,o}{=}\DUrole{default_value}{False}}}{}
\pysigstopsignatures
\sphinxAtStartPar
Build and return a csio that contains constraint constructed from
the StoichiometeryLP object
\begin{itemize}
\item {} 
\sphinxAtStartPar
\sphinxstyleemphasis{fba} an fba model object which has a stoichiometry

\item {} 
\sphinxAtStartPar
\sphinxstyleemphasis{use\_rational} write rational number output {[}default=False{]}

\end{itemize}

\end{fulllineitems}

\index{WriteFVAdata() (in module cbmpy.CBWrite)@\spxentry{WriteFVAdata()}\spxextra{in module cbmpy.CBWrite}}

\begin{fulllineitems}
\phantomsection\label{\detokenize{modules_doc:cbmpy.CBWrite.WriteFVAdata}}
\pysigstartsignatures
\pysiglinewithargsret{\sphinxbfcode{\sphinxupquote{WriteFVAdata}}}{\sphinxparam{\DUrole{n,n}{fva}}\sphinxparamcomma \sphinxparam{\DUrole{n,n}{names}}\sphinxparamcomma \sphinxparam{\DUrole{n,n}{fname}}\sphinxparamcomma \sphinxparam{\DUrole{n,n}{work\_dir}\DUrole{o,o}{=}\DUrole{default_value}{None}}\sphinxparamcomma \sphinxparam{\DUrole{n,n}{roundec}\DUrole{o,o}{=}\DUrole{default_value}{None}}\sphinxparamcomma \sphinxparam{\DUrole{n,n}{scale\_min}\DUrole{o,o}{=}\DUrole{default_value}{False}}\sphinxparamcomma \sphinxparam{\DUrole{n,n}{appendfile}\DUrole{o,o}{=}\DUrole{default_value}{False}}\sphinxparamcomma \sphinxparam{\DUrole{n,n}{info}\DUrole{o,o}{=}\DUrole{default_value}{None}}}{}
\pysigstopsignatures
\sphinxAtStartPar
INFO: this method will be deprecated please update your scripts to use “writeFVAdata()”

\end{fulllineitems}

\index{WriteFVAtoCSV() (in module cbmpy.CBWrite)@\spxentry{WriteFVAtoCSV()}\spxextra{in module cbmpy.CBWrite}}

\begin{fulllineitems}
\phantomsection\label{\detokenize{modules_doc:cbmpy.CBWrite.WriteFVAtoCSV}}
\pysigstartsignatures
\pysiglinewithargsret{\sphinxbfcode{\sphinxupquote{WriteFVAtoCSV}}}{\sphinxparam{\DUrole{n,n}{id}}\sphinxparamcomma \sphinxparam{\DUrole{n,n}{fva}}\sphinxparamcomma \sphinxparam{\DUrole{n,n}{names}}\sphinxparamcomma \sphinxparam{\DUrole{n,n}{Dir}\DUrole{o,o}{=}\DUrole{default_value}{None}}\sphinxparamcomma \sphinxparam{\DUrole{n,n}{fbaObj}\DUrole{o,o}{=}\DUrole{default_value}{None}}}{}
\pysigstopsignatures
\sphinxAtStartPar
INFO: this method will be deprecated please update your scripts to use “writeFVAtoCSV()”

\end{fulllineitems}

\index{WriteModelHFormatFBA() (in module cbmpy.CBWrite)@\spxentry{WriteModelHFormatFBA()}\spxextra{in module cbmpy.CBWrite}}

\begin{fulllineitems}
\phantomsection\label{\detokenize{modules_doc:cbmpy.CBWrite.WriteModelHFormatFBA}}
\pysigstartsignatures
\pysiglinewithargsret{\sphinxbfcode{\sphinxupquote{WriteModelHFormatFBA}}}{\sphinxparam{\DUrole{n,n}{fba}}\sphinxparamcomma \sphinxparam{\DUrole{n,n}{work\_dir}\DUrole{o,o}{=}\DUrole{default_value}{None}}\sphinxparamcomma \sphinxparam{\DUrole{n,n}{use\_rational}\DUrole{o,o}{=}\DUrole{default_value}{False}}\sphinxparamcomma \sphinxparam{\DUrole{n,n}{fullLP}\DUrole{o,o}{=}\DUrole{default_value}{True}}\sphinxparamcomma \sphinxparam{\DUrole{n,n}{format}\DUrole{o,o}{=}\DUrole{default_value}{\textquotesingle{}\%s\textquotesingle{}}}\sphinxparamcomma \sphinxparam{\DUrole{n,n}{infinity\_replace}\DUrole{o,o}{=}\DUrole{default_value}{None}}}{}
\pysigstopsignatures
\sphinxAtStartPar
INFO: this method will be deprecated please update your scripts to use “writeModelHFormatFBA2()”

\end{fulllineitems}

\index{WriteModelHFormatFBA2() (in module cbmpy.CBWrite)@\spxentry{WriteModelHFormatFBA2()}\spxextra{in module cbmpy.CBWrite}}

\begin{fulllineitems}
\phantomsection\label{\detokenize{modules_doc:cbmpy.CBWrite.WriteModelHFormatFBA2}}
\pysigstartsignatures
\pysiglinewithargsret{\sphinxbfcode{\sphinxupquote{WriteModelHFormatFBA2}}}{\sphinxparam{\DUrole{n,n}{fba}}\sphinxparamcomma \sphinxparam{\DUrole{n,n}{fname}\DUrole{o,o}{=}\DUrole{default_value}{None}}\sphinxparamcomma \sphinxparam{\DUrole{n,n}{work\_dir}\DUrole{o,o}{=}\DUrole{default_value}{None}}\sphinxparamcomma \sphinxparam{\DUrole{n,n}{use\_rational}\DUrole{o,o}{=}\DUrole{default_value}{False}}\sphinxparamcomma \sphinxparam{\DUrole{n,n}{fullLP}\DUrole{o,o}{=}\DUrole{default_value}{True}}\sphinxparamcomma \sphinxparam{\DUrole{n,n}{format}\DUrole{o,o}{=}\DUrole{default_value}{\textquotesingle{}\%s\textquotesingle{}}}\sphinxparamcomma \sphinxparam{\DUrole{n,n}{infinity\_replace}\DUrole{o,o}{=}\DUrole{default_value}{None}}}{}
\pysigstopsignatures
\sphinxAtStartPar
INFO: this method will be deprecated please update your scripts to use “writeModelHFormatFBA2()”

\end{fulllineitems}

\index{WriteModelLP() (in module cbmpy.CBWrite)@\spxentry{WriteModelLP()}\spxextra{in module cbmpy.CBWrite}}

\begin{fulllineitems}
\phantomsection\label{\detokenize{modules_doc:cbmpy.CBWrite.WriteModelLP}}
\pysigstartsignatures
\pysiglinewithargsret{\sphinxbfcode{\sphinxupquote{WriteModelLP}}}{\sphinxparam{\DUrole{n,n}{fba}}\sphinxparamcomma \sphinxparam{\DUrole{n,n}{work\_dir}\DUrole{o,o}{=}\DUrole{default_value}{None}}\sphinxparamcomma \sphinxparam{\DUrole{n,n}{fname}\DUrole{o,o}{=}\DUrole{default_value}{None}}\sphinxparamcomma \sphinxparam{\DUrole{n,n}{multisymb}\DUrole{o,o}{=}\DUrole{default_value}{\textquotesingle{} \textquotesingle{}}}\sphinxparamcomma \sphinxparam{\DUrole{n,n}{format}\DUrole{o,o}{=}\DUrole{default_value}{\textquotesingle{}\%s\textquotesingle{}}}\sphinxparamcomma \sphinxparam{\DUrole{n,n}{use\_rational}\DUrole{o,o}{=}\DUrole{default_value}{False}}\sphinxparamcomma \sphinxparam{\DUrole{n,n}{constraint\_mode}\DUrole{o,o}{=}\DUrole{default_value}{None}}\sphinxparamcomma \sphinxparam{\DUrole{n,n}{quiet}\DUrole{o,o}{=}\DUrole{default_value}{False}}}{}
\pysigstopsignatures
\sphinxAtStartPar
INFO: this method will be deprecated please update your scripts to use “writeModelLP()”

\end{fulllineitems}

\index{WriteModelLPOld() (in module cbmpy.CBWrite)@\spxentry{WriteModelLPOld()}\spxextra{in module cbmpy.CBWrite}}

\begin{fulllineitems}
\phantomsection\label{\detokenize{modules_doc:cbmpy.CBWrite.WriteModelLPOld}}
\pysigstartsignatures
\pysiglinewithargsret{\sphinxbfcode{\sphinxupquote{WriteModelLPOld}}}{\sphinxparam{\DUrole{n,n}{fba}}\sphinxparamcomma \sphinxparam{\DUrole{n,n}{work\_dir}\DUrole{o,o}{=}\DUrole{default_value}{None}}\sphinxparamcomma \sphinxparam{\DUrole{n,n}{multisymb}\DUrole{o,o}{=}\DUrole{default_value}{\textquotesingle{} \textquotesingle{}}}\sphinxparamcomma \sphinxparam{\DUrole{n,n}{lpt}\DUrole{o,o}{=}\DUrole{default_value}{True}}\sphinxparamcomma \sphinxparam{\DUrole{n,n}{constraint\_mode}\DUrole{o,o}{=}\DUrole{default_value}{\textquotesingle{}strict\textquotesingle{}}}\sphinxparamcomma \sphinxparam{\DUrole{n,n}{use\_rational}\DUrole{o,o}{=}\DUrole{default_value}{False}}\sphinxparamcomma \sphinxparam{\DUrole{n,n}{format}\DUrole{o,o}{=}\DUrole{default_value}{\textquotesingle{}\%s\textquotesingle{}}}}{}
\pysigstopsignatures
\sphinxAtStartPar
INFO: this method will be deprecated please update your scripts to use “writeModelLPOld()”

\end{fulllineitems}

\index{WriteModelRaw() (in module cbmpy.CBWrite)@\spxentry{WriteModelRaw()}\spxextra{in module cbmpy.CBWrite}}

\begin{fulllineitems}
\phantomsection\label{\detokenize{modules_doc:cbmpy.CBWrite.WriteModelRaw}}
\pysigstartsignatures
\pysiglinewithargsret{\sphinxbfcode{\sphinxupquote{WriteModelRaw}}}{\sphinxparam{\DUrole{n,n}{fba}}\sphinxparamcomma \sphinxparam{\DUrole{n,n}{work\_dir}\DUrole{o,o}{=}\DUrole{default_value}{None}}}{}
\pysigstopsignatures
\sphinxAtStartPar
INFO: this method will be deprecated please update your scripts to use “writeModelRaw()”

\end{fulllineitems}

\index{convertExcelToFloat() (in module cbmpy.CBWrite)@\spxentry{convertExcelToFloat()}\spxextra{in module cbmpy.CBWrite}}

\begin{fulllineitems}
\phantomsection\label{\detokenize{modules_doc:cbmpy.CBWrite.convertExcelToFloat}}
\pysigstartsignatures
\pysiglinewithargsret{\sphinxbfcode{\sphinxupquote{convertExcelToFloat}}}{\sphinxparam{\DUrole{n,n}{num}}}{}
\pysigstopsignatures
\sphinxAtStartPar
Converts an Excel “number” to a float
\begin{itemize}
\item {} 
\sphinxAtStartPar
\sphinxstyleemphasis{num} a number

\end{itemize}

\end{fulllineitems}

\index{convertFloatToExcel() (in module cbmpy.CBWrite)@\spxentry{convertFloatToExcel()}\spxextra{in module cbmpy.CBWrite}}

\begin{fulllineitems}
\phantomsection\label{\detokenize{modules_doc:cbmpy.CBWrite.convertFloatToExcel}}
\pysigstartsignatures
\pysiglinewithargsret{\sphinxbfcode{\sphinxupquote{convertFloatToExcel}}}{\sphinxparam{\DUrole{n,n}{num}}\sphinxparamcomma \sphinxparam{\DUrole{n,n}{roundoff}}}{}
\pysigstopsignatures
\sphinxAtStartPar
Converts a float to Excel compatible “number”
\begin{itemize}
\item {} 
\sphinxAtStartPar
\sphinxstyleemphasis{num} a number

\item {} 
\sphinxAtStartPar
\sphinxstyleemphasis{roundoff} the number of roundoff digits for round()

\end{itemize}

\end{fulllineitems}

\index{exportModel() (in module cbmpy.CBWrite)@\spxentry{exportModel()}\spxextra{in module cbmpy.CBWrite}}

\begin{fulllineitems}
\phantomsection\label{\detokenize{modules_doc:cbmpy.CBWrite.exportModel}}
\pysigstartsignatures
\pysiglinewithargsret{\sphinxbfcode{\sphinxupquote{exportModel}}}{\sphinxparam{\DUrole{n,n}{fba}}\sphinxparamcomma \sphinxparam{\DUrole{n,n}{fname}\DUrole{o,o}{=}\DUrole{default_value}{None}}\sphinxparamcomma \sphinxparam{\DUrole{n,n}{fmt}\DUrole{o,o}{=}\DUrole{default_value}{\textquotesingle{}lp\textquotesingle{}}}\sphinxparamcomma \sphinxparam{\DUrole{n,n}{work\_dir}\DUrole{o,o}{=}\DUrole{default_value}{None}}\sphinxparamcomma \sphinxparam{\DUrole{n,n}{use\_rational}\DUrole{o,o}{=}\DUrole{default_value}{\textquotesingle{}both\textquotesingle{}}}}{}
\pysigstopsignatures
\sphinxAtStartPar
Export the FBA model in different formats:
\begin{itemize}
\item {} 
\sphinxAtStartPar
\sphinxstyleemphasis{fba} the FBA model

\item {} 
\sphinxAtStartPar
\sphinxstyleemphasis{fname} {[}default=None{]} the exported filename if None then \sphinxtitleref{fba.getId()} is used

\item {} 
\sphinxAtStartPar
\sphinxstyleemphasis{fmt} {[}default=’lp’{]} the export format can be one of: ‘lp’ (CPLEX), ‘hformat’ (Polyhedra), ‘all’ (both)

\item {} 
\sphinxAtStartPar
\sphinxstyleemphasis{use\_rational} {[}default=’both’{]} if \sphinxstyleemphasis{all} or \sphinxstyleemphasis{hformat} is specified should hformat files be written using rational math or not. The default \sphinxstyleemphasis{both} is the legacy behaviour and writes both.

\end{itemize}

\sphinxAtStartPar
Note that ‘hformat’ ignores ‘fname’ and only uses fba.getId() this is a legacy behaviour

\end{fulllineitems}

\index{generateBGID() (in module cbmpy.CBWrite)@\spxentry{generateBGID()}\spxextra{in module cbmpy.CBWrite}}

\begin{fulllineitems}
\phantomsection\label{\detokenize{modules_doc:cbmpy.CBWrite.generateBGID}}
\pysigstartsignatures
\pysiglinewithargsret{\sphinxbfcode{\sphinxupquote{generateBGID}}}{\sphinxparam{\DUrole{n,n}{num}}\sphinxparamcomma \sphinxparam{\DUrole{n,n}{prefix}}}{}
\pysigstopsignatures\begin{description}
\sphinxlineitem{Create a BGID generator, which is \textless{}prefix\textgreater{}\textless{}num\textgreater{} where perfix is two letters num is padded to 6 figures}\begin{itemize}
\item {} 
\sphinxAtStartPar
\sphinxstyleemphasis{num} the starting number

\item {} 
\sphinxAtStartPar
\sphinxstyleemphasis{prefix} the two letter prefix

\end{itemize}

\end{description}

\end{fulllineitems}

\index{printFBASolution() (in module cbmpy.CBWrite)@\spxentry{printFBASolution()}\spxextra{in module cbmpy.CBWrite}}

\begin{fulllineitems}
\phantomsection\label{\detokenize{modules_doc:cbmpy.CBWrite.printFBASolution}}
\pysigstartsignatures
\pysiglinewithargsret{\sphinxbfcode{\sphinxupquote{printFBASolution}}}{\sphinxparam{\DUrole{n,n}{fba}}\sphinxparamcomma \sphinxparam{\DUrole{n,n}{include\_all}\DUrole{o,o}{=}\DUrole{default_value}{False}}}{}
\pysigstopsignatures
\sphinxAtStartPar
Prints the FBA optimal solution to the screen.
\begin{itemize}
\item {} 
\sphinxAtStartPar
\sphinxstyleemphasis{fba} an FBA model object

\item {} 
\sphinxAtStartPar
\sphinxstyleemphasis{include\_all} include all variables

\end{itemize}

\end{fulllineitems}

\index{saveModel() (in module cbmpy.CBWrite)@\spxentry{saveModel()}\spxextra{in module cbmpy.CBWrite}}

\begin{fulllineitems}
\phantomsection\label{\detokenize{modules_doc:cbmpy.CBWrite.saveModel}}
\pysigstartsignatures
\pysiglinewithargsret{\sphinxbfcode{\sphinxupquote{saveModel}}}{\sphinxparam{\DUrole{n,n}{model}}\sphinxparamcomma \sphinxparam{\DUrole{n,n}{filename}}\sphinxparamcomma \sphinxparam{\DUrole{n,n}{compress}\DUrole{o,o}{=}\DUrole{default_value}{False}}}{}
\pysigstopsignatures
\sphinxAtStartPar
Saves the model to an SBML file using the lates SBML3 FBC version.
\begin{itemize}
\item {} 
\sphinxAtStartPar
\sphinxstyleemphasis{model} the CBMPy model

\item {} 
\sphinxAtStartPar
\sphinxstyleemphasis{filename} the filename to write

\end{itemize}

\end{fulllineitems}

\index{writeCOBRASBML() (in module cbmpy.CBWrite)@\spxentry{writeCOBRASBML()}\spxextra{in module cbmpy.CBWrite}}

\begin{fulllineitems}
\phantomsection\label{\detokenize{modules_doc:cbmpy.CBWrite.writeCOBRASBML}}
\pysigstartsignatures
\pysiglinewithargsret{\sphinxbfcode{\sphinxupquote{writeCOBRASBML}}}{\sphinxparam{\DUrole{n,n}{fba}}\sphinxparamcomma \sphinxparam{\DUrole{n,n}{fname}}\sphinxparamcomma \sphinxparam{\DUrole{n,n}{directory}\DUrole{o,o}{=}\DUrole{default_value}{None}}}{}
\pysigstopsignatures
\sphinxAtStartPar
Takes an FBA model object and writes it to file as a COBRA compatible :
\begin{itemize}
\item {} 
\sphinxAtStartPar
\sphinxstyleemphasis{fba} an fba model object

\item {} 
\sphinxAtStartPar
\sphinxstyleemphasis{fname} the model will be written as XML to \sphinxstyleemphasis{fname}

\item {} 
\sphinxAtStartPar
\sphinxstyleemphasis{directory} {[}default=None{]} if defined it is prepended to fname

\end{itemize}

\end{fulllineitems}

\index{writeFVAdata() (in module cbmpy.CBWrite)@\spxentry{writeFVAdata()}\spxextra{in module cbmpy.CBWrite}}

\begin{fulllineitems}
\phantomsection\label{\detokenize{modules_doc:cbmpy.CBWrite.writeFVAdata}}
\pysigstartsignatures
\pysiglinewithargsret{\sphinxbfcode{\sphinxupquote{writeFVAdata}}}{\sphinxparam{\DUrole{n,n}{fvadata}}\sphinxparamcomma \sphinxparam{\DUrole{n,n}{names}}\sphinxparamcomma \sphinxparam{\DUrole{n,n}{fname}}\sphinxparamcomma \sphinxparam{\DUrole{n,n}{work\_dir}\DUrole{o,o}{=}\DUrole{default_value}{None}}\sphinxparamcomma \sphinxparam{\DUrole{n,n}{roundec}\DUrole{o,o}{=}\DUrole{default_value}{None}}\sphinxparamcomma \sphinxparam{\DUrole{n,n}{scale\_min}\DUrole{o,o}{=}\DUrole{default_value}{False}}\sphinxparamcomma \sphinxparam{\DUrole{n,n}{appendfile}\DUrole{o,o}{=}\DUrole{default_value}{False}}\sphinxparamcomma \sphinxparam{\DUrole{n,n}{info}\DUrole{o,o}{=}\DUrole{default_value}{None}}}{}
\pysigstopsignatures
\sphinxAtStartPar
Takes the resuls of a FluxVariabilityAnalysis method and writes it to a nice
csv file. Note this method replaces the glpk/cplx\_WriteFVAtoCSV methods. Data is output as a csv file
with columns: FluxName, FVA\_MIN, FVA\_MAX, OPT\_VAL, SPAN
\begin{itemize}
\item {} 
\sphinxAtStartPar
\sphinxstyleemphasis{fvadata} FluxVariabilityAnalysis() FVA OUTPUT\_ARRAY

\item {} 
\sphinxAtStartPar
\sphinxstyleemphasis{names} FluxVariabilityAnalysis() FVA OUTPUT\_NAMES

\item {} 
\sphinxAtStartPar
\sphinxstyleemphasis{fname} filename\_base for the CSV output

\item {} 
\sphinxAtStartPar
\sphinxstyleemphasis{work\_dir} {[}default=None{]} if set the output directory for the csv files

\item {} 
\sphinxAtStartPar
\sphinxstyleemphasis{roundec} {[}default=None{]} an integer indicating at which decimal to round off output. Default is no rounding.

\item {} 
\sphinxAtStartPar
\sphinxstyleemphasis{scale\_min} {[}default=False{]} normalise each flux such that that FVA\_MIN = 0.0

\item {} 
\sphinxAtStartPar
\sphinxstyleemphasis{appendfile} {[}default=False{]} instead of opening a new file try and append the data

\item {} 
\sphinxAtStartPar
\sphinxstyleemphasis{info} {[}default=None{]} a string added to the results as an extra column, useful with \sphinxtitleref{appendfile}

\end{itemize}

\end{fulllineitems}

\index{writeFVAtoCSV() (in module cbmpy.CBWrite)@\spxentry{writeFVAtoCSV()}\spxextra{in module cbmpy.CBWrite}}

\begin{fulllineitems}
\phantomsection\label{\detokenize{modules_doc:cbmpy.CBWrite.writeFVAtoCSV}}
\pysigstartsignatures
\pysiglinewithargsret{\sphinxbfcode{\sphinxupquote{writeFVAtoCSV}}}{\sphinxparam{\DUrole{n,n}{fvadata}}\sphinxparamcomma \sphinxparam{\DUrole{n,n}{names}}\sphinxparamcomma \sphinxparam{\DUrole{n,n}{fname}}\sphinxparamcomma \sphinxparam{\DUrole{n,n}{Dir}\DUrole{o,o}{=}\DUrole{default_value}{None}}\sphinxparamcomma \sphinxparam{\DUrole{n,n}{fbaObj}\DUrole{o,o}{=}\DUrole{default_value}{None}}}{}
\pysigstopsignatures
\sphinxAtStartPar
Takes the resuls of a FluxVariabilityAnalysis method and writes it to a nice
csv file. Note this method replaces the glpk/cplx\_WriteFVAtoCSV methods.
\begin{itemize}
\item {} 
\sphinxAtStartPar
\sphinxstyleemphasis{fvadata} FluxVariabilityAnalysis() OUTPUT\_ARRAY

\item {} 
\sphinxAtStartPar
\sphinxstyleemphasis{names} FluxVariabilityAnalysis() OUTPUT\_NAMES

\item {} 
\sphinxAtStartPar
\sphinxstyleemphasis{fname} filename\_base for the CSV output

\item {} 
\sphinxAtStartPar
\sphinxstyleemphasis{Dir} {[}default=None{]} if set the output directory for the csv files

\item {} 
\sphinxAtStartPar
\sphinxstyleemphasis{fbaObj} {[}default=None{]} if supplied adds extra model information into the output tables

\end{itemize}

\end{fulllineitems}

\index{writeMinDistanceLPwithCost() (in module cbmpy.CBWrite)@\spxentry{writeMinDistanceLPwithCost()}\spxextra{in module cbmpy.CBWrite}}

\begin{fulllineitems}
\phantomsection\label{\detokenize{modules_doc:cbmpy.CBWrite.writeMinDistanceLPwithCost}}
\pysigstartsignatures
\pysiglinewithargsret{\sphinxbfcode{\sphinxupquote{writeMinDistanceLPwithCost}}}{\sphinxparam{\DUrole{n,n}{fname}}\sphinxparamcomma \sphinxparam{\DUrole{n,n}{fbas}}\sphinxparamcomma \sphinxparam{\DUrole{n,n}{work\_dir}\DUrole{o,o}{=}\DUrole{default_value}{None}}\sphinxparamcomma \sphinxparam{\DUrole{n,n}{ignoreDistance}\DUrole{o,o}{=}\DUrole{default_value}{{[}{]}}}\sphinxparamcomma \sphinxparam{\DUrole{n,n}{constraint\_mode}\DUrole{o,o}{=}\DUrole{default_value}{\textquotesingle{}strict\textquotesingle{}}}}{}
\pysigstopsignatures
\sphinxAtStartPar
For backwards compatability only

\end{fulllineitems}

\index{writeModelHFormatFBA() (in module cbmpy.CBWrite)@\spxentry{writeModelHFormatFBA()}\spxextra{in module cbmpy.CBWrite}}

\begin{fulllineitems}
\phantomsection\label{\detokenize{modules_doc:cbmpy.CBWrite.writeModelHFormatFBA}}
\pysigstartsignatures
\pysiglinewithargsret{\sphinxbfcode{\sphinxupquote{writeModelHFormatFBA}}}{\sphinxparam{\DUrole{n,n}{fba}}\sphinxparamcomma \sphinxparam{\DUrole{n,n}{work\_dir}\DUrole{o,o}{=}\DUrole{default_value}{None}}\sphinxparamcomma \sphinxparam{\DUrole{n,n}{use\_rational}\DUrole{o,o}{=}\DUrole{default_value}{False}}\sphinxparamcomma \sphinxparam{\DUrole{n,n}{fullLP}\DUrole{o,o}{=}\DUrole{default_value}{True}}\sphinxparamcomma \sphinxparam{\DUrole{n,n}{format}\DUrole{o,o}{=}\DUrole{default_value}{\textquotesingle{}\%s\textquotesingle{}}}\sphinxparamcomma \sphinxparam{\DUrole{n,n}{infinity\_replace}\DUrole{o,o}{=}\DUrole{default_value}{None}}}{}
\pysigstopsignatures
\sphinxAtStartPar
Write an FBA\sphinxhyphen{}LP in polynomial H\sphinxhyphen{}Format file. This version has been replaced by \sphinxtitleref{writeModelHFormatFBA2()}
but is kept for backwards compatability.
\begin{itemize}
\item {} 
\sphinxAtStartPar
\sphinxstyleemphasis{fba} a PySCeS\sphinxhyphen{}CBM FBA object

\item {} 
\sphinxAtStartPar
\sphinxstyleemphasis{Work\_dir} {[}default=None{]} the output directory

\item {} 
\sphinxAtStartPar
\sphinxstyleemphasis{use\_rational} {[}default=false{]} use rational numbers in output (requires sympy)

\item {} 
\sphinxAtStartPar
\sphinxstyleemphasis{fullLP} {[}default=True{]} include the default objective function as a maximization target

\item {} 
\sphinxAtStartPar
\sphinxstyleemphasis{format} {[}default=’\%s’{]} the number format string

\item {} 
\sphinxAtStartPar
\sphinxstyleemphasis{infinity\_replace} {[}default=None{]} if defined this is the abs(value) of +\sphinxhyphen{}\textless{}infinity\textgreater{}

\end{itemize}

\end{fulllineitems}

\index{writeModelHFormatFBA2() (in module cbmpy.CBWrite)@\spxentry{writeModelHFormatFBA2()}\spxextra{in module cbmpy.CBWrite}}

\begin{fulllineitems}
\phantomsection\label{\detokenize{modules_doc:cbmpy.CBWrite.writeModelHFormatFBA2}}
\pysigstartsignatures
\pysiglinewithargsret{\sphinxbfcode{\sphinxupquote{writeModelHFormatFBA2}}}{\sphinxparam{\DUrole{n,n}{fba}}\sphinxparamcomma \sphinxparam{\DUrole{n,n}{fname}\DUrole{o,o}{=}\DUrole{default_value}{None}}\sphinxparamcomma \sphinxparam{\DUrole{n,n}{work\_dir}\DUrole{o,o}{=}\DUrole{default_value}{None}}\sphinxparamcomma \sphinxparam{\DUrole{n,n}{use\_rational}\DUrole{o,o}{=}\DUrole{default_value}{False}}\sphinxparamcomma \sphinxparam{\DUrole{n,n}{fullLP}\DUrole{o,o}{=}\DUrole{default_value}{True}}\sphinxparamcomma \sphinxparam{\DUrole{n,n}{format}\DUrole{o,o}{=}\DUrole{default_value}{\textquotesingle{}\%s\textquotesingle{}}}\sphinxparamcomma \sphinxparam{\DUrole{n,n}{infinity\_replace}\DUrole{o,o}{=}\DUrole{default_value}{None}}}{}
\pysigstopsignatures
\sphinxAtStartPar
Write an FBA\sphinxhyphen{}LP in polynomial H\sphinxhyphen{}Format file. This is an improved version of \sphinxtitleref{WriteModelHFormatFBA()}
which it replaces. Note that if a SymPy matrix is used as input then use\_rational is automatically enabled.
\begin{itemize}
\item {} 
\sphinxAtStartPar
\sphinxstyleemphasis{fba} a PySCeS\sphinxhyphen{}CBM FBA object

\item {} 
\sphinxAtStartPar
\sphinxstyleemphasis{fname} {[}default=None{]} the output filename, fba.getId() if not defined

\item {} 
\sphinxAtStartPar
\sphinxstyleemphasis{Work\_dir} {[}default=None{]} the output directory

\item {} 
\sphinxAtStartPar
\sphinxstyleemphasis{use\_rational} {[}default=false{]} use rational numbers in output (requires sympy)

\item {} 
\sphinxAtStartPar
\sphinxstyleemphasis{fullLP} {[}default=True{]} include the default objective function as a maximization target

\item {} 
\sphinxAtStartPar
\sphinxstyleemphasis{format} {[}default=’\%s’{]} the number format string

\item {} 
\sphinxAtStartPar
\sphinxstyleemphasis{infinity\_replace} {[}default=None{]} if defined this is the abs(value) of +\sphinxhyphen{}\textless{}infinity\textgreater{}

\end{itemize}

\end{fulllineitems}

\index{writeModelInfoToFile() (in module cbmpy.CBWrite)@\spxentry{writeModelInfoToFile()}\spxextra{in module cbmpy.CBWrite}}

\begin{fulllineitems}
\phantomsection\label{\detokenize{modules_doc:cbmpy.CBWrite.writeModelInfoToFile}}
\pysigstartsignatures
\pysiglinewithargsret{\sphinxbfcode{\sphinxupquote{writeModelInfoToFile}}}{\sphinxparam{\DUrole{n,n}{fba}}\sphinxparamcomma \sphinxparam{\DUrole{n,n}{fname}}\sphinxparamcomma \sphinxparam{\DUrole{n,n}{Dir}\DUrole{o,o}{=}\DUrole{default_value}{None}}\sphinxparamcomma \sphinxparam{\DUrole{n,n}{separator}\DUrole{o,o}{=}\DUrole{default_value}{\textquotesingle{},\textquotesingle{}}}\sphinxparamcomma \sphinxparam{\DUrole{n,n}{only\_exchange}\DUrole{o,o}{=}\DUrole{default_value}{False}}\sphinxparamcomma \sphinxparam{\DUrole{n,n}{met\_type}\DUrole{o,o}{=}\DUrole{default_value}{\textquotesingle{}all\textquotesingle{}}}}{}
\pysigstopsignatures
\sphinxAtStartPar
This function writes a CBModel to file
\begin{itemize}
\item {} 
\sphinxAtStartPar
\sphinxstyleemphasis{fba} an instance of an PySCeSCBM model

\item {} 
\sphinxAtStartPar
\sphinxstyleemphasis{fname} the output filename

\item {} 
\sphinxAtStartPar
\sphinxstyleemphasis{Dir} {[}default=None{]} use directory if not None

\item {} 
\sphinxAtStartPar
\sphinxstyleemphasis{separator} {[}default=’,’{]} the column separator

\item {} 
\sphinxAtStartPar
\sphinxstyleemphasis{only\_exchange} {[}default=False{]} only output fluxes labelled as exchange reactions

\item {} 
\sphinxAtStartPar
\sphinxstyleemphasis{type} {[}default=’all’{]} only output certain type of species: ‘all’,’boundary’ or ‘variable’

\end{itemize}

\end{fulllineitems}

\index{writeModelLP() (in module cbmpy.CBWrite)@\spxentry{writeModelLP()}\spxextra{in module cbmpy.CBWrite}}

\begin{fulllineitems}
\phantomsection\label{\detokenize{modules_doc:cbmpy.CBWrite.writeModelLP}}
\pysigstartsignatures
\pysiglinewithargsret{\sphinxbfcode{\sphinxupquote{writeModelLP}}}{\sphinxparam{\DUrole{n,n}{fba}}\sphinxparamcomma \sphinxparam{\DUrole{n,n}{work\_dir}\DUrole{o,o}{=}\DUrole{default_value}{None}}\sphinxparamcomma \sphinxparam{\DUrole{n,n}{fname}\DUrole{o,o}{=}\DUrole{default_value}{None}}\sphinxparamcomma \sphinxparam{\DUrole{n,n}{multisymb}\DUrole{o,o}{=}\DUrole{default_value}{\textquotesingle{} \textquotesingle{}}}\sphinxparamcomma \sphinxparam{\DUrole{n,n}{format}\DUrole{o,o}{=}\DUrole{default_value}{\textquotesingle{}\%s\textquotesingle{}}}\sphinxparamcomma \sphinxparam{\DUrole{n,n}{use\_rational}\DUrole{o,o}{=}\DUrole{default_value}{False}}\sphinxparamcomma \sphinxparam{\DUrole{n,n}{constraint\_mode}\DUrole{o,o}{=}\DUrole{default_value}{None}}\sphinxparamcomma \sphinxparam{\DUrole{n,n}{quiet}\DUrole{o,o}{=}\DUrole{default_value}{False}}}{}
\pysigstopsignatures
\sphinxAtStartPar
Writes an FBA object as an LP in CPLEX LP format
\begin{itemize}
\item {} 
\sphinxAtStartPar
\sphinxstyleemphasis{fba} an instantiated FBAmodel instance

\item {} 
\sphinxAtStartPar
\sphinxstyleemphasis{work\_dir} directory designated for output

\item {} 
\sphinxAtStartPar
\sphinxstyleemphasis{fname} the file name {[}default=fba.getId(){]}

\item {} 
\sphinxAtStartPar
\sphinxstyleemphasis{multisymb} the multiplication symbol (default: \textless{}space\textgreater{})

\item {} 
\sphinxAtStartPar
\sphinxstyleemphasis{format} the number format of the output

\item {} 
\sphinxAtStartPar
\sphinxstyleemphasis{use\_rational} output rational numbers {[}default=False{]}

\item {} 
\sphinxAtStartPar
\sphinxstyleemphasis{quiet} {[}default=False{]} supress information messages

\end{itemize}

\end{fulllineitems}

\index{writeModelLPOld() (in module cbmpy.CBWrite)@\spxentry{writeModelLPOld()}\spxextra{in module cbmpy.CBWrite}}

\begin{fulllineitems}
\phantomsection\label{\detokenize{modules_doc:cbmpy.CBWrite.writeModelLPOld}}
\pysigstartsignatures
\pysiglinewithargsret{\sphinxbfcode{\sphinxupquote{writeModelLPOld}}}{\sphinxparam{\DUrole{n,n}{fba}}\sphinxparamcomma \sphinxparam{\DUrole{n,n}{work\_dir}\DUrole{o,o}{=}\DUrole{default_value}{None}}\sphinxparamcomma \sphinxparam{\DUrole{n,n}{multisymb}\DUrole{o,o}{=}\DUrole{default_value}{\textquotesingle{} \textquotesingle{}}}\sphinxparamcomma \sphinxparam{\DUrole{n,n}{lpt}\DUrole{o,o}{=}\DUrole{default_value}{True}}\sphinxparamcomma \sphinxparam{\DUrole{n,n}{constraint\_mode}\DUrole{o,o}{=}\DUrole{default_value}{\textquotesingle{}strict\textquotesingle{}}}\sphinxparamcomma \sphinxparam{\DUrole{n,n}{use\_rational}\DUrole{o,o}{=}\DUrole{default_value}{False}}\sphinxparamcomma \sphinxparam{\DUrole{n,n}{format}\DUrole{o,o}{=}\DUrole{default_value}{\textquotesingle{}\%s\textquotesingle{}}}}{}
\pysigstopsignatures
\sphinxAtStartPar
Writes a fba as an LP/LPT
\begin{itemize}
\item {} 
\sphinxAtStartPar
\sphinxstyleemphasis{fba} an instantiated FBAmodel instance

\item {} 
\sphinxAtStartPar
\sphinxstyleemphasis{work\_dir} directory designated for output

\item {} 
\sphinxAtStartPar
\sphinxstyleemphasis{multisymb} the multiplication symbol (default: \textless{}space\textgreater{})

\item {} 
\sphinxAtStartPar
\sphinxstyleemphasis{lpt} the file format (default: True for lpt) or False for lp

\end{itemize}

\end{fulllineitems}

\index{writeModelRaw() (in module cbmpy.CBWrite)@\spxentry{writeModelRaw()}\spxextra{in module cbmpy.CBWrite}}

\begin{fulllineitems}
\phantomsection\label{\detokenize{modules_doc:cbmpy.CBWrite.writeModelRaw}}
\pysigstartsignatures
\pysiglinewithargsret{\sphinxbfcode{\sphinxupquote{writeModelRaw}}}{\sphinxparam{\DUrole{n,n}{fba}}\sphinxparamcomma \sphinxparam{\DUrole{n,n}{work\_dir}\DUrole{o,o}{=}\DUrole{default_value}{None}}}{}
\pysigstopsignatures
\sphinxAtStartPar
Writes a fba (actually just dumps it) to a text file.
\begin{itemize}
\item {} 
\sphinxAtStartPar
\sphinxstyleemphasis{fba} an instantiated FBAmodel instance

\item {} 
\sphinxAtStartPar
\sphinxstyleemphasis{work\_dir} directory designated for output

\end{itemize}

\end{fulllineitems}

\index{writeModelToCOMBINEarchive() (in module cbmpy.CBWrite)@\spxentry{writeModelToCOMBINEarchive()}\spxextra{in module cbmpy.CBWrite}}

\begin{fulllineitems}
\phantomsection\label{\detokenize{modules_doc:cbmpy.CBWrite.writeModelToCOMBINEarchive}}
\pysigstartsignatures
\pysiglinewithargsret{\sphinxbfcode{\sphinxupquote{writeModelToCOMBINEarchive}}}{\sphinxparam{\DUrole{n,n}{mod}}\sphinxparamcomma \sphinxparam{\DUrole{n,n}{fname}\DUrole{o,o}{=}\DUrole{default_value}{None}}\sphinxparamcomma \sphinxparam{\DUrole{n,n}{directory}\DUrole{o,o}{=}\DUrole{default_value}{None}}\sphinxparamcomma \sphinxparam{\DUrole{n,n}{sbmlname}\DUrole{o,o}{=}\DUrole{default_value}{None}}\sphinxparamcomma \sphinxparam{\DUrole{n,n}{withExcel}\DUrole{o,o}{=}\DUrole{default_value}{True}}\sphinxparamcomma \sphinxparam{\DUrole{n,n}{vc\_given}\DUrole{o,o}{=}\DUrole{default_value}{\textquotesingle{}CBMPy\textquotesingle{}}}\sphinxparamcomma \sphinxparam{\DUrole{n,n}{vc\_family}\DUrole{o,o}{=}\DUrole{default_value}{\textquotesingle{}Software\textquotesingle{}}}\sphinxparamcomma \sphinxparam{\DUrole{n,n}{vc\_email}\DUrole{o,o}{=}\DUrole{default_value}{\textquotesingle{}None\textquotesingle{}}}\sphinxparamcomma \sphinxparam{\DUrole{n,n}{vc\_org}\DUrole{o,o}{=}\DUrole{default_value}{\textquotesingle{}cbmpy.sourceforge.net\textquotesingle{}}}\sphinxparamcomma \sphinxparam{\DUrole{n,n}{add\_cbmpy\_annot}\DUrole{o,o}{=}\DUrole{default_value}{True}}\sphinxparamcomma \sphinxparam{\DUrole{n,n}{add\_cobra\_annot}\DUrole{o,o}{=}\DUrole{default_value}{True}}}{}
\pysigstopsignatures
\sphinxAtStartPar
Write a model in SBML and Excel format to a COMBINE archive using the following information:
\begin{itemize}
\item {} 
\sphinxAtStartPar
\sphinxstyleemphasis{mod} a model object

\item {} 
\sphinxAtStartPar
\sphinxstyleemphasis{fname} the output base filename, archive will be \textless{}fname\textgreater{}.zip

\item {} 
\sphinxAtStartPar
\sphinxstyleemphasis{directory} {[}default=None{]} created the combine archive ‘directory’

\item {} 
\sphinxAtStartPar
\sphinxstyleemphasis{sbmlname} {[}default=’None’{]} If \sphinxstyleemphasis{sbmlname} is defined then SBML file is \textless{}sbmlname\textgreater{}.xml otherwise sbml will be \textless{}fname\textgreater{}.xml.

\item {} 
\sphinxAtStartPar
\sphinxstyleemphasis{withExcel} {[}default=True{]} include a human readable Excel spreadsheet version of the model

\item {} 
\sphinxAtStartPar
\sphinxstyleemphasis{vc\_given} {[}default=’CBMPy’{]} first name

\item {} 
\sphinxAtStartPar
\sphinxstyleemphasis{vc\_family} {[}default=’Software’{]} family name

\item {} 
\sphinxAtStartPar
\sphinxstyleemphasis{vc\_email} {[}default=’None’{]} email

\item {} 
\sphinxAtStartPar
\sphinxstyleemphasis{vc\_org} {[}default=’None’{]} organisation

\item {} 
\sphinxAtStartPar
\sphinxstyleemphasis{add\_cbmpy\_annot} {[}default=True{]} add CBMPy KeyValueData annotation. Replaces \textless{}notes\textgreater{}

\item {} 
\sphinxAtStartPar
\sphinxstyleemphasis{add\_cobra\_annot} {[}default=True{]} add COBRA \textless{}notes\textgreater{} annotation

\end{itemize}

\end{fulllineitems}

\index{writeModelToExcel97() (in module cbmpy.CBWrite)@\spxentry{writeModelToExcel97()}\spxextra{in module cbmpy.CBWrite}}

\begin{fulllineitems}
\phantomsection\label{\detokenize{modules_doc:cbmpy.CBWrite.writeModelToExcel97}}
\pysigstartsignatures
\pysiglinewithargsret{\sphinxbfcode{\sphinxupquote{writeModelToExcel97}}}{\sphinxparam{\DUrole{n,n}{fba}}\sphinxparamcomma \sphinxparam{\DUrole{n,n}{filename}}\sphinxparamcomma \sphinxparam{\DUrole{n,n}{roundoff}\DUrole{o,o}{=}\DUrole{default_value}{6}}}{}
\pysigstopsignatures
\sphinxAtStartPar
Exports the model as an Excel 97 spreadsheet
\begin{itemize}
\item {} 
\sphinxAtStartPar
\sphinxstyleemphasis{fba} a CBMPy model instance

\item {} 
\sphinxAtStartPar
\sphinxstyleemphasis{filename} the filename of the workbook

\item {} 
\sphinxAtStartPar
\sphinxstyleemphasis{roundoff} {[}default=6{]} the number of digits to round off to

\end{itemize}

\end{fulllineitems}

\index{writeOptimalSolution() (in module cbmpy.CBWrite)@\spxentry{writeOptimalSolution()}\spxextra{in module cbmpy.CBWrite}}

\begin{fulllineitems}
\phantomsection\label{\detokenize{modules_doc:cbmpy.CBWrite.writeOptimalSolution}}
\pysigstartsignatures
\pysiglinewithargsret{\sphinxbfcode{\sphinxupquote{writeOptimalSolution}}}{\sphinxparam{\DUrole{n,n}{fba}}\sphinxparamcomma \sphinxparam{\DUrole{n,n}{fname}}\sphinxparamcomma \sphinxparam{\DUrole{n,n}{Dir}\DUrole{o,o}{=}\DUrole{default_value}{None}}\sphinxparamcomma \sphinxparam{\DUrole{n,n}{separator}\DUrole{o,o}{=}\DUrole{default_value}{\textquotesingle{},\textquotesingle{}}}\sphinxparamcomma \sphinxparam{\DUrole{n,n}{only\_exchange}\DUrole{o,o}{=}\DUrole{default_value}{False}}}{}
\pysigstopsignatures
\sphinxAtStartPar
This function writes the optimal solution to file
\begin{itemize}
\item {} 
\sphinxAtStartPar
\sphinxstyleemphasis{fba} an instance of an PySCeSCBM model

\item {} 
\sphinxAtStartPar
\sphinxstyleemphasis{fname} the output filename

\item {} 
\sphinxAtStartPar
\sphinxstyleemphasis{Dir} {[}default=None{]} use current directory if not None

\item {} 
\sphinxAtStartPar
\sphinxstyleemphasis{separator} {[}default=’,’{]} the column separator

\item {} 
\sphinxAtStartPar
\sphinxstyleemphasis{only\_exchange} {[}default=False{]} only output fluxes labelled as exchange reactions

\end{itemize}

\end{fulllineitems}

\index{writeProteinCostToCSV() (in module cbmpy.CBWrite)@\spxentry{writeProteinCostToCSV()}\spxextra{in module cbmpy.CBWrite}}

\begin{fulllineitems}
\phantomsection\label{\detokenize{modules_doc:cbmpy.CBWrite.writeProteinCostToCSV}}
\pysigstartsignatures
\pysiglinewithargsret{\sphinxbfcode{\sphinxupquote{writeProteinCostToCSV}}}{\sphinxparam{\DUrole{n,n}{fba}}\sphinxparamcomma \sphinxparam{\DUrole{n,n}{fname}}}{}
\pysigstopsignatures
\sphinxAtStartPar
Writes the protein costs ‘CBM\_PEPTIDE\_COST’ annotation toa csv file.
\begin{itemize}
\item {} 
\sphinxAtStartPar
\sphinxstyleemphasis{fba} an instantiated FBA object

\item {} 
\sphinxAtStartPar
\sphinxstyleemphasis{fname} the exported file name

\end{itemize}

\end{fulllineitems}

\index{writeReactionInfoToFile() (in module cbmpy.CBWrite)@\spxentry{writeReactionInfoToFile()}\spxextra{in module cbmpy.CBWrite}}

\begin{fulllineitems}
\phantomsection\label{\detokenize{modules_doc:cbmpy.CBWrite.writeReactionInfoToFile}}
\pysigstartsignatures
\pysiglinewithargsret{\sphinxbfcode{\sphinxupquote{writeReactionInfoToFile}}}{\sphinxparam{\DUrole{n,n}{fba}}\sphinxparamcomma \sphinxparam{\DUrole{n,n}{fname}}\sphinxparamcomma \sphinxparam{\DUrole{n,n}{Dir}\DUrole{o,o}{=}\DUrole{default_value}{None}}\sphinxparamcomma \sphinxparam{\DUrole{n,n}{separator}\DUrole{o,o}{=}\DUrole{default_value}{\textquotesingle{},\textquotesingle{}}}\sphinxparamcomma \sphinxparam{\DUrole{n,n}{only\_exchange}\DUrole{o,o}{=}\DUrole{default_value}{False}}}{}
\pysigstopsignatures
\sphinxAtStartPar
This function writes a CBModel to file
\begin{itemize}
\item {} 
\sphinxAtStartPar
\sphinxstyleemphasis{fba} an instance of an PySCeSCBM model

\item {} 
\sphinxAtStartPar
\sphinxstyleemphasis{fname} the output filename

\item {} 
\sphinxAtStartPar
\sphinxstyleemphasis{Dir} {[}default=None{]} use directory if not None

\item {} 
\sphinxAtStartPar
\sphinxstyleemphasis{separator} {[}default=’,’{]} the column separator

\item {} 
\sphinxAtStartPar
\sphinxstyleemphasis{only\_exchange} {[}default=False{]} only output fluxes labelled as exchange reactions

\end{itemize}

\end{fulllineitems}

\index{writeSBML2FBA() (in module cbmpy.CBWrite)@\spxentry{writeSBML2FBA()}\spxextra{in module cbmpy.CBWrite}}

\begin{fulllineitems}
\phantomsection\label{\detokenize{modules_doc:cbmpy.CBWrite.writeSBML2FBA}}
\pysigstartsignatures
\pysiglinewithargsret{\sphinxbfcode{\sphinxupquote{writeSBML2FBA}}}{\sphinxparam{\DUrole{n,n}{fba}}\sphinxparamcomma \sphinxparam{\DUrole{n,n}{fname}}\sphinxparamcomma \sphinxparam{\DUrole{n,n}{directory}\DUrole{o,o}{=}\DUrole{default_value}{None}}\sphinxparamcomma \sphinxparam{\DUrole{n,n}{sbml\_level\_version}\DUrole{o,o}{=}\DUrole{default_value}{None}}}{}
\pysigstopsignatures
\sphinxAtStartPar
Takes an FBA model object and writes it to file as SBML L2 with FBA annotations.
Note if you want to write BiGG/FAME style annotations then you must use \sphinxstyleemphasis{sbml\_level\_version=(2,1)}
\begin{itemize}
\item {} 
\sphinxAtStartPar
\sphinxstyleemphasis{fba} an fba model object

\item {} 
\sphinxAtStartPar
\sphinxstyleemphasis{fname} the model will be written as XML to \sphinxstyleemphasis{fname}

\item {} 
\sphinxAtStartPar
\sphinxstyleemphasis{sbml\_level\_version} {[}default=None{]} a tuple containing the SBML level and version e.g. (2,1)

\end{itemize}

\sphinxAtStartPar
This is a utility wrapper for the function \sphinxtitleref{CBXML.sbml\_writeSBML2FBA}

\end{fulllineitems}

\index{writeSBML3FBC() (in module cbmpy.CBWrite)@\spxentry{writeSBML3FBC()}\spxextra{in module cbmpy.CBWrite}}

\begin{fulllineitems}
\phantomsection\label{\detokenize{modules_doc:cbmpy.CBWrite.writeSBML3FBC}}
\pysigstartsignatures
\pysiglinewithargsret{\sphinxbfcode{\sphinxupquote{writeSBML3FBC}}}{\sphinxparam{\DUrole{n,n}{fba}}\sphinxparamcomma \sphinxparam{\DUrole{n,n}{fname}}\sphinxparamcomma \sphinxparam{\DUrole{n,n}{directory}\DUrole{o,o}{=}\DUrole{default_value}{None}}\sphinxparamcomma \sphinxparam{\DUrole{n,n}{gpr\_from\_annot}\DUrole{o,o}{=}\DUrole{default_value}{False}}\sphinxparamcomma \sphinxparam{\DUrole{n,n}{add\_groups}\DUrole{o,o}{=}\DUrole{default_value}{True}}\sphinxparamcomma \sphinxparam{\DUrole{n,n}{add\_cbmpy\_annot}\DUrole{o,o}{=}\DUrole{default_value}{True}}\sphinxparamcomma \sphinxparam{\DUrole{n,n}{add\_cobra\_annot}\DUrole{o,o}{=}\DUrole{default_value}{False}}\sphinxparamcomma \sphinxparam{\DUrole{n,n}{xoptions}\DUrole{o,o}{=}\DUrole{default_value}{\{\textquotesingle{}compress\_bounds\textquotesingle{}: True, \textquotesingle{}fbc\_version\textquotesingle{}: 1, \textquotesingle{}validate\textquotesingle{}: False\}}}}{}
\pysigstopsignatures
\sphinxAtStartPar
Takes an FBA model object and writes it to file as SBML L3 FBC:
\begin{itemize}
\item {} 
\sphinxAtStartPar
\sphinxstyleemphasis{fba} an fba model object

\item {} 
\sphinxAtStartPar
\sphinxstyleemphasis{fname} the model will be written as XML to \sphinxstyleemphasis{fname}

\item {} 
\sphinxAtStartPar
\sphinxstyleemphasis{directory} {[}default=None{]} if defined it is prepended to fname

\item {} 
\sphinxAtStartPar
\sphinxstyleemphasis{gpr\_from\_annot} {[}default=True{]} if enabled will attempt to add the gene protein associations from the annotations
if no gene protein association objects exist

\item {} 
\sphinxAtStartPar
\sphinxstyleemphasis{add\_groups} {[}default=True{]} add SBML3 groups (if supported by libSBML)

\item {} 
\sphinxAtStartPar
\sphinxstyleemphasis{add\_cbmpy\_annot} {[}default=True{]} add CBMPy KeyValueData annotation. Replaces \textless{}notes\textgreater{}

\item {} 
\sphinxAtStartPar
\sphinxstyleemphasis{add\_cobra\_annot} {[}default=True{]} add COBRA \textless{}notes\textgreater{} annotation

\item {} 
\sphinxAtStartPar
\sphinxstyleemphasis{xoptions} extended options
\begin{itemize}
\item {} 
\sphinxAtStartPar
\sphinxstyleemphasis{fbc\_version} {[}default=1{]} write SBML3FBC using version 1 (2013) or version 2 (2015)

\item {} 
\sphinxAtStartPar
\sphinxstyleemphasis{validate} {[}default=False{]} validate the output SBML file

\item {} 
\sphinxAtStartPar
\sphinxstyleemphasis{compress\_bounds} {[}default=False{]} try compress output flux bound parameters

\item {} 
\sphinxAtStartPar
\sphinxstyleemphasis{zip\_model} {[}default=False{]} compress the model using PKZIP encoding

\item {} 
\sphinxAtStartPar
\sphinxstyleemphasis{return\_model\_string} {[}default=False{]} return the SBML XML file as a string

\end{itemize}

\end{itemize}

\end{fulllineitems}

\index{writeSBML3FBCV2() (in module cbmpy.CBWrite)@\spxentry{writeSBML3FBCV2()}\spxextra{in module cbmpy.CBWrite}}

\begin{fulllineitems}
\phantomsection\label{\detokenize{modules_doc:cbmpy.CBWrite.writeSBML3FBCV2}}
\pysigstartsignatures
\pysiglinewithargsret{\sphinxbfcode{\sphinxupquote{writeSBML3FBCV2}}}{\sphinxparam{\DUrole{n,n}{fba}}\sphinxparamcomma \sphinxparam{\DUrole{n,n}{fname}}\sphinxparamcomma \sphinxparam{\DUrole{n,n}{directory}\DUrole{o,o}{=}\DUrole{default_value}{None}}\sphinxparamcomma \sphinxparam{\DUrole{n,n}{gpr\_from\_annot}\DUrole{o,o}{=}\DUrole{default_value}{False}}\sphinxparamcomma \sphinxparam{\DUrole{n,n}{add\_groups}\DUrole{o,o}{=}\DUrole{default_value}{True}}\sphinxparamcomma \sphinxparam{\DUrole{n,n}{add\_cbmpy\_annot}\DUrole{o,o}{=}\DUrole{default_value}{True}}\sphinxparamcomma \sphinxparam{\DUrole{n,n}{add\_cobra\_annot}\DUrole{o,o}{=}\DUrole{default_value}{False}}\sphinxparamcomma \sphinxparam{\DUrole{n,n}{validate}\DUrole{o,o}{=}\DUrole{default_value}{False}}\sphinxparamcomma \sphinxparam{\DUrole{n,n}{compress\_bounds}\DUrole{o,o}{=}\DUrole{default_value}{False}}\sphinxparamcomma \sphinxparam{\DUrole{n,n}{zip\_model}\DUrole{o,o}{=}\DUrole{default_value}{False}}\sphinxparamcomma \sphinxparam{\DUrole{n,n}{return\_model\_string}\DUrole{o,o}{=}\DUrole{default_value}{False}}}{}
\pysigstopsignatures
\sphinxAtStartPar
Takes an FBA model object and writes it to file as SBML L3 FBCv2 :
\begin{itemize}
\item {} 
\sphinxAtStartPar
\sphinxstyleemphasis{fba} an fba model object

\item {} 
\sphinxAtStartPar
\sphinxstyleemphasis{fname} the model will be written as XML to \sphinxstyleemphasis{fname}

\item {} 
\sphinxAtStartPar
\sphinxstyleemphasis{directory} {[}default=None{]} if defined it is prepended to fname

\item {} 
\sphinxAtStartPar
\sphinxstyleemphasis{gpr\_from\_annot} {[}default=False{]} if enabled will attempt to add the gene protein associations from the annotations

\item {} 
\sphinxAtStartPar
\sphinxstyleemphasis{add\_groups} {[}default=True{]} add SBML3 groups (if supported by libSBML)

\item {} 
\sphinxAtStartPar
\sphinxstyleemphasis{add\_cbmpy\_annot} {[}default=True{]} add CBMPy KeyValueData annotation. Replaces \textless{}notes\textgreater{}

\item {} 
\sphinxAtStartPar
\sphinxstyleemphasis{add\_cobra\_annot} {[}default=False{]} add COBRA \textless{}notes\textgreater{} annotation

\item {} 
\sphinxAtStartPar
\sphinxstyleemphasis{validate} {[}default=False{]} validate the output SBML file

\item {} 
\sphinxAtStartPar
\sphinxstyleemphasis{compress\_bounds} {[}default=True{]} try compress output flux bound parameters

\item {} 
\sphinxAtStartPar
\sphinxstyleemphasis{zip\_model} {[}default=False{]} compress the model using ZIP encoding

\item {} 
\sphinxAtStartPar
\sphinxstyleemphasis{return\_model\_string} {[}default=False{]} return the SBML XML file as a string

\end{itemize}

\end{fulllineitems}

\index{writeSBML3FBCV3() (in module cbmpy.CBWrite)@\spxentry{writeSBML3FBCV3()}\spxextra{in module cbmpy.CBWrite}}

\begin{fulllineitems}
\phantomsection\label{\detokenize{modules_doc:cbmpy.CBWrite.writeSBML3FBCV3}}
\pysigstartsignatures
\pysiglinewithargsret{\sphinxbfcode{\sphinxupquote{writeSBML3FBCV3}}}{\sphinxparam{\DUrole{n,n}{fba}}\sphinxparamcomma \sphinxparam{\DUrole{n,n}{fname}}\sphinxparamcomma \sphinxparam{\DUrole{n,n}{directory}\DUrole{o,o}{=}\DUrole{default_value}{None}}\sphinxparamcomma \sphinxparam{\DUrole{n,n}{gpr\_from\_annot}\DUrole{o,o}{=}\DUrole{default_value}{False}}\sphinxparamcomma \sphinxparam{\DUrole{n,n}{add\_groups}\DUrole{o,o}{=}\DUrole{default_value}{True}}\sphinxparamcomma \sphinxparam{\DUrole{n,n}{add\_cbmpy\_annot}\DUrole{o,o}{=}\DUrole{default_value}{True}}\sphinxparamcomma \sphinxparam{\DUrole{n,n}{add\_cobra\_annot}\DUrole{o,o}{=}\DUrole{default_value}{False}}\sphinxparamcomma \sphinxparam{\DUrole{n,n}{validate}\DUrole{o,o}{=}\DUrole{default_value}{False}}\sphinxparamcomma \sphinxparam{\DUrole{n,n}{compress\_bounds}\DUrole{o,o}{=}\DUrole{default_value}{False}}\sphinxparamcomma \sphinxparam{\DUrole{n,n}{zip\_model}\DUrole{o,o}{=}\DUrole{default_value}{False}}\sphinxparamcomma \sphinxparam{\DUrole{n,n}{return\_model\_string}\DUrole{o,o}{=}\DUrole{default_value}{False}}}{}
\pysigstopsignatures
\sphinxAtStartPar
Takes an FBA model object and writes it to file as SBML L3 FBCv3 :
\begin{itemize}
\item {} 
\sphinxAtStartPar
\sphinxstyleemphasis{fba} an fba model object

\item {} 
\sphinxAtStartPar
\sphinxstyleemphasis{fname} the model will be written as XML to \sphinxstyleemphasis{fname}

\item {} 
\sphinxAtStartPar
\sphinxstyleemphasis{directory} {[}default=None{]} if defined it is prepended to fname

\item {} 
\sphinxAtStartPar
\sphinxstyleemphasis{gpr\_from\_annot} {[}default=False{]} if enabled will attempt to add the gene protein associations from the annotations

\item {} 
\sphinxAtStartPar
\sphinxstyleemphasis{add\_groups} {[}default=True{]} add SBML3 groups (if supported by libSBML)

\item {} 
\sphinxAtStartPar
\sphinxstyleemphasis{add\_cbmpy\_annot} {[}default=True{]} add CBMPy KeyValueData annotation. Replaces \textless{}notes\textgreater{}

\item {} 
\sphinxAtStartPar
\sphinxstyleemphasis{add\_cobra\_annot} {[}default=False{]} add COBRA \textless{}notes\textgreater{} annotation

\item {} 
\sphinxAtStartPar
\sphinxstyleemphasis{validate} {[}default=False{]} validate the output SBML file

\item {} 
\sphinxAtStartPar
\sphinxstyleemphasis{compress\_bounds} {[}default=True{]} try compress output flux bound parameters

\item {} 
\sphinxAtStartPar
\sphinxstyleemphasis{zip\_model} {[}default=False{]} compress the model using ZIP encoding

\item {} 
\sphinxAtStartPar
\sphinxstyleemphasis{return\_model\_string} {[}default=False{]} return the SBML XML file as a string

\end{itemize}

\end{fulllineitems}

\index{writeSensitivitiesToCSV() (in module cbmpy.CBWrite)@\spxentry{writeSensitivitiesToCSV()}\spxextra{in module cbmpy.CBWrite}}

\begin{fulllineitems}
\phantomsection\label{\detokenize{modules_doc:cbmpy.CBWrite.writeSensitivitiesToCSV}}
\pysigstartsignatures
\pysiglinewithargsret{\sphinxbfcode{\sphinxupquote{writeSensitivitiesToCSV}}}{\sphinxparam{\DUrole{n,n}{sensitivities}}\sphinxparamcomma \sphinxparam{\DUrole{n,n}{fname}}}{}
\pysigstopsignatures
\sphinxAtStartPar
Write out a sensitivity report using the objective sensitivities and
bound sensitivity dictionaries created by e.g. cplx\_getSensitivities().
\begin{quote}
\begin{itemize}
\item {} 
\sphinxAtStartPar
\sphinxstyleemphasis{sensitivity} tuple containing

\end{itemize}
\begin{itemize}
\item {} 
\sphinxAtStartPar
\sphinxstyleemphasis{obj\_sens} dictionary of objective coefficient sensitivities (per flux)

\item {} 
\sphinxAtStartPar
\sphinxstyleemphasis{rhs\_sens} dictionary of constraint rhs sensitivities (per constraint)

\item {} 
\sphinxAtStartPar
\sphinxstyleemphasis{bound\_sens} dictionary of bound sensitivities (per flux)

\end{itemize}
\begin{itemize}
\item {} 
\sphinxAtStartPar
\sphinxstyleemphasis{fname} output filename e.g. fname.csv

\end{itemize}
\end{quote}

\end{fulllineitems}

\index{writeSolutions() (in module cbmpy.CBWrite)@\spxentry{writeSolutions()}\spxextra{in module cbmpy.CBWrite}}

\begin{fulllineitems}
\phantomsection\label{\detokenize{modules_doc:cbmpy.CBWrite.writeSolutions}}
\pysigstartsignatures
\pysiglinewithargsret{\sphinxbfcode{\sphinxupquote{writeSolutions}}}{\sphinxparam{\DUrole{n,n}{fname}}\sphinxparamcomma \sphinxparam{\DUrole{n,n}{sols}\DUrole{o,o}{=}\DUrole{default_value}{{[}{]}}}\sphinxparamcomma \sphinxparam{\DUrole{n,n}{sep}\DUrole{o,o}{=}\DUrole{default_value}{\textquotesingle{},\textquotesingle{}}}\sphinxparamcomma \sphinxparam{\DUrole{n,n}{extra\_output}\DUrole{o,o}{=}\DUrole{default_value}{None}}\sphinxparamcomma \sphinxparam{\DUrole{n,n}{fba}\DUrole{o,o}{=}\DUrole{default_value}{None}}}{}
\pysigstopsignatures
\sphinxAtStartPar
Write 2 or more solutions where a solution is a dictionary of flux:value pairs:
\begin{itemize}
\item {} 
\sphinxAtStartPar
\sphinxstyleemphasis{fname} the export filename

\item {} 
\sphinxAtStartPar
\sphinxstyleemphasis{sols} a list of dictionaries containing flux:value pairs (e.g. output by cmod.getReactionValues())

\item {} 
\sphinxAtStartPar
\sphinxstyleemphasis{sep} {[}default=’,’{]} the column separator

\item {} 
\sphinxAtStartPar
\sphinxstyleemphasis{extra\_output} {[}default=None{]} add detailed information to output e.g. reaction names by giving a CBModel object as an argument to \sphinxstyleemphasis{extra\_output}.

\item {} 
\sphinxAtStartPar
\sphinxstyleemphasis{fba} an fba model that canbe used for extra\_output

\end{itemize}

\end{fulllineitems}

\index{writeSpeciesInfoToFile() (in module cbmpy.CBWrite)@\spxentry{writeSpeciesInfoToFile()}\spxextra{in module cbmpy.CBWrite}}

\begin{fulllineitems}
\phantomsection\label{\detokenize{modules_doc:cbmpy.CBWrite.writeSpeciesInfoToFile}}
\pysigstartsignatures
\pysiglinewithargsret{\sphinxbfcode{\sphinxupquote{writeSpeciesInfoToFile}}}{\sphinxparam{\DUrole{n,n}{fba}}\sphinxparamcomma \sphinxparam{\DUrole{n,n}{fname}}\sphinxparamcomma \sphinxparam{\DUrole{n,n}{Dir}\DUrole{o,o}{=}\DUrole{default_value}{None}}\sphinxparamcomma \sphinxparam{\DUrole{n,n}{separator}\DUrole{o,o}{=}\DUrole{default_value}{\textquotesingle{},\textquotesingle{}}}\sphinxparamcomma \sphinxparam{\DUrole{n,n}{met\_type}\DUrole{o,o}{=}\DUrole{default_value}{\textquotesingle{}all\textquotesingle{}}}}{}
\pysigstopsignatures
\sphinxAtStartPar
This function writes a CBModel to file
\begin{itemize}
\item {} 
\sphinxAtStartPar
\sphinxstyleemphasis{fba} an instance of an PySCeSCBM model

\item {} 
\sphinxAtStartPar
\sphinxstyleemphasis{fname} the output filename

\item {} 
\sphinxAtStartPar
\sphinxstyleemphasis{Dir} {[}default=None{]} use directory if not None

\item {} 
\sphinxAtStartPar
\sphinxstyleemphasis{separator} {[}default=’,’{]} the column separator

\item {} 
\sphinxAtStartPar
\sphinxstyleemphasis{met\_type} {[}default=’all’{]} only output certain type of species: ‘all’,’boundary’ or ‘variable’

\end{itemize}

\end{fulllineitems}

\index{writeStoichiometricMatrix() (in module cbmpy.CBWrite)@\spxentry{writeStoichiometricMatrix()}\spxextra{in module cbmpy.CBWrite}}

\begin{fulllineitems}
\phantomsection\label{\detokenize{modules_doc:cbmpy.CBWrite.writeStoichiometricMatrix}}
\pysigstartsignatures
\pysiglinewithargsret{\sphinxbfcode{\sphinxupquote{writeStoichiometricMatrix}}}{\sphinxparam{\DUrole{n,n}{fba}}\sphinxparamcomma \sphinxparam{\DUrole{n,n}{fname}\DUrole{o,o}{=}\DUrole{default_value}{None}}\sphinxparamcomma \sphinxparam{\DUrole{n,n}{work\_dir}\DUrole{o,o}{=}\DUrole{default_value}{None}}\sphinxparamcomma \sphinxparam{\DUrole{n,n}{use\_rational}\DUrole{o,o}{=}\DUrole{default_value}{False}}\sphinxparamcomma \sphinxparam{\DUrole{n,n}{fullLP}\DUrole{o,o}{=}\DUrole{default_value}{True}}\sphinxparamcomma \sphinxparam{\DUrole{n,n}{format}\DUrole{o,o}{=}\DUrole{default_value}{\textquotesingle{}\%s\textquotesingle{}}}\sphinxparamcomma \sphinxparam{\DUrole{n,n}{infinity\_replace}\DUrole{o,o}{=}\DUrole{default_value}{None}}}{}
\pysigstopsignatures
\sphinxAtStartPar
Write an FBA\sphinxhyphen{}LP in polynomial H\sphinxhyphen{}Format file. This is an improved version of \sphinxtitleref{WriteModelHFormatFBA()}
which it replaces but is kept for backwards compatability.
\begin{itemize}
\item {} 
\sphinxAtStartPar
\sphinxstyleemphasis{fba} a PySCeS\sphinxhyphen{}CBM FBA object

\item {} 
\sphinxAtStartPar
\sphinxstyleemphasis{fname} {[}default=None{]} the output filename, fba.getId() if not defined

\item {} 
\sphinxAtStartPar
\sphinxstyleemphasis{Work\_dir} {[}default=None{]} the output directory

\item {} 
\sphinxAtStartPar
\sphinxstyleemphasis{use\_rational} {[}default=false{]} use rational numbers in output (requires sympy)

\item {} 
\sphinxAtStartPar
\sphinxstyleemphasis{fullLP} {[}default=True{]} include the default objective function as a maximization target

\item {} 
\sphinxAtStartPar
\sphinxstyleemphasis{format} {[}default=’\%s’{]} the number format string

\item {} 
\sphinxAtStartPar
\sphinxstyleemphasis{infinity\_replace} {[}default=None{]} if defined this is the abs(value) of +\sphinxhyphen{}\textless{}infinity\textgreater{}

\end{itemize}

\end{fulllineitems}

\phantomsection\label{\detokenize{modules_doc:module-cbmpy.CBWx}}\index{module@\spxentry{module}!cbmpy.CBWx@\spxentry{cbmpy.CBWx}}\index{cbmpy.CBWx@\spxentry{cbmpy.CBWx}!module@\spxentry{module}}

\section{CBMPy: CBWx module}
\label{\detokenize{modules_doc:cbmpy-cbwx-module}}
\sphinxAtStartPar
PySCeS Constraint Based Modelling (\sphinxurl{http://cbmpy.sourceforge.net})
Copyright (C) 2009\sphinxhyphen{}2024 Brett G. Olivier, VU University Amsterdam, Amsterdam, The Netherlands

\sphinxAtStartPar
This program is free software: you can redistribute it and/or modify
it under the terms of the GNU General Public License as published by
the Free Software Foundation, either version 3 of the License, or
(at your option) any later version.

\sphinxAtStartPar
This program is distributed in the hope that it will be useful,
but WITHOUT ANY WARRANTY; without even the implied warranty of
MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
GNU General Public License for more details.

\sphinxAtStartPar
You should have received a copy of the GNU General Public License
along with this program.  If not, see \textless{}\sphinxurl{http://www.gnu.org/licenses/}\textgreater{}

\sphinxAtStartPar
Author: Brett G. Olivier PhD
Contact developers: \sphinxurl{https://github.com/SystemsBioinformatics/cbmpy/issues}
Last edit: \$Author: bgoli \$ (\$Id: CBWx.py 710 2020\sphinxhyphen{}04\sphinxhyphen{}27 14:22:34Z bgoli \$)
\phantomsection\label{\detokenize{modules_doc:module-cbmpy.CBXML}}\index{module@\spxentry{module}!cbmpy.CBXML@\spxentry{cbmpy.CBXML}}\index{cbmpy.CBXML@\spxentry{cbmpy.CBXML}!module@\spxentry{module}}

\section{CBMPy: CBXML module}
\label{\detokenize{modules_doc:cbmpy-cbxml-module}}
\sphinxAtStartPar
PySCeS Constraint Based Modelling (\sphinxurl{http://cbmpy.sourceforge.net})
Copyright (C) 2009\sphinxhyphen{}2024 Brett G. Olivier, VU University Amsterdam, Amsterdam, The Netherlands

\sphinxAtStartPar
This program is free software: you can redistribute it and/or modify
it under the terms of the GNU General Public License as published by
the Free Software Foundation, either version 3 of the License, or
(at your option) any later version.

\sphinxAtStartPar
This program is distributed in the hope that it will be useful,
but WITHOUT ANY WARRANTY; without even the implied warranty of
MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
GNU General Public License for more details.

\sphinxAtStartPar
You should have received a copy of the GNU General Public License
along with this program.  If not, see \textless{}\sphinxurl{http://www.gnu.org/licenses/}\textgreater{}

\sphinxAtStartPar
Author: Brett G. Olivier PhD
Contact developers: \sphinxurl{https://github.com/SystemsBioinformatics/cbmpy/issues}
Last edit: \$Author: bgoli \$ (\$Id: CBXML.py 710 2020\sphinxhyphen{}04\sphinxhyphen{}27 14:22:34Z bgoli \$)
\index{MLStripper (class in cbmpy.CBXML)@\spxentry{MLStripper}\spxextra{class in cbmpy.CBXML}}

\begin{fulllineitems}
\phantomsection\label{\detokenize{modules_doc:cbmpy.CBXML.MLStripper}}
\pysigstartsignatures
\pysigline{\sphinxbfcode{\sphinxupquote{class\DUrole{w,w}{  }}}\sphinxbfcode{\sphinxupquote{MLStripper}}}
\pysigstopsignatures
\sphinxAtStartPar
Class for stripping a string of HTML/XML used from:
\sphinxurl{http://stackoverflow.com/questions/753052/strip-html-from-strings-in-python}

\end{fulllineitems}

\index{SBML\_NS (in module cbmpy.CBXML)@\spxentry{SBML\_NS}\spxextra{in module cbmpy.CBXML}}

\begin{fulllineitems}
\phantomsection\label{\detokenize{modules_doc:cbmpy.CBXML.SBML_NS}}
\pysigstartsignatures
\pysigline{\sphinxbfcode{\sphinxupquote{SBML\_NS}}\sphinxbfcode{\sphinxupquote{\DUrole{w,w}{  }\DUrole{p,p}{=}\DUrole{w,w}{  }{[}(\textquotesingle{}http://www.sbml.org/sbml/level3/version1/fbc/version3\textquotesingle{}, \textquotesingle{}L3V1FBC3\textquotesingle{}), (\textquotesingle{}http://www.sbml.org/sbml/level3/version1/fbc/version2\textquotesingle{}, \textquotesingle{}L3V1FBC2\textquotesingle{}), (\textquotesingle{}http://www.sbml.org/sbml/level3/version1/fbc/version1\textquotesingle{}, \textquotesingle{}L3V1FBC1\textquotesingle{}), (\textquotesingle{}http://www.sbml.org/sbml/level3/version2/fbc/version2\textquotesingle{}, \textquotesingle{}L3V2FBC2\textquotesingle{}), (\textquotesingle{}http://www.sbml.org/sbml/level3/version2/fbc/version1\textquotesingle{}, \textquotesingle{}L3V2FBC1\textquotesingle{}), (\textquotesingle{}http://www.sbml.org/sbml/level3/version2/core\textquotesingle{}, \textquotesingle{}L3V2core\textquotesingle{}), (\textquotesingle{}http://www.sbml.org/sbml/level3/version1/core\textquotesingle{}, \textquotesingle{}L3V1core\textquotesingle{}), (\textquotesingle{}http://www.sbml.org/sbml/level2/version4\textquotesingle{}, \textquotesingle{}L2\textquotesingle{}), (\textquotesingle{}http://www.sbml.org/sbml/level2\textquotesingle{}, \textquotesingle{}L2\textquotesingle{}){]}}}}
\pysigstopsignatures
\sphinxAtStartPar
print libsbml.BQB\_ENCODES            , 8  \# “encodes”,
print libsbml.BQB\_HAS\_PART           , 1  \# “hasPart”,
print libsbml.BQB\_HAS\_PROPERTY       , 10 \# “hasProperty”,
print libsbml.BQB\_HAS\_VERSION        , 4  \# “hasVersion”,
print libsbml.BQB\_IS                 , 0  \# “isA”,
print libsbml.BQB\_IS\_DESCRIBED\_BY    , 6  \# “isDescribedBy”,
print libsbml.BQB\_IS\_ENCODED\_BY      , 7  \# “isEncodedBy”,
print libsbml.BQB\_IS\_HOMOLOG\_TO      , 5  \# “isHomologTo”,
print libsbml.BQB\_IS\_PART\_OF         , 2  \# “isPartOf”,
print libsbml.BQB\_IS\_PROPERTY\_OF     , 11 \# “isPropertyOf”,
print libsbml.BQB\_IS\_VERSION\_OF      , 3  \# “isVersionOf”,
print libsbml.BQB\_OCCURS\_IN          , 9  \# “occursIn”,
print libsbml.BQB\_UNKNOWN            , 12 \# None

\sphinxAtStartPar
print libsbml.BQM\_IS                 , 0 \# None
print libsbml.BQM\_IS\_DERIVED\_FROM    , 2 \# None
print libsbml.BQM\_IS\_DESCRIBED\_BY    , 1 \# None
print libsbml.BQM\_UNKNOWN            , 3 \# None

\end{fulllineitems}

\index{sbml\_convertCOBRASBMLtoFBC() (in module cbmpy.CBXML)@\spxentry{sbml\_convertCOBRASBMLtoFBC()}\spxextra{in module cbmpy.CBXML}}

\begin{fulllineitems}
\phantomsection\label{\detokenize{modules_doc:cbmpy.CBXML.sbml_convertCOBRASBMLtoFBC}}
\pysigstartsignatures
\pysiglinewithargsret{\sphinxbfcode{\sphinxupquote{sbml\_convertCOBRASBMLtoFBC}}}{\sphinxparam{\DUrole{n,n}{fname}}\sphinxparamcomma \sphinxparam{\DUrole{n,n}{outname}\DUrole{o,o}{=}\DUrole{default_value}{None}}\sphinxparamcomma \sphinxparam{\DUrole{n,n}{work\_dir}\DUrole{o,o}{=}\DUrole{default_value}{None}}\sphinxparamcomma \sphinxparam{\DUrole{n,n}{output\_dir}\DUrole{o,o}{=}\DUrole{default_value}{None}}}{}
\pysigstopsignatures
\sphinxAtStartPar
Read in a COBRA SBML Level 2 file and return the name of the created SBML Level 3 with FBC
file that is created in the output directory
\begin{itemize}
\item {} 
\sphinxAtStartPar
\sphinxstyleemphasis{fname} is the filename

\item {} 
\sphinxAtStartPar
\sphinxstyleemphasis{outname} the name of the output file. If not specified then \textless{}filename\textgreater{}.l3fbc.xml is used as default

\item {} 
\sphinxAtStartPar
\sphinxstyleemphasis{work\_dir} {[}default=None{]} is the working directory

\item {} 
\sphinxAtStartPar
\sphinxstyleemphasis{output\_dir} {[}default=None{]} is the output directory (default is work\_dir)

\end{itemize}

\sphinxAtStartPar
This method is based on code from libSBML (\sphinxurl{http://sbml.org}) in the file “convertCobra.py”
written by Frank T. Bergmann.

\end{fulllineitems}

\index{sbml\_convertSBML3FBCToCOBRA() (in module cbmpy.CBXML)@\spxentry{sbml\_convertSBML3FBCToCOBRA()}\spxextra{in module cbmpy.CBXML}}

\begin{fulllineitems}
\phantomsection\label{\detokenize{modules_doc:cbmpy.CBXML.sbml_convertSBML3FBCToCOBRA}}
\pysigstartsignatures
\pysiglinewithargsret{\sphinxbfcode{\sphinxupquote{sbml\_convertSBML3FBCToCOBRA}}}{\sphinxparam{\DUrole{n,n}{fname}}\sphinxparamcomma \sphinxparam{\DUrole{n,n}{outname}\DUrole{o,o}{=}\DUrole{default_value}{None}}\sphinxparamcomma \sphinxparam{\DUrole{n,n}{work\_dir}\DUrole{o,o}{=}\DUrole{default_value}{None}}\sphinxparamcomma \sphinxparam{\DUrole{n,n}{output\_dir}\DUrole{o,o}{=}\DUrole{default_value}{None}}}{}
\pysigstopsignatures
\sphinxAtStartPar
Read in a SBML Level 3 file and return the name of the created COBRA
file that is created in the output directory
\begin{itemize}
\item {} 
\sphinxAtStartPar
\sphinxstyleemphasis{fname} is the filename

\item {} 
\sphinxAtStartPar
\sphinxstyleemphasis{outname} the name of the output file. If not specified then \textless{}filename\textgreater{}.cobra.xml is used as default

\item {} 
\sphinxAtStartPar
\sphinxstyleemphasis{work\_dir} {[}default=None{]} is the working directory

\item {} 
\sphinxAtStartPar
\sphinxstyleemphasis{output\_dir} {[}default=None{]} is the output directory (default is work\_dir)

\end{itemize}

\sphinxAtStartPar
This method is based on code from libSBML (\sphinxurl{http://sbml.org}) in the file “convertFbcToCobra.py”
written by Frank T. Bergmann.

\end{fulllineitems}

\index{sbml\_createAssociationFromAST() (in module cbmpy.CBXML)@\spxentry{sbml\_createAssociationFromAST()}\spxextra{in module cbmpy.CBXML}}

\begin{fulllineitems}
\phantomsection\label{\detokenize{modules_doc:cbmpy.CBXML.sbml_createAssociationFromAST}}
\pysigstartsignatures
\pysiglinewithargsret{\sphinxbfcode{\sphinxupquote{sbml\_createAssociationFromAST}}}{\sphinxparam{\DUrole{n,n}{node}}\sphinxparamcomma \sphinxparam{\DUrole{n,n}{out}}}{}
\pysigstopsignatures
\sphinxAtStartPar
Converts a GPR string ‘((g1 and g2) or g3)’ to an association via a Python AST.
In future I will get rid of all the string elements and work only with associations
and AST’s.
\begin{itemize}
\item {} 
\sphinxAtStartPar
\sphinxstyleemphasis{node} a Python AST note (e.g. body)

\item {} 
\sphinxAtStartPar
\sphinxstyleemphasis{out} a new shiny FBC V2 GeneProductAssociation

\end{itemize}

\end{fulllineitems}

\index{sbml\_createAssociationFromTreeV2() (in module cbmpy.CBXML)@\spxentry{sbml\_createAssociationFromTreeV2()}\spxextra{in module cbmpy.CBXML}}

\begin{fulllineitems}
\phantomsection\label{\detokenize{modules_doc:cbmpy.CBXML.sbml_createAssociationFromTreeV2}}
\pysigstartsignatures
\pysiglinewithargsret{\sphinxbfcode{\sphinxupquote{sbml\_createAssociationFromTreeV2}}}{\sphinxparam{\DUrole{n,n}{tree}}\sphinxparamcomma \sphinxparam{\DUrole{n,n}{out}}}{}
\pysigstopsignatures
\sphinxAtStartPar
Converts a GPR tree to an association
\begin{itemize}
\item {} 
\sphinxAtStartPar
\sphinxstyleemphasis{tree} a GPR dict tree

\item {} 
\sphinxAtStartPar
\sphinxstyleemphasis{out} a new shiny FBC V2 GeneProductAssociation

\end{itemize}

\end{fulllineitems}

\index{sbml\_createModelL2() (in module cbmpy.CBXML)@\spxentry{sbml\_createModelL2()}\spxextra{in module cbmpy.CBXML}}

\begin{fulllineitems}
\phantomsection\label{\detokenize{modules_doc:cbmpy.CBXML.sbml_createModelL2}}
\pysigstartsignatures
\pysiglinewithargsret{\sphinxbfcode{\sphinxupquote{sbml\_createModelL2}}}{\sphinxparam{\DUrole{n,n}{fba}}\sphinxparamcomma \sphinxparam{\DUrole{n,n}{level}\DUrole{o,o}{=}\DUrole{default_value}{2}}\sphinxparamcomma \sphinxparam{\DUrole{n,n}{version}\DUrole{o,o}{=}\DUrole{default_value}{1}}}{}
\pysigstopsignatures
\sphinxAtStartPar
Create an SBML model and document:
\begin{itemize}
\item {} 
\sphinxAtStartPar
\sphinxstyleemphasis{fba} a PySCeSCBM model instance

\item {} 
\sphinxAtStartPar
\sphinxstyleemphasis{level} always 2

\item {} 
\sphinxAtStartPar
\sphinxstyleemphasis{version} always 1

\end{itemize}

\sphinxAtStartPar
and returns:
\begin{itemize}
\item {} 
\sphinxAtStartPar
\sphinxstyleemphasis{model} an SBML model

\end{itemize}

\end{fulllineitems}

\index{sbml\_exportSBML2FBAModel() (in module cbmpy.CBXML)@\spxentry{sbml\_exportSBML2FBAModel()}\spxextra{in module cbmpy.CBXML}}

\begin{fulllineitems}
\phantomsection\label{\detokenize{modules_doc:cbmpy.CBXML.sbml_exportSBML2FBAModel}}
\pysigstartsignatures
\pysiglinewithargsret{\sphinxbfcode{\sphinxupquote{sbml\_exportSBML2FBAModel}}}{\sphinxparam{\DUrole{n,n}{document}}\sphinxparamcomma \sphinxparam{\DUrole{n,n}{filename}}\sphinxparamcomma \sphinxparam{\DUrole{n,n}{directory}\DUrole{o,o}{=}\DUrole{default_value}{None}}\sphinxparamcomma \sphinxparam{\DUrole{n,n}{return\_doc}\DUrole{o,o}{=}\DUrole{default_value}{False}}\sphinxparamcomma \sphinxparam{\DUrole{n,n}{remove\_note\_body}\DUrole{o,o}{=}\DUrole{default_value}{False}}}{}
\pysigstopsignatures
\sphinxAtStartPar
Writes an SBML model object to file. Note this is an internal SBML method use \sphinxtitleref{sbml\_writeSBML2FBA()} to write an FBA model:
\begin{itemize}
\item {} 
\sphinxAtStartPar
\sphinxstyleemphasis{model} a libSBML model instance

\item {} 
\sphinxAtStartPar
\sphinxstyleemphasis{filename} the output filename

\item {} 
\sphinxAtStartPar
\sphinxstyleemphasis{directory} {[}default=None{]} by default use filename otherwise join, \textless{}dir\textgreater{}\textless{}filename\textgreater{}

\item {} 
\sphinxAtStartPar
\sphinxstyleemphasis{return\_doc} {[}default=False{]} return the SBML document used to write the XML

\end{itemize}

\end{fulllineitems}

\index{sbml\_fileFindVersion() (in module cbmpy.CBXML)@\spxentry{sbml\_fileFindVersion()}\spxextra{in module cbmpy.CBXML}}

\begin{fulllineitems}
\phantomsection\label{\detokenize{modules_doc:cbmpy.CBXML.sbml_fileFindVersion}}
\pysigstartsignatures
\pysiglinewithargsret{\sphinxbfcode{\sphinxupquote{sbml\_fileFindVersion}}}{\sphinxparam{\DUrole{n,n}{f}}}{}
\pysigstopsignatures
\sphinxAtStartPar
Try and find the SBML version and FBC support
\begin{itemize}
\item {} 
\sphinxAtStartPar
\sphinxstyleemphasis{f} the SBML file

\end{itemize}

\end{fulllineitems}

\index{sbml\_fileValidate() (in module cbmpy.CBXML)@\spxentry{sbml\_fileValidate()}\spxextra{in module cbmpy.CBXML}}

\begin{fulllineitems}
\phantomsection\label{\detokenize{modules_doc:cbmpy.CBXML.sbml_fileValidate}}
\pysigstartsignatures
\pysiglinewithargsret{\sphinxbfcode{\sphinxupquote{sbml\_fileValidate}}}{\sphinxparam{\DUrole{n,n}{f}}\sphinxparamcomma \sphinxparam{\DUrole{n,n}{level}\DUrole{o,o}{=}\DUrole{default_value}{\textquotesingle{}normal\textquotesingle{}}}}{}
\pysigstopsignatures
\sphinxAtStartPar
Validate an SBML file and model
\begin{itemize}
\item {} 
\sphinxAtStartPar
\sphinxstyleemphasis{f} the SBML file

\item {} 
\sphinxAtStartPar
\sphinxstyleemphasis{level} {[}default=’normal’{]} the level of validation “normal” or “full”

\end{itemize}

\end{fulllineitems}

\index{sbml\_getCVterms() (in module cbmpy.CBXML)@\spxentry{sbml\_getCVterms()}\spxextra{in module cbmpy.CBXML}}

\begin{fulllineitems}
\phantomsection\label{\detokenize{modules_doc:cbmpy.CBXML.sbml_getCVterms}}
\pysigstartsignatures
\pysiglinewithargsret{\sphinxbfcode{\sphinxupquote{sbml\_getCVterms}}}{\sphinxparam{\DUrole{n,n}{sb}}\sphinxparamcomma \sphinxparam{\DUrole{n,n}{model}\DUrole{o,o}{=}\DUrole{default_value}{False}}}{}
\pysigstopsignatures
\sphinxAtStartPar
Get the MIRIAM compliant CV terms and return a MIRIAMAnnotation or None
\begin{itemize}
\item {} 
\sphinxAtStartPar
\sphinxstyleemphasis{sb} a libSBML SBase derived object

\item {} 
\sphinxAtStartPar
\sphinxstyleemphasis{model} is this a BQmodel term

\end{itemize}

\end{fulllineitems}

\index{sbml\_getGPRasDictFBCv1() (in module cbmpy.CBXML)@\spxentry{sbml\_getGPRasDictFBCv1()}\spxextra{in module cbmpy.CBXML}}

\begin{fulllineitems}
\phantomsection\label{\detokenize{modules_doc:cbmpy.CBXML.sbml_getGPRasDictFBCv1}}
\pysigstartsignatures
\pysiglinewithargsret{\sphinxbfcode{\sphinxupquote{sbml\_getGPRasDictFBCv1}}}{\sphinxparam{\DUrole{n,n}{node}}\sphinxparamcomma \sphinxparam{\DUrole{n,n}{out}}}{}
\pysigstopsignatures
\sphinxAtStartPar
Converts a GPR string ‘((g1 and g2) or g3)’ to a gprDict which is returned
\begin{itemize}
\item {} 
\sphinxAtStartPar
\sphinxstyleemphasis{node} a Python AST note (e.g. \sphinxtitleref{ast.parse(gprstring).body{[}0{]}})

\item {} 
\sphinxAtStartPar
\sphinxstyleemphasis{out} a new dictionary that will be be created in place

\end{itemize}

\end{fulllineitems}

\index{sbml\_getGPRasDictFBCv2() (in module cbmpy.CBXML)@\spxentry{sbml\_getGPRasDictFBCv2()}\spxextra{in module cbmpy.CBXML}}

\begin{fulllineitems}
\phantomsection\label{\detokenize{modules_doc:cbmpy.CBXML.sbml_getGPRasDictFBCv2}}
\pysigstartsignatures
\pysiglinewithargsret{\sphinxbfcode{\sphinxupquote{sbml\_getGPRasDictFBCv2}}}{\sphinxparam{\DUrole{n,n}{association}}\sphinxparamcomma \sphinxparam{\DUrole{n,n}{out}}\sphinxparamcomma \sphinxparam{\DUrole{n,n}{cntr}}}{}
\pysigstopsignatures
\sphinxAtStartPar
Walk through an SBML L3FBCV2 gene protein association and return a dictionary/tree representation

\end{fulllineitems}

\index{sbml\_getGeneRefs() (in module cbmpy.CBXML)@\spxentry{sbml\_getGeneRefs()}\spxextra{in module cbmpy.CBXML}}

\begin{fulllineitems}
\phantomsection\label{\detokenize{modules_doc:cbmpy.CBXML.sbml_getGeneRefs}}
\pysigstartsignatures
\pysiglinewithargsret{\sphinxbfcode{\sphinxupquote{sbml\_getGeneRefs}}}{\sphinxparam{\DUrole{n,n}{association}}\sphinxparamcomma \sphinxparam{\DUrole{n,n}{out}}}{}
\pysigstopsignatures
\sphinxAtStartPar
Walk through a gene association and extract GeneRefs inspired by Frank

\end{fulllineitems}

\index{sbml\_getNotes() (in module cbmpy.CBXML)@\spxentry{sbml\_getNotes()}\spxextra{in module cbmpy.CBXML}}

\begin{fulllineitems}
\phantomsection\label{\detokenize{modules_doc:cbmpy.CBXML.sbml_getNotes}}
\pysigstartsignatures
\pysiglinewithargsret{\sphinxbfcode{\sphinxupquote{sbml\_getNotes}}}{\sphinxparam{\DUrole{n,n}{obj}}}{}
\pysigstopsignatures
\sphinxAtStartPar
Returns the SBML objects notes
\begin{itemize}
\item {} 
\sphinxAtStartPar
\sphinxstyleemphasis{obj} an SBML object

\end{itemize}

\end{fulllineitems}

\index{sbml\_readCOBRANote() (in module cbmpy.CBXML)@\spxentry{sbml\_readCOBRANote()}\spxextra{in module cbmpy.CBXML}}

\begin{fulllineitems}
\phantomsection\label{\detokenize{modules_doc:cbmpy.CBXML.sbml_readCOBRANote}}
\pysigstartsignatures
\pysiglinewithargsret{\sphinxbfcode{\sphinxupquote{sbml\_readCOBRANote}}}{\sphinxparam{\DUrole{n,n}{s}}}{}
\pysigstopsignatures
\sphinxAtStartPar
Parses a COBRA style note from a XML string
\begin{itemize}
\item {} 
\sphinxAtStartPar
\sphinxstyleemphasis{s} an XML string

\end{itemize}

\end{fulllineitems}

\index{sbml\_readCOBRASBML() (in module cbmpy.CBXML)@\spxentry{sbml\_readCOBRASBML()}\spxextra{in module cbmpy.CBXML}}

\begin{fulllineitems}
\phantomsection\label{\detokenize{modules_doc:cbmpy.CBXML.sbml_readCOBRASBML}}
\pysigstartsignatures
\pysiglinewithargsret{\sphinxbfcode{\sphinxupquote{sbml\_readCOBRASBML}}}{\sphinxparam{\DUrole{n,n}{fname}}\sphinxparamcomma \sphinxparam{\DUrole{n,n}{work\_dir}\DUrole{o,o}{=}\DUrole{default_value}{None}}\sphinxparamcomma \sphinxparam{\DUrole{n,n}{return\_sbml\_model}\DUrole{o,o}{=}\DUrole{default_value}{False}}\sphinxparamcomma \sphinxparam{\DUrole{n,n}{delete\_intermediate}\DUrole{o,o}{=}\DUrole{default_value}{False}}\sphinxparamcomma \sphinxparam{\DUrole{n,n}{fake\_boundary\_species\_search}\DUrole{o,o}{=}\DUrole{default_value}{False}}\sphinxparamcomma \sphinxparam{\DUrole{n,n}{output\_dir}\DUrole{o,o}{=}\DUrole{default_value}{None}}\sphinxparamcomma \sphinxparam{\DUrole{n,n}{speciesAnnotationFix}\DUrole{o,o}{=}\DUrole{default_value}{True}}\sphinxparamcomma \sphinxparam{\DUrole{n,n}{skip\_genes}\DUrole{o,o}{=}\DUrole{default_value}{False}}}{}
\pysigstopsignatures
\sphinxAtStartPar
Read in a COBRA format SBML Level 2 file with FBA annotation where and return either a CBM model object
or a (cbm\_mod, sbml\_mod) pair if return\_sbml\_model=True
\begin{itemize}
\item {} 
\sphinxAtStartPar
\sphinxstyleemphasis{fname} is the filename

\item {} 
\sphinxAtStartPar
\sphinxstyleemphasis{work\_dir} is the working directory

\item {} 
\sphinxAtStartPar
\sphinxstyleemphasis{return\_sbml\_model} {[}default=False{]} return a a (cbm\_mod, sbml\_mod) pair

\item {} 
\sphinxAtStartPar
\sphinxstyleemphasis{delete\_intermediate} {[}default=False{]} delete the intermediate SBML Level 3 FBC file

\item {} 
\sphinxAtStartPar
\sphinxstyleemphasis{fake\_boundary\_species\_search} {[}default=False{]} after looking for the boundary\_condition of a species search for overloaded id’s \textless{}id\textgreater{}\_b

\item {} 
\sphinxAtStartPar
\sphinxstyleemphasis{output\_dir} {[}default=None{]} the directory to output the intermediate SBML L3 files (if generated) default to input directory

\item {} 
\sphinxAtStartPar
\sphinxstyleemphasis{speciesAnnotationFix} {[}default=True{]}

\item {} 
\sphinxAtStartPar
\sphinxstyleemphasis{skip\_genes} {[}default=False{]} convert GPR associations

\end{itemize}

\end{fulllineitems}

\index{sbml\_readFBCv3KeyValuePairs() (in module cbmpy.CBXML)@\spxentry{sbml\_readFBCv3KeyValuePairs()}\spxextra{in module cbmpy.CBXML}}

\begin{fulllineitems}
\phantomsection\label{\detokenize{modules_doc:cbmpy.CBXML.sbml_readFBCv3KeyValuePairs}}
\pysigstartsignatures
\pysiglinewithargsret{\sphinxbfcode{\sphinxupquote{sbml\_readFBCv3KeyValuePairs}}}{\sphinxparam{\DUrole{n,n}{fbcp}}}{}
\pysigstopsignatures
\sphinxAtStartPar
Reads FBCv3 KeyValue pair annotation and returns a dictionary of key:value pairs
\begin{itemize}
\item {} 
\sphinxAtStartPar
\sphinxstyleemphasis{fbcp} an FBC plugin

\end{itemize}

\end{fulllineitems}

\index{sbml\_readKeyValueDataAnnotation() (in module cbmpy.CBXML)@\spxentry{sbml\_readKeyValueDataAnnotation()}\spxextra{in module cbmpy.CBXML}}

\begin{fulllineitems}
\phantomsection\label{\detokenize{modules_doc:cbmpy.CBXML.sbml_readKeyValueDataAnnotation}}
\pysigstartsignatures
\pysiglinewithargsret{\sphinxbfcode{\sphinxupquote{sbml\_readKeyValueDataAnnotation}}}{\sphinxparam{\DUrole{n,n}{annotations}}}{}
\pysigstopsignatures
\sphinxAtStartPar
Reads KeyValueData annotation (\sphinxurl{http://pysces.sourceforge.net/KeyValueData}) and returns a dictionary of key:value pairs

\end{fulllineitems}

\index{sbml\_readSBML2FBA() (in module cbmpy.CBXML)@\spxentry{sbml\_readSBML2FBA()}\spxextra{in module cbmpy.CBXML}}

\begin{fulllineitems}
\phantomsection\label{\detokenize{modules_doc:cbmpy.CBXML.sbml_readSBML2FBA}}
\pysigstartsignatures
\pysiglinewithargsret{\sphinxbfcode{\sphinxupquote{sbml\_readSBML2FBA}}}{\sphinxparam{\DUrole{n,n}{fname}}\sphinxparamcomma \sphinxparam{\DUrole{n,n}{work\_dir}\DUrole{o,o}{=}\DUrole{default_value}{None}}\sphinxparamcomma \sphinxparam{\DUrole{n,n}{return\_sbml\_model}\DUrole{o,o}{=}\DUrole{default_value}{False}}\sphinxparamcomma \sphinxparam{\DUrole{n,n}{fake\_boundary\_species\_search}\DUrole{o,o}{=}\DUrole{default_value}{False}}}{}
\pysigstopsignatures
\sphinxAtStartPar
Read in an SBML Level 2 file with FBA annotation where and return either a CBM model object
or a (cbm\_mod, sbml\_mod) pair if return\_sbml\_model=True
\begin{itemize}
\item {} 
\sphinxAtStartPar
\sphinxstyleemphasis{fname} is the filename

\item {} 
\sphinxAtStartPar
\sphinxstyleemphasis{work\_dir} is the working directory (only used if not None)

\item {} 
\sphinxAtStartPar
\sphinxstyleemphasis{return\_sbml\_model} {[}default=False{]} return a a (cbm\_mod, sbml\_mod) pair

\item {} 
\sphinxAtStartPar
\sphinxstyleemphasis{fake\_boundary\_species\_search} {[}default=False{]} after looking for the boundary\_condition of a species search for overloaded id’s \textless{}id\textgreater{}\_b

\end{itemize}

\end{fulllineitems}

\index{sbml\_readSBML3FBC() (in module cbmpy.CBXML)@\spxentry{sbml\_readSBML3FBC()}\spxextra{in module cbmpy.CBXML}}

\begin{fulllineitems}
\phantomsection\label{\detokenize{modules_doc:cbmpy.CBXML.sbml_readSBML3FBC}}
\pysigstartsignatures
\pysiglinewithargsret{\sphinxbfcode{\sphinxupquote{sbml\_readSBML3FBC}}}{\sphinxparam{\DUrole{n,n}{fname}}\sphinxparamcomma \sphinxparam{\DUrole{n,n}{work\_dir}\DUrole{o,o}{=}\DUrole{default_value}{None}}\sphinxparamcomma \sphinxparam{\DUrole{n,n}{return\_sbml\_model}\DUrole{o,o}{=}\DUrole{default_value}{False}}\sphinxparamcomma \sphinxparam{\DUrole{n,n}{xoptions}\DUrole{o,o}{=}\DUrole{default_value}{\{\}}}}{}
\pysigstopsignatures
\sphinxAtStartPar
Read in an SBML Level 3 file with FBC annotation where and return either a CBM model object
or a (cbm\_mod, sbml\_mod) pair if return\_sbml\_model=True
\begin{itemize}
\item {} 
\sphinxAtStartPar
\sphinxstyleemphasis{fname} is the filename

\item {} 
\sphinxAtStartPar
\sphinxstyleemphasis{work\_dir} is the working directory

\item {} 
\sphinxAtStartPar
\sphinxstyleemphasis{return\_sbml\_model} {[}default=False{]} return a a (cbm\_mod, sbml\_mod) pair

\item {} 
\sphinxAtStartPar
\sphinxstyleemphasis{xoptions} special load options enable with option = True
\begin{itemize}
\item {} 
\sphinxAtStartPar
\sphinxstyleemphasis{nogenes} do not load/process genes

\item {} 
\sphinxAtStartPar
\sphinxstyleemphasis{noannot} do not load/process any annotations

\item {} 
\sphinxAtStartPar
\sphinxstyleemphasis{validate} validate model and display errors and warnings before loading

\item {} 
\sphinxAtStartPar
\sphinxstyleemphasis{readcobra} read the cobra annotation

\item {} 
\sphinxAtStartPar
\sphinxstyleemphasis{read\_model\_string} {[}default=False{]} read the model from a string (instead of a filename) containing an SBML document

\item {} 
\sphinxAtStartPar
\sphinxstyleemphasis{nmatrix\_type} {[}default=’normal’{]} define the type of stoichiometrich matrix to be built

\item {} 
\sphinxAtStartPar
\sphinxstyleemphasis{model\_extension\_class} extend CBModel class with new class (experimental, Python 3 only)

\item {} 
\sphinxAtStartPar
\sphinxstyleemphasis{model\_metaclass} add a custom metaclass to CBModel (experimental, Python 3 only)
\begin{itemize}
\item {} 
\sphinxAtStartPar
‘numpy’ dense numpy array (best performance)

\item {} 
\sphinxAtStartPar
‘scipy\_csr’ scipy sparse matrix (lower performance, low memory)

\item {} 
\sphinxAtStartPar
‘sympy’ a sympy rational matrix (low performance, high memory, cast to dense to analyse)

\item {} 
\sphinxAtStartPar
None do not build matrix

\end{itemize}

\end{itemize}

\end{itemize}

\end{fulllineitems}

\index{sbml\_setAnnotationsL3Fbc() (in module cbmpy.CBXML)@\spxentry{sbml\_setAnnotationsL3Fbc()}\spxextra{in module cbmpy.CBXML}}

\begin{fulllineitems}
\phantomsection\label{\detokenize{modules_doc:cbmpy.CBXML.sbml_setAnnotationsL3Fbc}}
\pysigstartsignatures
\pysiglinewithargsret{\sphinxbfcode{\sphinxupquote{sbml\_setAnnotationsL3Fbc}}}{\sphinxparam{\DUrole{n,n}{cbmo}}\sphinxparamcomma \sphinxparam{\DUrole{n,n}{sbmlo}}}{}
\pysigstopsignatures
\sphinxAtStartPar
Add CBMPy Fbase annotations to an SBML object, MIRIAM, SBO, Notes. Should
be called last when creating SBML objects.
\begin{quote}
\begin{itemize}
\item {} 
\sphinxAtStartPar
\sphinxstyleemphasis{cbmo} the CBMPy object

\item {} 
\sphinxAtStartPar
\sphinxstyleemphasis{sbmlo} SBML object

\end{itemize}

\sphinxAtStartPar
Note: this function should be used for new code, old code still needs to be
refactored.
\end{quote}

\end{fulllineitems}

\index{sbml\_setCVterms() (in module cbmpy.CBXML)@\spxentry{sbml\_setCVterms()}\spxextra{in module cbmpy.CBXML}}

\begin{fulllineitems}
\phantomsection\label{\detokenize{modules_doc:cbmpy.CBXML.sbml_setCVterms}}
\pysigstartsignatures
\pysiglinewithargsret{\sphinxbfcode{\sphinxupquote{sbml\_setCVterms}}}{\sphinxparam{\DUrole{n,n}{sb}}\sphinxparamcomma \sphinxparam{\DUrole{n,n}{uridict}}\sphinxparamcomma \sphinxparam{\DUrole{n,n}{model}\DUrole{o,o}{=}\DUrole{default_value}{False}}}{}
\pysigstopsignatures
\sphinxAtStartPar
Add MIRIAM compliant CV terms to a sbml object from a CBM object
\begin{itemize}
\item {} 
\sphinxAtStartPar
\sphinxstyleemphasis{sb} a libSBML SBase derived object

\item {} 
\sphinxAtStartPar
\sphinxstyleemphasis{uridict} a dictionary of uri’s as produced by getAllMIRIAMUris()

\item {} 
\sphinxAtStartPar
\sphinxstyleemphasis{model} is this a BQmodel term {[}deprecated attribute, ignored and autodetected{]}

\end{itemize}

\end{fulllineitems}

\index{sbml\_setCompartmentsL3() (in module cbmpy.CBXML)@\spxentry{sbml\_setCompartmentsL3()}\spxextra{in module cbmpy.CBXML}}

\begin{fulllineitems}
\phantomsection\label{\detokenize{modules_doc:cbmpy.CBXML.sbml_setCompartmentsL3}}
\pysigstartsignatures
\pysiglinewithargsret{\sphinxbfcode{\sphinxupquote{sbml\_setCompartmentsL3}}}{\sphinxparam{\DUrole{n,n}{model}}\sphinxparamcomma \sphinxparam{\DUrole{n,n}{fba}}}{}
\pysigstopsignatures
\sphinxAtStartPar
Sets the model compartments.
\begin{itemize}
\item {} 
\sphinxAtStartPar
\sphinxstyleemphasis{model} a libSBML model instance

\item {} 
\sphinxAtStartPar
\sphinxstyleemphasis{fba} a PySCeSCBM model instance

\end{itemize}

\end{fulllineitems}

\index{sbml\_setDescription() (in module cbmpy.CBXML)@\spxentry{sbml\_setDescription()}\spxextra{in module cbmpy.CBXML}}

\begin{fulllineitems}
\phantomsection\label{\detokenize{modules_doc:cbmpy.CBXML.sbml_setDescription}}
\pysigstartsignatures
\pysiglinewithargsret{\sphinxbfcode{\sphinxupquote{sbml\_setDescription}}}{\sphinxparam{\DUrole{n,n}{model}}\sphinxparamcomma \sphinxparam{\DUrole{n,n}{fba}}}{}
\pysigstopsignatures
\sphinxAtStartPar
Sets the model description as a \textless{}note\textgreater{} containing \sphinxtitleref{txt} in an HTML paragraph on the model object.
\begin{itemize}
\item {} 
\sphinxAtStartPar
\sphinxstyleemphasis{model} a libSBML model instance

\item {} 
\sphinxAtStartPar
\sphinxstyleemphasis{fba} a PySCeSCBM model instance

\end{itemize}

\end{fulllineitems}

\index{sbml\_setFBCv3KeyValuePairs() (in module cbmpy.CBXML)@\spxentry{sbml\_setFBCv3KeyValuePairs()}\spxextra{in module cbmpy.CBXML}}

\begin{fulllineitems}
\phantomsection\label{\detokenize{modules_doc:cbmpy.CBXML.sbml_setFBCv3KeyValuePairs}}
\pysigstartsignatures
\pysiglinewithargsret{\sphinxbfcode{\sphinxupquote{sbml\_setFBCv3KeyValuePairs}}}{\sphinxparam{\DUrole{n,n}{fbcp}}\sphinxparamcomma \sphinxparam{\DUrole{n,n}{kv\_pairs}}}{}
\pysigstopsignatures
\sphinxAtStartPar
Adds key value pairs to any FBCv3 SBase derived object
\begin{itemize}
\item {} 
\sphinxAtStartPar
\sphinxstyleemphasis{fbcp} an FBCv3 plugin

\item {} 
\sphinxAtStartPar
\sphinxstyleemphasis{kv\_pairs} a dictionary of CBMPy Key Value pairs (format will be upgraded over next few versions)

\end{itemize}

\end{fulllineitems}

\index{sbml\_setGroupsL3() (in module cbmpy.CBXML)@\spxentry{sbml\_setGroupsL3()}\spxextra{in module cbmpy.CBXML}}

\begin{fulllineitems}
\phantomsection\label{\detokenize{modules_doc:cbmpy.CBXML.sbml_setGroupsL3}}
\pysigstartsignatures
\pysiglinewithargsret{\sphinxbfcode{\sphinxupquote{sbml\_setGroupsL3}}}{\sphinxparam{\DUrole{n,n}{cs}}\sphinxparamcomma \sphinxparam{\DUrole{n,n}{fba}}}{}
\pysigstopsignatures
\sphinxAtStartPar
add groups to the SBML model
\begin{itemize}
\item {} 
\sphinxAtStartPar
\sphinxstyleemphasis{cs} a CBMLtoSBML instance

\item {} 
\sphinxAtStartPar
\sphinxstyleemphasis{fba} a CBMPy model instance

\end{itemize}

\end{fulllineitems}

\index{sbml\_setNotes3() (in module cbmpy.CBXML)@\spxentry{sbml\_setNotes3()}\spxextra{in module cbmpy.CBXML}}

\begin{fulllineitems}
\phantomsection\label{\detokenize{modules_doc:cbmpy.CBXML.sbml_setNotes3}}
\pysigstartsignatures
\pysiglinewithargsret{\sphinxbfcode{\sphinxupquote{sbml\_setNotes3}}}{\sphinxparam{\DUrole{n,n}{obj}}\sphinxparamcomma \sphinxparam{\DUrole{n,n}{s}}}{}
\pysigstopsignatures
\sphinxAtStartPar
Formats the CBMPy notes as an SBML note and adds it to the SBMl object
\begin{itemize}
\item {} 
\sphinxAtStartPar
\sphinxstyleemphasis{obj} an SBML object

\item {} 
\sphinxAtStartPar
\sphinxstyleemphasis{s} a string that should be added as a note

\end{itemize}

\end{fulllineitems}

\index{sbml\_setParametersL3Fbc() (in module cbmpy.CBXML)@\spxentry{sbml\_setParametersL3Fbc()}\spxextra{in module cbmpy.CBXML}}

\begin{fulllineitems}
\phantomsection\label{\detokenize{modules_doc:cbmpy.CBXML.sbml_setParametersL3Fbc}}
\pysigstartsignatures
\pysiglinewithargsret{\sphinxbfcode{\sphinxupquote{sbml\_setParametersL3Fbc}}}{\sphinxparam{\DUrole{n,n}{fbcmod}}\sphinxparamcomma \sphinxparam{\DUrole{n,n}{add\_cbmpy\_anno}\DUrole{o,o}{=}\DUrole{default_value}{True}}\sphinxparamcomma \sphinxparam{\DUrole{n,n}{fbc\_version}\DUrole{o,o}{=}\DUrole{default_value}{2}}}{}
\pysigstopsignatures
\sphinxAtStartPar
Add non fluxbound related parameters to the model
\begin{itemize}
\item {} 
\sphinxAtStartPar
\sphinxstyleemphasis{fbcmod} a CBM2SBML instance

\item {} 
\sphinxAtStartPar
\sphinxstyleemphasis{add\_cbmpy\_anno} {[}default=True{]} add CBMPy KeyValueData annotation.

\end{itemize}

\end{fulllineitems}

\index{sbml\_setReactionsL2() (in module cbmpy.CBXML)@\spxentry{sbml\_setReactionsL2()}\spxextra{in module cbmpy.CBXML}}

\begin{fulllineitems}
\phantomsection\label{\detokenize{modules_doc:cbmpy.CBXML.sbml_setReactionsL2}}
\pysigstartsignatures
\pysiglinewithargsret{\sphinxbfcode{\sphinxupquote{sbml\_setReactionsL2}}}{\sphinxparam{\DUrole{n,n}{model}}\sphinxparamcomma \sphinxparam{\DUrole{n,n}{fba}}\sphinxparamcomma \sphinxparam{\DUrole{n,n}{return\_dict}\DUrole{o,o}{=}\DUrole{default_value}{False}}}{}
\pysigstopsignatures
\sphinxAtStartPar
Add the FBA instance reactions to the SBML model
\begin{itemize}
\item {} 
\sphinxAtStartPar
\sphinxstyleemphasis{model} an SBML model instance

\item {} 
\sphinxAtStartPar
\sphinxstyleemphasis{fba} a PySCeSCBM model instance

\item {} 
\sphinxAtStartPar
\sphinxstyleemphasis{return\_dict} {[}default=False{]} if True do not add reactions to SBML document instead return a dictionary description of the reactions

\end{itemize}

\end{fulllineitems}

\index{sbml\_setReactionsL3Fbc() (in module cbmpy.CBXML)@\spxentry{sbml\_setReactionsL3Fbc()}\spxextra{in module cbmpy.CBXML}}

\begin{fulllineitems}
\phantomsection\label{\detokenize{modules_doc:cbmpy.CBXML.sbml_setReactionsL3Fbc}}
\pysigstartsignatures
\pysiglinewithargsret{\sphinxbfcode{\sphinxupquote{sbml\_setReactionsL3Fbc}}}{\sphinxparam{\DUrole{n,n}{fbcmod}}\sphinxparamcomma \sphinxparam{\DUrole{n,n}{return\_dict}\DUrole{o,o}{=}\DUrole{default_value}{False}}\sphinxparamcomma \sphinxparam{\DUrole{n,n}{add\_cobra\_anno}\DUrole{o,o}{=}\DUrole{default_value}{False}}\sphinxparamcomma \sphinxparam{\DUrole{n,n}{add\_cbmpy\_anno}\DUrole{o,o}{=}\DUrole{default_value}{True}}\sphinxparamcomma \sphinxparam{\DUrole{n,n}{fbc\_version}\DUrole{o,o}{=}\DUrole{default_value}{2}}}{}
\pysigstopsignatures
\sphinxAtStartPar
Add the FBA instance reactions to the SBML model
\begin{itemize}
\item {} 
\sphinxAtStartPar
\sphinxstyleemphasis{fbcmod} a CBM2SBML instance

\item {} 
\sphinxAtStartPar
\sphinxstyleemphasis{return\_dict} {[}default=False{]} if True do not add reactions to SBML document instead return a dictionary description of the reactions

\item {} 
\sphinxAtStartPar
\sphinxstyleemphasis{add\_cbmpy\_anno} {[}default=True{]} add CBMPy KeyValueData annotation. Replaces \textless{}notes\textgreater{}

\item {} 
\sphinxAtStartPar
\sphinxstyleemphasis{add\_cobra\_anno} {[}default=False{]} add COBRA \textless{}notes\textgreater{} annotation

\item {} 
\sphinxAtStartPar
\sphinxstyleemphasis{fbc\_version} {[}default=2{]} writes either FBC v1 (2013) or v2 (2015) or v3 (2023)

\end{itemize}

\end{fulllineitems}

\index{sbml\_setSpeciesL2() (in module cbmpy.CBXML)@\spxentry{sbml\_setSpeciesL2()}\spxextra{in module cbmpy.CBXML}}

\begin{fulllineitems}
\phantomsection\label{\detokenize{modules_doc:cbmpy.CBXML.sbml_setSpeciesL2}}
\pysigstartsignatures
\pysiglinewithargsret{\sphinxbfcode{\sphinxupquote{sbml\_setSpeciesL2}}}{\sphinxparam{\DUrole{n,n}{model}}\sphinxparamcomma \sphinxparam{\DUrole{n,n}{fba}}\sphinxparamcomma \sphinxparam{\DUrole{n,n}{return\_dicts}\DUrole{o,o}{=}\DUrole{default_value}{False}}}{}
\pysigstopsignatures
\sphinxAtStartPar
Add the species definitions to the SBML object:
\begin{itemize}
\item {} 
\sphinxAtStartPar
\sphinxstyleemphasis{model} {[}default=’’{]} a libSBML model instance or can be None if \sphinxstyleemphasis{return\_dicts} == True

\item {} 
\sphinxAtStartPar
\sphinxstyleemphasis{fba} a PySCeSCBM model instance

\item {} 
\sphinxAtStartPar
\sphinxstyleemphasis{return\_dicts} {[}default=False{]} only returns the compartment and species dictionaries without updated the SBML

\end{itemize}

\sphinxAtStartPar
returns:
\begin{itemize}
\item {} 
\sphinxAtStartPar
\sphinxstyleemphasis{compartments} a dictionary of compartments (except when give \sphinxstyleemphasis{return\_dicts} argument)

\end{itemize}

\end{fulllineitems}

\index{sbml\_setSpeciesL3() (in module cbmpy.CBXML)@\spxentry{sbml\_setSpeciesL3()}\spxextra{in module cbmpy.CBXML}}

\begin{fulllineitems}
\phantomsection\label{\detokenize{modules_doc:cbmpy.CBXML.sbml_setSpeciesL3}}
\pysigstartsignatures
\pysiglinewithargsret{\sphinxbfcode{\sphinxupquote{sbml\_setSpeciesL3}}}{\sphinxparam{\DUrole{n,n}{model}}\sphinxparamcomma \sphinxparam{\DUrole{n,n}{fba}}\sphinxparamcomma \sphinxparam{\DUrole{n,n}{return\_dicts}\DUrole{o,o}{=}\DUrole{default_value}{False}}\sphinxparamcomma \sphinxparam{\DUrole{n,n}{add\_cobra\_anno}\DUrole{o,o}{=}\DUrole{default_value}{False}}\sphinxparamcomma \sphinxparam{\DUrole{n,n}{add\_cbmpy\_anno}\DUrole{o,o}{=}\DUrole{default_value}{True}}\sphinxparamcomma \sphinxparam{\DUrole{n,n}{substance\_units}\DUrole{o,o}{=}\DUrole{default_value}{True}}\sphinxparamcomma \sphinxparam{\DUrole{n,n}{fbc\_version}\DUrole{o,o}{=}\DUrole{default_value}{2}}}{}
\pysigstopsignatures
\sphinxAtStartPar
Add the species definitions to the SBML object:
\begin{itemize}
\item {} 
\sphinxAtStartPar
\sphinxstyleemphasis{model} and SBML model instance or can be None if \sphinxstyleemphasis{return\_dicts} == True

\item {} 
\sphinxAtStartPar
\sphinxstyleemphasis{fba} a PySCeSCBM model instance

\item {} 
\sphinxAtStartPar
\sphinxstyleemphasis{return\_dicts} {[}default=False{]} only returns the compartment and species dictionaries without updating the SBML

\item {} 
\sphinxAtStartPar
\sphinxstyleemphasis{add\_cbmpy\_anno} {[}default=True{]} add CBMPy KeyValueData annotation. Replaces \textless{}notes\textgreater{}

\item {} 
\sphinxAtStartPar
\sphinxstyleemphasis{add\_cobra\_anno} {[}default=False{]} add COBRA \textless{}notes\textgreater{} annotation

\item {} 
\sphinxAtStartPar
\sphinxstyleemphasis{substance\_units} {[}default=True{]} defines the species in amounts rather than concentrations (necessary for default mmol/gdw.h)

\item {} 
\sphinxAtStartPar
\sphinxstyleemphasis{fbc\_version} {[}default=2{]} the FBC version to use

\end{itemize}

\sphinxAtStartPar
returns:
\begin{itemize}
\item {} 
\sphinxAtStartPar
\sphinxstyleemphasis{compartments} a dictionary of compartments (except when given \sphinxstyleemphasis{return\_dicts} argument)

\end{itemize}

\end{fulllineitems}

\index{sbml\_setUnits() (in module cbmpy.CBXML)@\spxentry{sbml\_setUnits()}\spxextra{in module cbmpy.CBXML}}

\begin{fulllineitems}
\phantomsection\label{\detokenize{modules_doc:cbmpy.CBXML.sbml_setUnits}}
\pysigstartsignatures
\pysiglinewithargsret{\sphinxbfcode{\sphinxupquote{sbml\_setUnits}}}{\sphinxparam{\DUrole{n,n}{model}}\sphinxparamcomma \sphinxparam{\DUrole{n,n}{units}\DUrole{o,o}{=}\DUrole{default_value}{None}}\sphinxparamcomma \sphinxparam{\DUrole{n,n}{give\_default}\DUrole{o,o}{=}\DUrole{default_value}{False}}\sphinxparamcomma \sphinxparam{\DUrole{n,n}{L3}\DUrole{o,o}{=}\DUrole{default_value}{True}}}{}
\pysigstopsignatures
\sphinxAtStartPar
Adds units to the model:
\begin{itemize}
\item {} 
\sphinxAtStartPar
\sphinxstyleemphasis{model} a libSBML model instance

\item {} 
\sphinxAtStartPar
\sphinxstyleemphasis{units} {[}default=None{]} a dictionary of units, if None default units are used

\item {} 
\sphinxAtStartPar
\sphinxstyleemphasis{give\_default} {[}default=False{]} if true method returns the default unit dictionary

\item {} 
\sphinxAtStartPar
\sphinxstyleemphasis{L3} {[}default=True{]} use the L3 defaults

\end{itemize}

\end{fulllineitems}

\index{sbml\_setValidationOptions() (in module cbmpy.CBXML)@\spxentry{sbml\_setValidationOptions()}\spxextra{in module cbmpy.CBXML}}

\begin{fulllineitems}
\phantomsection\label{\detokenize{modules_doc:cbmpy.CBXML.sbml_setValidationOptions}}
\pysigstartsignatures
\pysiglinewithargsret{\sphinxbfcode{\sphinxupquote{sbml\_setValidationOptions}}}{\sphinxparam{\DUrole{n,n}{D}}\sphinxparamcomma \sphinxparam{\DUrole{n,n}{level}}}{}
\pysigstopsignatures
\sphinxAtStartPar
set the validation level of an SBML document
\begin{quote}
\begin{itemize}
\item {} 
\sphinxAtStartPar
\sphinxstyleemphasis{D} an SBML document

\item {} 
\sphinxAtStartPar
\sphinxstyleemphasis{level} the level of consistency check can be either one of:

\end{itemize}
\begin{itemize}
\item {} 
\sphinxAtStartPar
‘normal’ basic id checking only

\item {} 
\sphinxAtStartPar
‘full’ all checks enabled

\item {} 
\sphinxAtStartPar
None disable all validation

\end{itemize}
\end{quote}

\end{fulllineitems}

\index{sbml\_validateDocument() (in module cbmpy.CBXML)@\spxentry{sbml\_validateDocument()}\spxextra{in module cbmpy.CBXML}}

\begin{fulllineitems}
\phantomsection\label{\detokenize{modules_doc:cbmpy.CBXML.sbml_validateDocument}}
\pysigstartsignatures
\pysiglinewithargsret{\sphinxbfcode{\sphinxupquote{sbml\_validateDocument}}}{\sphinxparam{\DUrole{n,n}{D}}\sphinxparamcomma \sphinxparam{\DUrole{n,n}{fullmsg}\DUrole{o,o}{=}\DUrole{default_value}{False}}\sphinxparamcomma \sphinxparam{\DUrole{n,n}{docread}\DUrole{o,o}{=}\DUrole{default_value}{False}}}{}
\pysigstopsignatures
\sphinxAtStartPar
Validates and SBML document returns three dictionaries, errors, warnings, other and a boolean indicating an invalid document:
\begin{itemize}
\item {} 
\sphinxAtStartPar
\sphinxstyleemphasis{D} and SBML document

\item {} 
\sphinxAtStartPar
\sphinxstyleemphasis{fullmsg} {[}default=False{]} optionally display the full error message

\end{itemize}

\end{fulllineitems}

\index{sbml\_writeAnnotationsAsCOBRANote() (in module cbmpy.CBXML)@\spxentry{sbml\_writeAnnotationsAsCOBRANote()}\spxextra{in module cbmpy.CBXML}}

\begin{fulllineitems}
\phantomsection\label{\detokenize{modules_doc:cbmpy.CBXML.sbml_writeAnnotationsAsCOBRANote}}
\pysigstartsignatures
\pysiglinewithargsret{\sphinxbfcode{\sphinxupquote{sbml\_writeAnnotationsAsCOBRANote}}}{\sphinxparam{\DUrole{n,n}{annotations}}}{}
\pysigstopsignatures
\sphinxAtStartPar
Writes the annotations dictionary as a COBRA compatible SBML \textless{}note\textgreater{}

\end{fulllineitems}

\index{sbml\_writeCOBRASBML() (in module cbmpy.CBXML)@\spxentry{sbml\_writeCOBRASBML()}\spxextra{in module cbmpy.CBXML}}

\begin{fulllineitems}
\phantomsection\label{\detokenize{modules_doc:cbmpy.CBXML.sbml_writeCOBRASBML}}
\pysigstartsignatures
\pysiglinewithargsret{\sphinxbfcode{\sphinxupquote{sbml\_writeCOBRASBML}}}{\sphinxparam{\DUrole{n,n}{fba}}\sphinxparamcomma \sphinxparam{\DUrole{n,n}{fname}}\sphinxparamcomma \sphinxparam{\DUrole{n,n}{directory}\DUrole{o,o}{=}\DUrole{default_value}{None}}}{}
\pysigstopsignatures
\sphinxAtStartPar
Takes an FBA model object and writes it to file as a COBRA compatible :
\begin{itemize}
\item {} 
\sphinxAtStartPar
\sphinxstyleemphasis{fba} an fba model object

\item {} 
\sphinxAtStartPar
\sphinxstyleemphasis{fname} the model will be written as XML to \sphinxstyleemphasis{fname}

\item {} 
\sphinxAtStartPar
\sphinxstyleemphasis{directory} {[}default=None{]} if defined it is prepended to fname

\end{itemize}

\end{fulllineitems}

\index{sbml\_writeKeyValueDataAnnotation() (in module cbmpy.CBXML)@\spxentry{sbml\_writeKeyValueDataAnnotation()}\spxextra{in module cbmpy.CBXML}}

\begin{fulllineitems}
\phantomsection\label{\detokenize{modules_doc:cbmpy.CBXML.sbml_writeKeyValueDataAnnotation}}
\pysigstartsignatures
\pysiglinewithargsret{\sphinxbfcode{\sphinxupquote{sbml\_writeKeyValueDataAnnotation}}}{\sphinxparam{\DUrole{n,n}{annotations}}}{}
\pysigstopsignatures
\sphinxAtStartPar
Writes the key:value annotations as a KeyValueData annotation (http://pysces.sourceforge.net/KeyValueData)

\end{fulllineitems}

\index{sbml\_writeSBML2FBA() (in module cbmpy.CBXML)@\spxentry{sbml\_writeSBML2FBA()}\spxextra{in module cbmpy.CBXML}}

\begin{fulllineitems}
\phantomsection\label{\detokenize{modules_doc:cbmpy.CBXML.sbml_writeSBML2FBA}}
\pysigstartsignatures
\pysiglinewithargsret{\sphinxbfcode{\sphinxupquote{sbml\_writeSBML2FBA}}}{\sphinxparam{\DUrole{n,n}{fba}}\sphinxparamcomma \sphinxparam{\DUrole{n,n}{fname}}\sphinxparamcomma \sphinxparam{\DUrole{n,n}{directory}\DUrole{o,o}{=}\DUrole{default_value}{None}}\sphinxparamcomma \sphinxparam{\DUrole{n,n}{sbml\_level\_version}\DUrole{o,o}{=}\DUrole{default_value}{None}}}{}
\pysigstopsignatures
\sphinxAtStartPar
Takes an FBA model object and writes it to file as SBML L3 FBA:
\begin{itemize}
\item {} 
\sphinxAtStartPar
\sphinxstyleemphasis{fba} an fba model object

\item {} 
\sphinxAtStartPar
\sphinxstyleemphasis{fname} the model will be written as XML to \sphinxstyleemphasis{fname}

\item {} 
\sphinxAtStartPar
\sphinxstyleemphasis{directory} {[}default=None{]} if defined it is prepended to fname

\item {} 
\sphinxAtStartPar
\sphinxstyleemphasis{sbml\_level\_version} {[}default=None{]} a tuple containing the SBML level and version e.g. (2,4) (ignored)

\end{itemize}

\end{fulllineitems}

\index{sbml\_writeSBML3FBC() (in module cbmpy.CBXML)@\spxentry{sbml\_writeSBML3FBC()}\spxextra{in module cbmpy.CBXML}}

\begin{fulllineitems}
\phantomsection\label{\detokenize{modules_doc:cbmpy.CBXML.sbml_writeSBML3FBC}}
\pysigstartsignatures
\pysiglinewithargsret{\sphinxbfcode{\sphinxupquote{sbml\_writeSBML3FBC}}}{\sphinxparam{\DUrole{n,n}{fba}}\sphinxparamcomma \sphinxparam{\DUrole{n,n}{fname}}\sphinxparamcomma \sphinxparam{\DUrole{n,n}{directory}\DUrole{o,o}{=}\DUrole{default_value}{None}}\sphinxparamcomma \sphinxparam{\DUrole{n,n}{sbml\_level\_version}\DUrole{o,o}{=}\DUrole{default_value}{(3, 1)}}\sphinxparamcomma \sphinxparam{\DUrole{n,n}{autofix}\DUrole{o,o}{=}\DUrole{default_value}{True}}\sphinxparamcomma \sphinxparam{\DUrole{n,n}{return\_fbc}\DUrole{o,o}{=}\DUrole{default_value}{False}}\sphinxparamcomma \sphinxparam{\DUrole{n,n}{gpr\_from\_annot}\DUrole{o,o}{=}\DUrole{default_value}{False}}\sphinxparamcomma \sphinxparam{\DUrole{n,n}{add\_groups}\DUrole{o,o}{=}\DUrole{default_value}{False}}\sphinxparamcomma \sphinxparam{\DUrole{n,n}{add\_cbmpy\_annot}\DUrole{o,o}{=}\DUrole{default_value}{True}}\sphinxparamcomma \sphinxparam{\DUrole{n,n}{add\_cobra\_annot}\DUrole{o,o}{=}\DUrole{default_value}{False}}\sphinxparamcomma \sphinxparam{\DUrole{n,n}{xoptions}\DUrole{o,o}{=}\DUrole{default_value}{\{\}}}}{}
\pysigstopsignatures
\sphinxAtStartPar
Takes an FBA model object and writes it to file as SBML L3 FBC:
\begin{itemize}
\item {} 
\sphinxAtStartPar
\sphinxstyleemphasis{fba} an fba model object

\item {} 
\sphinxAtStartPar
\sphinxstyleemphasis{fname} the model will be written as XML to \sphinxstyleemphasis{fname}

\item {} 
\sphinxAtStartPar
\sphinxstyleemphasis{directory} {[}default=None{]} if defined it is prepended to fname

\item {} 
\sphinxAtStartPar
\sphinxstyleemphasis{sbml\_level\_version} {[}default=(3,1){]} a tuple containing the SBML level and version e.g. (3,1)

\item {} 
\sphinxAtStartPar
\sphinxstyleemphasis{autofix} convert \textless{}\textgreater{} to \textless{}=\textgreater{}=

\item {} 
\sphinxAtStartPar
\sphinxstyleemphasis{return\_fbc} return the FBC converter instance

\item {} 
\sphinxAtStartPar
\sphinxstyleemphasis{gpr\_from\_annot} {[}default=False{]} if enabled will attempt to add the gene protein associations from the annotations
if no gene protein association objects exist

\item {} 
\sphinxAtStartPar
\sphinxstyleemphasis{add\_cbmpy\_annot} {[}default=True{]} add CBMPy KeyValueData annotation. Replaces \textless{}notes\textgreater{}

\item {} 
\sphinxAtStartPar
\sphinxstyleemphasis{add\_cobra\_annot} {[}default=True{]} add COBRA \textless{}notes\textgreater{} annotation

\item {} 
\sphinxAtStartPar
\sphinxstyleemphasis{xoptions} extended options
\begin{itemize}
\item {} 
\sphinxAtStartPar
\sphinxstyleemphasis{fbc\_version} {[}default=2{]} write SBML3FBC using version 1 (2013) or version 2 (2015) oe version (2023)

\item {} 
\sphinxAtStartPar
\sphinxstyleemphasis{validate} {[}default=False{]} validate the output SBML file

\item {} 
\sphinxAtStartPar
\sphinxstyleemphasis{compress\_bounds} {[}default=False{]} try compress output flux bound parameters

\item {} 
\sphinxAtStartPar
\sphinxstyleemphasis{zip\_model} {[}default=False{]} compress the model using ZIP encoding

\item {} 
\sphinxAtStartPar
\sphinxstyleemphasis{return\_model\_string} {[}default=False{]} return the SBML XML file as a string

\end{itemize}

\end{itemize}

\end{fulllineitems}

\index{setCBSBOterm() (in module cbmpy.CBXML)@\spxentry{setCBSBOterm()}\spxextra{in module cbmpy.CBXML}}

\begin{fulllineitems}
\phantomsection\label{\detokenize{modules_doc:cbmpy.CBXML.setCBSBOterm}}
\pysigstartsignatures
\pysiglinewithargsret{\sphinxbfcode{\sphinxupquote{setCBSBOterm}}}{\sphinxparam{\DUrole{n,n}{sbo}}\sphinxparamcomma \sphinxparam{\DUrole{n,n}{obj}}}{}
\pysigstopsignatures
\sphinxAtStartPar
Given an SBOterm from libSBML, add it to a CBMPy object
\begin{itemize}
\item {} 
\sphinxAtStartPar
\sphinxstyleemphasis{sbo} the sbo term string

\item {} 
\sphinxAtStartPar
\sphinxstyleemphasis{obj} the CBMPy Fbase derived object

\end{itemize}

\end{fulllineitems}

\index{with\_metaclass() (in module cbmpy.CBXML)@\spxentry{with\_metaclass()}\spxextra{in module cbmpy.CBXML}}

\begin{fulllineitems}
\phantomsection\label{\detokenize{modules_doc:cbmpy.CBXML.with_metaclass}}
\pysigstartsignatures
\pysiglinewithargsret{\sphinxbfcode{\sphinxupquote{with\_metaclass}}}{\sphinxparam{\DUrole{n,n}{meta}}\sphinxparamcomma \sphinxparam{\DUrole{o,o}{*}\DUrole{n,n}{bases}}}{}
\pysigstopsignatures
\sphinxAtStartPar
Create a base class with a metaclass.
Usage is: NewClass(with\_metaclass(MetaClass, BaseClass*)

\end{fulllineitems}

\index{xml\_addSBML2FBAFluxBound() (in module cbmpy.CBXML)@\spxentry{xml\_addSBML2FBAFluxBound()}\spxextra{in module cbmpy.CBXML}}

\begin{fulllineitems}
\phantomsection\label{\detokenize{modules_doc:cbmpy.CBXML.xml_addSBML2FBAFluxBound}}
\pysigstartsignatures
\pysiglinewithargsret{\sphinxbfcode{\sphinxupquote{xml\_addSBML2FBAFluxBound}}}{\sphinxparam{\DUrole{n,n}{document}}\sphinxparamcomma \sphinxparam{\DUrole{n,n}{rid}}\sphinxparamcomma \sphinxparam{\DUrole{n,n}{operator}}\sphinxparamcomma \sphinxparam{\DUrole{n,n}{value}}\sphinxparamcomma \sphinxparam{\DUrole{n,n}{fbid}\DUrole{o,o}{=}\DUrole{default_value}{None}}}{}
\pysigstopsignatures
\sphinxAtStartPar
Adds an SBML3FBA flux bound to the document:
\begin{itemize}
\item {} 
\sphinxAtStartPar
\sphinxstyleemphasis{document} a minidom XML document created by xml\_createSBML2FBADoc

\item {} 
\sphinxAtStartPar
\sphinxstyleemphasis{rid} the reaction id

\item {} 
\sphinxAtStartPar
\sphinxstyleemphasis{operator} one of {[}‘greater’,’greaterEqual’,’less’,’lessEqual’,’equal’,’\textgreater{}’,’\textgreater{}=’,’\textless{}’,’\textless{}=’,’=’{]}

\item {} 
\sphinxAtStartPar
\sphinxstyleemphasis{value} a float which will be cast to a string using str(value)

\item {} 
\sphinxAtStartPar
\sphinxstyleemphasis{fbid} the flux bound id, autogenerated by default

\end{itemize}

\end{fulllineitems}

\index{xml\_addSBML2FBAObjective() (in module cbmpy.CBXML)@\spxentry{xml\_addSBML2FBAObjective()}\spxextra{in module cbmpy.CBXML}}

\begin{fulllineitems}
\phantomsection\label{\detokenize{modules_doc:cbmpy.CBXML.xml_addSBML2FBAObjective}}
\pysigstartsignatures
\pysiglinewithargsret{\sphinxbfcode{\sphinxupquote{xml\_addSBML2FBAObjective}}}{\sphinxparam{\DUrole{n,n}{document}}\sphinxparamcomma \sphinxparam{\DUrole{n,n}{objective}}\sphinxparamcomma \sphinxparam{\DUrole{n,n}{active}\DUrole{o,o}{=}\DUrole{default_value}{True}}}{}
\pysigstopsignatures
\sphinxAtStartPar
Adds an objective element to the documents listOfObjectives and sets the active attribute:
\begin{itemize}
\item {} 
\sphinxAtStartPar
\sphinxstyleemphasis{document} a minidom XML document created by \sphinxtitleref{xml\_createSBML2FBADoc}

\item {} 
\sphinxAtStartPar
\sphinxstyleemphasis{objective} a minidom XML objective element created with \sphinxtitleref{xml\_createSBML2FBAObjective}

\item {} 
\sphinxAtStartPar
\sphinxstyleemphasis{active} {[}default=True{]} a boolean flag specifiying whether this objective is active

\end{itemize}

\end{fulllineitems}

\index{xml\_createListOfFluxObjectives() (in module cbmpy.CBXML)@\spxentry{xml\_createListOfFluxObjectives()}\spxextra{in module cbmpy.CBXML}}

\begin{fulllineitems}
\phantomsection\label{\detokenize{modules_doc:cbmpy.CBXML.xml_createListOfFluxObjectives}}
\pysigstartsignatures
\pysiglinewithargsret{\sphinxbfcode{\sphinxupquote{xml\_createListOfFluxObjectives}}}{\sphinxparam{\DUrole{n,n}{document}}\sphinxparamcomma \sphinxparam{\DUrole{n,n}{l}}}{}
\pysigstopsignatures
\sphinxAtStartPar
Create a list of fluxObjectives to add to an Objective:
\begin{itemize}
\item {} 
\sphinxAtStartPar
\sphinxstyleemphasis{document} a minidom XML document created by xml\_createSBML2FBADoc

\item {} 
\sphinxAtStartPar
\sphinxstyleemphasis{fluxobjs} a list of (rid, coefficient) tuples

\end{itemize}

\end{fulllineitems}

\index{xml\_createSBML2FBADoc() (in module cbmpy.CBXML)@\spxentry{xml\_createSBML2FBADoc()}\spxextra{in module cbmpy.CBXML}}

\begin{fulllineitems}
\phantomsection\label{\detokenize{modules_doc:cbmpy.CBXML.xml_createSBML2FBADoc}}
\pysigstartsignatures
\pysiglinewithargsret{\sphinxbfcode{\sphinxupquote{xml\_createSBML2FBADoc}}}{}{}
\pysigstopsignatures
\sphinxAtStartPar
Create a ‘document’ to store the SBML2FBA annotation, returns:
\begin{itemize}
\item {} 
\sphinxAtStartPar
\sphinxstyleemphasis{DOC} a minidom document

\end{itemize}

\end{fulllineitems}

\index{xml\_createSBML2FBAObjective() (in module cbmpy.CBXML)@\spxentry{xml\_createSBML2FBAObjective()}\spxextra{in module cbmpy.CBXML}}

\begin{fulllineitems}
\phantomsection\label{\detokenize{modules_doc:cbmpy.CBXML.xml_createSBML2FBAObjective}}
\pysigstartsignatures
\pysiglinewithargsret{\sphinxbfcode{\sphinxupquote{xml\_createSBML2FBAObjective}}}{\sphinxparam{\DUrole{n,n}{document}}\sphinxparamcomma \sphinxparam{\DUrole{n,n}{oid}}\sphinxparamcomma \sphinxparam{\DUrole{n,n}{sense}}\sphinxparamcomma \sphinxparam{\DUrole{n,n}{fluxObjectives}}}{}
\pysigstopsignatures
\sphinxAtStartPar
Create a list of fluxObjectives to add to an Objective:
\begin{itemize}
\item {} 
\sphinxAtStartPar
\sphinxstyleemphasis{document} a minidom XML document created by xml\_createSBML2FBADoc

\item {} 
\sphinxAtStartPar
\sphinxstyleemphasis{oid} the objective id

\item {} 
\sphinxAtStartPar
\sphinxstyleemphasis{sense} a string containing the objective sense either: \sphinxstylestrong{maximize} or \sphinxstylestrong{minimize}

\item {} 
\sphinxAtStartPar
\sphinxstyleemphasis{fluxObjectives} a list of (rid, coefficient) tuples

\end{itemize}

\end{fulllineitems}

\index{xml\_getSBML2FBAannotation() (in module cbmpy.CBXML)@\spxentry{xml\_getSBML2FBAannotation()}\spxextra{in module cbmpy.CBXML}}

\begin{fulllineitems}
\phantomsection\label{\detokenize{modules_doc:cbmpy.CBXML.xml_getSBML2FBAannotation}}
\pysigstartsignatures
\pysiglinewithargsret{\sphinxbfcode{\sphinxupquote{xml\_getSBML2FBAannotation}}}{\sphinxparam{\DUrole{n,n}{fba}}\sphinxparamcomma \sphinxparam{\DUrole{n,n}{fname}\DUrole{o,o}{=}\DUrole{default_value}{None}}}{}
\pysigstopsignatures
\sphinxAtStartPar
Takes an FBA model object and returns the SBML3FBA annotation as an XML string:
\begin{itemize}
\item {} 
\sphinxAtStartPar
\sphinxstyleemphasis{fba} an fba model object

\item {} 
\sphinxAtStartPar
\sphinxstyleemphasis{fname} {[}default=None{]} if supplied the XML will be written to file \sphinxstyleemphasis{fname}

\end{itemize}

\end{fulllineitems}

\index{xml\_stripTags() (in module cbmpy.CBXML)@\spxentry{xml\_stripTags()}\spxextra{in module cbmpy.CBXML}}

\begin{fulllineitems}
\phantomsection\label{\detokenize{modules_doc:cbmpy.CBXML.xml_stripTags}}
\pysigstartsignatures
\pysiglinewithargsret{\sphinxbfcode{\sphinxupquote{xml\_stripTags}}}{\sphinxparam{\DUrole{n,n}{html}}}{}
\pysigstopsignatures
\sphinxAtStartPar
Strip a string of HTML/XML, returns a string
\begin{itemize}
\item {} 
\sphinxAtStartPar
\sphinxstyleemphasis{html} the string containing html

\end{itemize}

\end{fulllineitems}

\index{xml\_viewSBML2FBAXML() (in module cbmpy.CBXML)@\spxentry{xml\_viewSBML2FBAXML()}\spxextra{in module cbmpy.CBXML}}

\begin{fulllineitems}
\phantomsection\label{\detokenize{modules_doc:cbmpy.CBXML.xml_viewSBML2FBAXML}}
\pysigstartsignatures
\pysiglinewithargsret{\sphinxbfcode{\sphinxupquote{xml\_viewSBML2FBAXML}}}{\sphinxparam{\DUrole{n,n}{document}}\sphinxparamcomma \sphinxparam{\DUrole{n,n}{fname}\DUrole{o,o}{=}\DUrole{default_value}{None}}}{}
\pysigstopsignatures
\sphinxAtStartPar
Print a minidom XML document to screen or file, arguments:
\begin{itemize}
\item {} 
\sphinxAtStartPar
\sphinxstyleemphasis{document} a minidom XML document

\item {} 
\sphinxAtStartPar
\sphinxstyleemphasis{fname} {[}default=None{]} by default print to screen or write to file fname

\end{itemize}

\end{fulllineitems}

\phantomsection\label{\detokenize{modules_doc:module-cbmpy._multicorefva}}\index{module@\spxentry{module}!cbmpy.\_multicorefva@\spxentry{cbmpy.\_multicorefva}}\index{cbmpy.\_multicorefva@\spxentry{cbmpy.\_multicorefva}!module@\spxentry{module}}

\section{CBMPy: MultiCoreFVA module}
\label{\detokenize{modules_doc:cbmpy-multicorefva-module}}
\sphinxAtStartPar
PySCeS Constraint Based Modelling (\sphinxurl{http://cbmpy.sourceforge.net})
Copyright (C) 2009\sphinxhyphen{}2024 Brett G. Olivier, VU University Amsterdam, Amsterdam, The Netherlands

\sphinxAtStartPar
This program is free software: you can redistribute it and/or modify
it under the terms of the GNU General Public License as published by
the Free Software Foundation, either version 3 of the License, or
(at your option) any later version.

\sphinxAtStartPar
This program is distributed in the hope that it will be useful,
but WITHOUT ANY WARRANTY; without even the implied warranty of
MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
GNU General Public License for more details.

\sphinxAtStartPar
You should have received a copy of the GNU General Public License
along with this program.  If not, see \textless{}\sphinxurl{http://www.gnu.org/licenses/}\textgreater{}

\sphinxAtStartPar
Author: Brett G. Olivier PhD
Contact developers: \sphinxurl{https://github.com/SystemsBioinformatics/cbmpy/issues}
Last edit: \$Author: bgoli \$ (\$Id: \_multicorefva.py 710 2020\sphinxhyphen{}04\sphinxhyphen{}27 14:22:34Z bgoli \$)
\phantomsection\label{\detokenize{modules_doc:module-cbmpy._multicoreenvfva}}\index{module@\spxentry{module}!cbmpy.\_multicoreenvfva@\spxentry{cbmpy.\_multicoreenvfva}}\index{cbmpy.\_multicoreenvfva@\spxentry{cbmpy.\_multicoreenvfva}!module@\spxentry{module}}

\section{CBMPy: MultiCoreEnvFVA module}
\label{\detokenize{modules_doc:cbmpy-multicoreenvfva-module}}
\sphinxAtStartPar
PySCeS Constraint Based Modelling (\sphinxurl{http://cbmpy.sourceforge.net})
Copyright (C) 2009\sphinxhyphen{}2024 Brett G. Olivier, VU University Amsterdam, Amsterdam, The Netherlands

\sphinxAtStartPar
This program is free software: you can redistribute it and/or modify
it under the terms of the GNU General Public License as published by
the Free Software Foundation, either version 3 of the License, or
(at your option) any later version.

\sphinxAtStartPar
This program is distributed in the hope that it will be useful,
but WITHOUT ANY WARRANTY; without even the implied warranty of
MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
GNU General Public License for more details.

\sphinxAtStartPar
You should have received a copy of the GNU General Public License
along with this program.  If not, see \textless{}\sphinxurl{http://www.gnu.org/licenses/}\textgreater{}

\sphinxAtStartPar
Author: Brett G. Olivier PhD
Contact developers: \sphinxurl{https://github.com/SystemsBioinformatics/cbmpy/issues}
Last edit: \$Author: bgoli \$ (\$Id: \_multicoreenvfva.py 710 2020\sphinxhyphen{}04\sphinxhyphen{}27 14:22:34Z bgoli \$)
\phantomsection\label{\detokenize{modules_doc:module-cbmpy.miriamids}}\index{module@\spxentry{module}!cbmpy.miriamids@\spxentry{cbmpy.miriamids}}\index{cbmpy.miriamids@\spxentry{cbmpy.miriamids}!module@\spxentry{module}}

\chapter{Indices and tables}
\label{\detokenize{cbmpy:indices-and-tables}}\begin{itemize}
\item {} 
\sphinxAtStartPar
\DUrole{xref,std,std-ref}{genindex}

\item {} 
\sphinxAtStartPar
\DUrole{xref,std,std-ref}{modindex}

\item {} 
\sphinxAtStartPar
\DUrole{xref,std,std-ref}{search}

\end{itemize}


\renewcommand{\indexname}{Python Module Index}
\begin{sphinxtheindex}
\let\bigletter\sphinxstyleindexlettergroup
\bigletter{c}
\item\relax\sphinxstyleindexentry{cbmpy.\_multicoreenvfva}\sphinxstyleindexpageref{modules_doc:\detokenize{module-cbmpy._multicoreenvfva}}
\item\relax\sphinxstyleindexentry{cbmpy.\_multicorefva}\sphinxstyleindexpageref{modules_doc:\detokenize{module-cbmpy._multicorefva}}
\item\relax\sphinxstyleindexentry{cbmpy.CBCommon}\sphinxstyleindexpageref{modules_doc:\detokenize{module-cbmpy.CBCommon}}
\item\relax\sphinxstyleindexentry{cbmpy.CBConfig}\sphinxstyleindexpageref{modules_doc:\detokenize{module-cbmpy.CBConfig}}
\item\relax\sphinxstyleindexentry{cbmpy.CBCPLEX}\sphinxstyleindexpageref{modules_doc:\detokenize{module-cbmpy.CBCPLEX}}
\item\relax\sphinxstyleindexentry{cbmpy.CBDataStruct}\sphinxstyleindexpageref{modules_doc:\detokenize{module-cbmpy.CBDataStruct}}
\item\relax\sphinxstyleindexentry{cbmpy.CBGUI}\sphinxstyleindexpageref{modules_doc:\detokenize{module-cbmpy.CBGUI}}
\item\relax\sphinxstyleindexentry{cbmpy.CBModel}\sphinxstyleindexpageref{modules_doc:\detokenize{module-cbmpy.CBModel}}
\item\relax\sphinxstyleindexentry{cbmpy.CBModelTools}\sphinxstyleindexpageref{modules_doc:\detokenize{module-cbmpy.CBModelTools}}
\item\relax\sphinxstyleindexentry{cbmpy.CBMultiCore}\sphinxstyleindexpageref{modules_doc:\detokenize{module-cbmpy.CBMultiCore}}
\item\relax\sphinxstyleindexentry{cbmpy.CBMultiEnv}\sphinxstyleindexpageref{modules_doc:\detokenize{module-cbmpy.CBMultiEnv}}
\item\relax\sphinxstyleindexentry{cbmpy.CBNetDB}\sphinxstyleindexpageref{modules_doc:\detokenize{module-cbmpy.CBNetDB}}
\item\relax\sphinxstyleindexentry{cbmpy.CBPlot}\sphinxstyleindexpageref{modules_doc:\detokenize{module-cbmpy.CBPlot}}
\item\relax\sphinxstyleindexentry{cbmpy.CBRead}\sphinxstyleindexpageref{modules_doc:\detokenize{module-cbmpy.CBRead}}
\item\relax\sphinxstyleindexentry{cbmpy.CBReadtxt}\sphinxstyleindexpageref{modules_doc:\detokenize{module-cbmpy.CBReadtxt}}
\item\relax\sphinxstyleindexentry{cbmpy.CBSolver}\sphinxstyleindexpageref{modules_doc:\detokenize{module-cbmpy.CBSolver}}
\item\relax\sphinxstyleindexentry{cbmpy.CBTools}\sphinxstyleindexpageref{modules_doc:\detokenize{module-cbmpy.CBTools}}
\item\relax\sphinxstyleindexentry{cbmpy.CBWrite}\sphinxstyleindexpageref{modules_doc:\detokenize{module-cbmpy.CBWrite}}
\item\relax\sphinxstyleindexentry{cbmpy.CBWx}\sphinxstyleindexpageref{modules_doc:\detokenize{module-cbmpy.CBWx}}
\item\relax\sphinxstyleindexentry{cbmpy.CBXML}\sphinxstyleindexpageref{modules_doc:\detokenize{module-cbmpy.CBXML}}
\item\relax\sphinxstyleindexentry{cbmpy.miriamids}\sphinxstyleindexpageref{modules_doc:\detokenize{module-cbmpy.miriamids}}
\end{sphinxtheindex}

\renewcommand{\indexname}{Index}
\printindex
\end{document}